---
title: "zymo reference build"
output:
  html_notebook:
    code_fold: hide
---

```{r setup, include=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(plyr) 
library(ggplot2) 
```

zymo infomation : `stLFR4META - 文档\Experiments\ZymoBIOMICS Microbial Community DNA Standard D6305`

Zymo mock strains:

| Species                  | Genomic DNA | 16S Only1 | 16S & 18S1 | Genome Copy2 | Cell Number3 |
| ------------------------ | ----------- | --------- | ---------- | ------------ | ------------ |
| Pseudomonas aeruginosa   | 12          | 4.2       | 3.6        | 6.1          | 6.1          | +
| Escherichia coli         | 12          | 10.1      | 8.9        | 8.5          | 8.5          | +
| Salmonella enterica      | 12          | 10.4      | 9.1        | 8.7          | 8.8          | +
| Lactobacillus fermentum  | 12          | 18.4      | 16.1       | 21.6         | 21.9         |
| Enterococcus faecalis    | 12          | 9.9       | 8.7        | 14.6         | 14.6         | +
| Staphylococcus aureus    | 12          | 15.5      | 13.6       | 15.2         | 15.3         | +
| Listeria monocytogenes   | 12          | 14.1      | 12.4       | 13.9         | 13.9         | +
| Bacillus subtilis        | 12          | 17.4      | 15.3       | 10.3         | 10.3         | +
| Saccharomyces cerevisiae | 2           | NA        | 9.3        | 0.57         | 0.29         |
| Cryptococcus neoformans  | 2           | NA        | 3.3        | 0.37         | 0.18         |


set global vars (bash):
```{bash, eval=FALSE}
SAM=RCA197B01_2

threads=16
DB=../../Source/REF/zymo
RES=../../Results/ZYMO
```
A list of manuals for check:
> Samtools: http://www.htslib.org/doc/samtools.html
> bwa: http://bio-bwa.sourceforge.net/bwa.shtml
> bed format: https://asia.ensembl.org/info/website/upload/bed.html

Mapping 16S, 28S sequences back to genome.
```
bwa mem -t $threads -a $DB/D6305.genomes.bwa $DB/D6305.ssrRNA.fasta |samtools view -b \
| samtools sort > $RES/D6305.ssrRNA2genome.bam

bwa mem -t $threads -a $DB/D6305.genomes.bwa $DB/D6305.markers.fa |samtools view -b \
| samtools sort > $RES/D6305.16S_23S.2genome.bam

samtools view $RES/D6305.16S_23S.2genome.bam| perl sam2bed.pl -s - -o $RES/D6305.ssrRNA2genome.bed
```

call snv:
```
grep --no-group-separator -A1 "16S_1" $DB/D6305.ssrRNA.fasta > $RES/D6305.16S_1.fasta
mafft --thread 32 --auto --inputorder $RES/D6305.16S_1.fasta > $RES/D6305.16S_1.mafft.fa
```


# Benchmark of zymo variances:

Find rDNA rigions from ref:
```{bash, eval=FALSE}
bwa bwasw ../../Source/REF/zymo/D6305.genomes.bwa ../../Source/REF/zymo/D6305.rRNA.fa \
 > ../../Source/REF/zymo/D6305.rRNA2genomes.bwa
```
Find E.coli SSU positions:
```{bash,eval=FALSE}
bwa mem ../../Source/REF/zymo/D6305.genomes.bwa ../../Source/REF/zymo/D6305.ssrRNA.fasta |awk '{$10="";print $0}'|grep "coli"
[M::bwa_idx_load_from_disk] read 0 ALT contigs
@SQ SN:Escherichia_coli_plasmid LN:110007
@SQ SN:Escherichia_coli_chromosome LN:4765434
[M::process] read 52 sequences (80526 bp)...
[M::mem_process_seqs] Processed 52 reads in 0.671 CPU sec, 0.672 real sec
Escherichia_coli_16S_1 16 Escherichia_coli_chromosome 3342577 0 1542M * 0 0  * NM:i:0 MD:Z:1542 AS:i:1542 XS:i:1542
Escherichia_coli_16S_2 16 Escherichia_coli_chromosome 3248951 0 1542M * 0 0  * NM:i:0 MD:Z:1542 AS:i:1542 XS:i:1542
Escherichia_coli_16S_3 16 Escherichia_coli_chromosome 3076422 0 1542M * 0 0  * NM:i:0 MD:Z:1542 AS:i:1542 XS:i:1542
Escherichia_coli_16S_4 0 Escherichia_coli_chromosome 4591467 10 1542M * 0 0  * NM:i:0 MD:Z:1542 AS:i:1542 XS:i:1532   XA:Z:Escherichia_coli_chromosome,-2428583,1542M,5;
Escherichia_coli_16S_5 16 Escherichia_coli_chromosome 3342577 0 1542M * 0 0  * NM:i:0 MD:Z:1542 AS:i:1542 XS:i:1542
Escherichia_coli_16S_7 16 Escherichia_coli_chromosome 3248951 0 1542M * 0 0  * NM:i:0 MD:Z:1542 AS:i:1542 XS:i:1542
Escherichia_coli_16S_8 0 Escherichia_coli_chromosome 3883127 0 1542M * 0 0  * NM:i:0 MD:Z:1542 AS:i:1542 XS:i:1542
[main] Version: 0.7.17-r1198-dirty
[main] CMD: bwa mem ../../Source/REF/zymo/D6305.genomes.bwa ../../Source/REF/zymo/D6305.ssrRNA.fasta
[main] Real time: 1.836 sec; CPU: 1.367 sec
```


Find E.coli LSU positions:
```{bash,eval=FALSE}
bwa mem -h 15 ../../Source/REF/zymo/D6305.genomes.bwa ../../Source/REF/zymo/D6305.markers.LSU.fa |awk '{$10="";print $0}'|grep "coli"
[M::bwa_idx_load_from_disk] read 0 ALT contigs
@SQ SN:Escherichia_coli_plasmid LN:110007
@SQ SN:Escherichia_coli_chromosome LN:4765434
[M::process] read 9 sequences (24313 bp)...
[M::mem_process_seqs] Processed 9 reads in 0.338 CPU sec, 0.348 real sec
Escherichia_coli_23S 16 Escherichia_coli_chromosome 3114568 0 2904M * 0 0  * NM:i:4 MD:Z:1039G134A6T548C1173 AS:i:2884 XS:i:2884 XA:Z:Escherichia_coli_chromosome,-3339319,2904M,4;Escherichia_coli_chromosome,+3885106,2904M,4;Escherichia_coli_chromosome,-2425233,2904M,4;Escherichia_coli_chromosome,-3245601,2904M,4;Escherichia_coli_chromosome,+4593440,2904M,4;Escherichia_coli_chromosome,-3073164,2904M,4;Salmonella_enterica_complete_genome,-2002972,1170M4I3M4D216M6I4M8D959M2D542M,79;Salmonella_enterica_complete_genome,+2775089,542M2D959M8D4M6I213M4D6M4I1170M,82;Salmonella_enterica_complete_genome,-1336705,1170M4I3M4D216M6I4M8D959M2D542M,82;Salmonella_enterica_complete_genome,+2524769,542M2D959M8D4M6I213M4D6M4I1170M,83;Salmonella_enterica_complete_genome,+2817845,542M2D959M8D4M6I213M4D6M4I1170M,83;Salmonella_enterica_complete_genome,+2619699,542M2D959M8D4M6I213M4D6M4I697M2D9M1I463M,86;
[main] Version: 0.7.17-r1198-dirty
[main] CMD: bwa mem -h 15 ../../Source/REF/zymo/D6305.genomes.bwa ../../Source/REF/zymo/D6305.markers.LSU.fa
[main] Real time: 1.283 sec; CPU: 0.960 sec
```

##gff-version 3
Escherichia_coli_chromosome     manually        rRNA_16S        2428583 2430125 0       -       .       Name=16S_rRNA;product=16S ribosomal RNA
Escherichia_coli_chromosome     manually        rRNA_23S        2425233 2428137 0       -       .       Name=23S_rRNA;product=23S ribosomal RNA
Escherichia_coli_chromosome     manually        rRNA_16S        3076422 3077964 0       -       .       Name=16S_rRNA;product=16S ribosomal RNA
Escherichia_coli_chromosome     manually        rRNA_23S        3073164 3076068 0       -       .       Name=23S_rRNA;product=23S ribosomal RNA
Escherichia_coli_chromosome     manually        rRNA_16S        3117826 3119368 0       -       .       Name=16S_rRNA;product=16S ribosomal RNA
Escherichia_coli_chromosome     manually        rRNA_23S        3114568 3117472 0       -       .       Name=23S_rRNA;product=23S ribosomal RNA
Escherichia_coli_chromosome     manually        rRNA_16S        3248951 3250493 0       -       .       Name=16S_rRNA;product=16S ribosomal RNA
Escherichia_coli_chromosome     manually        rRNA_23S        3245601 3248505 0       -       .       Name=23S_rRNA;product=23S ribosomal RNA
Escherichia_coli_chromosome     manually        rRNA_16S        3342577 3344119 0       -       .       Name=16S_rRNA;product=16S ribosomal RNA
Escherichia_coli_chromosome     manually        rRNA_23S        3339319 3342223 0       -       .       Name=23S_rRNA;product=23S ribosomal RNA
Escherichia_coli_chromosome     manually        rRNA_16S        3883127 3884669 0       +       .       Name=16S_rRNA;product=16S ribosomal RNA
Escherichia_coli_chromosome     manually        rRNA_23S        3885106 3888010 0       +       .       Name=23S_rRNA;product=23S ribosomal RNA
Escherichia_coli_chromosome     manually        rRNA_16S        4591467 4593009 0       +       .       Name=16S_rRNA;product=16S ribosomal RNA
Escherichia_coli_chromosome     manually        rRNA_23S        4593440 4596344 0       +       .       Name=23S_rRNA;product=23S ribosomal RNA

```{bash,eval=FALSE}
samtools view ../../Results/ZYMO/D6305.16S_23S.2genome.bam|perl script.bam2gff.pl > ../../Source/REF/zymo/D6305.rRNA.gff
```

```{bash, eval=FALSE}
for i in Z{1,2,3};do
  bwa mem -t 16 ../../Source/REF/zymo/D6305.genomes.bwa SAM/$i/clean/fastp.sort.{1,2}.fq.gz | samtools view -b -o SAM/$i/bwa/map2D6305.genome.bam
done

for i in Z{1,2,3};do
  samtools sort SAM/$i/bwa/map2D6305.genome.bam -o SAM/$i/bwa/map2D6305.genome.sort.bam &
done


for i in Z{1,2,3};do
  samtools index SAM/$i/bwa/map2D6305.genome.sort.bam &
done

for i in Z{1,2,3};do
  bcftools mpileup -f ../../Source/REF/zymo/D6305.genomes.fasta \
   SAM/$i/bwa/map2D6305.genome.sort.bam > SAM/$i/bwa/map2D6305.genome.vcf &
done


#
bcftools mpileup -d 10000 -f ../../Source/REF/zymo/D6305.genomes.fasta \
  -a DP,AD -Ov -o Z.map2D6305.genome.vcf SAM/Z{1,2,3}/bwa/map2D6305.genome.sort.bam 

#
bcftools call -c --ploidy 1 -Oz -o Z.map2D6305.genome.call.vcf Z.map2D6305.genome.vcf

bcftools mpileup --threads 8 -d 10000 -Ou -f ../../Source/REF/zymo/D6305.genomes.fasta \
  SAM/Z{1,2,3}/bwa/map2D6305.genome.sort.bam |bcftools call --threads 8 -c --ploidy 1 -Oz \
  -o Z.map2D6305.genome.call.vcf
  

bcftools mpileup  -d 10000 -f ../../Source/REF/zymo/D6305.genomes.fasta.gz \
  --threads 36 -a DP,AD -Ov -o vcf/zymo.mpileup.vcf \
  SAM/Z{1,2,3}/bwa/map2D6305.genome.sort.bam &> sam2vcf.log

```

## if filter
```{bash, eval=FALSE}
# if filter
for i in Z{1,2,3};do
  samtools view -@ 8 -e '[NM]<=15' SAM/$i/bwa/map2D6305.genome.sort.bam -o SAM/$i/bwa/map2D6305.genome.sort.filter.bam &
done
bcftools mpileup --threads 4 -d 10000 -Ou -f ../../Source/REF/zymo/D6305.genomes.fasta \
  SAM/Z{1,2,3}/bwa/map2D6305.genome.sort.filter.bam |bcftools call --threads 4 -c --ploidy 1 -Oz \
  -o Z.map2D6305.genome.call.vcf

bcftools mpileup  -d 10000 -f ../../Source/REF/zymo/D6305.genomes.fasta.gz \
  --threads 8 -a DP,AD -Ov -o vcf/zymo.mpileup2.vcf SAM/Z{1,2,3}/bwa/map2D6305.genome.sort.filter.bam &> sam2vcf2.log &
```

## for PBA alignment
```{bash, eval=FALSE}
for i in Z{1,2};do
  bwa bwasw -t 8 -b5 -q2 -r1 -z10 ../../Source/REF/zymo/D6305.genomes.bwa SAM/$i/summary.BI.megahit.clip.all.fasta > SAM/$i/bwa/PBA2genome.sam & 
done
  
for i in Z{1,2};do
  samtools sort SAM/$i/bwa/PBA2genome.sam -o SAM/$i/bwa/PBA2genome.sort.bam 
  samtools view -@ 8 -e '[NM]<=15' -o SAM/$i/bwa/PBA2genome.filter.bam SAM/$i/bwa/PBA2genome.sort.bam &
done

bcftools mpileup -d 10000 -f ../../Source/REF/zymo/D6305.genomes.fasta -a DP,AD \
  -Ov -o vcf/PBA2D6305.genome.mpileup.vcf SAM/Z{1,2}/bwa/PBA2genome.sort.bam 

bcftools mpileup -d 10000 -f ../../Source/REF/zymo/D6305.genomes.fasta \
  -a DP,AD -Ov -o vcf/PBA2D6305.genome.mpileup2.vcf SAM/Z{1,2}/bwa/PBA2genome.filter.bam 
```


More about [`vcfR`](https://cran.r-project.org/web/packages/vcfR/vignettes/intro_to_vcfR.html)
```{r}
library(vcfR)

vcf <- read.vcfR("vcf/Z.map2D6305.genome.call.vcf", verbose = FALSE )
gff <- read.table("../../Source/REF/zymo/D6305.Ecoli.rRNA.gff", skip=1,quote="",header=F)
dna <- ape::read.dna(gzfile("../../Source/REF/zymo/D6305.genomes.fasta.gz"), format = "fasta")
```

# Escherichia_coli
```{r}
#myChroms[grep("Escherichia",myChroms)]
EC.vcf <- vcf[getCHROM(vcf) == "Escherichia_coli_chromosome", ]
EC.gff <- gff[gff$V1=="Escherichia_coli_chromosome",]

#mySeqs[grep("Escherichia",mySeqs)]
EC.seq <- dna["Escherichia_coli_chromosome"]
EC.chrom <- create.chromR(name='Escherichia_coli', vcf=EC.vcf, seq=EC.seq,ann=EC.gff)
```
```{r}
#EC.chrom.f <- masker(EC.chrom, min_QUAL = 1, min_MQ=59.9, min_DP = 10, max_DP = 1000)
EC.chrom.p <- proc.chromR(EC.chrom, win.size=100,verbose=TRUE)

plot(EC.chrom.p)
```
```{r}
chromoqc(EC.chrom.p, dp.alpha=20)
```
## rRNA_1
```{r}
chromoqc(EC.chrom.p, xlim=c(2422000, 2431000))

```
## rRNA_2
```{r}
chromoqc(EC.chrom.p, xlim=c(3071164, 3078068))
```

## rRNA_3
```{r}
chromoqc(EC.chrom.p, xlim=c(3114068, 3119868))

```


## rRNA_4
```{r}
chromoqc(EC.chrom.p, xlim=c(3245001, 3250993))

```

## rRNA_5
```{r}
chromoqc(EC.chrom.p, xlim=c(3339019, 3344619))

```
## rRNA_6
```{r}
chromoqc(EC.chrom.p, xlim=c(3883027, 3888510))

```
## rRNA_7
```{r}
chromoqc(EC.chrom.p, xlim=c(4591067, 4596844))
```



# bedtools
```{bash,eval=FALSE}
bedtools genomecov -bga -ibam SAM/Z3/bwa/map2D6305.genome.sort.bam -scale 0.009941 | sort -k1,1 -k2,2n > Z3.bedGraph
```

# Public dataset

SRR10013610 16S pacbio

SRR7415647    Hiseq4000 data for E. coli
SRR7498044    PacBio RSII data for E. coli
SRR8137545    WGBS of E. coli
SRR8154667    PacBio RSII sequencing of E. coli K12
SRR8154668    PacBio RSII sequencing of E. coli K12
SRR8154669    PacBio RSII sequencing of E. coli K12
SRR8154670    R9 MinION sequencing of E. coli K12, lambda phage, and mouse: Mason Lab
SRR8154671    R9 MinION sequencing of E. coli K12, lambda phage, and mouse: Mason Lab
SRR8154672    WGBS of E. coli K12

```{bash,eval=FALSE}
module load bwa/0.7.17-gcc-11.2.0

salloc -N 1 -n 16 -p 64c512g
ssh xxx

# bwa index ../../Source/REF/zymo/D6305.genomes.fasta.gz
bwa bwasw -t 8 -b5 -q2 -r1 -z10 ../../Source/REF/zymo/D6305.genomes.fasta.gz SEQDATA/SRR10013610.fastq > bwa/pacbio2genome.sam


perl -ane 'BEGIN{$pre=""};if($_=~/^@/){print $_}else{$F[11]=~/AS:i:(\d+)/;
  $AS=$1; if($AS>999 && $pre ne $F[0]){print $_};$pre=$F[0]}' \
  bwa/pacbio2genome.sam | samtools view -b |\
  samtools sort -o bwa/pacbio2genome.bam

samtools index bwa/pacbio2genome.bam

bcftools mpileup -d 10000 -Ou -f ../../Source/REF/zymo/D6305.genomes.fasta.gz \
  bwa/pacbio2genome.bam | bcftools call -m --ploidy 1 -Oz -o bwa/pacbio2genome.call.m.vcf

bcftools mpileup -d 10000 -a AD -f ../../Source/REF/zymo/D6305.genomes.fasta.gz 

for i in `awk '/16S/{print $4"-"$4+50}' ../../Source/REF/zymo/D6305.Ecoli.rRNA.gff`;do samtools faidx  ../../Source/REF/zymo/D6305.genomes.fasta.gz Escherichia_coli_chromosome:$i;done
```

```{bash,eval=FALSE}
mkdir tmp
a=1
for i in SEQDATA/*.fastq;do
  b=$(basename -s .fastq $i); echo bwa bwasw -t 8 \
  ../../Source/REF/zymo/D6305.genomes.fasta.gz \
  SEQDATA/$b.fastq \> bwa/$b\ToGenome.sam > tmp/$a.sh; a=$((a+1));
done
 
sbatch scripts/bwa.slurm

for i in SEQDATA/*.fastq;do
  echo $i;
  b=$(basename -s .fastq $i); sh tmp/sam2vcf.sh $b;
done > sam2vcf.log &

# bcftools

```

`tmp/sam2vcf.sh`:
```{bash,eval=FALSE}
basename=$1

perl -ane 'BEGIN{$pre=""};if($_=~/^@/){print $_}else{$F[11]=~/AS:i:(\d+)/;
  $AS=$1; if($AS>999 && $pre ne $F[0]){print $_};$pre=$F[0]}' \
  bwa/$basename\ToGenome.sam | samtools view -b |\
  samtools sort -o bwa/$basename\ToGenome.bam

samtools view -b bwa/$basename\ToGenome.sam|\
  samtools sort -o bwa/$basename\ToGenome.bam
  
samtools index bwa/$basename\ToGenome.bam

bcftools mpileup -d 10000 -f ../../Source/REF/zymo/D6305.genomes.fasta.gz \
  --threads 8 -a DP,AD -Ov -o vcf/$basename.mpileup.vcf bwa/$basename\ToGenome.bam 

k=1
for i in {SRR10013610,SRR7415647,SRR7498044,SRR8137545,SRR8154667,SRR8154668,SRR8154669,SRR8154670,SRR8154671,SRR8154672};do
echo sh tmp/sam2vcf.sh $i > tmp/sort.$k.sh; k=$((k+1)); done

for i in {SRR10013610,SRR7415647,SRR7498044,SRR8137545,SRR8154667,SRR8154668,SRR8154669,SRR8154670,SRR8154671,SRR8154672};do
  perl script.vcf2tsv.pl ../../Source/REF/zymo/D6305.Ecoli.rRNA.gff vcf/$i.mpileup.vcf > vcf/$i.vcfStat.tsv &
done


```

or:
```{bash,eval=FALSE}
bcftools mpileup  -d 10000 -f ../../Source/REF/zymo/D6305.genomes.fasta.gz \
  --threads 36 -a DP,AD -Ov -o vcf/pub.mpileup.vcf bwa/*.bam &> sam2vcf.log

perl script.vcf2tsv.pl ../../Source/REF/zymo/D6305.rRNA.gff vcf/pub.mpileup.vcf > vcf/pub.vcfStat.tsv

perl script.vcf2tsv.pl ../../Source/REF/zymo/D6305.rRNA.gff vcf/zymo.mpileup.vcf > vcf/zymo.vcfStat.tsv

perl script.vcf2tsv.pl ../../Source/REF/zymo/D6305.rRNA.gff vcf/PBA2D6305.genome.mpileup.vcf > vcf/PBA2D6305.genome.vcfStat.vcf

# if filter 
for i in {SRR10013610,SRR7415647,SRR7498044,SRR8137545,SRR8154667,SRR8154668,SRR8154669,SRR8154670,SRR8154671,SRR8154672};do
  samtools view -e '[NM]<=15' bwa/$i\ToGenome.bam -o bwa/$i\ToGenome.filter.bam &
done

bcftools mpileup  -d 10000 -f ../../Source/REF/zymo/D6305.genomes.fasta.gz \
  --threads 36 -a DP,AD -Ov -o vcf/pub.mpileup2.vcf bwa/*.filter.bam &> sam2vcf2.log


perl script.vcf2tsv.pl ../../Source/REF/zymo/D6305.rRNA.gff vcf/pub.mpileup2.vcf > vcf/pub.vcfStat2.tsv

perl script.vcf2tsv.pl ../../Source/REF/zymo/D6305.rRNA.gff vcf/zymo.mpileup2.vcf > vcf/zymo.vcfStat2.tsv

perl script.vcf2tsv.pl ../../Source/REF/zymo/D6305.rRNA.gff vcf/PBA2D6305.genome.mpileup.vcf > vcf/PBA2D6305.genome.vcfStat2.vcf
```

```{bash,eval=FALSE}
CS <- function(x){
  sapply(x,FUN=function(y){
    z <- ifelse(y=="A","T",ifelse(y=="T","A",ifelse(y=="C","G",ifelse(y=="G","C",""))))
    return(z)
  })
}
```

# Load & Visualization

## No filter
```{r}
pub.stat.df <- read.table("vcf/pub.vcfStat.tsv",fill = NA,
                            col.names = c("sample","region","POS","rPOS","REF","ALT","TYPE","DP","AD"))
zymo.stat.df <- read.table("vcf/zymo.vcfStat.tsv",fill = NA,
                            col.names = c("sample","region","POS","rPOS","REF","ALT","TYPE","DP","AD"))

zymo.PBA.stat.df <- read.table("vcf/PBA2D6305.genome.vcfStat.vcf",fill = NA,
                            col.names = c("sample","region","POS","rPOS","REF","ALT","TYPE","DP","AD"))

rRNA.stat.df <- rbind(zymo.stat.df,pub.stat.df,zymo.PBA.stat.df)

sp1 <- strsplit(rRNA.stat.df$region,"|",fixed = T)
sp2 <- strsplit(sapply(sp1,"[",2),"_")

rRNA.stat.df$CHR <- sapply(sp1,"[",1)
rRNA.stat.df$UNIT <- sapply(sp2,"[",1)
rRNA.stat.df$UID <- sapply(sp2,"[",2)

rRNA.stat.df$sam <- as.factor(rRNA.stat.df$sample)
rRNA.stat.df$sample <- as.factor(rRNA.stat.df$sample)

levels(rRNA.stat.df$sample) <- c(
  "Z1(Reads)","Z1(PBA)","Z2(Reads)","Z2(PBA)","Z3(Reads)",
  "PacBio RSII(610)","HiSeq4000(647)","PacBio RSII(044)","WGBS(545)",
  "PacBio RSII(667)","PacBio RSII(668)","PacBio RSII(669)",
  "R9 MinION(670)","R9 MinION(671)", "WGBS(672)")

levels(rRNA.stat.df$sam) <- c(
  "stcLFR(Reads)","stcLFR(PBA)","stcLFR(Reads)","stcLFR(PBA)","stcLFR(Reads)",
  "PacBio RSII","HiSeq4000","PacBio RSII","HiSeq2500",
  "PacBio RSII","PacBio RSII","PacBio RSII",
  "R9 MinION","R9 MinION", "HiSeq2500"
)

rRNA.stat.df$sam <- factor(as.character(rRNA.stat.df$sam),levels = c(
  "stcLFR(Reads)","HiSeq2500","HiSeq4000",
  "stcLFR(PBA)","PacBio RSII","R9 MinION"
))
```

```{r}
rRNA.stat.df$win <- round(rRNA.stat.df$rPOS/50,0)*50
rRNA.stat.df3 <- ddply(rRNA.stat.df,c("sam","UNIT","win","TYPE"),summarize,
                       DP=sum(DP),AD=sum(AD))

ggplot(rRNA.stat.df3%>%filter(UNIT%in%c("16S","23S"))) +
  geom_hline(yintercept = -2.5, linetype=2) +
  geom_point(aes(y=log10(AD/DP),x=win,color=TYPE,shape=TYPE)) +
  facet_grid(UNIT~sam) + ylab("Sequence error rate(log10 scale)") + 
  xlab("Position") + 
  scale_shape_discrete(labels=c("Deletion","Insertions","substitutions")) +
  scale_y_continuous(breaks=c(-3,-2.5,-2,-1,0),
                     labels=round(10^c(-3,-2.5,-2,-1,0),3)) +
  theme_bw() 

if(do.write){
  ggsave("./error_rates_compared_with_pubdata.pdf",width=8.7,height=5)
}
```

## if filter
```{r}
pub.stat.df <- read.table("vcf/pub.vcfStat2.tsv",fill = NA,
                            col.names = c("sample","region","POS","rPOS","REF","ALT","TYPE","DP","AD"))
zymo.stat.df <- read.table("vcf/zymo.vcfStat2.tsv",fill = NA,
                            col.names = c("sample","region","POS","rPOS","REF","ALT","TYPE","DP","AD"))

zymo.PBA.stat.df <- read.table("vcf/PBA2D6305.genome.vcfStat2.vcf",fill = NA,
                            col.names = c("sample","region","POS","rPOS","REF","ALT","TYPE","DP","AD"))

rRNA.stat.df <- rbind(zymo.stat.df,pub.stat.df,zymo.PBA.stat.df)

rRNA.stat.df$UNIT <- sapply(strsplit(rRNA.stat.df$region,"_"),"[",1)
rRNA.stat.df$UID <- sapply(strsplit(rRNA.stat.df$region,"_"),"[",2)

rRNA.ref.seq <- unique(rRNA.stat.df[which(!is.na(rRNA.stat.df$REF)&rRNA.stat.df$REF!=""&rRNA.stat.df$TYPE=="SUB"),c("region","rPOS","REF","UNIT","UID")])
rRNA.ref.seq$CS <- ifelse(rRNA.ref.seq$UID<6,CS(rRNA.ref.seq$REF),rRNA.ref.seq$REF)
rRNA.ref.seq$rPOS2 <- ifelse(rRNA.ref.seq$UID>5,rRNA.ref.seq$rPOS+1,rRNA.ref.seq$rPOS)

rRNA.stat.df$rPOS2 <- ifelse(rRNA.stat.df$UID>5,rRNA.stat.df$rPOS+1,rRNA.stat.df$rPOS)

rRNA.stat.df$sam <- rRNA.stat.df$sample
rRNA.stat.df$sample <- as.factor(rRNA.stat.df$sample)


#levels(rRNA.stat.df$sample) <- c("PacBio(610)","HiSeq4000(647)","PacBio RSII(044)","WGBS(545)",
#                                 "PacBio RSII(667)","PacBio RSII(668)","PacBio RSII(669)",
#                                 "R9 MinION(670)","R9 MinION(671)","WGBS(672)",
#                                 "Z1","Z2","Z3")


levels(rRNA.stat.df$sample) <- c(
  "Z1(Reads)","Z1(PBA)","Z2(Reads)","Z2(PBA)","Z3(Reads)",
  "PacBio(610)","HiSeq4000(647)","WGBS(545)",
  "PacBio RSII(667)","PacBio RSII(668)","PacBio RSII(669)",
  "R9 MinION(670)","R9 MinION(671)", "WGBS(672)")

levels(rRNA.stat.df$sample) <- c(
  "Z1(Reads)","Z1(PBA)","Z2(Reads)","Z2(PBA)","Z3(Reads)",
  "PacBio(610)","HiSeq4000(647)","PacBio RSII(044)","WGBS(545)",
  "PacBio RSII(667)","PacBio RSII(668)","PacBio RSII(669)",
  "R9 MinION(670)","R9 MinION(671)", "WGBS(672)")

levels(rRNA.stat.df$sam) <- c(
  "stcLFR(Reads)","stcLFR(PBA)","stcLFR(Reads)","stcLFR(PBA)",
  "stcLFR(Reads)","PacBio","HiSeq4000",
  "PacBio RSII","WGBS",
  "PacBio RSII","PacBio RSII","PacBio RSII",
  "R9 MinION","R9 MinION", "WGBS")
#rRNA.stat.df <- rRNA.stat.df%>%filter(sample %in% c("Z1(Reads)","Z1(PBA)","HiSeq4000(647)","PacBio(610)","PacBio RSII(044)","R9 MinION(670)"))
```

```{r}
ggplot(rRNA.ref.seq,aes(x=rPOS2,y=UID)) + geom_tile(aes(fill=CS)) +
  facet_grid(UNIT~.,scale="free") + xlim(50,200)
```

```{r}
ggplot(rRNA.stat.df%>%filter(UNIT=="16S"),aes(x=rPOS2)) + geom_path(aes(y=DP,color=sample)) +
  facet_grid(sample~UID,scale="free")
```

```{r}
ggplot(rRNA.stat.df%>%filter(UNIT=="16S"),aes(x=rPOS2)) + geom_line(aes(y=AD/DP,color=sample)) +
  facet_grid(sample~UID,scale="free")
```


```{r}
library(plyr)
rRNA.stat.df2 <- ddply(rRNA.stat.df,c("sam","UNIT","rPOS2","TYPE"),summarize,
                       DP=sum(DP),AD=sum(AD))


ggplot(rRNA.stat.df2%>%filter(UNIT=="16S"),aes(x=rPOS2)) + geom_point(aes(y=DP,color=TYPE)) +
  facet_wrap(.~sam,scale="free")
```
```{r}
ggplot(rRNA.stat.df2%>%filter(DP>9)) +
  geom_segment(aes(yend=AD/DP,y=0,x=rPOS2,xend=rPOS2,color=TYPE)) +
  facet_grid(UNIT~sam,scale="free") +
  geom_hline(yintercept = seq(1,7)/7,linetype=2) + theme_bw()
```

```{r}
ggplot(rRNA.stat.df2%>%filter(DP>9)) +
  geom_segment(aes(yend=AD/DP,y=0,x=rPOS2,xend=rPOS2,color=sample)) +
  facet_grid(UNIT~sample,scale="free") + 
  scale_y_log10() + theme_bw()
```

```{r}
rRNA.stat.df$win <- round(rRNA.stat.df$rPOS/50,0)*50
rRNA.stat.df3 <- ddply(rRNA.stat.df,c("sam","UNIT","win","TYPE"),summarize,
                       DP=sum(DP),AD=sum(AD))

ggplot(rRNA.stat.df3%>%filter(DP>5)) +
  geom_point(aes(y=log2(AD/DP),x=win,color=TYPE,shape=TYPE)) +
  facet_grid(UNIT~sam) + 
  scale_y_continuous(breaks=seq(-10,0,2),labels = 2^(seq(-10,0,2))) + theme_bw()
```
