---
title: "zymo reference build"
output:
  html_notebook:
    code_fold: hide
---

zymo infomation : `stLFR4META - 文档\Experiments\ZymoBIOMICS Microbial Community DNA Standard D6305`

Zymo mock strains:

| Species                  | Genomic DNA | 16S Only1 | 16S & 18S1 | Genome Copy2 | Cell Number3 |
| ------------------------ | ----------- | --------- | ---------- | ------------ | ------------ |
| Pseudomonas aeruginosa   | 12          | 4.2       | 3.6        | 6.1          | 6.1          | +
| Escherichia coli         | 12          | 10.1      | 8.9        | 8.5          | 8.5          | +
| Salmonella enterica      | 12          | 10.4      | 9.1        | 8.7          | 8.8          | +
| Lactobacillus fermentum  | 12          | 18.4      | 16.1       | 21.6         | 21.9         |
| Enterococcus faecalis    | 12          | 9.9       | 8.7        | 14.6         | 14.6         | +
| Staphylococcus aureus    | 12          | 15.5      | 13.6       | 15.2         | 15.3         | +
| Listeria monocytogenes   | 12          | 14.1      | 12.4       | 13.9         | 13.9         | +
| Bacillus subtilis        | 12          | 17.4      | 15.3       | 10.3         | 10.3         | +
| Saccharomyces cerevisiae | 2           | NA        | 9.3        | 0.57         | 0.29         |
| Cryptococcus neoformans  | 2           | NA        | 3.3        | 0.37         | 0.18         |


set global vars (bash):
```{bash}
SAM=RCA197B01_2

threads=16
DB=../../Source/REF/zymo
RES=../../Results/ZYMO
```
A list of manuals for check:
> Samtools: http://www.htslib.org/doc/samtools.html
> bwa: http://bio-bwa.sourceforge.net/bwa.shtml
> bed format: https://asia.ensembl.org/info/website/upload/bed.html

Mapping 16S, 28S sequences back to genome.
```
bwa mem -t $threads -a $DB/D6305.genomes.bwa $DB/D6305.ssrRNA.fasta |samtools view -b \
| samtools sort > $RES/D6305.ssrRNA2genome.bam

bwa mem -t $threads -a $DB/D6305.genomes.bwa $DB/D6305.markers.fa |samtools view -b \
| samtools sort > $RES/D6305.16S_23S.2genome.bam

samtools view $RES/D6305.16S_23S.2genome.bam| perl sam2bed.pl -s - -o $RES/D6305.ssrRNA2genome.bed
```

call snv:
```
grep --no-group-separator -A1 "16S_1" $DB/D6305.ssrRNA.fasta > $RES/D6305.16S_1.fasta
mafft --thread 32 --auto --inputorder $RES/D6305.16S_1.fasta > $RES/D6305.16S_1.mafft.fa
```


# Benchmark of zymo variances:

Find rDNA rigions from ref:
```{bash}
bwa bwasw ../../Source/REF/zymo/D6305.genomes.bwa ../../Source/REF/zymo/D6305.rRNA.fa \
 > ../../Source/REF/zymo/D6305.rRNA2genomes.bwa
```
Find E.coli SSU positions:
```{bash,eval=FALSE}
bwa mem ../../Source/REF/zymo/D6305.genomes.bwa ../../Source/REF/zymo/D6305.ssrRNA.fasta |awk '{$10="";print $0}'|grep "coli"
[M::bwa_idx_load_from_disk] read 0 ALT contigs
@SQ SN:Escherichia_coli_plasmid LN:110007
@SQ SN:Escherichia_coli_chromosome LN:4765434
[M::process] read 52 sequences (80526 bp)...
[M::mem_process_seqs] Processed 52 reads in 0.671 CPU sec, 0.672 real sec
Escherichia_coli_16S_1 16 Escherichia_coli_chromosome 3342577 0 1542M * 0 0  * NM:i:0 MD:Z:1542 AS:i:1542 XS:i:1542
Escherichia_coli_16S_2 16 Escherichia_coli_chromosome 3248951 0 1542M * 0 0  * NM:i:0 MD:Z:1542 AS:i:1542 XS:i:1542
Escherichia_coli_16S_3 16 Escherichia_coli_chromosome 3076422 0 1542M * 0 0  * NM:i:0 MD:Z:1542 AS:i:1542 XS:i:1542
Escherichia_coli_16S_4 0 Escherichia_coli_chromosome 4591467 10 1542M * 0 0  * NM:i:0 MD:Z:1542 AS:i:1542 XS:i:1532   XA:Z:Escherichia_coli_chromosome,-2428583,1542M,5;
Escherichia_coli_16S_5 16 Escherichia_coli_chromosome 3342577 0 1542M * 0 0  * NM:i:0 MD:Z:1542 AS:i:1542 XS:i:1542
Escherichia_coli_16S_7 16 Escherichia_coli_chromosome 3248951 0 1542M * 0 0  * NM:i:0 MD:Z:1542 AS:i:1542 XS:i:1542
Escherichia_coli_16S_8 0 Escherichia_coli_chromosome 3883127 0 1542M * 0 0  * NM:i:0 MD:Z:1542 AS:i:1542 XS:i:1542
[main] Version: 0.7.17-r1198-dirty
[main] CMD: bwa mem ../../Source/REF/zymo/D6305.genomes.bwa ../../Source/REF/zymo/D6305.ssrRNA.fasta
[main] Real time: 1.836 sec; CPU: 1.367 sec
```


Find E.coli LSU positions:
```{bash,eval=FALSE}
bwa mem -h 15 ../../Source/REF/zymo/D6305.genomes.bwa ../../Source/REF/zymo/D6305.markers.LSU.fa |awk '{$10="";print $0}'|grep "coli"
[M::bwa_idx_load_from_disk] read 0 ALT contigs
@SQ SN:Escherichia_coli_plasmid LN:110007
@SQ SN:Escherichia_coli_chromosome LN:4765434
[M::process] read 9 sequences (24313 bp)...
[M::mem_process_seqs] Processed 9 reads in 0.338 CPU sec, 0.348 real sec
Escherichia_coli_23S 16 Escherichia_coli_chromosome 3114568 0 2904M * 0 0  * NM:i:4 MD:Z:1039G134A6T548C1173 AS:i:2884 XS:i:2884 XA:Z:Escherichia_coli_chromosome,-3339319,2904M,4;Escherichia_coli_chromosome,+3885106,2904M,4;Escherichia_coli_chromosome,-2425233,2904M,4;Escherichia_coli_chromosome,-3245601,2904M,4;Escherichia_coli_chromosome,+4593440,2904M,4;Escherichia_coli_chromosome,-3073164,2904M,4;Salmonella_enterica_complete_genome,-2002972,1170M4I3M4D216M6I4M8D959M2D542M,79;Salmonella_enterica_complete_genome,+2775089,542M2D959M8D4M6I213M4D6M4I1170M,82;Salmonella_enterica_complete_genome,-1336705,1170M4I3M4D216M6I4M8D959M2D542M,82;Salmonella_enterica_complete_genome,+2524769,542M2D959M8D4M6I213M4D6M4I1170M,83;Salmonella_enterica_complete_genome,+2817845,542M2D959M8D4M6I213M4D6M4I1170M,83;Salmonella_enterica_complete_genome,+2619699,542M2D959M8D4M6I213M4D6M4I697M2D9M1I463M,86;
[main] Version: 0.7.17-r1198-dirty
[main] CMD: bwa mem -h 15 ../../Source/REF/zymo/D6305.genomes.bwa ../../Source/REF/zymo/D6305.markers.LSU.fa
[main] Real time: 1.283 sec; CPU: 0.960 sec
```

Escherichia_coli_chromosome 3114568 0 2904M * 0 0  * XA:Z:
Escherichia_coli_chromosome,-3339319,2904M,4;
Escherichia_coli_chromosome,+3885106,2904M,4;
Escherichia_coli_chromosome,-2425233,2904M,4;
Escherichia_coli_chromosome,-3245601,2904M,4;
Escherichia_coli_chromosome,+4593440,2904M,4;
Escherichia_coli_chromosome,-3073164,2904M,4;
Salmonella_enterica_complete_genome,-2002972,1170M4I3M4D216M6I4M8D959M2D542M,79;Salmonella_enterica_complete_genome,+2775089,542M2D959M8D4M6I213M4D6M4I1170M,82;Salmonella_enterica_complete_genome,-1336705,1170M4I3M4D216M6I4M8D959M2D542M,82;Salmonella_enterica_complete_genome,+2524769,542M2D959M8D4M6I213M4D6M4I1170M,83;Salmonella_enterica_complete_genome,+2817845,542M2D959M8D4M6I213M4D6M4I1170M,83;Salmonella_enterica_complete_genome,+2619699,542M2D959M8D4M6I213M4D6M4I697M2D9M1I463M,86;

```{bash}
for i in Z{1,2,3};do
  bwa mem -t 16 ../../Source/REF/zymo/D6305.genomes.bwa SAM/$i/clean/fastp.sort.{1,2}.fq.gz | samtools view -b -o SAM/$i/bwa/map2D6305.genome.bam
done

for i in Z{1,2,3};do
  samtools sort SAM/$i/bwa/map2D6305.genome.bam -o SAM/$i/bwa/map2D6305.genome.sort.bam &
done


for i in Z{1,2,3};do
  samtools index SAM/$i/bwa/map2D6305.genome.sort.bam &
done

for i in Z{1,2,3};do
  bcftools mpileup -f ../../Source/REF/zymo/D6305.genomes.fasta \
   SAM/$i/bwa/map2D6305.genome.sort.bam > SAM/$i/bwa/map2D6305.genome.vcf &
done


#
bcftools mpileup -d 10000 -f ../../Source/REF/zymo/D6305.genomes.fasta \
  -Ov -o Z.map2D6305.genome.vcf SAM/Z{1,2,3}/bwa/map2D6305.genome.sort.bam 

bcftools call -c --ploidy 1 -Oz -o Z.map2D6305.genome.call.vcf Z.map2D6305.genome.vcf

bcftools mpileup --threads 8 -d 10000 -Ou -f ../../Source/REF/zymo/D6305.genomes.fasta \
  SAM/Z{1,2,3}/bwa/map2D6305.genome.sort.bam |bcftools call --threads 8 -c --ploidy 1 -Oz \
  -o Z.map2D6305.genome.call.vcf
```

More about [`vcfR`](https://cran.r-project.org/web/packages/vcfR/vignettes/intro_to_vcfR.html)
```{r}
library(vcfR)

vcf <- read.vcfR("Z.map2D6305.genome.call.vcf", verbose = FALSE )
gff <- read.table("../../Source/REF/zymo/D6305.Ecoli.rRNA.gff", sep="\t", quote="")
dna <- ape::read.dna("../../Source/REF/zymo/D6305.genomes.fasta", format = "fasta")
```

# Escherichia_coli
```{r}
myChroms[grep("Escherichia",myChroms)]
EC.vcf <- vcf[getCHROM(vcf) == "Escherichia_coli_chromosome", ]
EC.gff <- gff[gff$V1=="Escherichia_coli_chromosome",]

mySeqs[grep("Escherichia",mySeqs)]
EC.seq <- dna["Escherichia_coli_chromosome"]
EC.chrom <- create.chromR(name='Escherichia_coli', vcf=EC.vcf, seq=EC.seq,ann=EC.gff)
```
```{r}
#EC.chrom.f <- masker(EC.chrom, min_QUAL = 1, min_MQ=59.9, min_DP = 10, max_DP = 1000)
EC.chrom.p <- proc.chromR(EC.chrom, win.size=100,verbose=TRUE)

plot(EC.chrom.p)
```
```{r}
chromoqc(EC.chrom.p, dp.alpha=20)
```
## rRNA_1
```{r}
chromoqc(EC.chrom.p, xlim=c(2422000, 2431000))

```
## rRNA_2
```{r}
chromoqc(EC.chrom.p, xlim=c(3071164, 3078068))
```

## rRNA_3
```{r}
chromoqc(EC.chrom.p, xlim=c(3114068, 3119868))

```


## rRNA_4
```{r}
chromoqc(EC.chrom.p, xlim=c(3245001, 3250993))

```

## rRNA_5
```{r}
chromoqc(EC.chrom.p, xlim=c(3339019, 3344619))

```
## rRNA_6
```{r}
chromoqc(EC.chrom.p, xlim=c(3883027, 3888510))

```
## rRNA_7
```{r}
chromoqc(EC.chrom.p, xlim=c(4591067, 4596844))
```



# bedtools
```{bash,eval=FALSE}
bedtools genomecov -bga -ibam SAM/Z3/bwa/map2D6305.genome.sort.bam -scale 0.009941 | sort -k1,1 -k2,2n > Z3.bedGraph
```

