---
title: "AUG20 DATA ASSESSMENT: annotation part"
author: "Chao"
date: "3/22/2020"
output: 
  html_notebook:
    code_fold: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(reshape2)
library(ggplot2)
library(plyr)
library(dplyr)
library(RColorBrewer)
library(configr)
library(tools)
library(stringi)
source("../13.JAN20/lib.R")
```

# Bash prep
## Get valid sequences
```{bash, eval=FALSE}
# merge clips
mkdir -p SAM/SFR/ANNO
for i in {S,O}{4,5,6};do 
  awk -v n=$i '/>BI/{gsub(">",">"n)}{print}' SAM/$i/summary.BI.megahit.clip.all.fasta;
done > SAM/SFR/summary.BI.megahit.clip.all.fasta

for i in {S,O}{4,5,6};do 
  awk -v n=$i '{print n$0}' SAM/$i/ANNO/CLIP.map.merge.clip.anno;
done > SAM/SFR/ANNO/CLIP.map.merge.clip.anno

for i in {S,O}{4,5,6};do 
  awk -v n=$i '{print n$0}' SAM/$i/ANNO/CLIP.map.merge.bead.anno;
done > SAM/SFR/ANNO/CLIP.map.merge.bead.anno

#remove singleton clips
#metabbq smk -j -npk SAM/SFR/CLIP/id90def4.clust.fa
metabbq smk --configfile config.F.yaml -j -npk SAM/{S,O}{4,5,6}/CLIP/validSeqs.clust.fa

for i in {S,O}{4,5,6};do 
  awk -v n=$i '/>BI/{gsub(">",">"n)}{print}' SAM/$i/CLIP/validSeqs.clust.fa;
done > SAM/SFR/CLIP/validSeqs.clust.fa

# get PBA metadata
grep ">" SAM/SFR/CLIP/validSeqs.clust.fa|awk -F ">|_" '{
sub("flag=","",$5);sub("multi=","",$6);sub("C","",$8);print $2"."$4"."$8"\t"$5"\t"$6"\t"$9"\t"$10}'  \
>  SAM/SFR/CLIP/validSeqs.clust.metadat.tsv
```
## generate a cluster tree
```{bash,eval=FALSE}
metabbq IO makeclade -i SAM/SFR/CLIP/validSeqs.clust.fa -o SAM/SFR/CLIP/validSeqs.clade.uc -t 40 -v

metabbq IO clade2tree -i SAM/SFR/CLIP/validSeqs.clust.fa -a SAM/SFR/ANNO/CLIP.map.merge.clip.anno \
 -d $LFR/Source/REF/KRAKEN2/SUS -s SAM/SFR/CLIP/validSeqs.clade.uc -o SAM/SFR/KRAKEN/BEAD -p tax -v

rm SAM/SFR/KRAKEN/BEAD/{database*,*.k2d} 
metabbq buildKraken2db.sh SAM/SFR/KRAKEN/BEAD 4

perl -e 'while(<>){chomp;@s=split /\t/; @t=split /;/,$s[0]; %H=();%B=("T",0,"C",0);
for(@t){$_=~/^(\S+)__(.+)$/;($k,$v)=($1,$2);$H{$k}=$v;if($k=~/CLADE/){$B{C}++}else{$B{T}++}}; print "$s[1]\t$s[2]\t$s[3]\t$s[4]\t$s[5]\t$s[6]\t";
print "$H{domain}\t$H{phylum}\t$H{genus}\t$H{species}\t$B{T}\t$B{C}\n"}' SAM/SFR/KRAKEN/BEAD/taxonomy/paths.txt \
> SAM/SFR/KRAKEN/BEAD/data/topRank.map &

####
# mapping to merged db
for i in {S,O}{4,5,6};do
mkdir -p SAM/$i/KRAKEN
ln -s ../../SFR/KRAKEN/BEAD SAM/$i/KRAKEN/BEAD;
done

#generate bead/bin level bracken profile
metabbq smk --config kk2_db="BEAD" -j -npk SAM/{S,O}{4,5,6}/KRAKEN/reads2BEAD.bead.bracken

for i in {S,O}{4,5,6};do 
  awk -v n=$i 'FNR>1{print n"\t"$0}' SAM/$i/KRAKEN/reads2BEAD.bead.bracken 
done > SAM/SFR/KRAKEN/reads2BEAD.bead.bracken

# modify (for both bac and fungi)
for i in {S,O}{4,5,6};do 
  metabbq IO modKKreport -r Bacteria -m D1:P -i SAM/$i/KRAKEN/reads2BEAD.bead.kreport2 -o SAM/$i/KRAKEN/reads2BEAD.bead.R.kreport2 &
done

# get profile of each rank
for i in {S,O}{4,5,6};do for j in {P,F,G,S};do 
  bracken -d SAM/$i/KRAKEN/BEAD -i SAM/$i/KRAKEN/reads2BEAD.bead.R.kreport2 -l $j -o SAM/$i/KRAKEN/reads2BEAD.bead.$j.bracken > SAM/$i/KRAKEN/reads2BEAD.bead.$j.breport2 ;
done & done

for j in {P,G,F,S};do for i in {S,O}{1..6};do 
  awk -v n=$i 'FNR>1{print n"\t"$0}' SAM/$i/KRAKEN/reads2BEAD.bead.$j.bracken
done > ../../Results/AUG20/STAT/KRAKEN/SXR.RCA.reads2BEAD.bead.$j.bracken & done

```

## print taxname tree for merging to open-refs (BEAD+)
```{bash,eval=FALSE}
####
# print taxname tree for merging to open-refs
mkdir -p SAM/SFR/KRAKEN/BEAD+/{data,library}
cat $LFR/Source/REF/KRAKEN2/SUS/library/added.fna SAM/SFR/KRAKEN/BEAD/library/added.fna \
> SAM/SFR/KRAKEN/BEAD+/library/added.fna
#
cp SAM/SFR/KRAKEN/BEAD/data/added.txt SAM/SFR/KRAKEN/BEAD+/data/added.txt

cat $LFR/Source/REF/KRAKEN2/SUS/data/added.acc_taxid SAM/SFR/KRAKEN/BEAD/data/added.acc_taxid \
|sort > SAM/SFR/KRAKEN/BEAD+/data/added.acc_taxid

 rm SAM/SFR/KRAKEN/BEAD+/{database*,*.k2d} 
metabbq buildKraken2db.sh SAM/SFR/KRAKEN/BEAD+ 4

perl -e 'while(<>){chomp;@s=split /\t/; @t=split /;/,$s[0]; %H=();%B=("T",0,"C",0);
for(@t){$_=~/^(\S+)__(.+)$/;($k,$v)=($1,$2);$H{$k}=$v;if($k=~/CLADE/){$B{C}++}else{$B{T}++}}; 
print "$s[1]\t$s[2]\t$s[3]\t$s[4]\t$s[5]\t$s[6]\t";
print "$H{domain}\t$H{phylum}\t$H{genus}\t$H{species}\t$B{T}\t$B{C}\n"}' \
SAM/SFR/KRAKEN/BEAD+/taxonomy/paths.txt|gzip >  SAM/SFR/KRAKEN/BEAD+/data/topRank.map.gz


for i in {S,O}{4,5,6,A,B,C};do
ln -s ../../SFR/KRAKEN/BEAD+ SAM/$i/KRAKEN/BEAD+;
done

metabbq smk --config kk2_db="BEAD+" -j -npk SAM/{S,O}{4,5,6}/KRAKEN/reads2BEAD+.bead.bracken


for i in {S,O}{4,5,6,A,B,C};do 
  awk -v n=$i 'FNR>1{print n"\t"$0}' SAM/$i/KRAKEN/reads2BEAD+.bead.bracken 
done > SAM/SFR/KRAKEN/reads2BEAD+.bead.bracken

# modify
for i in {S,O}{4,5,6};do 
  metabbq IO modKKreport -r Bacteria -m D1:P -i SAM/$i/KRAKEN/reads2BEAD+.bead.kreport2 -o SAM/$i/KRAKEN/reads2BEAD+.bead.R.kreport2 &
done

# get profile of each rank
for i in {S,O}{4,5,6};do for j in {P,F,G,S};do 
  bracken -d SAM/$i/KRAKEN/BEAD+ -i SAM/$i/KRAKEN/reads2BEAD+.bead.R.kreport2 -l $j -o SAM/$i/KRAKEN/reads2BEAD+.bead.$j.bracken > SAM/$i/KRAKEN/reads2BEAD+.bead.$j.breport2 ;
done & done

for j in {P,G,F,S};do for i in {S,O}{1..6};do 
  awk -v n=$i 'FNR>1{print n"\t"$0}' SAM/$i/KRAKEN/reads2BEAD+.bead.$j.bracken
done > ../../Results/DEC20/STAT/KRAKEN/SXR.RCA.reads2BEAD+.bead.$j.bracken & done


```

## print taxname tree for competitable to open-refs (BEAD-)
```{bash,eval=FALSE}
####
mkdir -p SAM/SFR/KRAKEN/BEAD-/{data,library}

metabbq IO clade2tree -i SAM/SFR/CLIP/validSeqs.clust.fa -a SAM/SFR/ANNO/CLIP.map.merge.clip.anno \
 -d $LFR/Source/REF/KRAKEN2/SUS -s SAM/SFR/CLIP/validSeqs.clade.uc -o SAM/SFR/KRAKEN/BEAD- -p tax -v

# rm SAM/SFR/KRAKEN/BEAD+/{database*,*.k2d} 
metabbq buildKraken2db.sh SAM/SFR/KRAKEN/BEAD- 4

perl -e 'while(<>){chomp;@s=split /\t/; @t=split /;/,$s[0]; $name=($s[5])?$s[5]:$t[-1];
for(@t){$_=~/^(\w+)__(.+)$/;$H{$1}=$2}; print "$s[1]\t$s[2]\t$s[3]\t$s[4]\t$name\t$s[6]\t";
print "$H{domain}\t$H{phylum}\t$H{genus}\t$H{species}\n"}'  SAM/SFR/KRAKEN/BEAD-/taxonomy/paths.txt \
> SAM/SFR/KRAKEN/BEAD-/data/topRank.map &

for i in {S,O}{4,5,6,A,B,C};do
ln -s ../../SFR/KRAKEN/BEAD- SAM/$i/KRAKEN/BEAD-;
done

metabbq smk --config kk2_db="BEAD-" -j -npk SAM/{S,O}{4,5,6}/KRAKEN/reads2BEAD-.bead.bracken

####
for i in {S,O}{1,2,3,7,8,9};do 
  awk -v n=$i 'FNR>1{print n"\t"$0}' SAM/$i/KRAKEN/reads2BEAD-.bead.bracken;
done > SAM/SFR/KRAKEN/reads2BEAD-.bead.bracken
```
## get PBA metadata
```{bash,eval=FALSE}
awk '!/#skip/{if($2!=pB){if(FNR>1){print $1"\t"pB"\t"blen"\t"ctg"\t"clips};blen=$10;ctg=1;clips=1}else{blen=blen+$10;clips=clips+1;if($4!=pc){ctg=ctg+1}};pc=$4;pB=$2}END{print $1"\t"pB"\t"blen"\t"ctg"\t"clips}' SAM/SFR/summary.BI.megahit.clip.metadata.tsv > SAM/SFR/summary.BI.megahit.clip.metadata.cut.tsv

```



### observed OTUs estimation
1. link dbs
```{bash,eval=FALSE}
# mapping to above dbs
for db in {SILVA138.1SSU,SILVA138.1LSU,UNITE8.2,SUS};do
 for s in {S,O}{4,5,6,A,B,C};do
 ln -sf $LFR/Source/REF/KRAKEN2/$db SAM/$s/KRAKEN/
 done;
done
```
2. run bracken ( with fungi together )
```{bash,eval=FALSE}
for db in {SILVA138.1SSU,SILVA138.1LSU,UNITE8.2,SUS};do
for s in {S,O}{4,5,6,A,B,C}; do
 echo metabbq smk --config kk2_db=$db -j -npk SAM/$s/KRAKEN/reads2$db.bead.bracken
done;done > Soil.bracken.batch.sh

```
3. summary
```{bash,eval=FALSE}
for db in {SILVA138.1SSU,SILVA138.1LSU,UNITE8.2,SUS,BEAD};do
for i in {S,O}{4,5,6,A,B,C};do 
  line=`wc -l SAM/$i/KRAKEN/reads2BEAD.kraken2|cut -d " " -f1` && \
  mkdir -p SAM/$i/KRAKEN/rand && \
  for db in {SILVA138.1SSU,SILVA138.1LSU,UNITE8.2,SUS,BEAD};do
    file="SAM/$i/KRAKEN/reads2$db.kraken2";
    if [ -e $file ];then
      time=`expr $line / 10000000`;
      t=1; while(( $t <= $time ));do
        num=`expr $t \* 10000000`;
        
        for r in {0..9};do
          seed=`expr $line + $t + $r`;
          echo metabbq randomlyPickLines.pl -s $seed -n $num -m $line -v -i $file \| metabbq IO kk2prf -d SAM/$i/KRAKEN/$db -i - \| metabbq kraken-report --d=SAM/$i/KRAKEN/$db - -v \> SAM/$i/KRAKEN/rand/$db\_$num.$r.bead.kreport2
          echo bracken -d SAM/$i/KRAKEN/$db -i SAM/$i/KRAKEN/rand/$db\_$num.$r.bead.kreport2  -o SAM/$i/KRAKEN/rand/$db\_$num.$r.bead.bracken -l S
        done
        let "t++";
      done;
    fi
  done > tmp/$i.step2randBracken.sh &
done 

```
After run. Stat:
```{bash,eval=FALSE}
for s in {S,O}{4,5,6,A,B,C};do for i in SAM/$s/KRAKEN/rand/*.bead.kreport2;do 
perl -e 'exit unless $ARGV[0]=~/SAM\/(\S+)\/KRAKEN\/rand\/(\S+)_(\d+)\.(\d+)\.bead\.kreport2$/;
($s,$d,$t,$r)=($1,$2,$3,$4); open I,"<$ARGV[0]"; $c=0; while(<I>){ @s=split;next if $s[1] < 10;
if($s[3] eq "S"){;$c++; $sum += $s[1]}elsif($s[3] eq "U"){$u=$s[0]};
};close I; print join("\t",$s,$d,$t,$r,$c,$sum,$u)."\n" ' $i;
done  > SAM/$s/KRAKEN/rand.tax.count & done

#patch if any failed
for i in {S,O}{4,5,6,A,B,C};do 
 awk -v i=$i '$5==0{print $1".*"$2"_"$3"."$4" tmp/"i".step2randBracken.sh"}' SAM/$i/KRAKEN/rand.tax.count | xargs -n2 grep;
done > patch.step2randBracken.sh

# or
for db in {SILVA138.1SSU,SILVA138.1LSU,UNITE8.2,SUS,BEAD};do
  for sam in {S,O}{4,5,6,A,B,C};do
    metabbq IO observedOTUs -i SAM -d $db -o SAM/$sam/KRAKEN/$db.bead.sp.otu.count $sam &
  done;
done

#or accumulated stat
mkdir -p SAM/SFR/KRAKEN/rand
for db in {SILVA138.1SSU,SILVA138.1LSU,UNITE8.2,SUS,BEAD};do
  for rep in {0..9};do
    metabbq IO observedOTUs -i SAM -d $db -r $rep -o SAM/SFR/KRAKEN/rand/$db\_$rep.S.otu.count S{4,5,6,A,B,C} &
    metabbq IO observedOTUs -i SAM -d $db -r $rep -o SAM/SFR/KRAKEN/rand/$db\_$rep.O.otu.count O{4,5,6,A,B,C} &
  done;
done

# accumulated with specific size combined
mkdir -p SAM/SFR/KRAKEN/rand
for db in {SILVA138.1SSU,SILVA138.1LSU,UNITE8.2,SUS,BEAD};do
  for size in {1..10}0000000;do
    metabbq IO observedOTUs -i SAM -d $db -s $size -o SAM/SFR/KRAKEN/rand/SRS.$db\_$size.otu.count S{4,5,6} &
    metabbq IO observedOTUs -i SAM -d $db -s $size -o SAM/SFR/KRAKEN/rand/SLS.$db\_$size.otu.count S{A,B,C} &
    metabbq IO observedOTUs -i SAM -d $db -s $size -o SAM/SFR/KRAKEN/rand/ORS.$db\_$size.otu.count O{4,5,6} &
    metabbq IO observedOTUs -i SAM -d $db -s $size -o SAM/SFR/KRAKEN/rand/OLS.$db\_$size.otu.count O{A,B,C} &
  done;
done

#sum to 1 file
cat SAM/{S,O}{4,5,6,A,B,C}/KRAKEN/*.bead.sp.otu.count > STAT/KRAKEN/SFR.idvd.rand.observedOTUs.count
md5sum STAT/KRAKEN/SFR.idvd.rand.observedOTUs.count
cat SAM/SFR/KRAKEN/rand/*.otu.count > STAT/KRAKEN/SFR.cum.rand.observedOTUs.count
md5sum STAT/KRAKEN/SFR.cum.rand.observedOTUs.count
cat SAM/SFR/KRAKEN/rand/{S,O}{R,L}S.*.otu.count > STAT/KRAKEN/SFR.cumSize.rand.observedOTUs.count
md5sum STAT/KRAKEN/SFR.cumSize.rand.observedOTUs.count

```
md5: `0feb3fd26126ca1a653da9183fc61810`
md5: `4e5d12258a8d59ff28f3d2140a0c510e`
md5: `8902f83ed737919385dc9937ad6bc550`


### Read Bracken


```{r}
BEAD.B.topRank.map <- read.table(gzfile("../../Results/DEC20/SAM/SBR/KRAKEN/BEAD+/data/topRank.map.gz"),
  comment.char = "",quote="",sep="\t",fill=NA,as.is=T,
  col.names=c("taxID","treeRank","clade","members","taxon","PBA","domain","phylum","genus","species","treeDist","burgDist"))

PBA.B.metadata <- read.table("../../Results/DEC20/SAM/SBR/CLIP/validSeqs.clust.metadat.tsv",
  comment.char = "", sep="\t",col.names=c("PBA","flag","multi","adaptor","len"))

BEAD.F.topRank.map <- read.table(gzfile("../../Results/DEC20/SAM/SFR/KRAKEN/BEAD+/data/topRank.map.gz"),
  comment.char = "",quote="",sep="\t",fill=NA,as.is=T,
  col.names=c("taxID","treeRank","clade","members","taxon","PBA","domain","phylum","genus","species","treeDist","burgDist"))

PBA.F.metadata <- read.table("../../Results/DEC20/SAM/SFR/CLIP/validSeqs.clust.metadat.tsv",
  comment.char = "", sep="\t",col.names=c("PBA","flag","multi","adaptor","len"))

acc.group.info <- read.table("../../Results/AUG20/etc/supp.acc_taxid.stat",col.names = c("taxID","db","hit","all"))

ref.topRank.map <- rbind(
  cbind(DOM="BAC",merge(BEAD.B.topRank.map,PBA.B.metadata,by="PBA",all.x=T)),
  cbind(DOM="EUK",merge(BEAD.F.topRank.map,PBA.F.metadata,by="PBA",all.x=T))
)

ref.topRank.map <- merge(ref.topRank.map,acc.group.info[,1:2],by="taxID",all.x=T)
ref.topRank.map$taxon[which(ref.topRank.map$taxon=="")] <- "Unknown"

# BEAD+
SXR.r2BEADP.S.bracken <- read.table(
  md5("../../Results/DEC20/STAT/KRAKEN/SXR.RCA.reads2BEAD+.bead.S.bracken","470596f1c84417b27a91c45f46186241"),
  as.is = T,quote = "",comment.char = "", sep="\t",
  col.names = c("SID","name","taxID","rank","kraken_assigned_reads","added_reads","new_est_reads","fraction_total_reads"))

SXR.r2BEADP.G.bracken <- read.table(
  md5("../../Results/DEC20/STAT/KRAKEN/SXR.RCA.reads2BEAD+.bead.G.bracken","6f8cbeea4abc6b423b3bfd9c20c65fc6"),
  as.is = T,quote = "",comment.char = "", sep="\t",
  col.names = c("SID","name","taxID","rank","kraken_assigned_reads","added_reads","new_est_reads","fraction_total_reads"))

SXR.r2BEADP.P.bracken <- read.table(
  md5("../../Results/DEC20/STAT/KRAKEN/SXR.RCA.reads2BEAD+.bead.P.bracken","7e6f907c0f63ed9df762b332bc118970"),
  as.is = T,quote = "",comment.char = "", sep="\t",
  col.names = c("SID","name","taxID","rank","kraken_assigned_reads","added_reads","new_est_reads","fraction_total_reads"))


SXR.reads2multi.bracken <- rbind(
  cbind(DB="JOIN",profRank="Species",SXR.r2BEADP.S.bracken),
  cbind(DB="JOIN",profRank="Genus",SXR.r2BEADP.G.bracken),
  cbind(DB="JOIN",profRank="Phylum",SXR.r2BEADP.P.bracken)
)

SXR.reads2multi.bracken$SAM <- substr(SXR.reads2multi.bracken$SID,1,1)
SXR.reads2multi.bracken$REP <- substr(SXR.reads2multi.bracken$SID,2,2)
SXR.reads2multi.bracken$DOM <- ifelse(SXR.reads2multi.bracken$REP%in%c(1,2,3),"BAC","EUK")
SXR.reads2multi.bracken$REP <- as.factor(SXR.reads2multi.bracken$REP)
levels(SXR.reads2multi.bracken$REP) <- c("A","B","C","A","B","C","A","B","C","A","B","C")

SXR.reads2multi.bracken$UID <- paste0(
  substr(SXR.reads2multi.bracken$DOM,1,1),
  substr(SXR.reads2multi.bracken$DB,1,1),
  substr(SXR.reads2multi.bracken$profRank,1,1),
  SXR.reads2multi.bracken$SID)
  
SXR.reads2multi.bracken <- ddply(
  SXR.reads2multi.bracken,c("UID"),transform,
  pct.k=kraken_assigned_reads/sum(kraken_assigned_reads),
  pct.b=new_est_reads/sum(new_est_reads))

SXR.r2m.bracken.more <- merge(SXR.reads2multi.bracken,ref.topRank.map,by=c("taxID","DOM"),all.x=T)

SXR.r2m.bracken.more$obs <- as.numeric(SXR.r2m.bracken.more$kraken_assigned_reads>0)
SXR.r2m.bracken.more <- ddply(SXR.r2m.bracken.more,c("DOM","DB","profRank","taxID"),transform,sumObs=sum(obs))


save(SXR.r2m.bracken.more, file="../../Results/DEC20/STAT/KRAKEN/SXR.RCA.r2multi.load.RData")
tools::md5sum("../../Results/DEC20/STAT/KRAKEN/SXR.RCA.r2multi.load.RData")

```


```{r}
ggplot(SXR.r2BEAD.species.b.stat,aes(x=taxOrder,y=pct,fill=tech,color=LIB)) + geom_point() + 
  facet_grid(.~tech+CREF) + scale_y_log10()
ggplot(SXR.r2BEAD.species.b.stat,aes(x=taxOrder,y=pct,fill=UID)) + geom_bar(stat="identity",position="dodge") + coord_flip() + facet_grid(.~SAM) + scale_y_log10()
ggplot(SXR.r2BEAD.species.b.stat,aes(x=UID,y=pct,fill=taxOrder)) + geom_bar(stat="identity",position="fill")
ggplot(SXR.r2BEAD.species.b.stat%>%filter(!tax20%in%c("Others","bacterium")),aes(x=SID,y=pct,fill=taxOrder)) + 
  geom_bar(stat="identity",position="stack") +facet_grid(.~CREF+tech,scale="free")

```


### PCA

```{r}

SXR.otu.count.mtx <- dcast(SXR.r2BEAD.b.sp.more,taxID+CREF~UID,value.var = "kraken_assigned_reads",fill = 0,fun.aggregate = sum)
SBR.otu.count.mtx <- SXR.otu.count.mtx[,3:26]
SFR.otu.count.mtx <- SXR.otu.count.mtx[,27:50]
x.row.kept <- which(apply(SXR.otu.count.mtx[,c(3:5,27:29)],1,sum)>10)
x.tax.kept <- SXR.otu.count.mtx$taxID[x.row.kept]
SXR.otu.prf.mtx <- dcast(SXR.r2BEAD.b.sp.more,taxID+CREF~UID,value.var = "pct",fill = 0,fun.aggregate = sum)
taxName <- SXR.otu.prf.mtx$spName
#rownames(SXR.otu.prf.mtx) <- SXR.otu.prf.mtx$taxID
SXR.otu.cut.mtx <- SXR.otu.prf.mtx[x.row.kept,]
SXR.otu.cut.mtx.pca <- prcomp(t(as.matrix(SXR.otu.cut.mtx[,-c(1:2)])),center = T,scale. = T)
summary(SXR.otu.cut.mtx.pca)

B.row.kept <- which(apply(SBR.otu.count.mtx,1,sum)>10)
SBR.otu.cut.mtx <- SXR.otu.prf.mtx[B.row.kept,3:26]
B.taxName <- taxName[B.row.kept]
SBR.otu.cut.mtx.pca <- prcomp(t(as.matrix(SBR.otu.cut.mtx)),center = T,scale. = T)
summary(SBR.otu.cut.mtx.pca)

F.row.kept <- which(apply(SFR.otu.count.mtx,1,sum)>10)
SFR.otu.cut.mtx <- SXR.otu.prf.mtx[F.row.kept,27:50]
F.taxName <- taxName[F.row.kept]
SFR.otu.cut.mtx.pca <- prcomp(t(as.matrix(SFR.otu.cut.mtx)),center = T,scale. = T)
summary(SFR.otu.cut.mtx.pca)

group.df <- unique(SXR.r2BEAD.b.sp.more[,c("UID","CREF","SAM","SID","LIB","tech")])
SXR.otu.multi.pca.df <- merge(group.df,rbind(
  data.frame(type="ALL",SXR.otu.cut.mtx.pca$x[,1:5]),
  data.frame(type="BAC",SBR.otu.cut.mtx.pca$x[,1:5]),
  data.frame(type="FNG",SFR.otu.cut.mtx.pca$x[,1:5])
),by.y="row.names",by.x="UID")

ggplot(SXR.otu.multi.pca.df,aes(x=PC1,y=PC2,color=CREF,shape=interaction(tech,LIB))) +geom_point(size=3) + 
  #scale_shape_manual(values = c(21,22)) +
  geom_text(aes(label=SID),nudge_x = 1,nudge_y = 1)
pheatmap(cor(SBR.otu.cut.mtx,method = "s"),display_numbers = T,cluster_rows = F,cluster_cols = F)
pheatmap(cor(SFR.otu.cut.mtx,method = "s"),display_numbers = T,cluster_rows = F,cluster_cols = F)

```

```{r}
XR.r2BEAD.b.sp.more$findAnno <- ifelse(is.na(SXR.r2BEAD.b.sp.more$species),"Unknown","Known")

sFun <- function(d){
  res <- ddply(d,c("CREF","findAnno","tech","SID","rank","UID","SAM","REP","LIB","RANK","species"),summarise,
               kraken_assigned_reads=sum(kraken_assigned_reads),pct=sum(pct),pct.k=sum(pct.k))
  return(res)
}
cl <- makeCluster(6, type = "SOCK")
registerDoSNOW(cl)
SXR.r2BEAD.b.sp.more2 <- ddply(SXR.r2BEAD.b.sp.more,c("UID"),sFun,.parallel=T)
stopCluster(cl)

SXR.sp.count.mtx <- dcast(SXR.r2BEAD.b.sp.more2,findAnno+anno2+CREF~UID,value.var = "kraken_assigned_reads",fill = 0,fun.aggregate = sum)
SBR.sp.count.mtx <- SXR.sp.count.mtx[,4:27]
SFR.sp.count.mtx <- SXR.sp.count.mtx[,28:51]
x.row.kept <- which(apply(SXR.sp.count.mtx[,c(4:6,28:30)],1,sum)>10)
x.tax.kept <- SXR.sp.count.mtx$taxID[x.row.kept]
SXR.sp.prf.mtx <- dcast(SXR.r2BEAD.b.sp.more2,findAnno+anno2+CREF~UID,value.var = "pct",fill = 0,fun.aggregate = sum)
taxName <- SXR.sp.prf.mtx$anno2
#rownames(SXR.sp.prf.mtx) <- SXR.sp.prf.mtx$taxID
SXR.sp.cut.mtx <- SXR.sp.prf.mtx[x.row.kept,]
SXR.sp.cut.mtx.pca <- prcomp(t(as.matrix(SXR.sp.cut.mtx[,-c(1:3)])),center = T,scale. = T)
summary(SXR.sp.cut.mtx.pca)

B.row.kept <- which(apply(SBR.sp.count.mtx,1,sum)>10)
SBR.sp.cut.mtx <- SXR.sp.prf.mtx[B.row.kept,4:27]
B.taxName <- taxName[B.row.kept]
SBR.sp.cut.mtx.pca <- prcomp(t(as.matrix(SBR.sp.cut.mtx)),center = T,scale. = T)
summary(SBR.sp.cut.mtx.pca)

F.row.kept <- which(apply(SFR.sp.count.mtx,1,sum)>10)
SFR.sp.cut.mtx <- SXR.sp.prf.mtx[F.row.kept,28:51]
F.taxName <- taxName[F.row.kept]
SFR.sp.cut.mtx.pca <- prcomp(t(as.matrix(SFR.sp.cut.mtx)),center = T,scale. = T)
summary(SFR.sp.cut.mtx.pca)

group.df <- unique(SXR.r2BEAD.b.sp.more[,c("UID","CREF","SAM","SID","LIB","tech")])
SXR.sp.multi.pca.df <- merge(group.df,rbind(
  data.frame(type="ALL",SXR.sp.cut.mtx.pca$x[,1:5]),
  data.frame(type="BAC",SBR.sp.cut.mtx.pca$x[,1:5]),
  data.frame(type="FNG",SFR.sp.cut.mtx.pca$x[,1:5])
),by.y="row.names",by.x="UID")

ggplot(SXR.sp.multi.pca.df,aes(x=PC1,y=PC2,color=LIB,shape=interaction(tech,CREF))) +geom_point(size=3) + 
  #scale_shape_manual(values = c(21,22)) +
  geom_text(aes(label=SID),nudge_x = 1,nudge_y = 1)

pheatmap(cor(SXR.sp.cut.mtx[,-c(1:3)],method = "s"),display_numbers = T,cluster_rows = F,cluster_cols = F)
pheatmap(cor(SBR.sp.cut.mtx,method = "s"),display_numbers = T,cluster_rows = F,cluster_cols = F)
pheatmap(cor(SFR.sp.cut.mtx,method = "s"),display_numbers = T,cluster_rows = F,cluster_cols = F)

```



**QQ plot**
```{r}
SXR.stat.sp.df1s <- dcast(SXR.r2BEAD.b.sp.more2,CREF+SAM+LIB+findAnno+REP+anno2~tech,
                          value.var = "pct",fun.aggregate = sum,fill = 0)
SXR.stat.sp.df1s.cor <- ddply(SXR.stat.sp.df1s,c("CREF","SAM","LIB","REP"),summarise, cor=cor(READ,BEAD,method = "s") )
SXR.stat.sp.df1s.cor$corLabel <- paste0("r=",round(SXR.stat.sp.df1s.cor$cor,2))

ggplot(SXR.stat.sp.df1s,aes(x=READ,y=BEAD,color=factor(REP))) +geom_point() +
  geom_abline(intercept = c(-1,0,1),slope=1,linetype=2) + facet_grid(SAM+LIB~CREF+REP) +
  geom_text(data=SXR.stat.sp.df1s.cor,aes(x=1e-3,y=0.1,label=corLabel)) +
  scale_x_log10() + scale_y_log10() 

```

```{r}
SXR.stat.sp.df2s <- dcast(SXR.r2BEAD.b.sp.more2%>%filter(kraken_assigned_reads>10),CREF+SAM+tech+findAnno+REP+anno2~LIB,
                          value.var = "pct",fun.aggregate = sum,fill = 0)
SXR.stat.sp.df2s.cor <- ddply(SXR.stat.sp.df2s,c("CREF","SAM","tech","REP"),summarise, cor=cor(NORCA,RCA,method = "s") )
SXR.stat.sp.df2s.cor$corLabel <- paste0("r=",round(SXR.stat.sp.df2s.cor$cor,2))

ggplot(SXR.stat.sp.df2s,aes(x=NORCA,y=RCA,color=factor(REP))) +geom_point() +
  geom_abline(intercept = 0,slope=c(0.8,1,1.2),linetype=2) + facet_grid(SAM+tech~CREF+REP) +
  geom_text(data=SXR.stat.sp.df2s.cor,aes(x=1e-3,y=0.1,label=corLabel)) +
  scale_x_log10() + scale_y_log10() 

```
```{r}
SXR.stat.sp.df2s.cor2 <- ddply(SXR.stat.sp.df2s%>%filter(tech=="BEAD"),c("CREF","SAM","findAnno"),summarise, cor=cor(NORCA,RCA,method = "s") )
SXR.stat.sp.df2s.cor2$corLabel <- paste0("r=",round(SXR.stat.sp.df2s.cor2$cor,2))

ggplot(SXR.stat.sp.df2s%>%filter(tech=="BEAD"),
       aes(x=NORCA,y=RCA,color=factor(REP))) +geom_point(alpha=.3) +
  geom_abline(intercept = 0,slope=c(0.8,1,1.2),linetype=2) + facet_grid(findAnno~SAM+CREF) +
  geom_text(data=SXR.stat.sp.df2s.cor2,aes(x=1e-3,y=0.1,label=corLabel),color="black") +
  scale_x_log10() + scale_y_log10() 

```


```{r}
SXR.stat.sp.df3s <- dcast(SXR.r2BEAD.b.sp.more,CREF+SAM+tech+LIB+anno2~REP,
                          value.var = "pct",fill = 0,fun.aggregate = sum)

SXR.stat.sp.df3S <- NULL
for(i in 6:8){
  tmp <- SXR.stat.sp.df3s[,-i]
  colnames(tmp)[6:7] <- c("rep1","rep2")
  SXR.stat.sp.df3S <- rbind(SXR.stat.sp.df3S,data.frame(set=colnames(SXR.stat.sp.df3s)[i],tmp))
}
SXR.stat.sp.df3S.cor <- ddply(SXR.stat.sp.df3S,c("CREF","SAM","tech","LIB","set"),summarise,cor=cor(rep1,rep2,method="s"))
SXR.stat.sp.df3S.cor$corLabel <- paste0("r=",round(SXR.stat.sp.df3S.cor$cor,2))

ggplot(SXR.stat.sp.df3S,aes(x=rep1,y=rep2,color=factor(LIB))) +geom_point() +
  geom_abline(intercept = 0,slope=c(0.85,1,1.15),linetype=2) + facet_grid(set+LIB~CREF+tech+SAM) + 
  geom_text(data=SXR.stat.sp.df3S.cor,aes(x=1e-3,y=0.1,label=corLabel)) +
  scale_x_log10() + scale_y_log10()
```

```{r}
SXR.stat.sp.df3S.cor2 <- ddply(SXR.stat.sp.df3S%>%filter(tech=="BEAD"),c("CREF","SAM","LIB"),summarise,cor=cor(rep1,rep2,method="s"))
SXR.stat.sp.df3S.cor2$corLabel <- paste0("r=",round(SXR.stat.sp.df3S.cor2$cor,2))

ggplot(SXR.stat.sp.df3S%>%filter(tech=="BEAD"),aes(x=rep1,y=rep2,color=set)) +geom_point() +
  geom_abline(intercept = 0,slope=c(0.85,1,1.15),linetype=2) + facet_grid(LIB~CREF+SAM) + 
  geom_text(data=SXR.stat.sp.df3S.cor2,aes(x=1e-3,y=0.1,label=corLabel),color="black") +
  scale_x_log10() + scale_y_log10()
```


```{r}
SXR.stat.sp.df3sf <- dcast(SXR.r2BEAD.b.sp.more,CREF+SAM+tech+LIB+anno2+findAnno~REP,
                          value.var = "pct",fill = 0,fun.aggregate = sum)

SXR.stat.sp.df3SF <- NULL
for(i in 7:9){
  tmp <- SXR.stat.sp.df3sf[,-i]
  colnames(tmp)[7:8] <- c("rep1","rep2")
  SXR.stat.sp.df3SF <- rbind(SXR.stat.sp.df3SF,data.frame(set=colnames(SXR.stat.sp.df3sf)[i],tmp))
}
SXR.stat.sp.df3SF.cor <- ddply(SXR.stat.sp.df3SF%>%filter(tech=="BEAD"),c("CREF","SAM","LIB"),summarise,cor=cor(rep1,rep2,method="s"))
SXR.stat.sp.df3SF.cor$corLabel <- paste0("r=",round(SXR.stat.sp.df3SF.cor$cor,2))

ggplot(SXR.stat.sp.df3SF%>%filter(tech=="BEAD"),aes(x=rep1,y=rep2,color=set)) +geom_point() +
  geom_abline(intercept = 0,slope=c(0.85,1,1.15),linetype=2) + facet_grid(LIB~CREF+SAM) + 
  geom_text(data=SXR.stat.sp.df3SF.cor,aes(x=1e-3,y=0.1,label=corLabel),color="black") +
  scale_x_log10() + scale_y_log10()
```

```{r}
SXR.stat.sp.df3SF.cor2 <- ddply(SXR.stat.sp.df3SF%>%filter(tech=="BEAD"),c("CREF","SAM","LIB","findAnno"),summarise,cor=cor(rep1,rep2,method="s"))
SXR.stat.sp.df3SF.cor2$corLabel <- paste0("r=",round(SXR.stat.sp.df3SF.cor2$cor,2))

ggplot(SXR.stat.sp.df3SF%>%filter(tech=="BEAD"),aes(x=rep1,y=rep2,color=set)) +geom_point() +
  geom_abline(intercept = 0,slope=c(0.85,1,1.15),linetype=2) + facet_grid(LIB+findAnno~CREF+SAM) + 
  geom_text(data=SXR.stat.sp.df3SF.cor2,aes(x=1e-3,y=0.1,label=corLabel),color="black") +
  scale_x_log10() + scale_y_log10()
```

### use kraken count for quantification

```{r}
SXR.stat.sp.df3sf.k <- dcast(SXR.r2BEAD.b.sp.more,CREF+SAM+tech+LIB+anno2+findAnno~REP,
                          value.var = "pct.k",fill = 0,fun.aggregate = sum)

SXR.stat.sp.df3SF.k <- NULL
for(i in 7:9){
  tmp <- SXR.stat.sp.df3sf[,-i]
  colnames(tmp)[7:8] <- c("rep1","rep2")
  SXR.stat.sp.df3SF.k <- rbind(SXR.stat.sp.df3SF.k,data.frame(set=colnames(SXR.stat.sp.df3sf)[i],tmp))
}
SXR.stat.sp.df3SF.k.cor <- ddply(SXR.stat.sp.df3SF.k%>%filter(tech=="BEAD"),c("CREF","SAM","LIB"),summarise,cor=cor(rep1,rep2,method="s"))
SXR.stat.sp.df3SF.k.cor$corLabel <- paste0("r=",round(SXR.stat.sp.df3SF.k.cor$cor,2))

ggplot(SXR.stat.sp.df3SF.k%>%filter(tech=="BEAD"),aes(x=rep1,y=rep2,color=set)) +geom_point() +
  geom_abline(intercept = 0,slope=c(0.85,1,1.15),linetype=2) + facet_grid(LIB~CREF+SAM) + 
  geom_text(data=SXR.stat.sp.df3SF.k.cor,aes(x=1e-3,y=0.1,label=corLabel),color="black") +
  scale_x_log10() + scale_y_log10()
```

```{r}
SXR.stat.sp.df3SF.k.cor2 <- ddply(SXR.stat.sp.df3SF.k%>%filter(tech=="BEAD"),c("CREF","SAM","LIB","findAnno"),summarise,cor=cor(rep1,rep2,method="s"))
SXR.stat.sp.df3SF.k.cor2$corLabel <- paste0("r=",round(SXR.stat.sp.df3SF.k.cor2$cor,2))

ggplot(SXR.stat.sp.df3SF.k%>%filter(tech=="BEAD"),aes(x=rep1,y=rep2,color=set)) +geom_point() +
  geom_abline(intercept = 0,slope=c(0.85,1,1.15),linetype=2) + facet_grid(LIB+findAnno~CREF+SAM) + 
  geom_text(data=SXR.stat.sp.df3SF.k.cor2,aes(x=1e-3,y=0.1,label=corLabel),color="black") +
  scale_x_log10() + scale_y_log10()
```

** calculate GC content of each clade **
```{bash,eval=False}
metabbq beadsAnno.pl -m gc -o SAM/SFR/KRAKEN/BEAD/data/rank.GC.txt -v \
-i SAM/SFR/KRAKEN/BEAD/library/added.fna,SAM/SFR/KRAKEN/BEAD/taxonomy/paths.txt,SAM/SFR/KRAKEN/BEAD/seqid2taxid.map


```

# observed taxonomies
```{r}
SFCR.idvd.rand.tax.count <- read.table(
  md5("../../Results/AUG20/STAT/KRAKEN/SFR.idvd.rand.observedOTUs.count","0feb3fd26126ca1a653da9183fc61810"),
  sep="\t",col.names = c("db","rep","SID","sSize","cSize","OTUs","tBeads","rBeads","uBeads","taxSize"))
SFCR.cum.rand.tax.count <- read.table(
  md5("../../Results/AUG20/STAT/KRAKEN/SFR.cum.rand.observedOTUs.count","4e5d12258a8d59ff28f3d2140a0c510e"),
  sep="\t",col.names = c("db","rep","SID","sSize","cSize","OTUs","tBeads","rBeads","uBeads","taxSize"))
SFCR.cumSize.rand.tax.count <- read.table(
  md5("../../Results/AUG20/STAT/KRAKEN/SFR.cumSize.rand.observedOTUs.count","8902f83ed737919385dc9937ad6bc550"),
  sep="\t",col.names = c("db","rep","SID","sSize","cSize","OTUs","tBeads","rBeads","uBeads","taxSize"))

```

```{r}
SFCR.cumSize.rep.tax.count <- SFCR.cumSize.rand.tax.count%>%filter(SID%in%c("S6","O6","SC","OC"))

SFCR.rand.tax.count <- rbind(
  cbind(type="individual",SFCR.idvd.rand.tax.count),
  cbind(type="accumulate",SFCR.cum.rand.tax.count),
  cbind(type="cumSize",SFCR.cumSize.rep.tax.count)
)
SFCR.rand.tax.count$SAM <- substr(SFCR.rand.tax.count$SID,1,1)
SFCR.rand.tax.count$SREP <- substr(SFCR.rand.tax.count$SID,2,1)
SFCR.rand.tax.count$KDM <- ifelse(grepl("1|2|3|7|8|9",SFCR.rand.tax.count$SID),"BAC","FUNGI")
SFCR.rand.tax.count$LIB <- ifelse(grepl("[1-6]",SFCR.rand.tax.count$SID),"RCA","LFR")

ggplot(SFCR.rand.tax.count%>%filter(type=="individual"&rep==0),aes(x=cSize,y=tBeads)) + 
  geom_line(aes(color=SID,linetype=db),size=1) + geom_point(aes(color=SID)) + 
  facet_grid(type~SAM,scale="free_y")

ggplot(SFCR.rand.tax.count%>%filter(db=="BEAD"&rep==0),aes(x=cSize,y=tBeads)) + 
  geom_point(aes(color=SID,shape=db)) + 
  facet_grid(type~SAM,scale="free_y")

ggplot(SFCR.rand.tax.count%>%filter(SID%in%c("O4","O5","O6")),aes(x=cSize,y=tBeads,color=SID,group=db)) + 
  geom_line(size=1) + geom_point(aes(color=SID)) + facet_grid(type~SAM)

ggplot(SFCR.rand.tax.count,aes(x=tBeads,y=rBeads,color=SID,group=db)) + 
  geom_line(size=1) + geom_point(aes(color=SID)) + facet_grid(.~SAM)

SFCR.rand.tax.count.mean <- SFCR.rand.tax.count %>% 
  group_by(type,db,SAM,SID,LIB,sSize,cSize,tBeads) %>% 
  summarise(sd=sd(OTUs),OTUs=mean(OTUs))

ggplot(SFCR.rand.tax.count.mean,aes(x=cSize,y=OTUs,color=SID)) + 
  geom_line(aes(linetype=db),size=1) + geom_point(aes(color=SID)) + facet_grid(type~SAM,scale="free")
```

```{r}
SFCR.rand.tax.count.mean.stat <- SFCR.rand.tax.count.mean %>% 
  filter(LIB=="RCA"&type=="accumulate"&db%in%c("BEAD","SUS")) %>%
  group_by(type,db,SAM) %>% summarise(max=max(OTUs))

ggplot(SFCR.rand.tax.count.mean%>%filter(LIB=="RCA"&type!="individual"&db%in%c("BEAD","SUS")&SAM=="S"),
       aes(x=cSize,y=OTUs)) + 
  geom_hline(data=SFCR.rand.tax.count.mean.stat,aes(yintercept = max),linetype=2) +
  geom_hline(data=SFCR.rand.tax.count.mean.stat,aes(yintercept = 0.9*max),linetype=2,color="grey50") +
  geom_line(aes(group=interaction(db,type,SID)),color="grey50",size=1,alpha=.3) +
  geom_point(aes(shape=type,color=SID)) + facet_grid(.~SAM,scale="free",space="free_x") +
  scale_x_continuous(breaks = seq(0,4)*1e8) +
  xlab("# Sampling input reads") + ylab("Observed OTUs (clustered at 99%)")

```

```{r}

f.curve.p <- ggplot(SFCR.rand.tax.count.mean%>%filter(LIB=="RCA"&type=="cumSize"),
       aes(x=cSize,y=OTUs,color=db)) + 
  #geom_hline(data=SFCR.rand.tax.count.mean.stat,aes(yintercept = max),linetype=2) +
  #geom_hline(data=SFCR.rand.tax.count.mean.stat,aes(yintercept = 0.9*max),linetype=2,color="grey50") +
  geom_line(aes(group=interaction(db,type,SID)),size=1) +
  geom_point(aes(shape=SAM)) + #facet_grid(.~SAM,scale="free",space="free_x") +
  scale_x_continuous(breaks = seq(0,4)*1e8) + theme_bw() +
  xlab("# Sampling input reads") + ylab("Observed OTUs (clustered at 99%)")
f.curve.p
```
```{r}
if(do.write){
  ggsave(f.curve.p,width=4,height=4,filename = "../../Results/AUG20/FIG/sfr.rarefaction.boxplot.pdf")
}
```

#FIN.