---
title: "Beads annoatation assessment"
output:
  html_notebook:
    code_fold: hide
---

```{r setup, include=FALSE}
library(knitr)
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(plyr)
library(reshape2)
library(ggplot2)
library(RColorBrewer)
do.write<-F
```
# Questions to answer
1. What taxonomy are these beads belonging?
2. When will the beads tend to be hybridization?
3. What index dose the hybridization related with? Is it possible to filter and control in de novo stage?

# Preparation
1. Align the barcode identified fastq to reference via `bwa`.  
2. stat unique mapped annotation.
```{bash eval=FALSE}
SAM0=APR842_00
mkdir -p $SAM0/beadsAnno
#run once per sample
#bwa mem -t 48 -o $SAM0/beadsAnno/bwa.sam REF/fungal5REF.fa $SAM0/clean/fastp.sort.{1,2}.fq

perl -e 'while(<>){@a=split;next if $a[4]<60 || $a[1] > 1024;@b=split("/",$a[0]);if($b[1]=~/0000/){$b[1]="0000_0000_0000"};$HS{$b[1]}{$a[2]}++};print "Bead\tSaccharomyces_cerevisiae\tPleurotus_eryngii\tLentinula_edodes\tHypsizygus_marmoreus\tFlammulina_velutipes\n";foreach $b (sort keys %HS){print "$b\t$HS{$b}{Saccharomyces_cerevisiae}\t$HS{$b}{Pleurotus_eryngii}\t$HS{$b}{Lentinula_edodes}\t$HS{$b}{Hypsizygus_marmoreus}\t$HS{$b}{Flammulina_velutipes}\n"}' < $SAM0/beadsAnno/bwa.sam > $SAM0/beadsAnno/bwa.uniq.anno.count

perl -e '$once=1;while(<>){@a=split;if($a[0]=~/0000/){$b+=($a[2]-$a[1]+1)/4;next}elsif($once){print "0000_0000_0000\t$b\n";$once--};$b=($a[2]-$a[1]+1)/4;print "$a[0]\t$b\n"}' < $SAM0/clean/fastp.sort.1.fq.idx > $SAM0/beadsAnno/beads.num.count

perl -e '@names=("Saccharomyces_cerevisiae","Pleurotus_eryngii","Lentinula_edodes","Hypsizygus_marmoreus","Flammulina_velutipes"); open I,"<'$SAM0'/beadsAnno/beads.num.count"; while(<I>){@a=split;$HS{$a[0]}=$a[1]}; printf("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", "Bead","mainTax","mainPct","uniqPct","uniq","reads",@names);while(<>){next if $_ =~ /^Bead/; chomp; @a=split /\t/; $bb=shift @a; $sum=$a[1]+$a[2]+$a[3]+$a[4]+$a[0]; @b=sort {$a<=>$b} @a; if($b[-1] == $b[-2]){$maxTax="NA"}else{for($i=0;$i<5;$i++){$maxTax=$names[$i] if $a[$i]==$b[-1]}; printf("%s\t%s\t%.4f\t%.4f\t%d\t%d\t%d\t%d\t%d\t%d\t%d\n",$bb,$maxTax,$b[-1]/$sum,$sum/($HS{$bb}*2),$sum,$HS{$bb}*2,$a[0],$a[1],$a[2],$a[3],$a[4])}}' $SAM0/beadsAnno/bwa.uniq.anno.count > $SAM0/beadsAnno/bwa.uniq.anno.stat


#stat coverage
perl -e '$prv="";sub ccov{$k=shift;if($k>90){$k=91};return($k)}; while(<>){
  if($_=~/^\@SQ\tSN:(\S+)\tLN:(\d+)/){$REF{$1}=$2; next};
  @a=split; next if $a[1] > 1024 || $a[6] ne "=" || $a[8]<0;
  @b=split("/",$a[0]); 
  $c ++ if $b[1] ne $prv;
  if($b[1]=~/0000/){
    for($i=0;$i<$a[8];$i++){ $H0{$a[2]}{$a[3]+$i} ++;}
    $prv=$b[1];
  }else{
    if($b[1] ne $prv){
      #print STDERR "print $prv(!=$b[1] at $.) ; \$c=$c\n";
      foreach $t (sort keys %HS){
        print "$prv\t$t\t";for($i=1;$i<=$REF{$t};$i++){
          $HS{$t}{$i}||=0; print chr(33+&ccov($HS{$t}{$i}));
        }; print "\n";
      }; undef %HS;
    }; for($i=0;$i<$a[8];$i++){ $HS{$a[2]}{$a[3]+$i} ++}; $prv=$b[1];};
}; foreach $t (sort keys %HS){print "$prv\t$t\t";for($i=1;$i<=$REF{$t};$i++){
  $HS{$t}{$i}||=0; print chr(33+&ccov($HS{$t}{$i}));$HS{$t}{$i}=0};print "\n";};
foreach $t (sort keys %H0){print "0000_0000_0000\t$t\t";for($i=1;$i<=$REF{$t};$i++){
    $H0{$t}{$i}||=0; print chr(33+&ccov($H0{$t}{$i}))};print "\n";}
' < $SAM0/beadsAnno/bwa.sam > $SAM0/beadsAnno/bwa.pair.cov.stat

## unique
perl -e '$prv="";sub ccov{$k=shift;if($k>90){$k=91};return($k)}; while(<>){
  if($_=~/^\@SQ\tSN:(\S+)\tLN:(\d+)/){ $REF{$1}=$2; next};
  @a=split; next if $a[4]<60 || $a[1] > 1024 || $a[6] ne "=" || $a[8]<0;
  @b=split("/",$a[0]); 
  $c ++ if $b[1] ne $prv;
  if($b[1]=~/0000/){
    for($i=0;$i<$a[8];$i++){ $H0{$a[2]}{$a[3]+$i} ++;}
    $prv=$b[1];
  }else{
    if($b[1] ne $prv){
      #print STDERR "print $prv(!=$b[1] at $.) ; \$c=$c\n";
      foreach $t (sort keys %HS){
        print "$prv\t$t\t";for($i=1;$i<=$REF{$t};$i++){
          $HS{$t}{$i}||=0; print chr(33+&ccov($HS{$t}{$i}));
        }; print "\n";
      }; undef %HS;
    }; for($i=0;$i<$a[8];$i++){ $HS{$a[2]}{$a[3]+$i} ++}; $prv=$b[1];};
}; foreach $t (sort keys %HS){print "$prv\t$t\t";for($i=1;$i<=$REF{$t};$i++){
  $HS{$t}{$i}||=0; print chr(33+&ccov($HS{$t}{$i}));$HS{$t}{$i}=0};print "\n";};
foreach $t (sort keys %H0){print "0000_0000_0000\t$t\t";for($i=1;$i<=$REF{$t};$i++){
    $H0{$t}{$i}||=0; print chr(33+&ccov($H0{$t}{$i}))};print "\n";}
' < $SAM0/beadsAnno/bwa.sam > $SAM0/beadsAnno/bwa.uniq.cov.stat

###summary coverage
perl -e 'while(<>){ chomp; @a=split;
  for($i=0;$i<length($a[2]);$i++){ $HS{$a[1]}{$i}+=ord(substr($a[2],$i,1))-33;}
};foreach $t (sort keys %HS){foreach $i (sort {$a<=>$b} keys %{$HS{$t}}){
  print "$t\t$i\t$HS{$t}{$i}\n";
}}' < $SAM0/beadsAnno/bwa.uniq.cov.stat > $SAM0/beadsAnno/bwa.uniq.cov.sum1
perl -e 'while(<>){ chomp; @a=split;
  for($i=0;$i<length($a[2]);$i++){ $HS{$a[1]}{$i}+=ord(substr($a[2],$i,1))-33;}
};foreach $t (sort keys %HS){foreach $i (sort {$a<=>$b} keys %{$HS{$t}}){
  print "$t\t$i\t$HS{$t}{$i}\n";
}}' < $SAM0/beadsAnno/bwa.pair.cov.stat > $SAM0/beadsAnno/bwa.pair.cov.sum1
```

```{bash,eval=FALSE}
tag="SUB_2R20_0D04"  # 1st test
tag="SUB_2R100_0D10" # 2nd test
#tag="SUB_2R100_0D05" # 3rd test (no better than 2nd)
tag="SUB_2R50_0D10" # 4th test
perl -e 'open I,"<'$SAM0'/beadsAnno/bwa.uniq.anno.stat"; <I>;<I>;
while(<I>){@a=split; next if $a[4]==2;
  $HS{$a[0]}{tax}=$a[1];$HS{$a[0]}{pct}=$a[2];
};while(<>){@a=split; next if $a[0]==$a[1]; @b=sort ($HS{$a[0]}{tax},$HS{$a[1]}{tax});
  $min=($HS{$a[0]}{pct}<$HS{$a[1]}{pct})?$HS{$a[0]}{pct}:$HS{$a[1]}{pct};
  $HS2{$b[0]}{$b[1]}{sprintf("%.2f",$min)}{sprintf("%.2f",$a[2])} ++;
};print "tax1\ttax2\tmainPct\tdistance\tcount\n";
foreach $t1 (sort keys %HS2){foreach $t2 (sort keys %{$HS2{$t1}}){
    foreach $pct (sort keys %{$HS2{$t1}{$t2}}){
      foreach $dst (sort keys %{$HS2{$t1}{$t2}{$pct}}){
        printf("%s\t%s\t%f\t%f\t%d\n",$t1,$t2,$pct,$dst,$HS2{$t1}{$t2}{$pct}{$dst})
}}}}' < $SAM0/TT1M/$tag.mash/bMin2.raw.dist > $SAM0/TT1M/$tag.mash/bMin2.raw.dist.anno.stat
```

**Loading the stat results**
```{r}
SAM0="APR842_00"
bwa.uniq.anno.stat <- read.table("APR842_00/beadsAnno/bwa.uniq.anno.stat",
                                 header=T,sep="\t")
bwa.uniq.cov.sum1 <- read.table("APR842_00/beadsAnno/bwa.uniq.cov.sum1",
                                col.names=c("taxonomy","position","coverage"))
bwa.pair.cov.sum1 <- read.table("APR842_00/beadsAnno/bwa.pair.cov.sum1",
                                col.names=c("taxonomy","position","coverage"))
```

### Overview  
```{r results="asis"}
kable(colSums(bwa.uniq.anno.stat[,5:11]),caption="reads mapped to reference")
```
From all testing 8 million reads, 4.7m can uniquely hit a specific species, and the proportion of each species shows below:

```{r results="asis"}
kable(colSums(bwa.uniq.anno.stat[,7:11])/sum(colSums(bwa.uniq.anno.stat[,7:11])))
```
Above unique aligning hits indicate that *Hm* is almost lost in this mixture. *Le* is higher at 30%, while *Sc* is less with 20%.

```{r warning=FALSE}
sum.uniq.anno <- ddply(bwa.uniq.anno.stat,c("reads"),summarise,
                       uniq=sum(uniq),sumReads=sum(reads)/2,
                       Sc=sum(Saccharomyces_cerevisiae)/sum(uniq),
                       Pe=sum(Pleurotus_eryngii)/sum(uniq),
                       Le=sum(Lentinula_edodes)/sum(uniq),
                       Hm=sum(Hypsizygus_marmoreus)/sum(uniq),
                       Fv=sum(Flammulina_velutipes)/sum(uniq))
sum.uniq.anno$reads[nrow(sum.uniq.anno)] <- 0.2
sum.uniq.anno$reads <- sum.uniq.anno$reads/2
sum.uniq.anno$beads <- sum.uniq.anno$sumReads/sum.uniq.anno$reads
sum.uniq.anno2 <- melt(sum.uniq.anno,id.vars = c("uniq","reads","sumReads","beads"),
                       variable.name = "taxonomy", value.name = "ratio")
sum.uniq.anno2$sRp <- sum.uniq.anno2$sumReads/sum.uniq.anno2$sumReads[nrow(sum.uniq.anno)]
sum.uniq.anno2$sBp <- sum.uniq.anno2$beads/sum.uniq.anno2$beads[1]
ggplot(sum.uniq.anno2,aes(x=reads,y=ratio,color=taxonomy),alpha=.1) +
  geom_point(size=1,alpha=.5) + geom_line(linetype=1,alpha=.3) +
  geom_line(aes(y=sBp),color="grey50",linetype=2) +
  xlab("reads per bead") + annotation_logticks(sides = "b") +
  scale_x_log10(limits=c(0.1,1000)) + scale_y_continuous(breaks=seq(0,1,.2),limits=c(0,1))
```
Above figure shows the taxonomy abundance distribution over the layers of reads per bead(rpb).
Abundance is calculated by the reads proportion of unique hitted a specific taxonomy among each rbp layer. X-axis 0 is the reads failed detecting barcode.  
Dash line indicates the reads number in each layer.
When rbp larger than ~300, the distribution becomd unstable, might due to the lower beads number.

### coverage assessment
```{r}
bwa.comb.cov.sum1 <- rbind(
  cbind(type="uniq",bwa.uniq.cov.sum1),
  cbind(type="pair",bwa.pair.cov.sum1)
)

ggplot(bwa.comb.cov.sum1,aes(x=position,y=coverage)) + geom_line(aes(color=type)) +
  facet_wrap(~taxonomy,ncol=3,scale="free")
```
I've summarised all reads mapping to above 5 references, with directly hitting (which may contained hits that could also perfectly aligned to multiple other species) in green lines and uniquely hits in red lines. The coverage distribution shows that the sequence depth are not even among the ITS regions. *Hm* is lower hitted after ~1300 bp, while the depth of *Sc* decreased befor 1000 bp.  
Another point is that the unique hits are quite lower at specific region among those reference. *Hm* is almost missing unique hit records. And *Pe* is insufficient before 1200 bp. This characteristics of uneven depth will influence assembly performance.

```{r}
bwa.uniq.curate <- filter(bwa.uniq.anno.stat[-1,],reads>2 )
ggplot(bwa.uniq.curate,aes(x=log10(reads/2),y=mainPct*100)) +
  geom_point(alpha=.1) + annotation_logticks(sides = "b") +
  geom_vline(xintercept = log10(c(50,100)),linetype=2) +
  stat_density_2d(h=c(1,50)) + facet_wrap(~mainTax,ncol=3)
```
Each point indicats a bead with its belonging reads in x-axis (lg scale) and the percentage of the most abundant taxonomy annotation (mainPct) in y-axis, e.g, 100% mainPct of *Flammulina velutipes* means that all of the reads belonging to this bead are uniquely hitted this species. Otherwise, the bead may contained DNA source comming from other species, as known as the `hybridized bead`.  

Among all of these beads, those with 100% `mainPct` have:
```{r results='asis'}
kable(length(which(bwa.uniq.curate$mainPct==1))/nrow(bwa.uniq.curate))
```
Above result shows that **most (83.4%) of the beads** got unique annotation(100% mainPct), which indicating that the hybridization are actually not as common as we think previously.  
Besides, lower `mainPct` enriched in the area **when # reads(per bead) larger than 100 (right side of dash line)**.

**uniqPct**  
`uniqPct` means the percentage of reads uniquely mapping to the reference, without alternative hit position. The multi-position hitting happens when the references of different species shared homologous sequence. In this case, only **uniquely hitted recorded** are included to assess the hybridize degree.  
The distribution of `uniqPct` are shown below:
```{r}
ggplot(bwa.uniq.curate,aes(x=log10(reads/2),y=uniqPct*100)) +
  geom_point(alpha=.1) + annotation_logticks(sides = "b") +
  stat_density_2d(h=c(1,50)) + facet_wrap(~mainTax,ncol=3)
```
From above figure I guess that *Le* and *Pe* may be homologous in many ITS sequence region with each other, as their `uniqPct` both enriched at 50% in y-axis. While *Sc* or *Fv* maybe much more specific in sequence as most of their belonging beads got highly unique percentage. *Hm* is just at low content with weak signal.


**Exclue beads with 100% `mainPct`:**
```{r}
ggplot(bwa.uniq.curate%>%filter(mainPct<1),aes(x=log10(reads/2),y=uniqPct*100)) +
  geom_point(alpha=.1) + annotation_logticks(sides = "b") +
  stat_density_2d(h=c(1,30),alpha=.5) + facet_wrap(~mainTax,ncol=3)
```
After remove those 100% `mainPct` beads, we can see that the "potential hybrid" beads also enriched at relatively lower `uniqPct` (from 50% to 75%).

Anyway, it is impossible to know the `mainPct` or `uniqPct` previously in de novo stage, without mapping to the right reference. So the # reads per bead should be the important index to control the hybrid degree, such as discard beads with #reads more than 100. And cautions that the threshold maybe differed in different samples.

# merge cluster community
Since I've done the "beads cluster" process as I addressed in `clusterMethod.md`, it is possible to stat the performance of these communities.  
Here I test the communities clustered by rpb(reads per bead) between 2 to 100 and mash distance from 0 to 0.04 (boundary not included):

**Select the tag to read:**
```{r}
#tag="SUB_2R20_0D04"  # 1st test
tag="SUB_2R100_0D10" # 2nd test
#tag="SUB_2R100_0D05" # 3rd test
#tag="SUB_2R50_0D10" # 4th test
lv=1
cluster.select <- read.table(
  paste0("APR842_00/TT1M/",tag,".mash/lv",lv,"/tree.cluster.main"),
  col.names = c("cluster","Bead"))
bMin2.raw.dist.stat <- read.table(
  paste0("APR842_00/TT1M/",tag,".mash/bMin2.raw.dist.anno.stat"),sep="\t",header=T)

bMin2.831.dist.stat <- read.table(
  paste0("APR831_00/TT1M/","SUB_2R20_0D10",".mash/bMin2.raw.dist.anno.stat"),sep="\t",header=T)
```

```{r}
cluster.merge <- merge(cluster.select,bwa.uniq.curate,by="Bead")
ggplot(cluster.merge,aes(y=mainPct,x=reads/2,color=mainTax,size=1/uniqPct)) +
  geom_point(alpha=.1) + xlab("#reads/bead") +
  facet_wrap(~cluster,ncol=5)
```
Most of the abundant communities are belonging to a single specific species, which is good.    
Hybridization beads may included in the communities as you can see those colorful points distributed in the central area.
```{r}
cluster.sum <- ddply(cluster.merge%>%filter(mainPct>0.975),"cluster",summarise,reads=sum(reads)/2)
cluster.sum$pct <- cluster.sum$reads/sum(cluster.sum$reads)
cluster.merge$cluster <- reorder(cluster.merge$cluster,cluster.merge$reads,sum)
ggplot(cluster.merge%>%filter(mainPct>0.975),aes(x=factor(cluster))) +
  geom_bar(aes(fill=mainTax),position="fill") +
  geom_text(data=cluster.sum,aes(y=0.5,label=reads),angle=90) +
  xlab("Bead cluster ID") + ylab("percentage")
if(do.write){
  save(cluster.merge,file= paste0("APR842_00/TT1M/",tag,"/cluster.merge.anno.RData"))
}
```
Tile plot shows that cluster 5, 6, 7 and 39 can be the represented cluster for each species.

# stat hybridization
```{r}
shFun <- function(d){
  sp <- table(d$mainTax)
  iMax <- (rev(order(sp)))[1]
  i2nd <- (rev(order(sp)))[2]
  c.mainTax <- names(sp)[iMax]
  c.2ndTax <- names(sp)[i2nd]
  c.mainPct <- sp[iMax]/sum(sp)
  c.2ndPct <- sp[i2nd]/sum(sp)
  mainPct.100Pct <- length(which(d$mainPct==1))/nrow(d)
  return(data.frame(cluster=d[1,2],beadNum=nrow(d),main100Pct=mainPct.100Pct,
                    mainTax=c.mainTax,mainPct=c.mainPct,
                    m2ndTax=c.2ndTax,m2ndPct=c.2ndPct))
}
cluster.merge.stat <- ddply(cluster.merge,"cluster",shFun)
ggplot(cluster.merge.stat,aes(x=beadNum,color=mainTax),alpha=.3) +
  geom_point(aes(y=mainPct),shape=19) + geom_point(aes(y=main100Pct),shape=4) +
  theme_bw()
```
###### Each point indicates a cluster with the most abundant annotation colored. The cross indicate the percentage of beads with 100% `mainPct`.

This figure pointed that when #beads per cluster were too small, it's almost impossible to get a good enough scaffolds with unique annotation. 3 or 4 cluster are seeemed hybridized and all of them are mixed by *Fv* and *Le* (see below table). So my previous guess are totally worong :/  
It seems that the hybridization didn't occure randomly, but happened bewteen specific species?
```{r}
cluster.merge.stat%>%filter(beadNum>100&mainPct<0.85)
```
### Pick cluster coverage info
**stat code**
```{bash,eval=FALSE}
perl -e 'open I,"<'$SAM0/TT1M/$tag.mash/bMin2.bc.tree.target.cluster.main'";while(<I>){
  chomp;@a=split;$CL{$a[1]}=$a[0];} while(<>){chomp;@a=split;next unless $CL{$a[0]}; 
  for($i=0;$i<length($a[2]);$i++){ $HS{$CL{$a[0]}}{$a[1]}{$i}+=ord(substr($a[2],$i,1))-33;}
};foreach $c (sort keys %HS){foreach $t (sort keys %{$HS{$c}}){
  foreach $i (sort {$a<=>$b} keys %{$HS{$c}{$t}}){
    print "$c\t$t\t$i\t$HS{$c}{$t}{$i}\n";
}}}' < $SAM0/beadsAnno/bwa.uniq.cov.stat > $SAM0/TT1M/$tag.mash/bwa.uniq.cov.mainClusters

perl -e 'open I,"<'$SAM0/TT1M/$tag.mash/bMin2.bc.tree.target.cluster.main'";while(<I>){
  chomp;@a=split;$CL{$a[1]}=$a[0];} while(<>){chomp;@a=split;next unless $CL{$a[0]}; 
  for($i=0;$i<length($a[2]);$i++){ $HS{$CL{$a[0]}}{$a[1]}{$i}+=ord(substr($a[2],$i,1))-33;}
};foreach $c (sort keys %HS){foreach $t (sort keys %{$HS{$c}}){
  foreach $i (sort {$a<=>$b} keys %{$HS{$c}{$t}}){
    print "$c\t$t\t$i\t$HS{$c}{$t}{$i}\n";
}}}' < $SAM0/beadsAnno/bwa.pair.cov.stat > $SAM0/TT1M/$tag.mash/bwa.pair.cov.mainClusters
```
**Then load the results**
```{r}
bwa.uniq.cov.mainC <- read.table(paste0(SAM0,"/TT1M/",tag,".mash/bwa.uniq.cov.mainClusters"),
                        col.names = c("cluster","taxonomy","position","coverage"))
bwa.pair.cov.mainC <- read.table(paste0(SAM0,"/TT1M/",tag,".mash/bwa.pair.cov.mainClusters"),
                        col.names = c("cluster","taxonomy","position","coverage"))

bwa.cluster.cov.stat <- rbind(
  cbind(type="uniq",bwa.uniq.cov.mainC),
  cbind(type="pair",bwa.pair.cov.mainC)
)
```

**clusters selected**
```{r fig.width=5, fig.height=3}
ggplot(bwa.cluster.cov.stat%>%filter(cluster %in%c(2:7,39,99)),
#ggplot(bwa.cluster.cov.stat%>%filter(cluster %in%c(5,16,24,32)),       
       aes(x=position,y=coverage,color=type)) + geom_line() +
  facet_grid(cluster~taxonomy,scale="free")
```
This figure summarised the reads coverage belonging to each cluster. 5, 6, 7 and 39 is well uniquly mapping to a specific species. Cluster 2, 3 and 99 seem contained hybridized DNA sources.

# Distance assessment
```{r fig.width=5, fig.height=3}
dist.stat <- bMin2.raw.dist.stat%>%filter(tax1!=""&tax2!="")
dist.stat.sum <- ddply(dist.stat,c("tax1","tax2","distance"),summarise,count=sum(count))
dist.stat.sum$group <-ifelse(dist.stat.sum$tax1==dist.stat.sum$tax2,"same","diff")
ggplot(dist.stat.sum,aes(x=distance,y=count,color=interaction(tax1,tax2),linetype=group)) +
  geom_line() + geom_vline(xintercept = 0.1,linetype=2)
```
The mash distance between beads from the same species are distributed with their peak smaller than 0.1, while that from different species, the distance peak are larger than 0.1. Thus we can use 0.1 as the threshold to filter distance list in this dataset.

**sum them together**
```{r}
dist.stat.sum2 <- ddply(dist.stat.sum,c("distance","group"),summarise,count=sum(count))
ggplot(dist.stat.sum2,aes(x=distance,y=count,linetype=group),color="grey50",alpha=.3) +
  geom_line() + geom_point(size=.5) + geom_vline(xintercept = 0.1,linetype=2)
```
**ITS region detection**
```{r}
its.anno.sum <- read.table("../../Results/APR842_00/beadsAnno/its.anno.sum",
    col.names=c("BB","reads","mainTax","count","SSU","ITS1","S58","ITS2","LSU"))

its.anno.melt <- melt(its.anno.sum,c("BB","reads","mainTax","count"),
                      variable.name = "region",value.name = "Rcount")
its.anno.melt$rate <- its.anno.melt$Rcount/its.anno.melt$reads
ggplot(its.anno.melt,aes(x=rate,color=region)) + geom_density() + 
  scale_y_log10()
```
```{r}
bMin2.raw.dist.its.stat <- read.table(
  "../../Results/APR842_00/TT1M/mash/bMin2.raw.dist.stat",header = T,row.names = NULL)
bMin2.raw.dist.its.stat$n5.8s <- 
  (bMin2.raw.dist.its.stat$mainR1==9) + (bMin2.raw.dist.its.stat$mainR2==9)
bMin2.raw.dist.its.stat$taxGrp <- 
  ifelse(bMin2.raw.dist.its.stat$tax1==bMin2.raw.dist.its.stat$tax2,"same","diff")
bMin2.raw.dist.its.stat.sum1 <- ddply(bMin2.raw.dist.its.stat%>%filter(tax1!="*"&tax2!="*"),
                                     c("tax1","tax2","n5.8s","distance"),summarise,
                                     count=sum(count))
p1 <- ggplot(bMin2.raw.dist.its.stat.sum1,
       aes(x=distance,y=count,color=factor(n5.8s))) +
  geom_line(alpha=.3) + geom_point(size=.5,alpha=.5) + 
  geom_vline(xintercept = 0.1,linetype=2) + #scale_y_log10() +
  facet_wrap(~tax1+tax2,ncol=5,scales="free")
p1

bMin2.raw.dist.its.stat.sum2 <- ddply(bMin2.raw.dist.its.stat%>%filter(tax1!="*"&tax2!="*"),
                                     c("taxGrp","mainR1","mainR2","distance"),summarise,
                                     count=sum(count))
p2 <- ggplot(bMin2.raw.dist.its.stat.sum2,
       aes(x=distance,y=count,color=taxGrp)) +
  geom_line(alpha=.3) + geom_point(size=.5,alpha=.5) + 
  geom_vline(xintercept = 0.1,linetype=2) + #scale_y_log10() +
  facet_wrap(~mainR1+mainR2,ncol=5,scales = "free")
p2
```

```{r}
dist.sum.SSU <- 
  ddply(bMin2.raw.dist.its.stat%>%filter(tax1!="*"&mainR1==mainR2&mainR1==7),
        c("tax1","tax2","mainR1","distance"),summarise, count=sum(count))
p3 <- ggplot(bMin2.raw.dist.its.stat.sum3,
       aes(x=distance,y=count,color=tax1)) +
  geom_line(alpha=.3) + geom_point(size=.5,alpha=.5) + 
  geom_vline(xintercept = 0.1,linetype=2) + #scale_y_log10() +
  facet_grid(mainR1~tax1,scales = "free")
p3
```


**APR831**
```{r fig.width=5, fig.height=3}
dist.831.stat <- bMin2.831.dist.stat%>%filter(tax1!=""&tax2!="")
dist.831.stat.sum <- ddply(dist.831.stat,c("tax1","tax2","distance"),summarise,count=sum(count))
dist.831.stat.sum$group <-ifelse(dist.831.stat.sum$tax1==dist.831.stat.sum$tax2,"same","diff")
ggplot(dist.831.stat.sum,aes(x=distance,y=count,color=interaction(tax1,tax2),linetype=group)) +
  geom_line()
```
In this dataset, the peak of distance within the same species are less than 0.05, while others still larger than (or close to) 0.1, which indicated the better performance when using the distance range (0,0.04) for cluster.

###### #TO BE CONTINUED
