---
title: "RCA merged results"
author: "Chao"
date: "4/24/2019"
output: 
  html_notebook:
    code_fold: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(reshape2)
library(ggplot2)
library(plyr)
library(dplyr)
library(RColorBrewer)
```

# Background

Samples involved:

| Sample ID             | lane | barcode | Note    |
| --------------------- | ---- | ------- | ------- |
| ALS 16S               | M01  | 1       | RCA     |
| Zymo 16S              | M01  | 2       | RCA     |
| ALS ITS               | M01  | 3       | RCA     |
| 159:160:161=100:10:1  | M01  | 4       | RCA     |
| 163                   | M01  | 5       | RCA     |
| 164                   | M01  | 6       | RCA     |
| ALS ITS 10X dilute    | M01  | 7       | Non-RCA |
| Zymo 16S 10X dilute   | M01  | 8       | Non-RCA |
| Zymo 16S-2            | M02  | 1       | RCA     |
| Zymo 16S-3            | M02  | 2       | RCA     |
| ALS ITS-2             | M02  | 3       | RCA     |
| ALS ITS-3             | M02  | 4       | RCA     |
| Mock16_23S            | M02  | 5       | RCA     |
| Skin16_23S            | M02  | 6       | RCA     |
| ALS ITS 10X   dilute  | M02  | 7       | Non-RCA |
| Zymo 16S 10X   dilute | M02  | 8       | Non-RCA |   



Fungal mock strains:

| ID   | Prefix | Tag  | Taxonomy                    | Alternative                |
| ---- | ------ | ---- | --------------------------- | -------------------------- |
| 158  | JOMC   | Au   | Aspergillus ustus           | Aspergillus puniceus       |
| 159  | BCGH   | Tk   | Trichoderma koningii        | Trichoderma pseudokoningii |
| 160  | JQFZ   | Pe   | Penicillium expansum        |                            |
| 161  | AACD   | An   | Aspergillus nidulans        |                            |
| 162  | JMSF   | Pc   | Penicillium chrysogenum     |                            |
| 163  | MBDJ   | Tl   | Trichoderma longibrachiatum |                            |
| 164  | AAIL   | Tr   | Trichoderma viride          | Trichoderma reesei         |
| 165  | ABDG   | Ta   | Trichoderma atroviride      |                            |

find code note in `VSEARCH.md`

# 

# Load 
```{r}
SAM="RCA197M01_4"
pfx=paste0("../../Results/RCAs/RCA197FIN/",SAM,"/VSEARCH/summary.BI.megahit.contig.CLIP4")

name.uc <- c("RecordType","ClusterNO","size","Percentage","Orientation","NA1","NA2","CIGAR","query","centroid")
uc.derep <- read.table(paste0(pfx,".derep.full.uc"),col.names = name.uc)
uc.clust <- read.table(paste0(pfx,".derep.preclustered.uc"),col.names = name.uc)
uc.chime <- read.table(paste0(pfx,".derep.nonchimeras.LFRs.99.uc"),col.names = name.uc)
LFRID <- read.table(paste0(pfx,".derep.nonchimeras.LFRs.99.IDs"),col.names = c("LFR","BIC"))

#
LFRb6 <- read.table(paste0(pfx,".derep.nonchimeras.LFRs.98.b6"),
            col.names = c("LFR","Subject","pident","len","mismatch","gap","qs","qe","ss","se","eval","score","qLen"))
LFRb6m <- read.table(paste0(pfx,".derep.nonchimeras.LFRs.98.b6.more"),
            col.names = c("LFR","Subject","pident","len","mismatch","gap","qs","qe","ss","se","eval","score","qLen","region"))

```

# Curve
```{r}
LFRID$level <- "represent"
LFRD0 <- merge(LFRID[,1:2],(uc.chime%>%filter(RecordType=="H"))[,9:10],by.x="BIC",by.y="centroid")
LFRD0$level <- "centroid"

LFRD1 <- merge(LFRD0[,2:3],(uc.clust%>%filter(RecordType=="H"))[,9:10],by.x="query",by.y="centroid")
LFRD1$level <- "clusters"
LFRD2 <- merge(LFRD1[,2:3],(uc.derep%>%filter(RecordType=="H"))[,9:10],by.x="query.y",by.y="centroid")
LFRD2$level <- "duplicates"

colnames(LFRD0)[3] <- "BIC"
colnames(LFRD1)[3] <- "BIC"
colnames(LFRD2)[3] <- "BIC"

LFRDM <- rbind(LFRID,LFRD0[,-1],LFRD1[,-1],LFRD2[,-1])
tmp <- matrix(unlist(strsplit(as.character(LFRDM$BIC),"_")),ncol=8,byrow = T)
LFRDM$BI <- as.factor(tmp[,1])
LFRDM$CLIP <- as.factor(tmp[,6])
LFRDM$len <- as.numeric(tmp[,8])
LFR.stat <- ddply(LFRDM,c("LFR"),summarise,size=length(unique(BI)))
```
```{r}
LFR.map <- merge(LFR.stat,LFRb6m,by="LFR")
LFR.map$TaxID <- substr(LFR.map$Subject,1,4)
LFR.map$LFR <- reorder(LFR.map$LFR,LFR.map$size,sum)
```


```{r}
LFR.map.stat <- ddply(LFR.map,c("LFR","TaxID","Subject","region","qLen"),summarise,count=sum(size))
tmp <- matrix(unlist(strsplit(as.character(LFR.map.stat$region),"-")),ncol=2,byrow = T)
LFR.map.stat$regStart <- as.numeric(tmp[,1])
LFR.map.stat$regEnd <- as.numeric(tmp[,2])

LFR.map.stat$region <- reorder(LFR.map.stat$region,LFR.map.stat$regStart,mean)
LFR.map.stat$TaxID <- reorder(LFR.map.stat$TaxID,LFR.map.stat$count,sum)

```

```{r}
LFR.map.sum <- ddply(LFR.map.stat%>%filter(qLen>1999),c("TaxID","region"),summarise,count=sum(count))
ggplot(LFR.map.stat%>%filter(qLen>1999),aes(x=region,y=count)) + 
  geom_bar(stat="identity",aes(fill=LFR)) +
  geom_text(data=LFR.map.sum,aes(y=count,label=count)) +
  guides(fill=F) + facet_grid(TaxID~.,scale="free",space="free") + coord_flip()
```

```{r}
LFR.map.sum <- ddply(LFR.map.stat%>%filter(count>2&qLen>999&qLen<2000),c("TaxID","region"),summarise,count=sum(count))
ggplot(LFR.map.stat%>%filter(count>2&qLen>999&qLen<2000),aes(x=region,y=count)) + 
  geom_bar(stat="identity",aes(fill=LFR)) +
  geom_text(data=LFR.map.sum,aes(y=count,label=count)) +
  guides(fill=F) + facet_grid(TaxID~.,scale="free",space="free") + coord_flip()
```

```{r}
LFR.map.stat2 <- ddply(LFR.map.stat%>%filter(count>10),c("TaxID","region","regStart","regEnd"),
                       summarise,count=sum(count))
LFR.map.stat2 <- melt(LFR.map.stat2,id.vars = c("TaxID","region","count"),variable.name = "board",value.name = "pos")
LFR.map.stat2$shift <- ifelse(LFR.map.stat2$board=="regStart",-1,+1)
LFR.map.stat3 <- LFR.map.stat2
LFR.map.stat3$pos <- LFR.map.stat2$pos + LFR.map.stat2$shift
LFR.map.stat3$count <- 0 
LFR.map.stat4 <- rbind(LFR.map.stat2,LFR.map.stat3)
LFR.map.stat4 <- LFR.map.stat4[order(LFR.map.stat4$pos),]
ggplot(LFR.map.stat4,aes(x=pos,y=count)) + 
  geom_path() +
  guides(color=F) + facet_grid(~TaxID+region,scale="free",space="free")
```

#mapping all reads
```{r}
b2r <- read.table(paste0("../../Results/RCAs/RCA197FIN/",SAM,"/VSEARCH/test.B2R"),
           col.names = c("BB","LFR","UNIQ","MULT"))

b2rm <- merge(LFR.map.stat,b2r,by="LFR")
b2rm$LFR <- reorder(b2rm$LFR,as.numeric(b2rm$Subject),mean)
```

```{r}
ggplot(b2rm%>%filter(UNIQ>20),aes(x=region,y=UNIQ+MULT)) + 
  geom_bar(stat="identity",aes(fill=LFR)) +
  guides(fill=F) + facet_grid(TaxID~.,scale="free",space="free") + coord_flip()

b2rm.20.sum <- ddply(b2rm%>%filter(UNIQ>20),c("TaxID"),summarise,count=sum(UNIQ+MULT))
ggplot(b2rm%>%filter(UNIQ>20),aes(x=TaxID,y=UNIQ+MULT)) + 
  geom_bar(stat="identity",aes(fill=LFR)) +
  geom_text(data=b2rm.20.sum,aes(y=count,label=count)) +
  guides(fill=F) #+ facet_grid(TaxID~.,scale="free",space="free") + coord_flip()
```


#Quantification

```{r}
LFR.prop <- read.table(paste0(pfx,".derep.nonchimeras.LFRs.99.prop"), col.names = c("count","LFR"))

LFR.map.prop <- merge(LFR.map,LFR.prop,by="LFR")
```

```{r}
LFR.map.prop.sum <- ddply(LFR.map.prop%>%filter(count>999&qLen>1999),c("TaxID","region"),summarise,count=sum(count))
ggplot(LFR.map.prop%>%filter(count>999&qLen>1999),aes(x=region,y=count)) + 
  geom_bar(stat="identity",aes(fill=LFR)) +
  geom_text(data=LFR.map.prop.sum,aes(y=count,label=count)) +
  guides(fill=F) + facet_grid(TaxID~.,scale="free",space="free") + coord_flip()
```

#bbs
```{r}
pick.map <- unique((LFR.map.prop%>%filter(count>999&qLen>1999))[,c(1:3,16:17)])
pick.LFRs <- as.character((LFR.map.prop%>%filter(count>999&qLen>1999))$LFR)
bbs.dat <- read.table(paste0(pfx,".derep.nonchimeras.LFRs.99.bbs"),
                      col.names = c("mainTax","commTax","BBs","sumUniq","sumComm","pU","PM"))

bbs.dcast <- dcast(bbs.dat,mainTax~commTax,value.var = "PM") 
rownames(bbs.dcast) <- bbs.dcast$mainTax
bbs.dcast <- as.matrix(bbs.dcast[,-1])
bbs.select <- bbs.dcast[which(rownames(bbs.dcast)%in%pick.LFRs),which(colnames(bbs.dcast)%in%pick.LFRs)]
library(pheatmap)
pheatmap(bbs.select)
```






#FIN.