---
title: "RCA hbridize"
author: "Chao"
date: "4/24/2019"
output: 
  html_notebook:
    code_fold: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(reshape2)
library(ggplot2)
library(plyr)
library(dplyr)
library(RColorBrewer)
```

# Background

Samples involved:

| Sample ID             | lane | barcode | Note    |
| --------------------- | ---- | ------- | ------- |
| ALS 16S               | M01  | 1       | RCA     |
| Zymo 16S              | M01  | 2       | RCA     |
| ALS ITS               | M01  | 3       | RCA     |
| 159:160:161=100:10:1  | M01  | 4       | RCA     |
| 163                   | M01  | 5       | RCA     |
| 164                   | M01  | 6       | RCA     |
| ALS ITS 10X dilute    | M01  | 7       | Non-RCA |
| Zymo 16S 10X dilute   | M01  | 8       | Non-RCA |
| Zymo 16S-2            | M02  | 1       | RCA     |
| Zymo 16S-3            | M02  | 2       | RCA     |
| ALS ITS-2             | M02  | 3       | RCA     |
| ALS ITS-3             | M02  | 4       | RCA     |
| Mock16_23S            | M02  | 5       | RCA     |
| Skin16_23S            | M02  | 6       | RCA     |
| ALS ITS 10X   dilute  | M02  | 7       | Non-RCA |
| Zymo 16S 10X   dilute | M02  | 8       | Non-RCA |   



Zymo mock strains:

| Species                  | Genomic DNA | 16S Only1 | 16S & 18S1 | Genome Copy2 | Cell Number3 |
| ------------------------ | ----------- | --------- | ---------- | ------------ | ------------ |
| Pseudomonas aeruginosa   | 12          | 4.2       | 3.6        | 6.1          | 6.1          | +
| Escherichia coli         | 12          | 10.1      | 8.9        | 8.5          | 8.5          | +
| Salmonella enterica      | 12          | 10.4      | 9.1        | 8.7          | 8.8          | +
| Lactobacillus fermentum  | 12          | 18.4      | 16.1       | 21.6         | 21.9         |
| Enterococcus faecalis    | 12          | 9.9       | 8.7        | 14.6         | 14.6         | +
| Staphylococcus aureus    | 12          | 15.5      | 13.6       | 15.2         | 15.3         | +
| Listeria monocytogenes   | 12          | 14.1      | 12.4       | 13.9         | 13.9         | +
| Bacillus subtilis        | 12          | 17.4      | 15.3       | 10.3         | 10.3         | +
| Saccharomyces cerevisiae | 2           | NA        | 9.3        | 0.57         | 0.29         |
| Cryptococcus neoformans  | 2           | NA        | 3.3        | 0.37         | 0.18         |


# load 
```{r}

BB.stat <- read.table("../../Results/RCAs/RCA197FIN/A42/clean/BB.stat",
                         col.names = c("count","rpb","cumsum"))
contig.tsv <- read.table("../../Results/RCAs/RCA197FIN/A42/summary.BI.megahit.contig.tsv",
                         col.names = c("BI","id","flag","cov","len"))
B.kmer <- read.table("../../Results/RCAs/RCA197FIN/A02/mash/bMin2.msh.2.tsv",
                     col.names = c("BB","rpb","kmers"))
#clip.num <- read.table("../../Results/RCAs/RCA197FIN/A42/summary.BI.megahit.clip.num",                       col.names = c("BI","len"))

clip2zymo <- read.table("../../Results/RCAs/RCA197FIN/A42/clip2zymo.stat",
                       col.names = c("BI","id","flag","cov","len0","clip","len","ref","cigar","match","fLen","NM"))
clip2zymo.region <- read.table("../../Results/RCAs/RCA197FIN/A42/summary.BI.megahit.clip.zymo.region",
                       col.names = c("BI","id","flag","cov","len0","clip","len","ref","region","match","fLen","regionLen","regionPos","PosPct"))
rnahead.stat <- read.table("../../Results/RCAs/RCA197FIN/A42/summary.BI.megahit.rRNA.stat",
                       col.names = c("BI","id","len","start","end","unit"))
BI.lst <- read.table("../../Results/RCAs/RCA197FIN/A42/Assemble_BI/ID.lst",
                     col.names = c("BI0","BB","kmers","rpb"))

BI.freq.tsv <- read.table("../../Results/RCAs/RCA197FIN/A42/mash/BI.msh.freq.tsv", sep="\t",
                           col.names = c("BB","rpb","kmers","bins","hash","hash1","hash2"))
```
```{r}
BI.freq.tsv0 <- BI.freq.tsv
BI.freq.tsv0$rpbin <- round(BI.freq.tsv0$rpb/BI.freq.tsv0$bins,0)
BI.freq.tsv0$kCov <- BI.freq.tsv0$rpb*140/BI.freq.tsv0$kmers

ggplot(BI.freq.tsv0,aes(x=factor(bins),y=kmers)) + geom_boxplot()
ggplot(BI.freq.tsv0%>%filter(rpbin<100),aes(x=factor(rpbin),y=kCov)) + geom_boxplot()
```


# Curate
```{r}
BI.lst$BI <- sprintf("BI%08d",BI.lst$BI0)
#mdat.ctg <- merge(merge(BI.lst,contig.tsv,by="BI"),B.kmer[,-2],by="BB")
mdat.ctg <- merge(merge(BI.lst,contig.tsv,by="BI"),BI.freq.tsv[,-c(2:3)],by="BB")
mdat.ctg$cov0 <- round(mdat.ctg$cov/2-0.49,0) * 2
mdat.ctg$len0 <- round(mdat.ctg$len/500-0.49,0) * 500
mdat.ctg$kCov <- mdat.ctg$rpb*140/mdat.ctg$kmers
mdat.ctg$kCov0 <- round(mdat.ctg$kCov,0)
mdat.ctg$h12 <- round(mdat.ctg$hash1/mdat.ctg$hash2,0)
mdat.ctg$rpbin <- round(mdat.ctg$rpb/mdat.ctg$bins,0)

ggplot(mdat.ctg%>%filter(cov0<50),aes(x=factor(cov0),y=len,fill=factor(flag))) + geom_boxplot()
ggplot(mdat.ctg,aes(x=factor(len0),y=kCov,fill=factor(flag))) + geom_boxplot()
ggplot(mdat.ctg,aes(x=factor(kCov0),y=len,fill=factor(flag))) + geom_boxplot()
ggplot(mdat.ctg%>%filter(h12<20),aes(x=factor(h12),y=len,fill=factor(flag))) + geom_boxplot()
ggplot(mdat.ctg%>%filter(rpbin<100),aes(x=factor(rpbin),y=len,fill=factor(flag))) + geom_boxplot()
```
```{r}
ggplot(mdat.ctg%>%filter(bins<49),aes(x=factor(bins),y=len,fill=factor(flag))) + geom_boxplot()
```



```{r}
BI.num <- as.data.frame(table(mdat.ctg$BI))
colnames(BI.num) <- c("BI","member")
mdat.ctg3 <- merge(mdat.ctg,BI.num,by="BI")
mdat.ctg3$weight <- 1/mdat.ctg3$member
mdat.ctg3$member[which(mdat.ctg3$member>3)] <- "4&more"
mdat.ctg3$cov0 <- round(mdat.ctg3$cov,0)
mdat.ctg3$kCov0 <- round(mdat.ctg3$kCov,0)

ddply(mdat.ctg3,"member",summarise,num=sum(weight))

ggplot(mdat.ctg3%>%filter(cov0<30),aes(x=factor(cov0),y=len,fill=factor(flag))) + geom_boxplot() + 
  facet_grid(member~.)

```

```{r}
mdat.ctg3$kmer0 <- round(mdat.ctg3$kmers/100,0)*100
ggplot(mdat.ctg3%>%filter(kmers<3999),aes(x=factor(kmer0),y=len,fill=factor(flag))) + geom_boxplot() + 
  facet_grid(member~.)

mdat.ctg3$rpb0 <- round(mdat.ctg3$rpb/10,0)*10
ggplot(mdat.ctg3%>%filter(rpb<299),aes(x=factor(rpb0),y=len,fill=factor(flag))) + geom_boxplot() + 
  facet_grid(member~.)
```

结合clip比对信息：
```{r}
cmdat <- merge(merge(mdat.ctg[,-c(9,11,13)],clip2zymo[,c(1,2,5:7,8:12)],by=c("BI","id")),
               clip2zymo.region[,c("BI","id","region")],by=c("BI","id"))
cmdat$pct <- cmdat$match/cmdat$len
cmdat$len100 <- round(cmdat$len/100,0)*100
```
集合barrnap预测信息：
```{r}
rnahead.stat$rnaLen <- rnahead.stat$end-rnahead.stat$start+1
rnahead.stat$rnaPct <- rnahead.stat$rnaLen/rnahead.stat$len

cmdat <- merge(cmdat,rnahead.stat[,c(1,2,6:8)],by=c("BI","id"),all.x=T)
cmdat$rnaPct[which(is.na(cmdat$unit))] <- 0
levels(cmdat$unit) <- c("16S","23S","5S","mis")
cmdat$unit[which(is.na(cmdat$unit))] <- "mis"
```

```{r}
ggplot(cmdat,aes(x=pct,fill=factor(len100))) + geom_histogram() + scale_x_continuous(breaks = seq(0,1,0.2))
```

根据比对coverage来判断


```{r}
ggplot(cmdat%>%filter(pct<0.9),aes(x=len)) + geom_histogram()

```

```{r}
hybFun <- function(d){
  rN <- length(unique(d$species))
  res <- d[1,c("BI"),F]
  res$Rpb <- rN
  return(res)
}

cmdat$species <- sapply(strsplit(as.character(cmdat$ref),".",fixed = T),function(x) return(x[[1]]))
hybStat <- ddply(cmdat,"BI",hybFun)
hybStat$hyb <- ifelse(hybStat$Rpb>1,T,F)
cmstat <- merge(cmdat,hybStat,by="BI")
cmstat$kCov0 <- round(cmstat$rpb*140/cmstat$kmers,0)
cmstat$rpb0 <- round(cmstat$rpb/10,0) * 10
cmstat$kmers00 <- round(cmstat$kmers/100,0) * 100
cmstat$regUnit <- substr(cmstat$region,1,3)

ggplot(cmstat,aes(x=species,fill=factor(Rpb))) + geom_bar(stat="count") + theme(axis.text.x = element_text(angle=90,vjust=0))
ggplot(cmstat,aes(x=species,fill=factor(Rpb))) + geom_bar(stat="count",position="fill") + theme(axis.text.x = element_text(angle=90,vjust=0))
ggplot(cmstat,aes(x=unit,fill=factor(Rpb))) + geom_bar(stat="count") + theme(axis.text.x = element_text(angle=90,vjust=0))
ggplot(cmstat,aes(x=unit,fill=factor(Rpb))) + geom_bar(stat="count",position="fill") + theme(axis.text.x = element_text(angle=90,vjust=0))
ggplot(cmstat%>%filter(rpb0<300),aes(x=factor(rpb0),fill=factor(Rpb))) + geom_bar(stat="count",position="fill") + theme(axis.text.x = element_text(angle=90,vjust=0))

ggplot(cmstat,aes(x=fLen,color=regUnit,linetype=hyb)) + geom_density(size=1,alpha=.7) + facet_grid(regUnit~.,scale="free_y")
ggplot(cmstat,aes(x=fLen,color=unit,linetype=hyb)) + geom_density(size=1,alpha=.7) + facet_grid(unit~.,scale="free_y")
ggplot(cmstat,aes(x=rnaPct,color=unit,linetype=hyb)) + geom_density(size=1,alpha=.7) + facet_grid(unit~.,scale="free_y")
ggplot(cmstat,aes(x=pct,color=unit,linetype=hyb)) + geom_density(size=1,alpha=.7) + facet_grid(unit~.,scale="free_y") + scale_x_continuous(breaks=seq(0,1,.2))
ggplot(cmstat,aes(x=pct,color=unit,linetype=hyb)) + geom_density(size=1,alpha=.7) + facet_grid(unit~.,scale="free_y") + scale_x_continuous(breaks=seq(0,1,.2))


ggplot(cmstat,aes(x=kCov0,fill=hyb)) + geom_bar(stat="count",position="stack")
ggplot(cmstat,aes(x=kCov0,fill=hyb)) + geom_bar(stat="count",position="fill")
ggplot(cmstat%>%filter(cov<30),aes(x=cov0,fill=hyb)) + geom_bar(stat="count",position="stack")
ggplot(cmstat%>%filter(cov<30),aes(x=cov0,fill=hyb)) + geom_bar(stat="count",position="fill")

ggplot(cmstat,aes(x=factor(rpb0),fill=hyb)) + geom_bar(stat="count",position="fill") + 
  theme(axis.text.x = element_text(angle=90,vjust=0))

```

taxonomy percentage:
```{r}
zymo.copy.num <- data.frame(species=c("Pseudomonas_aeruginosa","Escherichia_coli","Salmonella_enterica","Lactobacillus_fermentum",
  "Enterococcus_faecalis","Staphylococcus_aureus","Listeria_monocytogenes","Bacillus_subtilis","Saccharomyces_cerevisiae","Cryptococcus_neoformans"),
  TC16S18S=c(3.6,8.9,9.1,16.1,8.7,13.6,12.4,15.3,9.3,3.3),copyNum=c(4,7,7,5,4,6,6,10,109,60))

cmstat$species <- as.factor(cmstat$species)
levels(cmstat$species) <- c(NA,"Bacillus_subtilis","Cryptococcus_neoformans","Enterococcus_faecalis","Escherichia_coli","Escherichia_coli",
                            "Lactobacillus_fermentum","Listeria_monocytogenes","Pseudomonas_aeruginosa","Saccharomyces_cerevisiae",
                            "Salmonella_enterica","Staphylococcus_aureus")
ggplot(zymo.copy.num,aes(x="zymo",y=TC16S18S,fill=species)) + geom_bar(stat="identity",position="fill") + ylab("16S & 18S")
ggplot(cmstat%>%filter(rpb<300),aes(x=factor(rpb0),fill=species)) + geom_bar(stat="count",position="fill") + 
  theme(axis.text.x = element_text(angle=90,vjust=0))


ggplot(zymo.copy.num%>%filter(!species%in%c("Saccharomyces_cerevisiae","Cryptococcus_neoformans")),
       aes(x="zymo",y=TC16S18S,fill=species)) + geom_bar(stat="identity",position="fill") + ylab("16S & 18S")
ggplot(cmstat%>%filter(rpb<300&!species%in%c("Saccharomyces_cerevisiae","Cryptococcus_neoformans")),aes(x=factor(rpb0),fill=species)) +
  geom_bar(stat="count",position="fill") + 
  theme(axis.text.x = element_text(angle=90,vjust=0))
ggplot(cmstat%>%filter(rpb<300&!species%in%c("Saccharomyces_cerevisiae","Cryptococcus_neoformans")),aes(x=factor(rpb0))) + 
  geom_bar(stat="count") + theme(axis.text.x = element_text(angle=90,vjust=0))

ggplot(cmstat%>%filter(unit!="5S"&rpb<300&!species%in%c("Saccharomyces_cerevisiae","Cryptococcus_neoformans")),aes(x=factor(rpb0),fill=species)) +
  geom_bar(stat="count",position="fill") + theme(axis.text.x = element_text(angle=90,vjust=0)) + facet_grid(unit~.)

ggplot(cmstat%>%filter(unit!="5S"&!species%in%c("Saccharomyces_cerevisiae","Cryptococcus_neoformans")),aes(x="all",fill=species)) +
  geom_bar(stat="count",position="fill") + theme(axis.text.x = element_text(angle=90,vjust=0)) + facet_grid(unit~.)

```


```{r}
cmstat.clean <- cmstat%>%filter(ref2!="hybridize")
cmstat.clean$kCov0 <- round(cmstat.clean$rpb*140/cmstat.clean$kmers,0)
cmstat.clean$kmers0 <- round(cmstat.clean$kmers/100,0) * 100
cmstat.clean$rpb0 <- round(cmstat.clean$rpb/10,0) * 10

ggplot(cmstat.clean%>%filter(kCov0<20),aes(x=factor(kCov0),fill=factor(ref2))) + geom_bar(stat="count")
ggplot(cmstat.clean%>%filter(kCov0<20),aes(x=factor(kCov0),fill=factor(ref2))) + geom_bar(stat="count",position="fill")
ggplot(cmstat.clean%>%filter(kmers<3000),aes(x=factor(kmers0),fill=factor(ref2))) + geom_bar(stat="count",position="fill")
ggplot(cmstat.clean%>%filter(rpb<300),aes(x=factor(rpb0),fill=factor(ref2))) + geom_bar(stat="count")
ggplot(cmstat.clean%>%filter(rpb<300),aes(x=factor(rpb0),fill=factor(ref2))) + geom_bar(stat="count",position="fill")

```

```{r}
cmdat2 <- cmdat
cmdat2$kCov0 <- round(cmdat2$rpb*140/cmdat2$kmers,0)
cmdat2$rpb0 <- round(cmdat2$rpb/10,0) * 10
cmdat2$len00 <- round(cmdat2$len/100,0) * 100
cmdat2$kmers00 <- round(cmdat2$kmers/100,0) * 100
cmdat2$kmerLen <- substr(cmdat2$id,1,4)

ggplot(cmdat2%>%filter(kCov0<20),aes(x=factor(kCov0),fill=factor(ref2))) + geom_bar(stat="count",position="fill")
ggplot(cmdat2%>%filter(rpb<300),aes(x=factor(rpb0),fill=factor(ref2))) + geom_bar(stat="count",position="fill")
ggplot(cmdat2%>%filter(rpb<300),aes(x=factor(ref2),fill=factor(rpb0))) + geom_bar(stat="count",position="fill")
ggplot(cmdat2,aes(x=factor(len00),fill=factor(ref2))) + geom_bar(stat="count")
ggplot(cmdat2,aes(x=factor(len00),fill=factor(ref2))) + geom_bar(stat="count",position="fill")

ggplot(cmdat2,aes(x=factor(len00),y=pct)) + geom_boxplot()
ggplot(cmdat2,aes(x=factor(len00),y=pct,fill=factor(flag))) + geom_boxplot() + scale_y_continuous(breaks=seq(0,1,0.2))
ggplot(cmdat2,aes(x=factor(cov0),y=pct,fill=factor(flag))) + geom_boxplot() + scale_y_continuous(breaks=seq(0,1,0.2))

```

```{r}
ggplot(cmdat2%>%filter(rpb<300),aes(x=factor(rpb0),fill=factor(ref2))) + geom_bar(stat="count",position="fill")

```


模型
```{r}
Exp.rpb <- sum(BB.stat$rpb*BB.stat$count)/sum(BB.stat$count)
Exp.rpb0 <- round(Exp.rpb,0)

BB.stat$prob <- BB.stat$count/(1.2*sum(BB.stat$count))
mock.dat <- data.frame(rpb=seq(0,49),
                       pois=(dpois(seq(0,49),1)+dpois(seq(0,49),2)+dpois(seq(0,49),3))/3,
                       norm=dnorm(seq(0,49),mean=5,sd=15))
mock.dat$p2 <- mock.dat$pois*0.8+mock.dat$norm*0.2
mock.dat <- merge(mock.dat,BB.stat[,c(2,4)],by="rpb",all.x=T)
mmock <- melt(mock.dat,id.vars = "rpb",variable.name = "method",value.name = "p")
ggplot(mmock,aes(x=rpb,y=p,color=method)) + geom_line() + geom_point()
```

```{r}
sdat.ctg <- ddply(mdat.ctg,c("BI","rpb","flag"),summarise,len=sum(len))
sdat.ctg$len0 <- round(sdat.ctg$len/500-0.49,0) * 500

sdat.ctg2 <- ddply(sdat.ctg,c("rpb","flag","len0"),summarise,count=length(BI))
sdat.ctg2 <- ddply(sdat.ctg2,c("flag","len0"),transform,frac=count/sum(count))

ggplot(sdat.ctg2%>%filter(len0>499&len0<2000),aes(x=rpb,y=frac,color=factor(len0))) + geom_line() + facet_grid(flag~.,scale="free") + xlim(0,200)
```

```{r}
ggplot(B.kmer,aes(x=kmers)) + geom_histogram() + scale_x_log10() + annotation_logticks(sides="b")

B.kmer2 <- B.kmer
B.kmer2$rpbL <- round(B.kmer2$rpb/10-0.49,0) *10
B.kmer2$rpbL[which(B.kmer2$rpbL>=150)] <- ">=150"
B.kmer2$rpbL <- factor(B.kmer2$rpbL,levels=c(seq(0,140,10),">=150"))
B.kmer2$kL <- round(B.kmer2$kmers/100-0.49,0) *100
B.kmer2$kL[which(B.kmer2$kL>=3000)] <- 3000
B.kmer2$kL[which(B.kmer2$kL<1000)] <- 900
B.kmer2$kL <- factor(B.kmer2$kL,levels=c(seq(900,3000,100)))

ggplot(B.kmer2,aes(x=rpb,fill=factor(kL))) + geom_histogram(position="stack") +
  scale_x_log10() + annotation_logticks(sides="b")

ggplot(B.kmer2,aes(x=rpb,fill=factor(kL))) + geom_histogram(position="fill") +
  scale_x_log10() + annotation_logticks(sides="b")
```

```{r}
ggplot(B.kmer2,aes(x=factor(rpbL),y=kmers)) + geom_boxplot() + 
  scale_y_continuous(breaks=seq(0,1e4,2000),limits = c(0,1e4))
```

```{r}
B.kmer2$kCov <- B.kmer2$rpb*160/B.kmer2$kmers

ggplot(B.kmer2,aes(x=factor(rpbL),y=kCov)) + geom_boxplot() + 
  scale_y_continuous(breaks=seq(0,30,2))

ggplot(B.kmer2,aes(x=kmers,y=kCov)) + 
  geom_density_2d()
```

```{r}
ggplot(B.kmer2,aes(x=factor(kL),y=kCov)) + geom_boxplot() + 
  scale_y_continuous(breaks=seq(0,10,2),limits = c(0,10))
```

```{r}
B.kmer2S <- ddply(B.kmer2,c("kL","rpbL"),summarise,count=length(BB))
ggplot(B.kmer2S,aes(x=rpbL,y=kL,fill=log10(count))) + 
  geom_tile() + scale_fill_gradientn(colors=rev(brewer.pal(11,"Spectral")))
```

```{r}
ggplot(mdat2,aes(x=slen,fill=factor(num))) + geom_histogram(position="stack")
```

```{r}
ggplot(mdat2%>%filter(num%in%seq(1,2)),aes(x=rpb.x,y=kmers,color=factor(num))) + 
  geom_density_2d() + 
  scale_y_continuous(breaks=seq(0,8000,1000)) +
  scale_x_continuous(breaks=seq(100,500,20))
```

```{r}
mdat2$slen <- mdat2$mlen*mdat2$num
ggplot(mdat2%>%filter(num%in%seq(1,2)),aes(x=kmers,y=slen,color=factor(num))) + 
  geom_density_2d()
```

```{r}
mdat2$kCov <- mdat2$rpb.x*160/mdat2$kmers
ggplot(mdat2%>%filter(num%in%seq(1,3)),aes(x=kmers,y=kCov,color=factor(num))) + 
  geom_density_2d() + scale_y_continuous(breaks=seq(4,12))
```

1000 < kmers < 3000 & kCov > 6


```{r}
mdat3 <- B.kmer2%>%filter(kCov>6&kmers>1000)

ggplot(mdat3,aes(x=rpb,y=kmers)) + geom_density_2d()
```






#FIN.