---
title: "RCA test"
author: "Chao"
date: "4/24/2019"
output: 
  html_notebook:
    code_fold: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(reshape2)
library(ggplot2)
library(plyr)
library(dplyr)
library(RColorBrewer)
```

#Background

Samples involved:

  Sample ID           	sample barcode
  ALS 16S             	1             
  Zymo 16S            	2             
  ALS ITS             	3             
  159:160:161=100:10:1	4             
  163                 	5             
  164                 	6             
  ALS ITS 10X dilute  	7             
  Zymo 16S 10X dilute 	8             

#init
```{bash,eval=FALSE}
sDir=*** # source directory, find it in email
#batch 01 #chip: CL100124160
chip=CL100124159
for i in {1,2};do for j in {1..8};do mkdir -p RCA197L0$i\_$j/input;done;done
#batch 02 #chip: CL100124159
chip=CL100124159
for i in {1,2};do for j in {1..8};do let k=$i+2; mkdir -p RCA197L0$k\_$j/input;done;done
for h in {1,2};do for i in {1,2};do for j in {1..8};do let k=$i+2; ln -sf $sDir/$chip/L0$i/$chip\_L0$i\_$j\_$h.fq.gz RCA197L0$k\_$j/input/rawSeq_$h.fq.gz;done;done;done
```

#Run
```{bash,eval=FALSE}
#batch 01 #chip: CL100124160
#batch 02 #chip: CL100124159
metabbq smk -j 32 -np RCA197L0{1,2}_{1..8}/BB.stat
```

#combine same libraries
```{bash,eval=FALSE}
for j in {1..8};do mkdir -p RCA197M01\_$j/clean;done;done
for j in {1..8};
do
  echo 'cat RCA197L0{1,2}\_'$j'/clean/fastp.sort.1.fq | paste - - - - | sort -T ./tmp -k2,2 -t "/" | tr "\t" "\n" > RCA197M01\_'$j'/clean/fastp.sort.1.fq'
  echo 'cat RCA197L0{1,2}\_'$j'/clean/fastp.sort.2.fq | paste - - - - | sort -T ./tmp -k2,2 -t "/" | tr "\t" "\n" > RCA197M01\_'$j'/clean/fastp.sort.2.fq'
done > sort.sh

for j in {1..8};do mkdir -p RCA197M02\_$j/clean;done
for j in {1..8};
do
  echo 'cat RCA197L0{3,4}\_'$j'/clean/fastp.sort.1.fq | paste - - - - | sort -T ./tmp -k2,2 -t "/" | tr "\t" "\n" > RCA197M02\_'$j'/clean/fastp.sort.1.fq'
  echo 'cat RCA197L0{3,4}\_'$j'/clean/fastp.sort.2.fq | paste - - - - | sort -T ./tmp -k2,2 -t "/" | tr "\t" "\n" > RCA197M02\_'$j'/clean/fastp.sort.2.fq'
done >> sort.sh

```


#blast select scaf to nt
```{bash,eval=FALSE}
SAM="RCA197ALS16S"
#make a primer lsit first

#make db
makeblastdb -in $SAM/summary.BI.megahit.contig.fasta -input_type fasta -dbtype nucl \
-title megahit.contig -parse_seqids -out $SAM/blast/summary.BI.megahit.contig.fasta
#command refer to http://bioweb.cbm.uam.es/courses/MasterVirol2013/alignment/BLAST-NCBI-programs.htm
blastn -num_threads 8 -query primers.fa -db $SAM/blast/summary.BI.megahit.contig.fasta \
-out $SAM/blast/summary.BI.megahit.contig.b6 -outfmt 6 -word_size 7 -evalue 10
```

#Load
```{r}
SAM="RCA197ALS16S"
primers.b6 <- rbind(
  cbind(seqID="RCA197L01_1",read.table("../../Results/RCAs/RCA197FIN/RCA197L01_1/blast/summary.BI.megahit.contig.b6")),
  cbind(seqID="RCA197L01_2",read.table("../../Results/RCAs/RCA197FIN/RCA197L01_2/blast/summary.BI.megahit.contig.b6")),
  cbind(seqID="RCA197L01_3",read.table("../../Results/RCAs/RCA197FIN/RCA197L01_3/blast/summary.BI.megahit.contig.b6")),
  cbind(seqID="RCA197L01_4",read.table("../../Results/RCAs/RCA197FIN/RCA197L01_4/blast/summary.BI.megahit.contig.b6")),
  cbind(seqID="RCA197L01_5",read.table("../../Results/RCAs/RCA197FIN/RCA197L01_5/blast/summary.BI.megahit.contig.b6")),
  cbind(seqID="RCA197L01_6",read.table("../../Results/RCAs/RCA197FIN/RCA197L01_6/blast/summary.BI.megahit.contig.b6")),
  #cbind(seqID="RCA197L01_7",read.table("../../Results/RCAs/RCA197FIN/RCA197L01_7/blast/summary.BI.megahit.contig.b6")),
  cbind(seqID="RCA197L01_8",read.table("../../Results/RCAs/RCA197FIN/RCA197L01_8/blast/summary.BI.megahit.contig.b6")),
  cbind(seqID="RCA197L02_1",read.table("../../Results/RCAs/RCA197FIN/RCA197L02_1/blast/summary.BI.megahit.contig.b6")),
  cbind(seqID="RCA197L02_2",read.table("../../Results/RCAs/RCA197FIN/RCA197L02_2/blast/summary.BI.megahit.contig.b6")),
  cbind(seqID="RCA197L02_3",read.table("../../Results/RCAs/RCA197FIN/RCA197L02_3/blast/summary.BI.megahit.contig.b6")),
  cbind(seqID="RCA197L02_4",read.table("../../Results/RCAs/RCA197FIN/RCA197L02_4/blast/summary.BI.megahit.contig.b6")),
  cbind(seqID="RCA197L02_5",read.table("../../Results/RCAs/RCA197FIN/RCA197L02_5/blast/summary.BI.megahit.contig.b6")),
  cbind(seqID="RCA197L02_6",read.table("../../Results/RCAs/RCA197FIN/RCA197L02_6/blast/summary.BI.megahit.contig.b6")),
  #cbind(seqID="RCA197L02_7",read.table("../../Results/RCAs/RCA197FIN/RCA197L02_7/blast/summary.BI.megahit.contig.b6")),
  cbind(seqID="RCA197L02_8",read.table("../../Results/RCAs/RCA197FIN/RCA197L02_8/blast/summary.BI.megahit.contig.b6"))
)
colnames(primers.b6) <- c("tag","qseqid","sseqid","pident","length","mismatch","gapopen",
              "qstart","qend","sstart","send","evalue","bitscore")
```

#curate
```{r}
tmp <- matrix(unlist(strsplit(as.character(primers.b6$sseqid),"_")),ncol=5,byrow=T)
primers.b6$scafLen <- as.numeric(tmp[,5])
primers.b6$coverage<- as.numeric(tmp[,4])
primers.b6$BI <- tmp[,1]

scanFun <- function(d){
  d <- d[order(d$sstart),]
  find.p <- paste(d$qseqid,collapse = "_")
  d <- d[1,-1]
  d$primers <- find.p
  return(d)
}
primers.b6s <- ddply(primers.b6,c("tag","sseqid"),scanFun)
```

```{r}
freqTab <- table(primers.b6s$primers,primers.b6s$tag)
```

```{r}
top10 <- names(rev(sort(table(primers.b6s$primers)))[1:10])
primers.b6s$p10 <- ifelse(primers.b6s$primers%in%top10,primers.b6s$primers,"others")

primers.b6s$p10 <- reorder(primers.b6s$p10,primers.b6s$scafLen,median)
ggplot(primers.b6s,aes(x=p10,y=scafLen)) + geom_boxplot() + geom_jitter(alpha=.3) + coord_flip()
ggplot(primers.b6s,aes(x=tag,fill=primers,y=..count..)) + geom_bar(position="fill") + guides(fill=F)
```

```{r}
ggplot(primers.b6s,aes(x=scafLen)) + geom_density(aes(color=tag)) + facet_grid(primers~.)
```










#FIN.
