---
title: "JAN20 DATA ASSESSMENT"
author: "Chao"
date: "8/30/2020"
output:
  html_notebook:
    code_fold: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(reshape2)
library(ggplot2)
library(plyr)
library(dplyr)
library(RColorBrewer)
library(configr)
```

# Background
### blastn database source:
1. SILVA release 138.1 SSU:
```{bash, eval=FALSE}
mkdir -p $LFR/Source/REF/SILVA/138.1
makeblastdb -dbtype nucl -in $MDB/silva/release_138.1/SILVA_138.1_SSURef_tax_silva_trunc.fasta \
-title nucl -out $LFR/Source/REF/SILVA/138.1/SSURef_tax_silva_trunc
ln -s $MDB/silva/release_138.1/taxmap_slv_ssu_ref_138.1.txt $LFR/Source/REF/SILVA/138.1/taxmap_slv_ssu_ref_138.1.txt

awk -F "\t" 'FNR>1{b=$5;gsub(" \\(.+\\)","",b); if(b == "unidentified"){split($4,a,";");b=a[length(a)-1]};print $1"."$2"."$3"\t"b}' $MDB/silva/release_138.1/taxmap_slv_ssu_ref_138.1.txt | taxonkit name2taxid -i 2 \
| taxonkit lineage -r -i 3 | taxonkit reformat -i 4 -f "{k};{p};{c};{o};{f};{g};{s};{S}" \
| awk -F "\t" '{print $1"\t"$2"\t"$3"\t"$5"\t"$4"\t"$6}' > $LFR/Source/REF/SILVA/138.1/taxmap_slv_ssu_ref_138.1.tax
```
2. SILVA release 138.1 LSU:
```{bash, eval=FALSE}
mkdir -p $LFR/Source/REF/SILVA/138.1
makeblastdb -dbtype nucl -in $MDB/silva/release_138.1/SILVA_138.1_LSURef_tax_silva_trunc.fasta \
-title nucl -out $LFR/Source/REF/SILVA/138.1/LSURef_tax_silva_trunc
ln -s $MDB/silva/release_138.1/taxmap_slv_lsu_ref_138.1.txt $LFR/Source/REF/SILVA/138.1/taxmap_slv_lsu_ref_138.1.txt
awk -F "\t" 'FNR>1{b=$5;gsub(" \\(.+\\)","",b); if(b == "unidentified"){split($4,a,";");b=a[length(a)-1]};print $1"."$2"."$3"\t"b}' $MDB/silva/release_138.1/taxmap_slv_lsu_ref_138.1.txt| taxonkit name2taxid -i 2 \
| taxonkit lineage -r -i 3 | taxonkit reformat -i 4 -f "{k};{p};{c};{o};{f};{g};{s};{S}" \
| awk -F "\t" '{print $1"\t"$2"\t"$3"\t"$5"\t"$4"\t"$6}' > $LFR/Source/REF/SILVA/138.1/taxmap_slv_lsu_ref_138.1.tax

#check board by ITSx
cd $LFR/Source/REF/SILVA/138.1
ITSx -i $MDB/silva/release_138.1/SILVA_138.1_LSURef_tax_silva_trunc.fasta \
-o ITSx/SILVA_138.1_LSURef_tax_silva_trunc -t F --cpu 8
```

### cluster with trust reference

step 0. Add ref to database
For bacteria:
```{bash,eval=False}
#manually edit $LFR/Source/REF/zymo/D6305.rRNA.tax.txt

cd $LFR/Assay/15.AUG20/SAM/MBR/KRAKEN/refbead/data

cat $LFR/Source/REF/zymo/D6305.rRNA.fa $MDB/silva/release_138.1/SILVA_138.1_SSURef_tax_silva_trunc.fasta > 138.1_SSURef_tax_silva_trunc+.fasta &
cat $LFR/Source/REF/zymo/D6305.rRNA.fa $MDB/silva/release_138.1/SILVA_138.1_LSURef_tax_silva_trunc.fasta > 138.1_LSURef_tax_silva_trunc+.fasta &
cat $LFR/Source/REF/zymo/D6305.rRNA.fa $LFR/Source/REF/KRAKEN2/bak/rDNA/data/UNITE_ITS_ONLY.fasta > UNITE_ITS_ONLY+.fasta &
for i in {138.1_SSURef_tax_silva_trunc+.fasta,138.1_LSURef_tax_silva_trunc+.fasta,UNITE_ITS_ONLY+.fasta};do
  echo makeblastdb -dbtype nucl -in $i -title nucl -out ${i/.fasta/} \&
done
cat $LFR/Source/REF/zymo/D6305.rRNA.tax.txt $MDB/silva/release_138.1/taxmap_slv_ssu_ref_138.1.txt > taxmap_slv_ssu_ref_138.1+.tax
cat $LFR/Source/REF/zymo/D6305.rRNA.tax.txt $LFR/Source/REF/UNITE/sh_general_release_all_04.02.2020/taxmap_taxonkit_merge_ref.tax > taxmap_unite_ref+.tax
cat $LFR/Source/REF/zymo/D6305.rRNA.tax.txt $MDB/silva/release_138.1/taxmap_slv_lsu_ref_138.1.txt > taxmap_slv_lsu_ref_138.1+.tax
```

For fungi
```{bash,eval=False}
cd $LFR/Source/REF/fungi/sanger/
# get subunits
perl onetime.cutSubunits.pl
# manually edit sanger.mock7.tax.txt

cd $LFR/Assay/15.AUG20/SAM/MFR/KRAKEN/refbead/data

cat $LFR/Source/REF/fungi/sanger/sanger.mock7.rRNA.fa $MDB/silva/release_138.1/SILVA_138.1_SSURef_tax_silva_trunc.fasta > 138.1_SSURef_tax_silva_trunc+.fasta &
cat $LFR/Source/REF/fungi/sanger/sanger.mock7.rRNA.fa $MDB/silva/release_138.1/SILVA_138.1_LSURef_tax_silva_trunc.fasta > 138.1_LSURef_tax_silva_trunc+.fasta &
cat $LFR/Source/REF/fungi/sanger/sanger.mock7.rRNA.fa $LFR/Source/REF/KRAKEN2/bak/rDNA/data/UNITE_ITS_ONLY.fasta > UNITE_ITS_ONLY+.fasta & 
for i in {138.1_SSURef_tax_silva_trunc+.fasta,138.1_LSURef_tax_silva_trunc+.fasta,UNITE_ITS_ONLY+.fasta};do
  makeblastdb -dbtype nucl -in $i -title nucl -out ${i/.fasta/} &
done
cat $LFR/Source/REF/fungi/sanger/sanger.mock7.tax.txt $MDB/silva/release_138.1/taxmap_slv_ssu_ref_138.1.txt > taxmap_slv_ssu_ref_138.1+.tax
cat $LFR/Source/REF/fungi/sanger/sanger.mock7.tax.txt $LFR/Source/REF/UNITE/sh_general_release_all_04.02.2020/taxmap_taxonkit_merge_ref.tax > taxmap_unite_ref+.tax
cat $LFR/Source/REF/fungi/sanger/sanger.mock7.tax.txt $MDB/silva/release_138.1/taxmap_slv_lsu_ref_138.1.txt > taxmap_slv_lsu_ref_138.1+.tax
```


### KRAKEN2 database source:

1. SILVA release 138.1 SSU
prepare a `config.cfg` file with following context:
```
path:ssu = $MDB/silva/release_138.1/tax_slv_ssu_138.1.txt
acc:ssu = $MDB/silva/release_138.1/tax_slv_ssu_138.1.acc_taxid
map:ssu = $MDB/silva/release_138.1/taxmap_slv_ssu_ref_138.1.txt
fna:ssu = $MDB/silva/release_138.1/SILVA_138.1_SSURef_tax_silva_trunc.fasta
```
and make sure above files are accessible.
```{bash, eval=FALSE}
export DB=$LFR/Source/REF/KRAKEN2/SILVA138.1SSU
mkdir -p $DB/{data,library}
cd $DB/data
vi config.cfg
metabbq mergeTaxonPath.pl added &> added.log
mv added.fasta ../library/added.fna
metabbq buildKraken2db.sh .. 4
```

2.SILVA release 138.1 LSU
prepare a `config.cfg` file with following context:
```
path:lsu = $MDB/silva/release_138.1/tax_slv_lsu_138.1.txt
acc:lsu = $MDB/silva/release_138.1/tax_slv_lsu_138.1.acc_taxid
map:lsu = $MDB/silva/release_138.1/taxmap_slv_lsu_ref_138.1.txt
fna:lsu = $MDB/silva/release_138.1/SILVA_138.1_LSURef_tax_silva_trunc.fasta
```
and make sure above files are accessible.
```{bash, eval=FALSE}
export DB=$LFR/Source/REF/KRAKEN2/SILVA138.1LSU
mkdir -p $DB/{data,library}
cd $DB/data
vi config.cfg
metabbq mergeTaxonPath.pl added &> added.log
mv added.fasta ../library/added.fna
metabbq buildKraken2db.sh .. 4
```

3. UNITE release 8.2
prepare a `config.cfg` file with following context:
```
map:unite = $MDB/UNITE/8.2/sh_general_release_dynamic_all.tax
fna:unite = $MDB/UNITE/8.2/sh_general_release_dynamic_all.fasta
```
and make sure above files are accessible.
```{bash, eval=FALSE}
export DB=$LFR/Source/REF/KRAKEN2/UNITE8.2
mkdir -p $DB/{data,library}
cd $DB/data
vi config.cfg
metabbq mergeTaxonPath.pl added &> added.log
sed -e '/^>/!y/U/T/' added.fasta > ../library/added.fna
metabbq buildKraken2db.sh .. 4
```

merge above 3 databases:
prepare a `config.cfg` file with following context:
```
path:ssu = tax_slv_ssu_138.1.txt
path:lsu = tax_slv_lsu_138.1.txt
acc:ssu = tax_slv_ssu_138.1.acc_taxid
acc:lsu = tax_slv_lsu_138.1.acc_taxid
map:ssu = taxmap_slv_ssu_ref_138.1.txt
map:lsu = taxmap_slv_lsu_ref_138.1.txt
map:unite = sh_general_release_dynamic_all.tax
fna:ssu = SILVA_138.1_SSURef_tax_silva_trunc.fasta
fna:lsu = SILVA_138.1_LSURef_tax_silva_trunc.fasta
fna:unite = UNITE_ITS_ONLY.fasta
```
and make sure above files are accessible.
```{bash, eval=FALSE}
export DB=$LFR/Source/REF/KRAKEN2/SUS
mkdir -p $DB/{data,library}
cd $DB/data
vi config.cfg
metabbq mergeTaxonPath.pl added &> added.log
mv added.fasta ../library/added.fna
metabbq buildKraken2db.sh .. 4
```


**stat**
```{bash, eval=FALSE}
cd $LFR/Source/REF/KRAKEN2/SILVA138.1LSU
metabbq beadsAnno.pl -m otupair -i library/lsu.fna,data/lsu.txt,data/lsu.acc_taxid -o data/vsearch.allpair.txt -v
perl -e 'while(<>){@s=split /\t/;@t=split /;/,$s[0];print "$t[0]\t$t[-1]\n"}' \
< data/lsu.txt > data/domain.lst

cd $LFR/Source/REF/KRAKEN2/SILVA138SSU
metabbq beadsAnno.pl -m otupair -i data/SILVA_138_SSURef_tax_silva_trunc.fasta,data/ssu.txt,data/ssu.acc_taxid -o data/vsearch.allpair.txt -v
perl -e 'while(<>){@s=split /\t/;@t=split /;/,$s[0];print "$t[0]\t$t[-1]\n"}' \
< data/ssu.txt > data/domain.lst

cd $LFR/Source/REF/KRAKEN2/UNITE/
metabbq beadsAnno.pl -m otupair -i library/unite.fna,data/unite.txt,data/unite.acc_taxid -o data/vsearch.allpair.txt -v
perl -e 'while(<>){@s=split /\t/;@t=split /;/,$s[0];print "$t[0]\t$t[-1]\n"}' \
< data/unite.txt > data/kingdom.lst

cd $LFR/Source/REF/KRAKEN2/rDNA
perl -e 'while(<>){@s=split /\t/;@t=split /;/,$s[0];print "$t[0]\t$t[-1]\n"}' \
< data/added.txt > data/domain.lst
# use contigs
perl -e 'while(<>){chomp;@s=split;$HS{$s[0]}=$s[1]};open I,"< ANNO/CLIP.map.merge.clip.anno";
while(<I>){@s=split /\t/;next unless $s[2]>99;next unless $HS{$s[18]}; $s[0]=~/^(.*BI\d+)_k\d+_(\d+)_.*C([0-9]+|-)_/;print "$1.$2.$3\t$HS{$s[18]}\n"}' \
< $LFR/Source/REF/KRAKEN2/rDNA/seqid2taxid.map > ANNO/clip2rDNA.seqid2taxid.map
cd $LFR/Results/JAN20/SAM/SFR/
#SSU
metabbq beadsAnno.pl -m otupair -i CLIP/SSU.barrnap.fa,$LFR/Source/REF/KRAKEN2/rDNA/taxonomy/paths.txt,ANNO/clip2rDNA.seqid2taxid.map -o CLIP/vsearch.SSU.allpair.txt -v
#LSU
metabbq beadsAnno.pl -m otupair -i CLIP/LSU.barrnap.fa,$LFR/Source/REF/KRAKEN2/rDNA/taxonomy/paths.txt,ANNO/clip2rDNA.seqid2taxid.map -o CLIP/vsearch.LSU.allpair.txt -v
#CLIP
metabbq beadsAnno.pl -m otupair -i CLIP/id90def4.clust.fa,$LFR/Source/REF/KRAKEN2/rDNA/taxonomy/paths.txt,ANNO/clip2rDNA.seqid2taxid.map -o CLIP/vsearch.CLIP.allpair.txt -v



```

# load
```{r}
its.allpair.df <- read.table(
  "../../Source/REF/KRAKEN2/UNITE/data/vsearch.allpair.txt",sep="\t",
  col.names = c("rank","tax","seqs","minIdent","idents"))
tmp <- as.data.frame(
  matrix(as.numeric(unlist(sapply(strsplit(as.character(its.allpair.df$idents),","),"[",))),ncol=7,byrow = T))
colnames(tmp) <- c("q01","q05","q25","q50","q75","q95","q99")
its.allpair.df <- cbind(its.allpair.df,tmp)
its.allpair.df$rank <- factor(its.allpair.df$rank,levels=c("domain","major_clade","kingdom","superphylum","phylum","class","subclass","order","suborder","family","genus","species","subspecies"))
################################################################################
lsu.allpair.df <- read.table(
  "../../Source/REF/KRAKEN2/SILVA132LSU/data/vsearch.allpair.txt",sep="\t",quote = "",comment.char = "",
  col.names = c("rank","tax","seqs","minIdent","idents"))
tmp <- as.data.frame(
  matrix(as.numeric(unlist(sapply(strsplit(as.character(lsu.allpair.df$idents),","),"[",))),ncol=7,byrow = T))
colnames(tmp) <- c("q01","q05","q25","q50","q75","q95","q99")
lsu.allpair.df <- cbind(lsu.allpair.df,tmp)
lsu.allpair.df$rank <- factor(lsu.allpair.df$rank,levels=c("domain","major_clade","kingdom","superphylum","phylum","class","subclass","order","suborder","family","genus","species","subspecies"))
lsu.domain.lst <- read.table(
    "../../Source/REF/KRAKEN2/SILVA132LSU/data/domain.lst",sep="\t",
  comment.char = "",quote = "",col.names = c("domain","tax"))
lsu.allpair.mdf <- merge(lsu.allpair.df,lsu.domain.lst,by="tax")

################################################################################
ssu.allpair.df <- read.table(
  "../../Source/REF/KRAKEN2/SILVA138SSU/data/vsearch.allpair.txt",sep="\t",quote = "",comment.char = "",
  col.names = c("rank","tax","seqs","minIdent","idents"))
tmp <- as.data.frame(
  matrix(as.numeric(unlist(sapply(strsplit(as.character(ssu.allpair.df$idents),","),"[",))),ncol=7,byrow = T))
colnames(tmp) <- c("q01","q05","q25","q50","q75","q95","q99")
ssu.allpair.df <- cbind(ssu.allpair.df,tmp)
ssu.allpair.df$rank <- factor(ssu.allpair.df$rank,levels=c("domain","major_clade","kingdom","superphylum","phylum","class","subclass","order","suborder","family","genus","species","subspecies"))
ssu.domain.lst <- read.table(
    "../../Source/REF/KRAKEN2/SILVA138SSU/data/domain.lst",sep="\t",
  comment.char = "",quote = "",col.names = c("domain","tax"))
ssu.allpair.mdf <- merge(ssu.allpair.df,ssu.domain.lst,by="tax",all=T)
```

```{r}
ggplot(lsu.allpair.df,aes(x=minIdent,color=rank)) + geom_density() +
  scale_x_continuous(breaks=seq(0,100,2)) + ylim(c(0,0.5))
ggplot(lsu.allpair.df,aes(x=q01,color=rank)) + geom_density() +
    scale_x_continuous(breaks=seq(0,100,2)) + ylim(c(0,0.5))
ggplot(lsu.allpair.df,aes(x=q05,color=rank)) + geom_density() +
    scale_x_continuous(breaks=seq(0,100,2)) + ylim(c(0,0.5))
ggplot(lsu.allpair.df,aes(x=q25,color=rank)) + geom_density() +
    scale_x_continuous(breaks=seq(0,100,2)) + ylim(c(0,0.5))
ggplot(lsu.allpair.df,aes(x=q50,color=rank)) + geom_density() +
      scale_x_continuous(breaks=seq(0,100,2)) + ylim(c(0,0.5))
ggplot(lsu.allpair.df,aes(x=q75,color=rank)) + geom_density() +
      scale_x_continuous(breaks=seq(0,100,2)) + ylim(c(0,0.5))
ggplot(lsu.allpair.df,aes(x=q95,color=rank)) + geom_density() +
      scale_x_continuous(breaks=seq(0,100,2)) + ylim(c(0,0.5))
ggplot(lsu.allpair.df,aes(x=q99,color=rank)) + geom_density() +
      scale_x_continuous(breaks=seq(0,100,2)) + ylim(c(0,0.5))
```
```{r}
ggplot(lsu.allpair.df,aes(x=q50,color=rank)) + geom_density() +
      scale_x_continuous(breaks=seq(0,100,2)) + ylim(c(0,0.5))
```


```{r}
ggplot(ssu.allpair.df,aes(x=minIdent,color=rank)) + geom_density() +
  scale_x_continuous(breaks=seq(0,100,2))
ggplot(ssu.allpair.df,aes(x=q01,color=rank)) + geom_density() +
    scale_x_continuous(breaks=seq(0,100,2))
ggplot(ssu.allpair.df,aes(x=q05,color=rank)) + geom_density() +
    scale_x_continuous(breaks=seq(0,100,2))
ggplot(ssu.allpair.df,aes(x=q25,color=rank)) + geom_density() +
    scale_x_continuous(breaks=seq(0,100,2))
ggplot(ssu.allpair.df,aes(x=q50,color=rank)) + geom_density() +
      scale_x_continuous(breaks=seq(0,100,2))
ggplot(ssu.allpair.df,aes(x=q75,color=rank)) + geom_density() +
      scale_x_continuous(breaks=seq(0,100,2))
ggplot(ssu.allpair.df,aes(x=q95,color=rank)) + geom_density() +
      scale_x_continuous(breaks=seq(0,100,2))
ggplot(ssu.allpair.df,aes(x=q99,color=rank)) + geom_density() +
      scale_x_continuous(breaks=seq(0,100,2))
```
```{r}
ggplot(ssu.allpair.df%>%filter(rank!="subspecies"),aes(x=q50,color=rank)) + geom_density() +
      scale_x_continuous(breaks=seq(0,100,2)) + ylim(c(0,0.2))
ssu.allpair.mdf <- merge(ssu.allpair.df,ssu.domain.lst,by="tax")
ggplot(ssu.allpair.mdf%>%filter(!rank%in%c("subclass","suborder","subphylum","superphylum")),
       aes(x=rank,y=q05,fill=domain)) + geom_boxplot() +
      scale_y_continuous(breaks=seq(0,100,2)) + coord_flip()

lsu.allpair.mdf <- merge(lsu.allpair.df,lsu.domain.lst,by="tax")
ggplot(lsu.allpair.mdf%>%filter(!rank%in%c("subclass","suborder","subphylum","superphylum")),
       aes(x=rank,y=q05,fill=domain)) + geom_boxplot() +
      scale_y_continuous(breaks=seq(0,100,2)) + coord_flip()

ggplot(its.allpair.df%>%filter(!rank%in%c("subclass","suborder","subphylum","superphylum")),
       aes(x=rank,y=q05)) + geom_boxplot() +
      scale_y_continuous(breaks=seq(0,100,2)) + coord_flip()
```

```{r}
ggplot(ssu.allpair.mdf%>%filter(!rank%in%c("subclass","suborder","subphylum","superphylum")),
       aes(x=rank,y=q05,fill=domain)) + geom_boxplot() +
      scale_y_continuous(breaks=seq(0,100,2)) + coord_flip()

```


```{r}
ssu.allpair.melt <- melt(ssu.allpair.df,measure.vars = c("q01","q05","q25","q50","q75","q95","q99"),
                        variable.name = "quant",value.name = "identity")
lsu.allpair.melt <- melt(lsu.allpair.df,measure.vars = c("q01","q05","q25","q50","q75","q95","q99"),
                        variable.name = "quant",value.name = "identity")
its.allpair.melt <- melt(its.allpair.df,measure.vars = c("q01","q05","q25","q50","q75","q95","q99"),
                        variable.name = "quant",value.name = "identity")
comb.allpair.melt <- rbind(
  cbind(unit="ssu",ssu.allpair.melt),
  cbind(unit="lsu",lsu.allpair.melt),
  cbind(unit="its",its.allpair.melt)
)
comb.allpair.melt$rank <- factor(comb.allpair.melt$rank,levels=rev(levels(comb.allpair.melt$rank)))
ggplot(comb.allpair.melt%>%filter(rank%in%c("domain","kingdom","phylum","class","order","family","genus","species","subspecies")),
       aes(x=identity,color=quant)) + geom_density() +
  scale_x_continuous(breaks=seq(0,100,2)) +
  facet_wrap(~rank+unit,ncol=3,scale="free_y")

ggplot(comb.allpair.melt%>%filter(rank%in%c("domain","kingdom","phylum","class","order","family","genus","species")&quant=="q05"),
       aes(x=identity,color=rank)) + geom_density() +
  scale_x_continuous(breaks=seq(0,100,2)) +
  facet_wrap(~unit,ncol=3,scale="free") + ylim(c(0,0.2))

ggplot(comb.allpair.melt%>%filter(rank%in%c("domain","kingdom","phylum","class","order","family","genus","species")&quant=="q50"),
       aes(x=identity,color=rank)) + geom_density() +
  scale_x_continuous(breaks=seq(0,100,2)) +
  facet_wrap(~unit,ncol=3,scale="free")

#comb.allpair.melt.mean <- ddply(comb.allpair.melt[,c(1:3,7,8)],c("rank","tax","quant"),summarise,mident=mean(identity))

#ggplot(comb.allpair.melt.mean,aes(x=mident,color=quant)) + geom_density() +
#        scale_x_continuous(breaks=seq(0,100,2),limits = c(70,100)) +
#  facet_wrap(~rank,ncol=2,scale="free")
```
```{r}
comb.allpair.melt$unit <- factor(comb.allpair.melt$unit,levels=c("ssu","lsu","its"))
p.PUB.seqs.identityMedian.line <- 
  ggplot(comb.allpair.melt%>%filter(rank%in%c("domain","kingdom","phylum","class","order","family","genus","species")&quant=="q50"),
       aes(x=identity,color=rank)) + geom_vline(xintercept = c(95,97,99),linetype=2) +
  geom_density() + scale_x_continuous(breaks=seq(0,100,2)) + #ylim(c(0,0.15)) +
  facet_wrap(~unit,ncol=1,scale="free_y")  +
  theme_bw()

p.PUB.seqs.identityMedian.line
```

```{r}
its.allpair.mdf <- its.allpair.df
its.allpair.mdf$domain <- "domain__Eukaryota"

ssu.allpair.meltm <- melt(ssu.allpair.mdf,measure.vars = c("q01","q05","q25","q50","q75","q95","q99"),
                        variable.name = "quant",value.name = "identity")
lsu.allpair.meltm <- melt(lsu.allpair.mdf,measure.vars = c("q01","q05","q25","q50","q75","q95","q99"),
                        variable.name = "quant",value.name = "identity")
its.allpair.meltm <- melt(its.allpair.mdf,measure.vars = c("q01","q05","q25","q50","q75","q95","q99"),
                        variable.name = "quant",value.name = "identity")
comb.allpair.meltm <- rbind(
  cbind(unit="ssu",ssu.allpair.meltm),
  cbind(unit="lsu",lsu.allpair.meltm),
  cbind(unit="its",its.allpair.meltm)
)

ggplot(comb.allpair.meltm%>%filter(rank%in%c("domain","kingdom","phylum","class","order","family","genus","species")&quant=="q50"&!is.na(domain)),
       aes(x=rank,y=identity,fill=unit)) + geom_boxplot() +
  scale_y_continuous(breaks=seq(0,100,2)) + coord_flip() + facet_grid(domain~.)

```

# ITS

```{r}
ggplot(its.allpair.df,aes(x=minIdent,color=rank)) + geom_density() +
  scale_x_continuous(breaks=seq(0,100,2))
ggplot(its.allpair.df,aes(x=q01,color=rank)) + geom_density() +
    scale_x_continuous(breaks=seq(0,100,2))
ggplot(its.allpair.df,aes(x=q05,color=rank)) + geom_density() +
    scale_x_continuous(breaks=seq(0,100,2))
ggplot(its.allpair.df,aes(x=q25,color=rank)) + geom_density() +
    scale_x_continuous(breaks=seq(0,100,2))
ggplot(its.allpair.df,aes(x=q50,color=rank)) + geom_density() +
      scale_x_continuous(breaks=seq(0,100,2))
ggplot(its.allpair.df,aes(x=q75,color=rank)) + geom_density() +
      scale_x_continuous(breaks=seq(0,100,2))
ggplot(its.allpair.df,aes(x=q95,color=rank)) + geom_density() +
      scale_x_continuous(breaks=seq(0,100,2))
ggplot(its.allpair.df,aes(x=q99,color=rank)) + geom_density() +
    scale_x_continuous(breaks=seq(0,100,2))

```

```{r}
its.allpair.df$unkSp <- ifelse(grepl(" sp",its.allpair.df$tax),T,F)
ggplot(its.allpair.df%>%filter(unkSp==F),aes(x=minIdent,color=rank)) + geom_density() +
  scale_x_continuous(breaks=seq(0,100,2))
ggplot(its.allpair.df%>%filter(unkSp==F),aes(x=q01,color=rank)) + geom_density() +
    scale_x_continuous(breaks=seq(0,100,2))
ggplot(its.allpair.df%>%filter(unkSp==F),aes(x=q05,color=rank)) + geom_density() +
    scale_x_continuous(breaks=seq(0,100,2))
ggplot(its.allpair.df%>%filter(unkSp==F),aes(x=q25,color=rank)) + geom_density() +
    scale_x_continuous(breaks=seq(0,100,2))
ggplot(its.allpair.df%>%filter(unkSp==F),aes(x=q50,color=rank)) + geom_density() +
      scale_x_continuous(breaks=seq(0,100,2))
ggplot(its.allpair.df%>%filter(unkSp==F),aes(x=q75,color=rank)) + geom_density() +
      scale_x_continuous(breaks=seq(0,100,2))
ggplot(its.allpair.df%>%filter(unkSp==F),aes(x=q95,color=rank)) + geom_density() +
      scale_x_continuous(breaks=seq(0,100,2))
ggplot(its.allpair.df%>%filter(unkSp==F),aes(x=q99,color=rank)) + geom_density() +
    scale_x_continuous(breaks=seq(0,100,2))

```

```{r}
ggplot(its.allpair.df%>%filter(seqs>9),aes(x=q50,color=rank)) + geom_density() +
      scale_x_continuous(breaks=seq(0,100,2))
```


# Section for clips alignemnt results
# load
```{r}
all.clip.pair.df <- read.table(
  "../../Results/JAN20/SAM/SFR/CLIP/vsearch.CLIP.allpair.txt",sep="\t",
  col.names = c("rank","tax","seqs","minIdent","idents"))
tmp <- as.data.frame(
  matrix(as.numeric(unlist(sapply(strsplit(as.character(all.clip.pair.df$idents),","),"[",))),ncol=7,byrow = T))
colnames(tmp) <- c("q01","q05","q25","q50","q75","q95","q99")
all.clip.pair.df <- cbind(all.clip.pair.df,tmp)
all.clip.pair.df$rank <- factor(all.clip.pair.df$rank,levels=c("domain","major_clade","kingdom","superphylum","phylum","class","subclass","order","suborder","family","genus","species","subspecies"))
################################################################################
lsu.clip.pair.df <- read.table(
  "../../Results/JAN20/SAM/SFR/CLIP/vsearch.LSU.allpair.txt",sep="\t",
  col.names = c("rank","tax","seqs","minIdent","idents"))
tmp <- as.data.frame(
  matrix(as.numeric(unlist(sapply(strsplit(as.character(lsu.clip.pair.df$idents),","),"[",))),ncol=7,byrow = T))
colnames(tmp) <- c("q01","q05","q25","q50","q75","q95","q99")
lsu.clip.pair.df <- cbind(lsu.clip.pair.df,tmp)
lsu.clip.pair.df$rank <- factor(lsu.clip.pair.df$rank,levels=c("domain","major_clade","kingdom","superphylum","phylum","class","subclass","order","suborder","family","genus","species","subspecies"))
################################################################################
ssu.clip.pair.df <- read.table(
  "../../Results/JAN20/SAM/SFR/CLIP/vsearch.SSU.allpair.txt",sep="\t",quote = "",
  col.names = c("rank","tax","seqs","minIdent","idents"))
tmp <- as.data.frame(
  matrix(as.numeric(unlist(sapply(strsplit(as.character(ssu.clip.pair.df$idents),","),"[",))),ncol=7,byrow = T))
colnames(tmp) <- c("q01","q05","q25","q50","q75","q95","q99")
ssu.clip.pair.df <- cbind(ssu.clip.pair.df,tmp)
ssu.clip.pair.df$rank <- factor(ssu.clip.pair.df$rank,levels=c("domain","major_clade","kingdom","superphylum","phylum","class","subclass","order","suborder","family","genus","species","subspecies"))
```


```{r}
ssu.clip.pair.melt <- melt(ssu.clip.pair.df,measure.vars = c("q01","q05","q25","q50","q75","q95","q99"),
                        variable.name = "quant",value.name = "identity")
lsu.clip.pair.melt <- melt(lsu.clip.pair.df,measure.vars = c("q01","q05","q25","q50","q75","q95","q99"),
                        variable.name = "quant",value.name = "identity")
all.clip.pair.melt <- melt(all.clip.pair.df,measure.vars = c("q01","q05","q25","q50","q75","q95","q99"),
                        variable.name = "quant",value.name = "identity")
comb.clip.pair.melt <- rbind(
  cbind(unit="ssu",ssu.clip.pair.melt),
  cbind(unit="lsu",lsu.clip.pair.melt),
  cbind(unit="clip",all.clip.pair.melt)
)
ggplot(comb.clip.pair.melt%>%filter(rank%in%c("domain","kingdom","phylum","class","order","family","genus","species")&unit!="clip"),
       aes(x=identity,color=quant)) + geom_density() +
  scale_x_continuous(breaks=seq(0,100,2)) +
  facet_grid(rank~unit,scale="free")

ggplot(comb.clip.pair.melt%>%filter(rank%in%c("domain","kingdom","phylum","class","order","family","genus","species")&quant=="q05"),
       aes(x=identity,color=rank)) + geom_density() +
  scale_x_continuous(breaks=seq(0,100,2)) +
  facet_wrap(~unit,ncol=3,scale="free") + ylim(c(0,0.2))

ggplot(comb.clip.pair.melt%>%filter(rank%in%c("domain","kingdom","phylum","class","order","family","genus","species")&quant=="q50"&unit!="clip"),
       aes(x=identity,color=rank)) + geom_density() +
  scale_x_continuous(breaks=seq(0,100,2)) +
  facet_wrap(~unit,ncol=2,scale="free")

ggplot(comb.clip.pair.melt%>%filter(rank%in%c("domain","kingdom","phylum","class","order","family","genus","species")&quant=="q50"&unit!="clip"),
       aes(x=identity,color=unit)) + geom_density() +
  scale_x_continuous(breaks=seq(0,100,10)) +
  facet_wrap(~rank,ncol=4,scale="free")
```


compare of SSU
```{r}
compare.ssu.df <- rbind(
  cbind(source="SILVA",ssu.allpair.melt),
  cbind(source="CLIP",ssu.clip.pair.melt)
)
compare.ssu.ddf <- dcast(compare.ssu.df%>%filter(quant=="q50"),rank+tax~source,value.var = "identity",fun.aggregate = mean)
compare.ssu.ddf$delta <- compare.ssu.ddf$CLIP/compare.ssu.ddf$SILVA
ggplot(compare.ssu.ddf%>%filter(rank!="subspecies"),aes(x=delta,color=rank)) +geom_density()
```

# EXPORT
```{r,eval=FALSE}
ggsave(p.PUB.seqs.identityMedian.line,width=8,height=8,
       file="../../Results/AUG20/FIG/PUB.seqs.identityMedian.line.pdf")
```

#FIN.
