---
title: "AUG20 DATA ASSESSMENT"
author: "Chao"
date: "1/21/2020"
output: 
  html_notebook:
    code_fold: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(reshape2)
library(ggplot2)
library(plyr)
library(dplyr)
library(RColorBrewer)
library(configr)
```

# Background

Samples involved:

| ID | lane | barcode | Kindom | RCA | Location    |
| -- | ---- | ------- | -------| --- | ----------- |
| U1 | FF1  | 1,17    | Bac    | RCA | Urine       |
| S1 | FF1  | 3,4     | Bac    | RCA | Soil spot A |
| O1 | FF1  | 5,22    | Bac    | RCA | Soil spot B |
| M1 | FF1  | 7       | Bac    | RCA | Mock        |
| Z1 | FF1  | 8       | Bac    | RCA | ZYMO 8      |

Zymo mock strains:

| Species                  | Genomic DNA | 16S Only1 | 16S & 18S1 | Genome Copy2 | Cell Number3 |
| ------------------------ | ----------- | --------- | ---------- | ------------ | ------------ |
| Pseudomonas aeruginosa   | 12          | 4.2       | 3.6        | 6.1          | 6.1          | +
| Escherichia coli         | 12          | 10.1      | 8.9        | 8.5          | 8.5          | +
| Salmonella enterica      | 12          | 10.4      | 9.1        | 8.7          | 8.8          | +
| Lactobacillus fermentum  | 12          | 18.4      | 16.1       | 21.6         | 21.9         |
| Enterococcus faecalis    | 12          | 9.9       | 8.7        | 14.6         | 14.6         | +
| Staphylococcus aureus    | 12          | 15.5      | 13.6       | 15.2         | 15.3         | +
| Listeria monocytogenes   | 12          | 14.1      | 12.4       | 13.9         | 13.9         | +
| Bacillus subtilis        | 12          | 17.4      | 15.3       | 10.3         | 10.3         | +
| Saccharomyces cerevisiae | 2           | NA        | 9.3        | 0.57         | 0.29         |
| Cryptococcus neoformans  | 2           | NA        | 3.3        | 0.37         | 0.18         |

**stat on HPC**
```{bash, eval=False}


for i in SAM/??/mash/BI.msh.tsv;do 
  awk '{b=sprintf("%0.0f",$3/100)*100;print $2"\t"b}'  $i|sort|uniq -c|sort -nk2 -nk3 > ${i/mash/stat} &
done


#sum
for i in `ls SAM`;do 
 awk -v n=$i '{print n"\t"$0}' SAM/$i/stat/BI.msh.tsv;
done > STAT/BI.msh.tsv

```


**For the first time, grab data from HPC**
```{bash,eval=FALSE}
#get BB.stat
res=../../Results/AUG20/ORI
rsync -atvP --files-from=$res/tmp.file.lst ngb8:$remote/Results/AUG20/ORI/ $res/
```

```{bash,eval=FALSE}
res=../../Results/AUG20/ORI
ls $res/|grep FF
```

```{bash,eval=FALSE}
res=../../Results/AUG20/ORI
for i in `ls $res|grep FF`;do 
  awk -v i=$i '{print i"\t"$0}' $res/$i/clean/BB.stat;
done > $res/SUM.BB.stat


```

# load 

```{r}
metadata <- read.csv("../../Results/JAN20/metadata.csv")
```

```{r}
metadata.df <- (read.table("../../Results/JAN20/dataInfo.tsv",
                          col.names=c("SID","barcode","lane","tag","path")))[,1:4]

basic.ORI.tsv <- read.table("../../Results/AUG20/STAT/ORI.basic.log",header = T)
basic.SAM.S.tsv <- read.table("../../Results/AUG20/STAT/SAM.basic.single.log",header = T)
basic.SAM.R.tsv <- read.table("../../Results/AUG20/STAT/SAM.basic.RCA.log",header = T)
basic.SAM.N.tsv <- read.table("../../Results/AUG20/STAT/SAM.basic.NOR.log",header = T)

ORI.BB.stat <- read.table("../../Results/AUG20/STAT/ORI.BB.stat",
                         col.names = c("tag","count","rpb","cumsum"))
BB.stat <- ORI.BB.stat
SAM.BB.stat <- read.table("../../Results/AUG20/STAT/SAM.BB.stat",
                         col.names = c("tag","count","rpb","cumsum"))
BI.msh.tsv <- read.table("../../Results/AUG20/STAT/BI.msh.tsv",
                         col.names = c("SID","count","rpb","kmers"))

```


# Curate
```{r}
metadata.df$SAM <- substr(metadata.df$SID,1,1)
metadata.df$lane <- substr(metadata.df$SID,2,2)
metadata.df$RCA <- ifelse(metadata.df$lane%in%seq(1,6),T,F)
metadata.df$kingdom <- ifelse(metadata.df$lane%in%c(1,2,3,7,8,9),"B","F")

metadata$SAM <- substr(metadata$SID,1,1)
metadata$lane <- substr(metadata$SID,2,2)
metadata$RCA <- ifelse(metadata$lane%in%seq(1,6),T,F)
metadata$kingdom <- ifelse(metadata$lane%in%c(1,2,3,7,8,9),"B","F")

SAM.BB.reads <- ddply(SAM.BB.stat,"tag",summarise,BB.reads=sum(count*rpb))
```

```{r}
basic.ori.df2 <- merge(metadata.df,basic.ORI.tsv,by.x="tag",by.y=".ID")
rn <- colnames(basic.SAM.R.tsv)
basic.SAM.N.tsv$ASM.beads <- NA
basic.sam.df2 <- merge(unique(metadata.df[,c(1,3,5:7)]),
  rbind(basic.SAM.S.tsv[,rn],basic.SAM.R.tsv,basic.SAM.N.tsv),
  by.x="SID",by.y=".ID")
basic.sam.df2 <- merge(basic.sam.df2,SAM.BB.reads,by.x="SID",by.y="tag")

basic.sam.df2%>%filter(RCA==T&SAM%in%c("S","O"))
ddply(basic.sam.df2%>%filter(RCA==T&SAM%in%c("S","O")),c("kingdom"),summarise,count=sum(BB.reads))
```
The MOCK bacteria data info:
```{r}
ZBR.info <- basic.sam.df2%>%filter(RCA==T&SAM%in%c("Z"))
ZBR.info
colSums(ZBR.info[,6:10])
```
The MOCK fungal data info:
```{r}
MFR.info <- basic.sam.df2%>%filter(RCA==T&SID%in%c("M4","M5","M6"))
MFR.info
colSums(MFR.info[,6:10])
```

The Soil bacterial data info:
```{r}
SBR.info <- basic.sam.df2%>%filter(SID%in%c("S1","S2","S3","O1","O2","O3"))
SBR.info
colSums(SBR.info[,6:10])
```
The Soil fungal data info:
```{r}
SFR.info <- basic.sam.df2%>%filter(SID%in%c("S4","S5","S6","O4","O5","O6"))
SFR.info
ddply(SFR.info,c("SAM"),summarise,ASM.beads=sum(ASM.beads),BB.beads=sum(BB.beads),QC.reads=sum(QC.reads),BB.reads=sum(BB.reads))
colSums(SFR.info[,6:10])
```
```{r p1}
p1.1 <- ggplot(basic.ori.df2,aes(x=SID)) +
  geom_bar(aes(y=fastp.before_filtering.total_reads),position="stack",stat="identity",fill=NA,color="black") +
  geom_bar(aes(y=fastp.after_filtering.total_reads),position="stack",stat="identity",fill="grey50",alpha=.5) +
  ggtitle("Beofre/after filter reads number")
p1.1

ggplot(basic.ori.df2,aes(x=SID)) +
  geom_bar(aes(y=fastp.before_filtering.total_reads),position="stack",stat="identity",fill=NA,color="black") +
  geom_bar(aes(y=fastp.after_filtering.total_reads),position="stack",stat="identity",fill="grey50",alpha=.5) +
  ggtitle("Beofre/after filter reads number") + facet_grid(.~kingdom,scale="free")

rm.name <- c("fastp.before_filtering.read1_mean_length","fastp.before_filtering.read2_mean_length")
basic.ori.df3 <- melt(basic.ori.df2[,-which(colnames(basic.ori.df2)%in%rm.name)],
                      id.vars = colnames(basic.ori.df2)[1:7],
                      variable.name = "var",value.name = "val")
ggplot(basic.ori.df3,aes(x=kingdom,y=val)) + geom_boxplot() + coord_flip() +
  facet_wrap(~var,ncol=5,scale="free")
```

```{r p2}
selectSID <- c("Z1","Z2","Z3","M4","M5","M6",paste0(rep(c("S","O"),each=6),rep(seq(1,6),2)))
bacMockSID <- c("Z1","Z2","Z3")
p2.1 <- ggplot(basic.sam.df2%>%filter(SID%in%selectSID),aes(x=SID)) +
  geom_bar(aes(y=QC.reads/1e6),stat="identity",fill=NA,color="black") +
  geom_bar(aes(y=BB.reads/1e6),stat="identity",fill="grey50",color=NA) +
  geom_bar(aes(y=BB.singleton/1e6),stat="identity",fill="brown1",color=NA) +
  ylab("# reads/ Mb")
p2.1

bsd2.stat <- ddply(basic.sam.df2%>%filter(SID%in%selectSID),
   "RCA",summarise,QC.reads=mean(QC.reads),BB.reads=mean(BB.reads),BB.singleton=mean(BB.singleton),
   read.pct=mean(BB.reads)/mean(QC.reads),singleton.pct=mean(BB.singleton)/mean(QC.reads))
bsd2.stat
p2.2 <- ggplot(bsd2.stat,aes(x="mean")) +
  geom_bar(aes(y=QC.reads/1e6),stat="identity",fill=NA,color="black") +
  geom_bar(aes(y=BB.reads/1e6),stat="identity",fill="grey50",color=NA) +
  geom_bar(aes(y=BB.singleton/1e6),stat="identity",fill="brown1",color=NA,width=.85) +
  ylab("# reads/ Mb")
p2.2

basic.sam.df3 <- melt(basic.sam.df2,
   id.vars = colnames(basic.sam.df2)[1:5], variable.name = "var",value.name = "val")

ggplot(basic.sam.df2,aes(x=SID)) +
  geom_bar(aes(y=QC.reads),stat="identity",fill=NA,color="black") +
  geom_bar(aes(y=BB.reads),stat="identity",fill="grey50",color=NA) +
  geom_bar(aes(y=BB.singleton),stat="identity",fill="brown1",color=NA)


BB.stat <- melt(ddply(basic.sam.df2,"SID",summarise,BB.detect=BB.reads/QC.reads,BB.sing=BB.singleton/BB.reads),
                id.vars = "SID",variable.name = "var",value.name = "pecentage")
ddply(BB.stat,"var",summarise,pct=mean(pecentage))
ddply(basic.sam.df2,c("kingdom","SID"),summarise,ASM.pct=ASM.beads/BB.beads)
```

```{r}
zymo.RCA.df <- ddply(basic.sam.df2%>%filter(SID%in%c("Z1","Z2","Z3")),"kingdom",summarise,
                     QC.reads=mean(QC.reads),BB.reads=mean(BB.reads),BB.singleton=mean(BB.singleton))
zymo.RCA.df.pct <- zymo.RCA.df[,2:4]/zymo.RCA.df$QC.reads
ggplot(zymo.RCA.df,aes(x="Zymo")) +
  geom_bar(aes(y=QC.reads/1e6),stat="identity",fill=NA,color="black") +
  geom_bar(aes(y=BB.reads/1e6),stat="identity",fill="lightseagreen",color=NA) +
  geom_bar(aes(y=BB.singleton/1e6),stat="identity",fill="grey50",color=NA) +
  ylab("# reads/ Mb")
```

```{r}
ggplot(basic.sam.df2,aes(x=SID)) +
  geom_bar(aes(y=ASM.beads),stat="identity",fill=NA,color="black")
  #geom_bar(aes(y=ASM.clips),stat="identity",fill="grey50",color=NA)

ggplot(basic.sam.df2%>%filter(lane%in%seq(1,6)),aes(x=SID)) +
  geom_bar(aes(y=BB.beads),stat="identity",fill=NA,color="black") +
  geom_bar(aes(y=ASM.beads),stat="identity",fill="grey50",color=NA)
  #geom_bar(aes(y=ASM.clips),stat="identity",fill="brown1",color=NA)

ggplot(ddply(basic.sam.df2%>%filter(lane%in%seq(1,6)),c("kingdom","SID"),summarise,pct=ASM.beads/BB.beads),
       aes(x=kingdom,y=pct)) +geom_boxplot()
```

#BB.stat
```{r bb}
# Get unbarcoded reads number
basic.sam.df2$UB.reads <- basic.sam.df2$QC.reads - basic.sam.df2$BB.reads
sam.BB0.dat <- merge(metadata,basic.sam.df2[,c("SID","UB.reads","UB.reads")],by="SID")
colnames(sam.BB0.dat)[8:9] <- c("count","rpb")
sam.BB0.dat$rpb <- 0
sam.BB0.dat$cumsum <- NA
#
sam.BB.mdat <- rbind(sam.BB0.dat,merge(metadata,SAM.BB.stat,by.x="SID",by.y="tag"))
sam.BB.mdat$reads <- ifelse(sam.BB.mdat$rpb<2,sam.BB.mdat$count,sam.BB.mdat$count*sam.BB.mdat$rpb)
#
p5.1 <- ggplot(sam.BB.mdat%>%filter(SID%in%selectSID),aes(x=rpb,y=count,color=SAM,linetype=kingdom,group=SID)) + geom_line(size=1,alpha=.5) + 
  scale_y_log10() + scale_x_log10() + annotation_logticks() +
  xlab("reads/bead") + ylab("# Beads")
p5.1

p5.2 <- ggplot(sam.BB.mdat%>%filter(SID%in%selectSID&rpb!=0),aes(x=rpb,y=reads,color=SAM,linetype=kingdom,group=SID)) + geom_line(size=1,alpha=.5) + 
  scale_x_log10() +annotation_logticks(sides = "b") +
  xlab("reads/bead") + ylab("# reads") + theme_bw()
p5.2


```

accumulated counts
```{r}
revOrder <- function(d){
  d <- d[order(d$rpb,decreasing=T),]
  d$cumRd <- cumsum(d$reads)
  return(d)
}

sam.BB.mdat2 <- ddply(sam.BB.mdat,c("SID"),revOrder)
p5.3 <- ggplot(sam.BB.mdat2%>%filter(SID%in%bacMockSID&rpb>0),aes(x=rpb,y=cumsum,color=SID,group=SID)) +
  geom_line(size=1,alpha=.5) + 
  geom_point(data=sam.BB.mdat%>%filter(SID%in%bacMockSID&rpb>0&rpb<11),alpha=.5) +
  scale_y_log10() + scale_x_log10() + annotation_logticks() +
  xlab("reads per barcode") + ylab("count of barcodes")
p5.3
p5.4 <- ggplot(sam.BB.mdat2%>%filter(SID%in%bacMockSID&rpb>0),aes(x=rpb,y=cumRd,color=SID,group=SID)) +
  geom_line(size=1,alpha=.5) + 
  geom_point(data=sam.BB.mdat2%>%filter(SID%in%bacMockSID&rpb>0&rpb<11),alpha=.5) +
  scale_y_log10() + scale_x_log10() + annotation_logticks() +
  xlab("reads per barcode") + ylab("count of barcodes")
p5.4
```


**kmers**
```{r p6}
BI.msh.df <- merge(metadata,BI.msh.tsv,by="SID")
quanFun <- function(d,x){
  if(nrow(d)>1){
    var <- d[,which(colnames(d)==x)]
    d <- d[order(var),]
    n=sum(d$count)
    d$cumsum <- cumsum(d$count)
    i05 <- which(d$cumsum>n/40)[1] -1 
    i25 <- which(d$cumsum>n/4)[1] -1 
    i50 <- which(d$cumsum>n/2)[1] -1
    i75 <- which(d$cumsum>n*3/4)[1] -1 
    i95 <- which(d$cumsum>n*39/40)[1] -1 
    res <- d[1,]
    if(i05==0){i05=1}
    if(i25==0){i25=1}
    if(i50==0){i50=1}
    if(i75==0){i75=1}
    if(i95==0){i95=1}
    res$kmer.min <- var[i05]
    res$kmer.q25 <- var[i25]
    res$kmer.q50 <- var[i50]
    res$kmer.q75 <- var[i75]
    res$kmer.max <- var[i95]
    return(res)
  }
}
if(file.exists("./cache.BI.msh.RData")){
  load("./cache.BI.msh.RData")
}else{
  BI.msh.stat <- ddply(BI.msh.df%>%filter(SID%in%selectSID),colnames(BI.msh.df)[-6],summarise,count=sum(count))
  save(BI.msh.stat,file = "./cache.BI.msh.RData")
}
BI.msh.stat$rpbs <- ifelse(BI.msh.stat$rpb<100,round(BI.msh.stat$rpb/5,0)*5,ifelse(
BI.msh.stat$rpb<1000,round(BI.msh.stat$rpb/50,0)*50,">999"))
BI.msh.stat.sam <- ddply(BI.msh.stat,c("RCA","kindom","rpbs","kmers"),summarise,count=sum(count))
BI.msh.stat.sam.qt <- ddply(BI.msh.stat.sam,c("RCA","kindom","rpbs"),quanFun,"kmers")
BI.msh.stat.sam.qt$rpbs <- factor(BI.msh.stat.sam.qt$rpbs,levels=c(seq(1,999),">999"))
p6.1 <- ggplot(BI.msh.stat.sam.qt%>%filter(rpbs!=">999"),aes(x=rpbs,y=kmers,fill=kindom)) + 
  geom_boxplot(aes(ymin = kmer.min, lower = kmer.q25, middle = kmer.q50, upper = kmer.q75, ymax = kmer.max),
   stat = "identity") + theme(axis.text.x = element_text(angle = -90))
p6.1

BI.msh.stat$cpk <- round(BI.msh.stat$rpb*170/BI.msh.stat$kmers,0)
BI.msh.stat.sam2 <- ddply(BI.msh.stat,c("RCA","kindom","rpbs","cpk"),summarise,count=sum(count))
BI.msh.stat.sam2.qt <- ddply(BI.msh.stat.sam2,c("RCA","kindom","rpbs"),quanFun,"cpk")
BI.msh.stat.sam2.qt$rpbs <- factor(BI.msh.stat.sam.qt$rpbs,levels=c(seq(1,999),">999"))
p6.2 <- ggplot(BI.msh.stat.sam2.qt%>%filter(rpbs!=">999"),aes(x=rpbs,y=cpk,fill=kindom)) + 
  geom_boxplot(aes(ymin = kmer.min, lower = kmer.q25, middle = kmer.q50, upper = kmer.q75, ymax = kmer.max),
   stat = "identity") + theme(axis.text.x = element_text(angle = -90))
p6.2
```












#FIN.