---
title: "JAN20 DATA ASSESSMENT: annotation part"
author: "Chao"
date: "3/22/2020"
output: 
  html_notebook:
    code_fold: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(reshape2)
library(ggplot2)
library(plyr)
library(dplyr)
library(RColorBrewer)
library(configr)
library(tools)
library(stringi)

#functions

#https://www.biostars.org/p/9335/
matcher <- function(pattern, x) {

  ind = gregexpr(pattern, x)[[1]]
  start = as.numeric(ind)
  if(length(start)==1 && start <1){
    return(NA)
  }else{
    end = start + attr(ind, "match.length") - 2
    res <- as.numeric(apply(cbind(start,end), 1, function(y) substr(x, start=y[1], stop=y[2])))
    res[which(is.na(res))] <- 1
    return(res)
  }
}

doone <- function(c, cigar) {
  pat <- paste("\\d*", c , sep="")
  sum(as.numeric(matcher(pat, cigar)), na.rm=T)
}


## takes a cigar string and parses it, not very fast but...
cigarsums <- function(cigar, chars=c("M","N","D","I","S","H", "P", "X", "=")) {
   sapply (chars, doone, cigar)
}


n50 <- function(x){
  x1 <- rev(sort(x))
  cumsum.x1 <- cumsum(x1)
  find.n50 <- which(cumsum.x1 >= sum(x1)/2)
  return(x1[find.n50[1]])
}

md5 <- function(f,m){
  if(tools::md5sum(f)==m){
    warning("MD5 check passed.\n")
    return(f)
  }else{
    stop("MD5 check failed. Please check the version and content!\n")
  }
}

getMapPropFun <- function(d){
  uOrder <- c("ALL","SSU","ITS1","5.8S","ITS2","LSU")
  ulist <- data.frame(key=c("ALL","SSU","16S","ITS","ITS1","5.8S","ITS2","LSU","23S"),
                      ind=c(1,2,2,3,3,4,5,6,6))
  getM <- stri_match_all_regex(d,"(ALL|23S|LSU|ITS|ITS1|ITS2|5.8S|SSU|16S)\\((\\d+)%\\)")
  getN <- length(getM)
  res <- matrix(0,ncol = 6,nrow = getN)
  for(i in 1:length(getM)){
    geti <- getM[[i]]
    for(j in 1:nrow(geti)){
      res[i,ulist$ind[which(ulist$key==geti[j,2])]] <- geti[j,3]
    }
  }
  res <- matrix(as.numeric(res),ncol=6)
  colnames(res) <- uOrder
  return(res)
}
```

# Background

Samples involved:

| ID | lane | barcode | Kindom | RCA | Location    |
| -- | ---- | ------- | -------| --- | ----------- |
| U1 | FF1  | 1,17    | Bac    | RCA | Urine       |
| S1 | FF1  | 3,4     | Bac    | RCA | Soil spot A |
| O1 | FF1  | 5,22    | Bac    | RCA | Soil spot B |
| M1 | FF1  | 7       | Bac    | RCA | Mock        |
| Z1 | FF1  | 8       | Bac    | RCA | ZYMO 8      |

Zymo mock strains:

| Species                  | Genomic DNA | 16S Only1 | 16S & 18S1 | Genome Copy2 | Cell Number3 | Gram | 16/18 Copy Num |
| ------------------------ | ----------- | --------- | ---------- | ------------ | ------------ | ---- | -------------- |
| Pseudomonas aeruginosa   | 12          | 4.2       | 3.6        | 6.1          | 6.1          | -    | 4              |
| Escherichia coli         | 12          | 10.1      | 8.9        | 8.5          | 8.5          | -    | 7              |
| Salmonella enterica      | 12          | 10.4      | 9.1        | 8.7          | 8.8          | -    | 7              |
| Lactobacillus fermentum  | 12          | 18.4      | 16.1       | 21.6         | 21.9         | +    | 5              |
| Enterococcus faecalis    | 12          | 9.9       | 8.7        | 14.6         | 14.6         | +    | 4              |
| Staphylococcus aureus    | 12          | 15.5      | 13.6       | 15.2         | 15.3         | +    | 6              |
| Listeria monocytogenes   | 12          | 14.1      | 12.4       | 13.9         | 13.9         | +    | 6              |
| Bacillus subtilis        | 12          | 17.4      | 15.3       | 10.3         | 10.3         | +    | 10             |
| Saccharomyces cerevisiae | 2           | NA        | 9.3        | 0.57         | 0.29         | Y    | 109            |
| Cryptococcus neoformans  | 2           | NA        | 3.3        | 0.37         | 0.18         | Y    | 60             |

# check whether adaptors can be found from each BI and whether it is hybridized
```{bash, eval=FALSE}
vsearch --allpair_global SAM/Z1/summary.BI.megahit.clip.all.fasta 


bwa mem -t 8 -R '@RG\tID:Z1' $LFR/Source/REF/zymo/D6305.rRNA.fa SAM/Z1/summary.BI.megahit.clip.all.fasta \
| samtools view -b -o SAM/Z1/CLIP/clips2closeRef.bwa.bam

sed 's/ /_/g' SAM/Z1/CLIP/LOTU.fa| bwa mem -t 8 -k 121 -R '@RG\tID:Z1' $LFR/Source/REF/zymo/D6305.rRNA.fa - \
| samtools view -b -o SAM/Z1/CLIP/LOTU2closeRef.bwa.bam
sed 's/ /_/g' SAM/Z1/CLIP/LOTU.fa| blastn -num_threads 16 -perc_identity 95 -word_size 121 \
 -db $LFR/Source/REF/zymo/D6305.rRNA.fa -query - -outfmt '6 std qlen slen' -out SAM/Z1/CLIP/clips2closeRef.m6

```


# test call SNV from clips to Refs
```{bash, eval=FALSE}
# learn from https://www.htslib.org/workflow/
samtools sort -O bam -o SAM/Z1/CLIP/LOTU2closeRef.bwa.sort.bam SAM/Z1/CLIP/LOTU2closeRef.bwa.bam 
bcftools mpileup --threads 8 -d 9999 -Ob -o SAM/Z1/CLIP/LOTU2REF.bcf -f $LFR/Source/REF/zymo/D6305.rRNA.fa SAM/Z1/CLIP/LOTU2closeRef.bwa.sort.bam 
bcftools call --ploidy 1 -vmO z -o SAM/Z1/CLIP/LOTU2REF.vcf.gz SAM/Z1/CLIP/LOTU2REF.bcf
tabix -p vcf SAM/Z1/CLIP/LOTU2REF.vcf.gz
bcftools stats -F $LFR/Source/REF/zymo/D6305.rRNA.fa -s - SAM/Z1/CLIP/LOTU2REF.vcf.gz > SAM/Z1/CLIP/LOTU2REF.vcf.gz.stats
mkdir SAM/Z1/CLIP/LOTU2REF.vcf_plots
plot-vcfstats -p  SAM/Z1/CLIP/LOTU2REF.vcf_plots  SAM/Z1/CLIP/LOTU2REF.vcf.gz.stats
# try by clips
bwa mem -k 121 -t 32 -R '@RG\tID:foo\tSM:bar\tLB:library1' $LFR/Source/REF/zymo/D6305.rRNA.fa SAM/Z1/summary.BI.megahit.clip.all.fasta \
| samtools sort -O bam -o SAM/Z1/CLIP/clips2REF.bam -
bcftools mpileup --threads 8 -Ob -o SAM/Z1/CLIP/clips2REF.bcf -f $LFR/Source/REF/zymo/D6305.rRNA.fa SAM/Z1/CLIP/clips2REF.bam
bcftools view SAM/Z1/CLIP/clips2REF.bcf|metabbq readBCF2tsv > SAM/Z1/CLIP/clips2REF.bcf.tsv
bcftools call --ploidy 1 -vmO z -o SAM/Z1/CLIP/clips2REF.vcf.gz SAM/Z1/CLIP/clips2REF.bcf
tabix -p vcf SAM/Z1/CLIP/clips2REF.vcf.gz
bcftools stats -F $LFR/Source/REF/zymo/D6305.rRNA.fa -s - SAM/Z1/CLIP/clips2REF.vcf.gz > SAM/Z1/CLIP/clips2REF.vcf.gz.stats
mkdir SAM/Z1/CLIP/clips2REF.vcf_plots
plot-vcfstats -p  SAM/Z1/CLIP/clips2REF.vcf_plots  SAM/Z1/CLIP/clips2REF.vcf.gz.stats

```

# stat assembly clips
```{bash, eval=FALSE}
for i in Z{1,2,3};do 
 awk -v n=$i '{print n"\t"$0}' SAM/$i/summary.BI.megahit.clip.metadata.tsv;
done > STAT/CLIP/ZBR.clip.metadata.tsv

```


# map clips and OTUs to zymo ref
```{bash,eval=FALSE}
mkdir -p ../../Results/JAN20/STAT/ANNO

metabbq smk --force -j -npk SAM/Z{1,2,3}/CLIP/LOTU.fa
for i in Z{1,2,3};do sed 's/\S\+=//g' SAM/$i/CLIP/LOTU.log|awk -v n=$i '{print n" "$0}' ;done > STAT/CLIP/ZBR.LOTU.log
for i in Z{1,2,3};do grep "^S" SAM/$i/CLIP/all.clust.uc |sed 's/_/\t/g;s/\S\+=//g'\
 | awk -v n=$i '{print n"\t"$9"\t"$10"\t"$11"\t"$12"\t"$13"\t"$2}';
done > STAT/CLIP/ZBR.LOTU.centroids.info

for i in Z{1,2,3};do grep "^H" SAM/$i/CLIP/all.clust.uc | awk -v n=$i '{print n"\t"$2"\t"$4"\t"$8"\t"$9"\t"$10}';
done > STAT/CLIP/ZBR.LOTU.members.info

#closed mock ref mapping
metabbq smk --force -j -npk SAM/Z{1,2,3}/ANNO/clip2MockRef.clip.anno
for i in Z{1,2,3};do awk -v n=$i '{print n"\t"$0}' SAM/$i/ANNO/clip2MockRef.clip.anno;done > STAT/ANNO/clip2MockRef.clip.anno
awk '{if(p!=$2){print $0};p=$2}' STAT/ANNO/clip2MockRef.clip.anno > STAT/ANNO/clip2MockRef.top1.anno
for i in Z{1,2,3};do awk -v n=$i '{print n"\t"$0}' SAM/$i/ANNO/clip2MockRef.bead.anno;done > STAT/ANNO/ZBR.clip2MockRef.bead.anno

metabbq smk --force -j -npk SAM/Z{1,2,3}/ANNO/LOTU.to.MockRef.bead.anno
for i in Z{1,2,3};do awk -v n=$i '{print n"\t"$0}' SAM/$i/ANNO/LOTU.to.MockRef.bead.anno;done > STAT/ANNO/ZBR.LOTU.to.MockRef.bead.anno

#open silva ref mapping
metabbq smk -j -pk SAM/Z{1,2,3}/ANNO/CLIP.map.merge.anno
for i in Z{1,2,3};do awk -F "\t" -v n=$i 'BEGIN{OFS = "\t"}{gsub("_","\t",$1);print n"\t"$0}' SAM/$i/ANNO/CLIP.map.merge.anno ;done > STAT/ANNO/ZBR.CLIP.map.merge.anno
metabbq smk -j -npk SAM/Z{1,2,3}/ANNO/LOTU.map.merge.anno
for i in Z{1,2,3};do sed 's/\S\+=//g' SAM/$i/ANNO/LOTU.map.merge.anno|awk -v n=$i '{print n"\t"$0}' ;done > STAT/ANNO/ZBR.LOTU.map.merge.anno

metabbq smk -j -pk SAM/Z{1,2,3}/ANNO/CLIP.map.merge.bead.anno
for i in Z{1,2,3};do awk -v n=$i '{print n"\t"$0}' SAM/$i/ANNO/CLIP.map.merge.bead.anno;done > STAT/ANNO/CLIP.map.merge.bead.anno

#SNV
metabbq smk --force -j -pk SAM/Z{1,2,3}/CLIP/clips2REF.bcf.tsv
for i in Z{1,2,3};do awk -v n=$i '{print n"\t"$0}' SAM/$i/CLIP/clips2REF.bcf.tsv;done > STAT/CLIP/ZBR.clips2REF.bcf.tsv


#quantification
metabbq smk -j -npk SAM/Z{2,3}/PROFILE/LOTU.bwa.prop

# write to tmp/mock.LOTU.quantification.sh
SAM=$1
j=$2
mkdir -p $SAM/PROFILE/RAND
bwa mem -t 8 -h 10000 $SAM/CLIP/all.LOTU.fa $SAM/clean/fastp.sort.{1,2}_$j.fq.gz \
| samtools view -b |tee $SAM/PROFILE/RAND/all.LOTU.bwa.map_$j.bam \
| samtools view | metabbq beadStat sam -i - -o $SAM/PROFILE/RAND/LOTU.bwa.stat_$j -v
metabbq beadStat sam2b -i $SAM/PROFILE/RAND/LOTU.bwa.stat_$j -o $SAM/PROFILE/RAND/LOTU.bwa.sbb1_$j -v
#

for i in {1,2,3} ;do SAM=SAM/Z$i; for j in {00..02};do echo sh tmp/mock.LOTU.quantification.sh $SAM $j \&;done;done > tmp/mock.batch.quantification.sh

mkdir -p STAT/PROFILE
for i in {1,2,3} ;do SAM=Z$i; for j in {00..02};do 
awk -v s=$SAM -v r=$j '{print s"\t"r"\t"$1"\t"$2"\t"$3}' SAM/$SAM/PROFILE/RAND/LOTU.bwa.sbb1_$j.p;
done;done > STAT/PROFILE/ZBR.LOTU.bwa.sbb1_rand3


```




Load data:
```{r}
metadata <- read.csv("../../Results/JAN20/metadata.csv")

ZBR.LOTU.info <- read.table(md5("../../Results/JAN20/STAT/CLIP/ZBR.LOTU.log","b10505df928124f0d4d80586f7ce47ac"),sep=" ", 
    col.names = c("SID","LOTU","size","S75","CL","MAXLEN","MEANLEN","SBP","P1D","D9P","COR","PMR"))

ZBR.LOTU.centroids.info <- read.table(md5("../../Results/JAN20/STAT/CLIP/ZBR.LOTU.centroids.info",
  "97a02fd00947dafdaeda5c3c65ba6ddf"), sep="\t",
  col.names = c("SID","BID","kmer","BNO","flag","multi","cluster"))

ZBR.LOTU.members.info <- read.table(md5("../../Results/JAN20/STAT/CLIP/ZBR.LOTU.members.info",
  "b85aa7e1272cc5a741ad4cb4dd28eb57"), sep="\t",
  col.names = c("SID","CID","ident","CIART","member","centroid"))

ZBR.clip2closeRef.top1.anno <- read.table(md5("../../Results/JAN20/STAT/ANNO/clip2MockRef.top1.anno",
                                "325268f87f50515c5ee2d1b62da2dd6e"),
                           sep="\t", col.names =c("SID","CLIP","tax","idents","lens","mis","gap","qs","qe","ss","se","eval","bit","qlen")
)

ZBR.bead.anno <- read.table(md5("../../Results/JAN20/STAT/ANNO/ZBR.clip2MockRef.bead.anno","7b8dc3b06cbef1a19946cd368b2aed2d"),
    sep="\t", col.names = c("SID","BID","BLen","hits","hits.2","tax.clips","tax.pieces","tax",
                 "ident","qMapLen","mismatch","gap","bitScore","bitScore2nd",
                 "sMapPieces","sMapLen","MAPDESC","strands","qMaps","sMaps"
))

ZBR.mock.anno <- read.table(md5("../../Results/JAN20/STAT/ANNO/ZBR.LOTU.to.MockRef.bead.anno",
                                "ea648ae5671a5b67ef838f2a87dd2bd5"),
                           sep="\t", col.names = c("SID","LOTU","BLen","hits","hits.2","tax.clips","tax.pieces","tax",
                 "ident","qMapLen","mismatch","gap","bitScore","bitScore2nd",
                 "sMapPieces","sMapLen","MAPDESC","strands","qMaps","sMaps"
))

#c("SID","BID","k","CID","flag","multi","contigLen","cType","ad","clipLen","tax","idents","lens","mis","gap","bit","qlen","qcov","scov","scov.ITS","scov.SSU","scov.LSU","taxID","rank","taxPath")

ZBR.silva.bead.anno <- read.table(md5("../../Results/JAN20/STAT/ANNO/CLIP.map.merge.bead.anno","653c55f9a09c293912103924f362f800"),comment.char = "",sep="\t", col.names = c("SID","BINFO","tax","ident","mapLen","mismatch","gap","qs","qe","ss","se","eval","bitscore","qlen","slen","strands","qDesc","sDesc","covs")
)

ZBR.silva.anno <- read.table(md5("../../Results/JAN20/STAT/ANNO/ZBR.LOTU.map.merge.anno",
      "5d079afbd752d1f2de3ee7fb2d2c1e52"), sep="\t",comment.char = "",
   col.names = c("SID","LOTU","tax","idents","lens","mis","gap","bit","qlen","qcov","scov","scov.ITS","scov.SSU","scov.LSU","taxID","rank","taxPath")
)

ZBR.clip2REF.bcf <- read.table(md5("../../Results/JAN20/STAT/CLIP/ZBR.clips2REF.bcf.tsv",
  "844ff306eaa5985685ef6885d32fb5fa"), sep="\t",col.names = c("SID","CHR","POS","REF","ALT","QUAL","TYPE","IDV","DP","DR","DA","BQR","BQA","MQR","MQA","MDR","MDA","BQB","MQSB"))

ZBR.prop <- read.table(md5("../../Results/JAN20/STAT/PROFILE/ZBR.LOTU.bwa.sbb1_rand3","12564411f41155ab37b0a07f92d0c57c"),sep="\t",
                      col.names = c("SID","rand","count0","count","LOTU"))
#

zymo.copy.num <- data.frame(species=c("Pseudomonas aeruginosa","Escherichia coli","Salmonella enterica","Lactobacillus fermentum",
  "Enterococcus faecalis","Staphylococcus aureus","Listeria monocytogenes","Bacillus subtilis","Saccharomyces cerevisiae","Cryptococcus neoformans"),
  TC16S=c(4.2,10.1,10.4,18.4,9.9,15.5,14.1,17.4,NA,NA),
  TC16S18S=c(3.6,8.9,9.1,16.1,8.7,13.6,12.4,15.3,9.3,3.3),copyNum=c(4,7,7,5,4,6,6,10,109,60))
zymo.copy.num$genus <- sapply(strsplit(as.character(zymo.copy.num$species)," "),"[",1)
zymo.bac <- c("Pseudomonas aeruginosa","Escherichia coli","Salmonella enterica","Lactobacillus fermentum",
  "Enterococcus faecalis","Staphylococcus aureus","Listeria monocytogenes","Bacillus subtilis")
zymo.euk <- c("Saccharomyces cerevisiae","Cryptococcus neoformans")

```


#Curation
```{r}
###
ZBR.bead.df <- merge(metadata,ZBR.bead.anno,by="SID")
ZBR.bead.df <- cbind(ZBR.bead.df,getMapPropFun(ZBR.bead.df$MAPDESC))
ZBR.bead.df$mainUnit <- ifelse(ZBR.bead.df$SSU>ZBR.bead.df$LSU,"SSU","LSU")
ZBR.bead.df$unit <- factor(ifelse(ZBR.bead.df$SSU>50,1,0) + ifelse(ZBR.bead.df$LSU>50,2,0) + ifelse(ZBR.bead.df$ITS1+ZBR.bead.df$ITS2>50,4,0),
                           levels=seq(0,7))
levels(ZBR.bead.df$unit) <- c("partial(<50%)","SSU","LSU","Both-ITS","ITS","SSU+ITS","LSU+ITS","All")

###
ZBR.LOTU.centroids.info$LOTU <- sprintf("LOTU_%06d",ZBR.LOTU.centroids.info$cluster+1)
ZBR.LOTU.centroids.info$KMER <- as.numeric(sub("k","",ZBR.LOTU.centroids.info$kmer))
ZBR.badf <- merge(merge(metadata,ZBR.mock.anno,by="SID"),ZBR.LOTU.centroids.info,by=c("SID","LOTU"))
#badf$delta.clips <- badf$clips-badf$tax.clips

ZBR.badf <- cbind(ZBR.badf,getMapPropFun(ZBR.badf$MAPDESC))
ZBR.badf$mainUnit <- ifelse(ZBR.badf$SSU>ZBR.badf$LSU,"SSU","LSU")
ZBR.badf$unit <- as.factor(ifelse(ZBR.badf$SSU>50,1,0) + ifelse(ZBR.badf$LSU>50,2,0)) 
levels(ZBR.badf$unit) <- c("partial(<50%)","SSU","LSU","Both")
ZBR.badfm <- merge(ZBR.LOTU.info,ZBR.badf,by=c("SID","LOTU"),all.x=T)
ZBR.badfm$D9P[which(is.na(ZBR.badfm$D9P))] <- 0
```

# clip annotation info:
```{r}
n50(ZBR.bead.df$BLen)
ggplot(ZBR.bead.df,aes(x=BLen,fill=unit)) +geom_histogram(position="stack",binwidth=100) +
  theme(legend.position = c(0.99,.9),legend.justification = c(1,1)) +
  scale_x_continuous(breaks=seq(0,10000,1000),limits=c(0,6000))

ggplot(ZBR.bead.df,aes(x=ident,fill=unit)) +geom_histogram(position="stack",binwidth=.5) +
  theme(legend.position = c(.1,.9),legend.justification = c(0,1)) +
  scale_x_continuous(breaks=seq(90,100,1))

ggplot(ZBR.bead.df,aes(x=tax,fill=unit)) +geom_bar(position="stack") + coord_flip()
ggplot(ZBR.bead.df,aes(x=ident,color=tax)) +geom_density() +
    scale_x_continuous(breaks=seq(90,100,1))


ggplot(ZBR.bead.df,aes(x=BLen,y=ident)) + 
  stat_density_2d(aes(fill = stat(nlevel)), geom = "polygon") + 
  scale_fill_gradientn(colours=rev(brewer.pal(11,"Spectral"))) +
  ggtitle("Polygon between BLen and ident") +
  scale_y_continuous(breaks=seq(99,100,0.2)) +
  scale_x_continuous(breaks=seq(0,4000,500))


```

```{r}
ZBR.bead.99.stat <- ddply(ZBR.bead.df[,c("SID","BID","ident","tax")]%>%filter(ident>99),c("SID","tax"),summarise,count=length(BID))
ZBR.bead.99.stat <- ddply(ZBR.bead.99.stat,c("SID"),transform,pct=count/sum(count)*100)
ZBR.bead.99.stat$tax <- sub("_"," ",ZBR.bead.99.stat$tax)
zymo.merge.df <- merge(zymo.copy.num,ZBR.bead.99.stat,by=)
```


# basic:
```{r}
paste0("N50: ",n50(ZBR.badfm$COR))
```

```{r}
ggplot(ZBR.badfm,aes(x=COR,y=ident)) + 
  stat_density_2d(aes(fill = stat(nlevel)), geom = "polygon") + 
  scale_fill_gradientn(colours=rev(brewer.pal(11,"Spectral"))) +
  facet_grid(.~SID) + ggtitle("Polygon between COR and ident")

ZBR.badfm$CORG <- round(ZBR.badfm$COR/500 + 0.49,0) *500
ZBR.badfm$CORG <- ifelse(ZBR.badfm$CORG>=3000,3000,ZBR.badfm$CORG)
ggplot(ZBR.badfm,aes(x=ident)) + geom_histogram() + facet_wrap(~CORG)
```

```{r}

ggplot(ZBR.badfm,aes(x=qMapLen,fill=SID)) +geom_histogram(position="dodge") + scale_x_log10()
ggplot(ZBR.badfm,aes(x=ident,fill=SID)) +geom_histogram(position="dodge")
ggplot(ZBR.badfm,aes(x=ident,color=SID)) +geom_density()
ZBR.badfm$qMap1k <- round(ZBR.badfm$qMapLen/1000-0.49,0)*1000
ggplot(ZBR.badfm,aes(x=ident,color=factor(qMap1k),linetype=SID)) +geom_density()
ggplot(ZBR.badfm,aes(x=ident,fill=factor(qMap1k))) +geom_histogram(position = "dodge") + facet_grid(SID~.)

ZBR.badfm$S75.2 <- 2^round(log2(ZBR.badfm$S75),0)

ggplot(ZBR.badfm,aes(x=S75,fill=SID)) +geom_histogram(position="dodge") +
  scale_x_log10() + annotation_logticks(side="b")
ggplot(ZBR.badfm,aes(y=ident,x=factor(S75.2),fill=SID)) +geom_boxplot(aes(fill=SID)) +
  scale_y_continuous(breaks=seq(94,102,1),limits=c(94,100))
ggplot(ZBR.badfm%>%filter(S75<20),aes(y=ident,x=factor(S75),fill=SID)) +geom_boxplot(aes(fill=SID)) +
  scale_y_continuous(breaks=seq(94,102,1),limits=c(94,100))

ZBR.badfm$SBP.2 <- 2^round(log2(ZBR.badfm$SBP),0)
ggplot(ZBR.badfm,aes(y=ident,x=factor(SBP.2),fill=SID)) +geom_boxplot(aes(fill=SID)) +
  scale_y_continuous(breaks=seq(94,102,1),limits=c(94,100))
ggplot(ZBR.badfm%>%filter(SBP<32),aes(y=ident,x=factor(SBP))) +geom_boxplot() +
  scale_y_continuous(breaks=seq(94,102,1),limits=c(94,100))

ZBR.badfm$P1D.2 <- 2^round(log2(ZBR.badfm$P1D),0)
ggplot(ZBR.badfm,aes(y=ident,x=factor(P1D.2),fill=SID)) +geom_boxplot(aes(fill=SID)) +
  scale_y_continuous(breaks=seq(94,102,1),limits=c(94,100))
ggplot(ZBR.badfm%>%filter(P1D<32),aes(y=ident,x=factor(P1D))) +geom_boxplot() +
  scale_y_continuous(breaks=seq(94,102,1),limits=c(94,100))

ZBR.badfm$D9P.2 <- 2^round(log2(ZBR.badfm$D9P),0)
ggplot(ZBR.badfm,aes(y=ident,x=factor(D9P.2),fill=SID)) +geom_boxplot(aes(fill=SID)) +
  scale_y_continuous(breaks=seq(94,102,1),limits=c(94,100))

ggplot(ZBR.badfm,aes(x=CL,fill=unit)) +geom_histogram(position="stack",binwidth=50)  + facet_grid(.~SID)
ggplot(ZBR.badfm,aes(x=COR,fill=unit)) +geom_histogram(position="stack",binwidth=50)  + facet_grid(.~SID)
ggplot(ZBR.badfm%>%filter(CL>499),aes(x=CL,fill=unit)) +geom_histogram(position="stack",binwidth=100) +
  theme(legend.position = c(0.99,.9),legend.justification = c(1,1))

```

```{r}
ZBR.badfm$KMERG <- ifelse(ZBR.badfm$KMER<=111,"<=111",ZBR.badfm$KMER)

ggplot(ZBR.badfm%>%filter(!is.na(flag)),aes(x=multi,y=ident)) + 
  stat_density_2d(aes(fill = stat(nlevel)), geom = "polygon") + 
  scale_fill_gradientn(colours=rev(brewer.pal(11,"Spectral"))) +
  facet_grid(SID~flag,scale="free_x")

```


```{r}
ZBR.badfm$ident5 <- round(ZBR.badfm$ident+0.49,)-0.5
ZBR.badfm$ident5 <- ifelse(ZBR.badfm$ident5<95,94.5,ZBR.badfm$ident5)

ggplot(ZBR.badfm%>%filter(COR>999&flag!=0&D9P/COR<0.01),aes(x=ident5)) +
  geom_bar(aes(fill=unit),stat="count",position="stack") + 
  scale_x_continuous(breaks=seq(94,102,1),limits=c(94,100)) +
  coord_flip() + facet_grid(.~SID)

badf.stat2 <- ddply(
  ddply(ZBR.badfm%>%filter(COR>999&flag!=0&D9P/COR<0.01),c("SID","ident5"),summarise,count=length(SID)),
  c("ident5"),summarise,mean.count=mean(count),sd=sd(count))

badf.stat3 <- ddply(
  ddply(ZBR.badfm%>%filter(COR>999&flag!=0&D9P/COR<0.01),c("SID","ident5","unit"),summarise,count=length(SID)),
  c("ident5","unit"),summarise,mean.count=mean(count))

ggplot(badf.stat3,aes(x=ident5,y=mean.count)) + geom_bar(aes(fill=unit),stat="identity",position="stack") + 
  geom_errorbar(data=badf.stat2,aes(ymax=mean.count+sd,ymin=mean.count-sd),width=.1) + 
  scale_x_continuous(breaks=seq(94,102,1),limits=c(94,100)) +
  coord_flip()

ZBR.badfm$CORG <- round(ZBR.badfm$COR/500-0.49,0)*500

badf.stat4 <- ddply(
  ddply(ZBR.badfm%>%filter(COR>999&flag!=0&D9P/COR<0.01),c("SID","CORG","ident5","unit"),summarise,count=length(SID)),
  c("CORG","ident5","unit"),summarise,mean.count=mean(count))

ggplot(badf.stat4,aes(x=ident5,y=mean.count)) + geom_bar(aes(fill=unit),stat="identity",position="stack") + 
  coord_flip() + facet_wrap(~CORG,scale="free")

badf.stat5 <- ddply(
  ddply(ZBR.badfm%>%filter(COR>999&flag!=0&D9P/COR<0.01),c("SID","flag","ident5","unit"),summarise,count=length(SID)),
  c("flag","ident5","unit"),summarise,mean.count=mean(count))

ggplot(badf.stat5,aes(x=ident5,y=mean.count)) + geom_bar(aes(fill=unit),stat="identity",position="stack") + 
  coord_flip() + facet_wrap(~flag)
```

# SNV calling
```{r}
ZBR.clip2REF.bcf$DAP <- ZBR.clip2REF.bcf$DA/ZBR.clip2REF.bcf$DP *100
ggplot(ZBR.clip2REF.bcf,aes(x=POS)) +geom_line(aes(y=DP,color=SID)) +
  facet_wrap(~CHR,ncol=2)

ggplot(ZBR.clip2REF.bcf%>%filter(TYPE=="INDEL"),aes(x=POS)) +
  geom_bar(aes(y=IDV/DP,fill=SID),position="dodge",stat="identity") +
  facet_wrap(~CHR,ncol=2)

ggplot(ZBR.clip2REF.bcf%>%filter(TYPE=="SUB"),aes(x=POS)) +
  geom_bar(aes(y=DAP,fill=SID),position="dodge",stat="identity") +
  facet_wrap(~CHR,ncol=2)

ggplot(ZBR.clip2REF.bcf,aes(x=POS)) +geom_ribbon(aes(ymin=0,ymax=DP),alpha=.1,color="grey50") +
  geom_bar(data=ZBR.clip2REF.bcf%>%filter(TYPE=="SUB"),aes(y=DA),color="cyan4",position="dodge",stat="identity") +
  geom_bar(data=ZBR.clip2REF.bcf%>%filter(TYPE=="INDEL"),aes(y=IDV),color="brown",position="stack",stat="identity",size=1) +
  facet_grid(CHR~SID)

pick.b <- c("Bacillus_subtilis","Escherichia_coli")
ggplot(ZBR.clip2REF.bcf%>%filter(SID=="Z1"&CHR%in%pick.b),aes(x=POS)) +geom_ribbon(aes(ymin=0,ymax=DP),alpha=.1,color="grey50") +
  geom_bar(data=ZBR.clip2REF.bcf%>%filter(SID=="Z1"&CHR%in%pick.b&TYPE=="SUB"),aes(y=DA),color="cyan4",position="dodge",stat="identity") +
  geom_bar(data=ZBR.clip2REF.bcf%>%filter(SID=="Z1"&CHR%in%pick.b&TYPE=="INDEL"),aes(y=IDV),color="brown",position="stack",stat="identity",size=1) +
  facet_grid(CHR~.,scale="free_y")

```

# Check annotation consistence between cluster members to their centroid
```{r}
ZBR.clip2closeRef.merge.df <- merge(ZBR.LOTU.members.info,ZBR.clip2closeRef.top1.anno[,c("SID","CLIP","tax","idents","bit")],
                                    by.x=c("SID","centroid"),by.y=c("SID","CLIP"),all.x=T)
ZBR.clip2closeRef.merge.df <- merge(ZBR.clip2closeRef.merge.df,ZBR.clip2closeRef.top1.anno[,c("SID","CLIP","tax","idents","bit")],
                                    by.x=c("SID","member"),by.y=c("SID","CLIP"),all.x=T)

consistStat <- function(d){
  nrow <- nrow(d)
  hits <- length(which(!is.na(d$idents.y)))
  cluster.ident <- c(d$idents.y,d$idents.x[1])
  consisPct <- ifelse(hits==0,0,length(which(d$tax.x==d$tax.y))/hits)
  mean.ident <- mean(cluster.ident[which(!is.na(cluster.ident))])
  sd.ident <- sd(cluster.ident[which(!is.na(cluster.ident))])
  res <- data.frame(SID=d$SID[1],CID=d$CID[1],mcount=nrow,hits=hits,consisPct=consisPct,
                    c.ident=d$idents.x[1],mean.ident=mean.ident,sd.ident=sd.ident)
  return(res)
}
ZBR.clip2closeRef.stat.df <- ddply(ZBR.clip2closeRef.merge.df,c("SID","centroid"),consistStat)

ggplot(ZBR.clip2closeRef.stat.df,aes(x=consisPct,color=SID)) + geom_density()

ggplot(ZBR.clip2closeRef.stat.df,aes(x=c.ident,y=mean.ident)) + 
  stat_density_2d(aes(fill = stat(nlevel)), geom = "polygon") + 
  scale_fill_gradientn(colours=rev(brewer.pal(11,"Spectral"))) +
  facet_grid(.~SID,scale="free_x")

ZBR.clip2closeRef.LOTU.df2 <- merge(unique(ZBR.LOTU.members.info[,c(1,2,6)]),ZBR.clip2closeRef.top1.anno[,c("SID","CLIP","tax","idents","bit")],
                                    by.x=c("SID","centroid"),by.y=c("SID","CLIP"))
ZBR.clip2closeRef.LOTU.df2$LOTU <- sprintf("LOTU_%06d",ZBR.clip2closeRef.LOTU.df2$CID+1)
ZBR.clip2closeRef.LOTU.df2 <- merge(ZBR.clip2closeRef.LOTU.df2,ZBR.badfm,
                                    by=c("SID","LOTU"),all.x=T)
ZBR.clip2closeRef.LOTU.df2$tax.y <- sub(" ","_",ZBR.clip2closeRef.LOTU.df2$tax.y)
ZBR.clip2closeRef.LOTU.df2$consist <- ifelse(as.character(ZBR.clip2closeRef.LOTU.df2$tax.x)==as.character(ZBR.clip2closeRef.LOTU.df2$tax.y),"SAME","DIFF")

ggplot(ZBR.clip2closeRef.LOTU.df2%>%filter(!is.na(consist)),aes(x=idents  ,y=ident)) + 
  stat_density_2d(aes(fill = stat(nlevel)), geom = "polygon") + 
  scale_fill_gradientn(colours=rev(brewer.pal(11,"Spectral"))) +
  facet_grid(consist~SID,scale="free_x")


```

# silva mapping annotation
```{r}
getMapPropFun2 <- function(d){
  uOrder <- c("ALL","SSU","ITS1","5.8S","ITS2","LSU")
  ulist <- data.frame(key=c("ALL","SSU","16S","ITS","ITS1","5.8S","ITS2","LSU","23S"),
                      ind=c(1,2,2,3,3,4,5,6,6))
  getM <- stri_match_all_regex(d,"(ALL|23S|LSU|ITS|ITS1|ITS2|5.8S|SSU|16S)\\((\\d+)\\)")
  getN <- length(getM)
  res <- matrix(0,ncol = 6,nrow = getN)
  for(i in 1:length(getM)){
    geti <- getM[[i]]
    for(j in 1:nrow(geti)){
      res[i,ulist$ind[which(ulist$key==geti[j,2])]] <- geti[j,3]
    }
  }
  res <- matrix(as.numeric(res),ncol=6)
  colnames(res) <- uOrder
  return(res)
}
sbinfo <- matrix(unlist(strsplit(as.character(ZBR.silva.bead.anno$BINFO),"|",fixed = T)),ncol=4,byrow = T)
sbinfodf <- data.frame(BID=sbinfo[,1],clips=as.numeric(sbinfo[,3]),hits=as.numeric(sbinfo[,4]))
covdf <- getMapPropFun2(as.character(ZBR.silva.bead.anno$covs))

ZBR.slv.bead.df <- cbind(sbinfodf,ZBR.silva.bead.anno[,-c(2,16:19)],covdf[,c(2,6)])
ZBR.slv.bead.df$unit <- factor(ifelse(ZBR.slv.bead.df$SSU>0,1,0) + ifelse(ZBR.slv.bead.df$LSU>0,2,0),levels = seq(0,3))
levels(ZBR.slv.bead.df$unit) <- c("NONE","SSU","LSU","BOTH")
```

```{r}
n50(ZBR.slv.bead.df$qlen)
ggplot(ZBR.slv.bead.df%>%filter(SSU>0),aes(x=SSU,fill=SID)) +geom_histogram(position="dodge")
ggplot(ZBR.slv.bead.df%>%filter(LSU>0),aes(x=LSU,fill=SID)) +geom_histogram(position="dodge")

ggplot(ZBR.slv.bead.df,aes(x=qlen,fill=unit)) +geom_histogram(position="stack",binwidth=100) +
  theme(legend.position = c(0.99,.9),legend.justification = c(1,1)) +
  scale_x_continuous(breaks=seq(0,10000,1000),limits=c(0,6000))

ggplot(ZBR.slv.bead.df,aes(x=ident,fill=unit)) +geom_histogram(position="stack",binwidth=.5) +
  theme(legend.position = c(.1,.9),legend.justification = c(0,1)) +
  scale_x_continuous(breaks=seq(90,100,1))

topTax <- head(rev(sort(table(ZBR.slv.bead.df$tax))),10)
ZBR.slv.bead.df$topTax <- ifelse(as.character(ZBR.slv.bead.df$tax)%in%names(topTax),as.character(ZBR.slv.bead.df$tax),"rest")
ggplot(ZBR.slv.bead.df,aes(x=topTax,fill=unit)) +geom_bar(position="stack") + coord_flip()

ggplot(ZBR.slv.bead.df,aes(x=ident,color=topTax)) +geom_density() +
    scale_x_continuous(breaks=seq(90,100,1))


ggplot(ZBR.slv.bead.df,aes(x=qlen,y=ident)) + 
  stat_density_2d(aes(fill = stat(nlevel)), geom = "polygon") + 
  scale_fill_gradientn(colours=rev(brewer.pal(11,"Spectral"))) +
  ggtitle("Polygon between qlen and ident") +
  scale_y_continuous(breaks=seq(99,100,0.2)) +
  scale_x_continuous(breaks=seq(0,4000,500))

```



# compare annotation : mock ref vs. silva (bead scale)
```{r}
levels(ZBR.badfm$tax) <- sub("_"," ",levels(ZBR.badfm$tax))
ZBR.compare.anno <- merge(ZBR.badfm,ZBR.silva.anno,by=c("SID","LOTU"),all.x=T)
ZBR.compare.anno <- merge(ZBR.badfm%>%filter(COR>999&flag!=0&D9P/COR<0.01),ZBR.silva.anno,by=c("SID","LOTU"),all.x=T)
checkTaxConsitRank <- function(d){
  tax.x <- strsplit(as.character(d$tax.x)," ")
  tax.y <- strsplit(as.character(d$tax.y)," ")
  tax.y.spg <- sapply(strsplit(as.character(d$taxPath),";"),"[",7)
  spName <- ifelse(is.na(sapply(tax.y,"[",2)),"",sapply(tax.y,"[",2))
  #consitLv <- ifelse(sapply(tax.x,"[",1)==sapply(tax.y,"[",1),1,0) + ifelse(d$tax.x==tax.y.spg,2,0) + ifelse(sapply(tax.x,"[",2)==spName,4,0)
  consitLv <- ifelse(sapply(tax.x,"[",1)==sapply(tax.y,"[",1),1,0) +ifelse(sapply(tax.x,"[",2)==spName,2,0)
  return(consitLv)
}
ZBR.compare.anno$consitLv <- as.factor((ddply(ZBR.compare.anno[,c("SID","LOTU","tax.x","tax.y","taxPath")],c("SID","LOTU"),checkTaxConsitRank))$V1)
#levels(ZBR.compare.anno$consitLv) <- c("different","genus","species group","species","species")
levels(ZBR.compare.anno$consitLv) <- c("different","genus","species","species")

```

```{r}
ZBR.compare.anno$hitDB <- as.factor(ifelse(ZBR.compare.anno$scov.SSU==0,0,1) + ifelse(ZBR.compare.anno$scov.LSU==0,0,2))
levels(ZBR.compare.anno$hitDB) <- c("SILVA.SSU","SILVA.LSU","SILVA.LSU+SSU")

ZBR.compare.anno$reachLimit <- as.factor(ifelse(ZBR.compare.anno$scov.ITS>99,T,ifelse(ZBR.compare.anno$scov.SSU>99,T,ifelse(ZBR.compare.anno$scov.LSU>99,T,0))))

ZBR.compare.anno$idents5 <- round(ZBR.compare.anno$idents+0.49,0)
ZBR.compare.anno$validAnno <- ifelse(grepl("metagenome|uncultured ",ZBR.compare.anno$tax.y),"unknown",
                                     ifelse(grepl("sp. ",ZBR.compare.anno$tax.y),"sp. ID","valid"))


ggplot(ZBR.compare.anno,aes(x=SID,fill=consitLv)) + geom_bar(position="stack") + facet_wrap(~ident5,ncol=5)
ggplot(ZBR.compare.anno,aes(x=SID,fill=consitLv)) + geom_bar(position="fill") + facet_wrap(~ident5,ncol=5)

ggplot(ZBR.compare.anno%>%filter(ident>99),aes(x=SID,fill=consitLv)) + geom_bar(position="stack") + facet_wrap(~idents5,ncol=5)
ggplot(ZBR.compare.anno%>%filter(ident>99),aes(x=SID,fill=consitLv)) + geom_bar(position="fill") + facet_wrap(~idents5,ncol=5)

ZBR.compare.anno$scovG <- round(ZBR.compare.anno$scov/10-0.49,0)*10

ggplot(ZBR.compare.anno%>%filter(ident>99),aes(x=SID,fill=consitLv)) + geom_bar(position="fill")+
  facet_wrap(~scovG,ncol=5)

ZBR.compare.anno$bitG <- round(ZBR.compare.anno$bit/1000,0)*1000

ggplot(ZBR.compare.anno%>%filter(COR>999&ident>99&idents>99),aes(x=consitLv,fill=validAnno)) + 
  geom_bar(position="stack") + facet_grid(.~SID,scale="free")

ggplot(ZBR.compare.anno%>%filter(COR>999),aes(x=unit,fill=consitLv)) + geom_bar(position="fill") +
  facet_wrap(~ident5,ncol=5) + coord_flip()

ggplot(ZBR.compare.anno%>%filter(COR>999),aes(x=ident,y=idents)) + 
  stat_density_2d(aes(fill = stat(nlevel)), geom = "polygon") + 
  scale_fill_gradientn(colours=rev(brewer.pal(11,"Spectral"))) +
  facet_grid(SID~.,scale="free_x")

ggplot(ZBR.compare.anno%>%filter(COR>999),aes(x=SBP,y=idents)) + 
  stat_density_2d(aes(fill = stat(nlevel)), geom = "polygon") + 
  scale_fill_gradientn(colours=rev(brewer.pal(11,"Spectral"))) +
  facet_grid(SID~.,scale="free_x")

ggplot(ZBR.compare.anno,aes(x=idents,color=consitLv)) + geom_density()
ggplot(ZBR.compare.anno%>%filter(COR>999),aes(x=idents,color=consitLv)) + geom_density()
ggplot(ZBR.compare.anno%>%filter(COR>999&multi>20),aes(x=idents,color=consitLv)) + geom_density() +facet_grid(SID~validAnno)
ggplot(ZBR.compare.anno%>%filter(COR>999),aes(x=SBP,color=consitLv)) + geom_density()



ggplot(ZBR.compare.anno%>%filter(COR>999),aes(x=idents-ident,color=consitLv)) + geom_density() 

ggplot(ZBR.compare.anno,aes(x=qMapLen,y=lens,color=consitLv)) + geom_point()
ggplot(ZBR.compare.anno,aes(x=bitScore,y=bit,color=consitLv)) + geom_point(alpha=.5)
ggplot(ZBR.compare.anno,aes(x=bitScore,y=bit,color=hitDB,shape=reachLimit)) +
  geom_abline(slope = 1,intercept = 0, linetype=2,alpha=.5,size=2,color="grey50") +
  geom_point(alpha=.5) + scale_shape_manual(values=c(16,4))
ggplot(ZBR.compare.anno,aes(x=bitScore,y=bit,color=hitDB,shape=reachLimit)) + geom_point(alpha=.5) +facet_grid(.~consitLv) +
    geom_abline(slope = 1,intercept = 0, linetype=2,alpha=.5,size=2,color="grey50") +
  geom_point(alpha=.5) + scale_shape_manual(values=c(16,4))

ggplot(ZBR.compare.anno,aes(x=hitDB,fill=consitLv)) + geom_bar(position="fill")
ggplot(ZBR.compare.anno%>%filter(COR>999),aes(x=hitDB,fill=consitLv)) + geom_bar(position="stack") +facet_grid(.~SID)
ggplot(ZBR.compare.anno%>%filter(COR>999),aes(x=hitDB,fill=interaction(consitLv,validAnno))) + geom_bar(position="fill") + facet_grid(.~SID)


ggplot(ZBR.compare.anno%>%filter(!is.na(consitLv)),aes(x=idents,color=consitLv)) + geom_density(size=1,alpha=.5) +
  theme(legend.position = c(0.1,.99),legend.justification = c(0,1)) + facet_grid(hitDB~.)
#ggplot(ZBR.compare.anno,aes(x=qlen,fill=hitDB)) + geom_bar() +facet_grid(.~consitLv) #+ theme(axis.text.x = element_text(angle=90))
ggplot(ZBR.compare.anno,aes(x=bitScore,y=bit,color=mainUnit)) + geom_point(alpha=.5) +facet_grid(.~consitLv)

```

completness of subunits
```{r}
ZBR.stat.comp <- melt(ZBR.compare.anno%>%filter(!is.na(consitLv)),measure.vars = c("ALL","ITS1","SSU","LSU"),variable.name = "subunits", value.name = "completness")
ZBR.stat.comp$qlen.seg <- round(ZBR.stat.comp$qlen/500,0)*500
ggplot(ZBR.stat.comp%>%filter(qlen.seg>400&qlen.seg<5500),aes(x=completness,fill=subunits)) + geom_histogram(position="dodge") +
  facet_grid(qlen.seg~.,scale="free")

```

comapre with mock ref annotation
```{r}
selectTax <- c(zymo.bac)
selectGenus <- sapply(strsplit(selectTax," "),"[",1)

ZBR.qdf <- merge(ZBR.prop,ZBR.compare.anno,by=c("SID","LOTU"),all.x=T)
ZBR.qdf$species <- sapply(strsplit(as.character(ZBR.qdf$taxPath),";"),"[",7)
ZBR.qdf$genus <- sapply(strsplit(as.character(ZBR.qdf$taxPath),";"),"[",6)
ZBR.qdf$selectTax <- ifelse(ZBR.qdf$genus%in%selectGenus,as.character(ZBR.qdf$genus),"Others.above.Genus")
ZBR.qdf$checkLv <- ifelse(ZBR.qdf$species%in%selectTax,"Same Sp.",ifelse(ZBR.qdf$genus%in%selectGenus,"Same Genus","Others.above.Genus"))

ZBR.qdf$subunit <- as.factor(ifelse(ZBR.qdf$scov.ITS>25,1,0)+ifelse(ZBR.qdf$scov.SSU>25,2,0)+ifelse(ZBR.qdf$scov.LSU>25,4,0))
ZBR.stat <- ddply(ZBR.qdf%>%filter(LOTU!='*'),c("SID","rand","tax.x"),summarise,count=sum(count))
ZBR.stat <- ddply(ZBR.stat,c("SID","rand"),transform,pct=count/sum(count)*100)
ZBR.stat1 <- merge(ZBR.stat,zymo.copy.num[,c("species","TC16S")],by.x="tax.x",by.y="species",all.x=T)
ggplot(ZBR.stat1,aes(x=tax.x,y=pct,fill=factor(rand))) + 
  geom_bar(aes(y=TC16S),fill=NA,color="grey50",stat="identity",position="dodge") +
  geom_bar(alpha=.5,stat = 'identity', position = 'dodge') +
  facet_grid(SID~.) + coord_flip()

ZBR.stat2 <- ddply(ZBR.qdf%>%filter(ident>99),c("SID","rand","tax.x"),summarise,count=sum(count))
ZBR.stat2 <- ddply(ZBR.stat2,c("SID","rand"),transform,pct=count/sum(count)*100)
ZBR.stat2 <- merge(ZBR.stat2,zymo.copy.num[,c("species","TC16S")],by.x="tax.x",by.y="species",all.x=T)
ggplot(ZBR.stat2,aes(x=tax.x,y=pct,fill=factor(rand))) + 
  geom_bar(aes(y=TC16S),fill=NA,color="grey50",stat="identity",position="dodge") +
  geom_bar(alpha=.5,stat = 'identity', position = 'dodge') +
  facet_grid(SID~.) + coord_flip()

ZBR.stat2$pct.delta <- ZBR.stat2$pct-ZBR.stat2$TC16S

ZBR.mean2.df <- ddply(ZBR.stat2,c("tax.x"),summarise,sd=sd(pct.delta),pct=mean(pct.delta))
ZBR.mean2.df <- ZBR.mean2.df%>%filter(tax.x%in%selectTax)
colnames(ZBR.mean2.df)[1] <- "selectTax"


ZBR.mean2.df$selectTax <- as.factor(as.character(ZBR.mean2.df$selectTax))

ZBR.mean2.df <- ZBR.mean2.df[order(ZBR.mean2.df$pct),]
ZBR.mean2.df$selectTax <- reorder(ZBR.mean2.df$selectTax,ZBR.mean2.df$pct,mean)
ggplot(ZBR.mean2.df,aes(x=selectTax,y=pct,fill=selectTax)) + geom_hline(yintercept = 0,linetype=2,color="grey50",size=2) +
  geom_bar(stat = 'identity', position = 'dodge') + geom_errorbar(aes(ymax=pct+sd,ymin=pct-sd),width=.4) + coord_flip() +
  ylab("delta of quantificatin \nover theoretical value (%)") + guides(fill=F)


```

```{r}
ZBR.qdf$selectTax <- ifelse(ZBR.qdf$species%in%selectTax,as.character(ZBR.qdf$species),ifelse(ZBR.qdf$genus%in%selectGenus,"Others.in.Genus","Others.above.Genus"))

ZBR.stat3 <- ddply(ZBR.qdf%>%filter(idents>99.5),c("SID","rand","selectTax"),summarise,count=sum(count))
ZBR.stat3 <- merge(ddply(ZBR.stat3,c("SID","rand"),transform, pct=count/sum(count)*100),
                  fungi.copy.num[,c("species","TC16S")],by.x="selectTax",by.y="species",all.x=T)
ZBR.stat3$TC16S[which(is.na(ZBR.stat3$TC16S))] <- 0
ZBR.stat3$pct.delta <- ZBR.stat3$pct-ZBR.stat3$TC16S
ggplot(ZBR.stat3,aes(x=SID,y=pct,fill=factor(rand))) +geom_bar(stat="identity",position="dodge") +facet_grid(.~selectTax)
ggplot(ZBR.stat3,aes(x=SID,y=pct,fill=factor(rand))) + geom_bar(aes(y=TC16S),fill=NA,color="grey50",stat="identity",position="dodge") +
  geom_bar(stat="identity",position="dodge",alpha=.5) +facet_grid(.~selectTax)

ZBR.mean.df <- ddply(ZBR.stat3,c("selectTax"),summarise,sd=sd(pct.delta),pct=mean(pct.delta))
ZBR.mean.df <- ZBR.mean.df%>%filter(selectTax%in%selectTax)
colnames(ZBR.mean.df)[1] <- "selectTax"


ZBR.mean.df$selectTax <- as.factor(as.character(ZBR.mean.df$selectTax))

ZBR.mean.df <- ZBR.mean.df[order(ZBR.mean.df$pct),]
ZBR.mean.df$selectTax <- reorder(ZBR.mean.df$selectTax,ZBR.mean.df$pct,mean)
ggplot(ZBR.mean.df,aes(x=selectTax,y=pct,fill=selectTax)) + geom_hline(yintercept = 0,linetype=2,color="grey50",size=2) +
  geom_bar(stat = 'identity', position = 'dodge') + geom_errorbar(aes(ymax=pct+sd,ymin=pct-sd),width=.4) + coord_flip() +
  ylab("delta of quantificatin \nover theoretical value (%)") + guides(fill=F)

```

```{r}

ZBR.stat3 <- ddply(ZBR.qdf%>%filter(idents>99),c("SID","rand","checkLv","selectTax"),summarise,count=sum(count))
ZBR.stat3 <- merge(ddply(ZBR.stat3,c("SID","rand"),transform, pct=count/sum(count)*100),
                  fungi.copy.num[,c("genus","TC16S")],by.x="selectTax",by.y="genus",all.x=T)
ZBR.stat3$TC16S[which(is.na(ZBR.stat3$TC16S))] <- 0
ZBR.stat3$pct.delta <- ZBR.stat3$pct-ZBR.stat3$TC16S
ggplot(ZBR.stat3,aes(x=SID,y=pct,fill=factor(rand))) +geom_bar(stat="identity",position="dodge") +facet_grid(.~selectTax)
ggplot(ZBR.stat3,aes(x=SID,y=pct,fill=factor(rand))) + geom_bar(aes(y=TC16S),fill=NA,color="grey50",stat="identity",position="dodge") +
  geom_bar(stat="identity",position="dodge",alpha=.5) +facet_grid(.~selectTax)

ZBR.mean.df <- ddply(ZBR.stat3,c("selectTax"),summarise,sd=sd(pct.delta),pct=mean(pct.delta))

ZBR.mean.df1 <- ddply(ddply(ddply(ZBR.qdf%>%filter(idents>99),c("SID","rand","checkLv","selectTax"),summarise,count=sum(count)),
                          c("SID","rand","checkLv"),transform, pct=count/sum(count)*100),
                    c("selectTax","checkLv"),summarise,pct=mean(pct))
ZBR.mean.df1 <- merge(ZBR.mean.df1,fungi.copy.num[,c("genus","TC16S")],by.x="selectTax",by.y="genus",all.x=T)
ZBR.mean.df1$pct.delta <- ZBR.mean.df1$pct-ZBR.mean.df1$TC16S

#ZBR.mean.df <- ZBR.mean.df%>%filter(selectTax%in%selectTax)

ZBR.mean.df$selectTax <- as.factor(as.character(ZBR.mean.df$selectTax))

ZBR.mean.df <- ZBR.mean.df[order(ZBR.mean.df$pct),]
ZBR.mean.df$selectTax <- reorder(ZBR.mean.df$selectTax,ZBR.mean.df$pct,mean)
ggplot(ZBR.mean.df,aes(x=selectTax,y=pct,fill=selectTax)) + geom_hline(yintercept = 0,linetype=2,color="grey50",size=2) +
  geom_bar(data=ZBR.mean.df1,stat = 'identity', position = 'stack') + geom_errorbar(aes(ymax=pct+sd,ymin=pct-sd),width=.4) + coord_flip() +
  ylab("delta of quantification over theoretical value (%)")

selectTax.list <- selectTax
ggplot(ZBR.filter.df.ref%>%filter(selectTax%in%selectTax.list),aes(x=rand,y=pct,fill=selectTax)) + geom_bar(stat = 'identity', position = 'fill')

```

# Find the ZYMO taxonomies in DATABASE
```{bash, eval=FALSE}
for i in {'Pseudomonas aeruginosa','Escherichia coli','Salmonella enterica','Lactobacillus fermentum', 'Enterococcus faecalis','Staphylococcus aureus','Listeria monocytogenes','Bacillus subtilis','Saccharomyces cerevisiae','Cryptococcus neoformans'};do grep $i $STL/Source/REF/silva132/SSU/taxmap_embl_ssu_ref_132.tax;done|les

```

```{r}
fungi.copy.num <- data.frame(species=c("Pseudomonas aeruginosa","Escherichia coli","Salmonella enterica","Lactobacillus fermentum",
  "Enterococcus faecalis","Staphylococcus aureus","Listeria monocytogenes","Bacillus subtilis","Saccharomyces cerevisiae","Cryptococcus neoformans"),
  TC16S=c(4.2,10.1,10.4,18.4,9.9,15.5,14.1,17.4,NA,NA),
  TC16S18S=c(3.6,8.9,9.1,16.1,8.7,13.6,12.4,15.3,9.3,3.3),copyNum=c(4,7,7,5,4,6,6,10,109,60))

ZS.qdf$selectTax <- ifelse(ZS.qdf$species%in%selectTax,as.character(ZS.qdf$species),ifelse(ZS.qdf$genus%in%selectGenus,as.character(ZS.qdf$genus),"Others"))
ZS.stat3 <- ddply(ZS.qdf%>%filter(idents>99.5),c("SID","rand","selectTax"),summarise,count=sum(count))
ZS.stat3 <- merge(ddply(ZS.stat3,c("SID","rand"),transform, pct=count/sum(count)*100),
                  fungi.copy.num[,c("species","TC16S")],by.x="selectTax",by.y="species",all.x=T)
ZS.stat3$pct.delta <- ZS.stat3$pct-ZS.stat3$TC16S
ggplot(ZS.stat3,aes(x=SID,y=pct,fill=factor(rand))) +geom_bar(stat="identity",position="dodge") +facet_grid(.~selectTax)
ggplot(ZS.stat3,aes(x=SID,y=pct,fill=factor(rand))) + geom_bar(aes(y=TC16S),fill=NA,color="grey50",stat="identity",position="dodge") +
  geom_bar(stat="identity",position="dodge",alpha=.5) +facet_grid(.~selectTax)

ZS.mean.df <- ddply(ZS.stat3,c("selectTax"),summarise,sd=sd(pct.delta),pct=mean(pct.delta))
ZS.mean.df <- ZS.mean.df%>%filter(selectTax%in%selectTax)
colnames(ZS.mean.df)[1] <- "selectTax"


ZS.mean.df$selectTax <- as.factor(as.character(ZS.mean.df$selectTax))

ZS.mean.df <- ZS.mean.df[order(ZS.mean.df$pct),]
ZS.mean.df$selectTax <- reorder(ZS.mean.df$selectTax,ZS.mean.df$pct,mean)
ggplot(ZS.mean.df,aes(x=selectTax,y=pct,fill=selectTax)) + geom_hline(yintercept = 0,linetype=2,color="grey50",size=2) +
  geom_bar(stat = 'identity', position = 'dodge') + geom_errorbar(aes(ymax=pct+sd,ymin=pct-sd),width=.4) + coord_flip() +
  ylab("delta of quantification over theoretical value (%)")

selectTax.list <- selectTax
ggplot(ZS.filter.df.ref%>%filter(selectTax%in%selectTax.list),aes(x=rand,y=pct,fill=selectTax)) + geom_bar(stat = 'identity', position = 'fill')


```


```{r}
ZS.qdf$selectTax <- ifelse(ZS.qdf$species%in%selectTax,as.character(ZS.qdf$species),ifelse(ZS.qdf$genus%in%selectGenus,as.character(ZS.qdf$genus),"Others"))
ZS.stat4 <- ddply(ZS.qdf%>%filter(idents>99.5&!is.na(consitLv)),c("SID","rand","consitLv","selectTax"),summarise,count=sum(count))
ZS.stat4 <- merge(ddply(ZS.stat4,c("SID","rand","consitLv"),transform, pct=count/sum(count)*100),
                  fungi.copy.num[,c("species","TC16S")],by.x="selectTax",by.y="species",all.x=T)
ZS.stat4$pct.delta <- ZS.stat4$pct-ZS.stat4$TC16S

ggplot(ZS.stat4,aes(x=selectTax,y=pct,fill=interaction(rand,SID))) +geom_bar(stat="identity",position="dodge") +facet_grid(consitLv~.,scale="free") +
  coord_flip()

ZS.mean.df.4 <- ddply(ZS.stat4,c("selectTax","consitLv"),summarise,sd=sd(pct.delta),pct=mean(pct.delta))
ZS.mean.df.4 <- ZS.mean.df.4%>%filter(selectTax%in%selectTax)

ZS.mean.df.4$selectTax <- as.factor(as.character(ZS.mean.df.4$selectTax))

ZS.mean.df.4 <- ZS.mean.df.4[order(ZS.mean.df.4$pct),]
ZS.mean.df.4$selectTax <- reorder(ZS.mean.df.4$selectTax,ZS.mean.df.4$pct,mean)
ggplot(ZS.mean.df.4,aes(x=selectTax,y=pct,fill=selectTax)) + geom_hline(yintercept = 0,linetype=2,color="grey50",size=2) +
  geom_bar(stat = 'identity', position = 'dodge') + geom_errorbar(aes(ymax=pct+sd,ymin=pct-sd),width=.4) + coord_flip() +
  facet_grid(.~consitLv) + ylab("delta of quantification over theoretical value (%)")

```


```{r}
ZS.filter.df.ref2 <- merge(fungi.copy.num.pair[which(!is.na(fungi.copy.num.pair$count)),c("selectTax","count")],
                          ZS.stat[which(ZS.stat$subunit!="*"),select.col],by=c("selectTax"))
#ZS.filter.df.ref2$pct.STANDARD <- ZS.filter.df.ref2$count.x
ZS.filter.df.ref2 <- ddply(ZS.filter.df.ref2,c("SID","rand","subunit"),transform,pct=count.y/sum(count.y)*100)
ZS.filter.df.ref2$pct.delta <- ZS.filter.df.ref2$pct - ZS.filter.df.ref2$count.x
ggplot(ZS.filter.df.ref2,aes(x=SID,y=pct.delta,fill=selectTax,group=selectTax)) + 
  geom_boxplot()+facet_grid(.~selectTax)

ZS.filter.df.ref2$method <- ifelse(ZS.filter.df.ref2$SID%in%c("Z1","Z2","Z3"),"RCA","LFR")
ZS.filter.df.ref2.sum <- ddply(ZS.filter.df.ref2,c("selectTax","method","subunit"),summarise,delta.mean=mean(pct.delta),pct.se=sd(pct.delta))
ggplot(ZS.filter.df.ref2.sum,aes(x=subunit,y=delta.mean,fill=method)) +geom_bar(stat="identity",position="dodge") + 
  geom_errorbar(aes(ymax=delta.mean+pct.se,ymin=delta.mean-pct.se),width=.2) + 
  theme(axis.text.x = element_text(angle = -45,hjust=0)) + ylab("Delta of percentage (%)") +
  facet_grid(.~selectTax)
```


```{r}
#Z2.anno2 <- merge(Z2.faID,Z2.anno,by="LOTU",all.x=T)
Z2.prop2 <- merge(Z,Z2.anno2,by="LOTU")
Z2.stat2 <- ddply(Z2.prop2,"rTax",summarise,count=sum(count))
top20 <- as.character(Z2.stat2$rTax[rev(order(Z2.stat2$count))[1:13]])
Z2.stat2$tax20 <- as.character(Z2.stat2$rTax)
Z2.stat2$tax20 <- ifelse(Z2.stat2$tax20%in%top20,Z2.stat2$tax20,"Others")
ggplot(Z2.stat2,aes(x="1",y=count,fill=tax20)) + geom_bar(stat = 'identity', position = 'stack')

Z2.filter.df.ref2 <- merge(Z2.stat2,fungi.copy.num,by.x="rTax",by.y="species")
Z2.filter.df.ref2$prop <- Z2.filter.df.ref2$count/sum(Z2.filter.df.ref2$count)*100
Z2.filter.df.ref3 <- melt(Z2.filter.df.ref2,measure.vars = c("prop","TC16S"),variable.name = "group",value.name = "pct")
ggplot(Z2.filter.df.ref3,aes(x=group,y=pct,fill=tax20)) + geom_bar(stat = 'identity', position = 'stack')

```


# FUNGI

```{bash,eval=FALSE}
for i in `cat fungi.lst`;do 
  grep $i $STL/Source/REF/silva132/SSU/taxmap_embl_ssu_ref_132.r7.tax|awk -F "\t" '{print "SSU\t"$1"\t"$2"\t"$3"\t"$4}';
  grep $i $STL/Source/REF/silva132/LSU/taxmap_embl_lsu_ref_132.r7.tax|awk -F "\t" '{print "LSU\t"$1"\t"$2"\t"$3"\t"$4}' ;
  grep $i $LFR/Source/REF/UNITE/sh_general_release_all_04.02.2020/taxmap_taxonkit_merge_ref.r7.tax| awk -F "\t" '{print "ITS\t"$1"\t"$2"\t"$3"\t"$4}';
done > $LFR/Results/FUNGI/mock.tax.compare.in.database.tsv

# TRIM unite database to ITS region
cd $LFR/Source/REF/UNITE/sh_general_release_all_04.02.2020/
split -d -l 10000 sh_general_release_dynamic_all.fasta split/sp.
for i in split/sp.{00..20}; do 
  ITSx -t F --cpu 2 -i $i -o $i.ITSx &> $i.TISx.log &
done

#write a script to trim:
############################
#/usr/env/perl
my %REGION;
open POS,"cat split/sp.??.ITSx.positions.txt|";
while(<POS>){
  my @s = split /\t/;
  $s[1] =~ /(\d+) bp./; my $len = $1;
  my ($i1s,$i1e,$i2s,$i2e) = (0,$len,0,$len);
  if($s[3] =~ /ITS1: (\d+)-(\d+)/){ $i1s = $1; $i1e = $2;}
  if($s[5] =~ /ITS2: (\d+)-(\d+)/){ $i2s = $1; $i2e = $2;}
  #next if $i1s == 1 && $i2e == $len;
  print "$s[1]\n" if $i2s > $len;
  $REGION{$s[0]}{S} = $i1s - 1;
  $REGION{$s[0]}{L} = $i2e - $i1s + 1;
}
close POS;
open FA, "<sh_general_release_dynamic_all.fasta";
$/=">";
while (<FA>){
  chomp; next unless $_;
  my @s = split "\n";
  my $id = shift @s;
  my $str = join ("",@s);
  if(exists $REGION{$id}){
    if(length($str) < $REGION{$id}{L}){
        $str = substr($str,$REGION{$id}{S},$REGION{$id}{L});
    }
    print ">$id\n$str\n";
  }
}
close FA;
exit;
############################
perl onetime.trim.ITSs.pl > UNITE_ITS_ONLY.fasta
makeblastdb -dbtype nucl -in UNITE_ITS_ONLY.fasta

```

#close ref
## map clips and OTUs to zymo ref
```{bash,eval=FALSE}
metabbq smk --force -j -npk SAM/M{4,5,6}/CLIP/LOTU.fa
for i in M{4,5,6};do sed 's/\S\+=//g' SAM/$i/CLIP/LOTU.log|awk -v n=$i '{print n" "$0}' ;done > STAT/CLIP/MFR.LOTU.log
for i in M{4,5,6};do grep "^S" SAM/$i/CLIP/all.clust.uc |sed 's/_/\t/g;s/\S\+=//g'\
 | awk -v n=$i '{print n"\t"$9"\t"$10"\t"$11"\t"$12"\t"$13"\t"$2}';
done > STAT/CLIP/MFR.LOTU.centroids.info

for i in M{4,5,6};do grep "^H" SAM/$i/CLIP/all.clust.uc | awk -v n=$i '{print n"\t"$2"\t"$4"\t"$8"\t"$9"\t"$10}';
done > STAT/CLIP/MFR.LOTU.members.info

#closed mock ref mapping
metabbq smk --configfile config.F.yaml -j -npk SAM/M{4,5,6}/ANNO/clip2MockRef.clip.anno
for i in M{4,5,6};do awk -v n=$i '{print n"\t"$0}' SAM/$i/ANNO/clip2MockRef.clip.anno;done > STAT/ANNO/MFR.clip2MockRef.clip.anno
awk '{if(p!=$2){print $0};p=$2}' STAT/ANNO/clip2MockRef.clip.anno > STAT/ANNO/clip2MockRef.top1.anno
for i in M{4,5,6};do awk -v n=$i '{print n"\t"$0}' SAM/$i/ANNO/clip2MockRef.bead.anno;done > STAT/ANNO/MFR.clip2MockRef.bead.anno

metabbq smk --configfile config.F.yaml -j -npk SAM/M{4,5,6}/ANNO/LOTU.to.MockRef.bead.anno
for i in M{4,5,6};do awk -v n=$i '{print n"\t"$0}' SAM/$i/ANNO/LOTU.to.MockRef.bead.anno;done > STAT/ANNO/MFR.LOTU.to.MockRef.bead.anno

metabbq smk --configfile config.F.yaml -j -npk SAM/M{4,5,6}/ANNO/LOTU.map.merge.anno
for i in M{4,5,6};do sed 's/\S\+=//g' SAM/$i/ANNO/LOTU.map.merge.anno|awk -v n=$i '{print n"\t"$0}' ;done > STAT/ANNO/MFR.LOTU.map.merge.anno

for i in M{4,5,6};do sort -k1,1 -k3,3nr -k4,4nr -k12,12nr SAM/$i/ANNO/LOTU.map.UNITE.m6.more |awk -v n=$i '{if(p!=$1){print n"\t"$0};p=$1}' ;done > STAT/ANNO/MFR.LOTU.map.unite.sort1.anno

metabbq smk --configfile config.F.yaml -j -npk SAM/M{4,5,6}/ANNO/CLIP.map.merge.bead.anno
for i in Z{1,2,3};do awk -v n=$i '{print n"\t"$0}' SAM/$i/ANNO/CLIP.map.merge.bead.anno;done > STAT/ANNO/CLIP.map.merge.bead.anno


#SNV
metabbq smk --configfile config.F.yaml --force -j -npk SAM/M{4,5,6}/CLIP/clips2REF.bcf.tsv
for i in M{4,5,6};do awk -v n=$i '{print n"\t"$0}' SAM/$i/CLIP/clips2REF.bcf.tsv;done > STAT/CLIP/MFR.clips2REF.bcf.tsv


#quantification
metabbq smk --configfile config.F.yaml -j -npk SAM/M{4,5,6}/PROFILE/LOTU.fa.ann
for i in {4,5,6} ;do SAM=SAM/M$i; for j in {00..02};do echo sh tmp/mock.LOTU.quantification.sh $SAM $j \&;done;done > tmp/mock.batch.quantification.sh

mkdir -p STAT/PROFILE
for i in {1,2,3} ;do SAM=Z$i; for j in {00..02};do 
awk -v s=$SAM -v r=$j '{print s"\t"r"\t"$1"\t"$2"\t"$3}' SAM/$SAM/PROFILE/RAND/LOTU.bwa.sbb1_$j.p;
done;done > STAT/PROFILE/ZBR.LOTU.bwa.sbb1_rand3


# merge samples for cluster LOTU
mkdir -p SAM/MFR/CLIP/summary.BI.megahit.clip.all.fasta

```




Load data:
```{r}
metadata <- read.csv("../../Results/JAN20/metadata.csv")

MFR.LOTU.info <- read.table(md5("../../Results/JAN20/STAT/CLIP/MFR.LOTU.log","2c3f0cc2f3fa032c7af1092789a1d0a1"),sep=" ", 
    col.names = c("SID","LOTU","size","S75","CL","MAXLEN","MEANLEN","SBP","P1D","D9P","COR","START","END","PMR"))

MFR.clip.anno <- read.table(md5("../../Results/JAN20/STAT/ANNO/MFR.clip2MockRef.bead.anno","f841aac18724f7049803387641e54e56"),
    sep="\t", col.names = c("SID","BID","BLen","hits","hits.2","tax.clips","tax.pieces","tax",
                 "ident","qMapLen","mismatch","gap","bitScore","bitScore2nd",
                 "sMapPieces","sMapLen","MAPDESC","strands","qMaps","sMaps"
))


MFR.mock.anno <- read.table(md5("../../Results/JAN20/STAT/ANNO/MFR.LOTU.to.MockRef.bead.anno","4c3381a72eca88aa8c4d5ba451488a6c"),
    sep="\t", col.names = c("SID","LOTU","BLen","hits","hits.2","tax.clips","tax.pieces","tax",
                 "ident","qMapLen","mismatch","gap","bitScore","bitScore2nd",
                 "sMapPieces","sMapLen","MAPDESC","strands","qMaps","sMaps"
))

MFR.silva.anno <- read.table(
  md5("../../Results/JAN20/STAT/ANNO/MFR.LOTU.map.merge.anno","b1c2b04880a10fa1a01f261be67cfe68"), sep="\t",comment.char = "",quote = "",
   col.names = c("SID","LOTU","tax","idents","lens","mis","gap","bit","qlen","qcov","scov","scov.ITS","scov.SSU","scov.LSU","taxID","rank","taxPath")
)
MFR.unite.1.anno <- read.table(
  md5("../../Results/JAN20/STAT/ANNO/MFR.LOTU.map.unite.sort1.anno","a3848f957b92d67b2647239db40009c3"), sep="\t",comment.char = "",quote = "",
   col.names = c("SID","LOTU","refID","idents","lens","mis","gap","qs","qe","ss","se","eval","bit","qlen","slen","tax","taxID","rank","taxPath")
)
MFR.clip2REF.bcf <- read.table(md5("../../Results/JAN20/STAT/CLIP/MFR.clips2REF.bcf.tsv",
  "853fd715c32eb7838fd474d049b2ccea"), sep="\t",col.names = c("SID","CHR","POS","REF","ALT","QUAL","TYPE","IDV","DP","DR","DA","BQR","BQA","MQR","MQA","MDR","MDA","BQB","MQSB"))



MFR.prop <- read.table("../../Results/JAN20/STAT/CLIP/MFR.rand3.LOTU.bwa.sbb1.p",sep="\t",
                      col.names = c("SID","rand","count0","count","LOTU"))


```

#Curation
```{r}
MFR.clip.df <- merge(metadata,MFR.clip.anno,by="SID")
MFR.clip.df <- cbind(MFR.clip.df,getMapPropFun(MFR.clip.df$MAPDESC))
MFR.clip.df$mainUnit <- ifelse(MFR.clip.df$SSU>MFR.clip.df$LSU,"SSU","LSU")
MFR.clip.df$unit <- as.factor(ifelse(MFR.clip.df$SSU>50,1,0) + ifelse(MFR.clip.df$LSU>50,2,0) + ifelse(MFR.clip.df$ITS1+MFR.clip.df$ITS2>50,4,0))
levels(MFR.clip.df$unit) <- c("partial(<50%)","SSU","LSU","Both-ITS","ITS","SSU+ITS","LSU+ITS","All")

########
MFR.badf <- merge(metadata,MFR.mock.anno,by="SID")

MFR.badf <- cbind(MFR.badf,getMapPropFun(MFR.badf$MAPDESC))
MFR.badf$mainUnit <- ifelse(MFR.badf$SSU>MFR.badf$LSU,"SSU","LSU")
MFR.badf$unit <- as.factor(ifelse(MFR.badf$SSU>50,1,0) + ifelse(MFR.badf$LSU>50,2,0) + ifelse(MFR.badf$ITS1+MFR.badf$ITS2>50,4,0))
levels(MFR.badf$unit) <- c("partial(<50%)","SSU","LSU","Both-ITS","ITS","SSU+ITS","LSU+ITS","All")
MFR.badfm <- merge(MFR.LOTU.info,MFR.badf,by=c("SID","LOTU"),all.x=T)
```
clip annotation info:
```{r}
n50(MFR.clip.df$BLen)
ggplot(MFR.clip.df,aes(x=BLen,fill=unit)) +geom_histogram(position="stack",binwidth=100) +
  theme(legend.position = c(0.99,.9),legend.justification = c(1,1)) +
  scale_x_continuous(breaks=seq(0,10000,1000),limits=c(0,6000))

ggplot(MFR.clip.df,aes(x=ident,fill=unit)) +geom_histogram(position="stack",binwidth=.5) +
  theme(legend.position = c(.1,.9),legend.justification = c(0,1)) +
  scale_x_continuous(breaks=seq(90,100,1))

ggplot(MFR.clip.df,aes(x=tax,fill=unit)) +geom_bar(position="stack") + coord_flip()
ggplot(MFR.clip.df,aes(x=ident,color=tax)) +geom_density() +
    scale_x_continuous(breaks=seq(90,100,1))


ggplot(MFR.clip.df,aes(x=BLen,y=ident)) + 
  stat_density_2d(aes(fill = stat(nlevel)), geom = "polygon") + 
  scale_fill_gradientn(colours=rev(brewer.pal(11,"Spectral"))) +
  ggtitle("Polygon between BLen and ident") +
  scale_y_continuous(breaks=seq(99,100,0.2)) +
  scale_x_continuous(breaks=seq(0,4000,500))

```

basic:
```{r}
n50(MFR.badfm$CL)
ggplot(MFR.badf,aes(x=qMapLen,fill=SID)) +geom_histogram(position="dodge") + scale_x_log10()
ggplot(MFR.badf,aes(x=ident,fill=SID)) +geom_histogram(position="dodge") + scale_x_log10()

MFR.badfm$S75.2 <- 2^round(log2(MFR.badfm$S75),0)

ggplot(MFR.badfm,aes(x=S75,fill=SID)) +geom_histogram(position="dodge") +
  scale_x_log10() + annotation_logticks(side="b")
ggplot(MFR.badfm,aes(y=ident,x=factor(S75.2),fill=SID)) +geom_boxplot(aes(fill=SID)) +
  scale_y_continuous(breaks=seq(94,102,1),limits=c(94,100))

ggplot(MFR.badfm,aes(x=CL,fill=unit)) +geom_histogram(position="stack",binwidth=50)  + facet_grid(.~SID)
ggplot(MFR.badfm%>%filter(CL>499),aes(x=CL,fill=unit)) +geom_histogram(position="stack",binwidth=100) +
  theme(legend.position = c(0.99,.9),legend.justification = c(1,1))
```


```{r}
MFR.badfm$ident5 <- round(MFR.badfm$ident+0.49,)-0.5
MFR.badfm$ident5 <- ifelse(MFR.badfm$ident5<95,94.5,MFR.badfm$ident5)

ggplot(MFR.badfm,aes(x=ident5)) + geom_bar(aes(fill=unit),stat="count",position="stack") + 
  scale_x_continuous(breaks=seq(94,102,1),limits=c(94,100)) +
  coord_flip() + facet_grid(.~SID)

MFR.badf.stat2 <- ddply(
  ddply(MFR.badfm%>%filter(CL>499),c("SID","ident5"),summarise,count=length(SID)),
  c("ident5"),summarise,mean.count=mean(count),sd=sd(count))
MFR.badf.stat3 <- ddply(
  ddply(MFR.badfm%>%filter(CL>499),c("SID","ident5","unit"),summarise,count=length(SID)),
  c("ident5","unit"),summarise,mean.count=mean(count))

ggplot(MFR.badf.stat3,aes(x=ident5,y=mean.count)) + geom_bar(aes(fill=unit),stat="identity",position="stack") + 
  geom_errorbar(data=MFR.badf.stat2,aes(ymax=mean.count+sd,ymin=mean.count-sd),width=.1) + 
  scale_x_continuous(breaks=seq(94,102,1),limits=c(94,100)) +
  coord_flip()

MFR.badf.stat4 <- ddply(
  ddply(MFR.badfm%>%filter(COR>999),c("SID","ident5"),summarise,count=length(SID)),
  c("ident5"),summarise,mean.count=mean(count),sd=sd(count))
MFR.badf.stat5 <- ddply(
  ddply(MFR.badfm%>%filter(COR>999),c("SID","ident5","unit"),summarise,count=length(SID)),
  c("ident5","unit"),summarise,mean.count=mean(count))

ggplot(MFR.badf.stat5,aes(x=ident5,y=mean.count)) + geom_bar(aes(fill=unit),stat="identity",position="stack") + 
  geom_errorbar(data=MFR.badf.stat4,aes(ymax=mean.count+sd,ymin=mean.count-sd),width=.1) + 
  scale_x_continuous(breaks=seq(94,102,1),limits=c(94,100)) +
  coord_flip()
```



# compare annotation : mock ref vs. silva
```{r}
levels(MFR.badfm$tax) <- gsub("_"," ",levels(MFR.badfm$tax))
levels(MFR.silva.anno$tax) <- gsub("_"," ",levels(MFR.silva.anno$tax))

MFR.compare.anno <- merge(MFR.badfm,MFR.silva.anno,by=c("SID","LOTU"),all.x=T)

MFR.compare.anno$consitLv <- as.factor((ddply(MFR.compare.anno[,c("SID","LOTU","tax.x","tax.y","taxPath")],c("SID","LOTU"),checkTaxConsitRank))$V1)
levels(MFR.compare.anno$consitLv) <- c("different","genus","species")

```


```{r}
MFR.compare.anno$hitDB <- factor(ifelse(MFR.compare.anno$scov.ITS==0,0,1)  + ifelse(MFR.compare.anno$scov.SSU==0,0,2) + 
  ifelse(MFR.compare.anno$scov.LSU==0,0,4),levels=seq(0,7))
levels(MFR.compare.anno$hitDB) <- c("NONE","ITS","SSU","SSU+ITS","LSU","LSU+ITS","SSU+LSU","ALL")

MFR.compare.anno$reachLimit <- as.factor(ifelse(MFR.compare.anno$scov.ITS>99,T,ifelse(MFR.compare.anno$scov.SSU>99,T,ifelse(MFR.compare.anno$scov.LSU>99,T,0))))

ggplot(MFR.compare.anno,aes(x=ident,y=idents,color=consitLv)) + geom_point()
ggplot(MFR.compare.anno,aes(x=idents-ident,color=consitLv)) + geom_density()

ggplot(MFR.compare.anno%>%filter(COR>999&ident>99),aes(x=SID,fill=consitLv)) + geom_bar(position="stack")
ggplot(MFR.compare.anno%>%filter(COR>999&ident>99),aes(x=SID,fill=consitLv)) + geom_bar(position="stack") + facet_grid(.~tax.x)


ggplot(MFR.compare.anno,aes(x=bitScore,y=bit,color=hitDB,shape=reachLimit)) +
  geom_abline(slope = 1,intercept = 0, linetype=2,alpha=.5,size=2,color="grey50") +
  geom_point(alpha=.5) + scale_shape_manual(values=c(16,4))

ggplot(MFR.compare.anno%>%filter(!is.na(consitLv)),aes(x=idents,color=consitLv)) + geom_density(size=1,alpha=.5) +
  theme(legend.position = c(0.9,.98),legend.justification = c(1,1)) + facet_grid(hitDB~.)
ggplot(MFR.compare.anno,aes(x=mainUnit,fill=mainUnit)) + geom_bar() +facet_grid(.~consitLv) 
ggplot(MFR.compare.anno,aes(x=bitScore,y=bit,color=mainUnit,linetype=reachLimit)) + geom_point(alpha=.5) +facet_grid(.~consitLv)


MFR.compare.anno$idents5 <- round(MFR.compare.anno$idents*2,0)/2
ggplot(MFR.compare.anno,aes(x=SID,fill=consitLv)) + geom_bar(position="fill") +
  facet_wrap(~ident5,ncol=5)

MFR.compare.anno$bitG <- round(MFR.compare.anno$bit/1000,0)*1000
ggplot(MFR.compare.anno%>%filter(ident>99),aes(x=SID,fill=consitLv)) + geom_bar(position="fill") +facet_grid(.~bitG)

ggplot(MFR.compare.anno%>%filter(COR>999),aes(x=unit,fill=consitLv)) + geom_bar(position="fill") +
  facet_wrap(~ident5,ncol=5) + coord_flip()

ggplot(MFR.compare.anno%>%filter(COR>999),aes(x=ident,y=idents)) + 
  stat_density_2d(aes(fill = stat(nlevel)), geom = "polygon") + 
  scale_fill_gradientn(colours=rev(brewer.pal(11,"Spectral"))) +
  facet_grid(SID~.,scale="free_x")

ggplot(MFR.compare.anno%>%filter(COR>999),aes(x=SBP,y=idents)) + 
  stat_density_2d(aes(fill = stat(nlevel)), geom = "polygon") + 
  scale_fill_gradientn(colours=rev(brewer.pal(11,"Spectral"))) +
  facet_grid(SID~.,scale="free_x")

ggplot(MFR.compare.anno,aes(x=idents,color=consitLv)) + geom_density()
ggplot(MFR.compare.anno%>%filter(COR>999),aes(x=idents,color=consitLv)) + geom_density()
ggplot(MFR.compare.anno%>%filter(COR>999),aes(x=SBP,color=consitLv)) + geom_density()



ggplot(MFR.compare.anno%>%filter(COR>999),aes(x=idents-ident,color=consitLv)) + geom_density() 

ggplot(MFR.compare.anno,aes(x=qMapLen,y=lens,color=consitLv)) + geom_point()
ggplot(MFR.compare.anno,aes(x=bitScore,y=bit,color=consitLv)) + geom_point(alpha=.5)
ggplot(MFR.compare.anno,aes(x=bitScore,y=bit,color=hitDB,shape=reachLimit)) +
  geom_abline(slope = 1,intercept = 0, linetype=2,alpha=.5,size=2,color="grey50") +
  geom_point(alpha=.5) + scale_shape_manual(values=c(16,4))
ggplot(MFR.compare.anno,aes(x=bitScore,y=bit,color=hitDB,shape=reachLimit)) + geom_point(alpha=.5) +facet_grid(.~consitLv) +
    geom_abline(slope = 1,intercept = 0, linetype=2,alpha=.5,size=2,color="grey50") +
  geom_point(alpha=.5) + scale_shape_manual(values=c(16,4))

ggplot(MFR.compare.anno,aes(x=hitDB,fill=consitLv)) + geom_bar(position="fill")
ggplot(MFR.compare.anno%>%filter(COR>999&ident>99),aes(x=hitDB,fill=consitLv)) + geom_bar(position="stack") +facet_grid(.~SID) + coord_flip()
ggplot(MFR.compare.anno%>%filter(COR>999&ident>99&lens>999),aes(x=hitDB,fill=consitLv)) + geom_bar(position="stack") +facet_grid(.~SID) + coord_flip()
ggplot(MFR.compare.anno%>%filter(COR>999&idents>97),aes(x=tax.x,fill=consitLv)) + geom_bar(position="fill") + facet_grid(hitDB~.) + coord_flip()


ggplot(MFR.compare.anno%>%filter(!is.na(consitLv)),aes(x=idents,color=consitLv)) + geom_density(size=1,alpha=.5) +
  theme(legend.position = c(0.1,.99),legend.justification = c(0,1)) + facet_grid(hitDB~.)
#ggplot(MFR.compare.anno,aes(x=qlen,fill=hitDB)) + geom_bar() +facet_grid(.~consitLv) #+ theme(axis.text.x = element_text(angle=90))
ggplot(MFR.compare.anno,aes(x=bitScore,y=bit,color=mainUnit)) + geom_point(alpha=.5) +facet_grid(.~consitLv)

```

# compare annotation : mock ref vs. unite only annotation results
```{r}
levels(MFR.unite.1.anno$tax) <- gsub("_"," ",levels(MFR.unite.1.anno$tax))

MFR.compare.unite.anno <- merge(MFR.badfm,MFR.unite.1.anno,by=c("SID","LOTU"),all.x=T)

MFR.compare.unite.anno$consitLv <- factor((ddply(MFR.compare.unite.anno[,c("SID","LOTU","tax.x","tax.y","taxPath")],c("SID","LOTU"),checkTaxConsitRank))$V1,
                                          levels=seq(0,3))
levels(MFR.compare.unite.anno$consitLv) <- c("different","genus","species","species")

ggplot(MFR.compare.unite.anno%>%filter(COR>999&ident>99),aes(x=SID,fill=consitLv)) + geom_bar(position="stack")
ggplot(MFR.compare.unite.anno%>%filter(COR>999&ident>99),aes(x=SID,fill=consitLv)) + geom_bar(position="stack") + facet_grid(.~tax.x)

```

completness of subunits
```{r}
MFR.stat.comp <- melt(MFR.compare.anno%>%filter(!is.na(consitLv)),measure.vars = c("ALL","ITS1","SSU","LSU"),variable.name = "subunits", value.name = "completness")
MFR.stat.comp$qlen.seg <- round(MFR.stat.comp$qlen/500,0)*500
ggplot(MFR.stat.comp%>%filter(qlen.seg>400&qlen.seg<3500),aes(x=completness,fill=subunits)) + geom_histogram(position="dodge") +
  facet_grid(qlen.seg~.,scale="free")

```

comapre with mock ref annotation
```{r}
fungi.copy.num <- as.data.frame(table(MFR.qdf$tax.x))
colnames(fungi.copy.num) <- c("species","Tval")
fungi.copy.num$Tval <- 100/7
fungi.copy.num$genus <- sapply(strsplit(as.character(fungi.copy.num$species)," "),"[",1)
selectTax <- as.character(fungi.copy.num$species)
selectGenus <- sapply(strsplit(selectTax," "),"[",1)

MFR.qdf <- merge(MFR.prop,MFR.compare.anno,by=c("SID","LOTU"),all.x=T)
MFR.qdf$species <- MFR.qdf$tax.y
MFR.qdf$genus <- sapply(strsplit(as.character(MFR.qdf$taxPath),";"),"[",6)
MFR.qdf$selectTax <- ifelse(MFR.qdf$species%in%selectTax,as.character(MFR.qdf$species),"Others")
MFR.qdf$checkLv <- ifelse(MFR.qdf$species%in%selectTax,"Same Sp.",ifelse(MFR.qdf$genus%in%selectGenus,"Same Genus","Others.above.Genus"))

MFR.qdf$subunit <- as.factor(ifelse(MFR.qdf$scov.ITS>25,1,0)+ifelse(MFR.qdf$scov.SSU>25,2,0)+ifelse(MFR.qdf$scov.LSU>25,4,0))
MFR.stat <- ddply(MFR.qdf%>%filter(LOTU!='*'),c("SID","rand","tax.x"),summarise,count=sum(count))
MFR.stat <- ddply(MFR.stat,c("SID","rand"),transform,pct=count/sum(count)*100)
MFR.stat1 <- merge(MFR.stat,fungi.copy.num[,c("species","TC16S")],by.x="tax.x",by.y="species",all.x=T)
ggplot(MFR.stat1%>%filter(SID!="Z3"),aes(x=tax.x,y=pct,fill=factor(rand))) + 
  geom_bar(aes(y=TC16S),fill=NA,color="grey50",stat="identity",position="dodge") +
  geom_bar(alpha=.5,stat = 'identity', position = 'dodge') +
  facet_grid(SID~.) + coord_flip()

MFR.stat2 <- ddply(MFR.qdf%>%filter(ident>99.5),c("SID","rand","tax.x"),summarise,count=sum(count))
MFR.stat2 <- ddply(MFR.stat2,c("SID","rand"),transform,pct=count/sum(count)*100)
MFR.stat2 <- merge(MFR.stat2,fungi.copy.num[,c("species","TC16S")],by.x="tax.x",by.y="species",all.x=T)
ggplot(MFR.stat2%>%filter(SID!="Z3"),aes(x=tax.x,y=pct,fill=factor(rand))) + 
  geom_bar(aes(y=TC16S),fill=NA,color="grey50",stat="identity",position="dodge") +
  geom_bar(alpha=.5,stat = 'identity', position = 'dodge') +
  facet_grid(SID~.) + coord_flip()


```
# SNV calling
```{r}
MFR.clip2REF.bcf$DAP <- MFR.clip2REF.bcf$DA/MFR.clip2REF.bcf$DP *100
ggplot(MFR.clip2REF.bcf,aes(x=POS)) +geom_line(aes(y=DP,color=SID)) +
  facet_wrap(~CHR,ncol=2)

ggplot(MFR.clip2REF.bcf%>%filter(TYPE=="INDEL"),aes(x=POS)) +
  geom_bar(aes(y=IDV/DP,fill=SID),position="dodge",stat="identity") +
  facet_wrap(~CHR,ncol=2)

ggplot(MFR.clip2REF.bcf%>%filter(TYPE=="SUB"),aes(x=POS)) +
  geom_bar(aes(y=DAP,fill=SID),position="dodge",stat="identity") +
  facet_wrap(~CHR,ncol=2)

ggplot(MFR.clip2REF.bcf,aes(x=POS)) +geom_ribbon(aes(ymin=0,ymax=DP),alpha=.1,color="grey50") +
  geom_bar(data=MFR.clip2REF.bcf%>%filter(TYPE=="SUB"),aes(y=DA),color="cyan4",position="dodge",stat="identity") +
  geom_bar(data=MFR.clip2REF.bcf%>%filter(TYPE=="INDEL"),aes(y=IDV),color="brown",position="stack",stat="identity",size=1) +
  facet_grid(CHR~SID)

pick.f <- c("Aspergillus_nidulans","Penicillium_chrysogenum")
ggplot(MFR.clip2REF.bcf%>%filter(CHR%in%pick.f&SID=="M4"),aes(x=POS)) +geom_ribbon(aes(ymin=0,ymax=DP),alpha=.1,color="grey50") +
  geom_bar(data=MFR.clip2REF.bcf%>%filter(CHR%in%pick.f&SID=="M4"&TYPE=="SUB"),aes(y=DA),color="cyan4",position="dodge",stat="identity",alpha=.5) +
  geom_bar(data=MFR.clip2REF.bcf%>%filter(CHR%in%pick.f&SID=="M4"&TYPE=="INDEL"),aes(y=IDV),color="brown",position="stack",stat="identity",size=1) +
  facet_grid(CHR~.)
```

```{r}
MFR.qdf$selectTax <- ifelse(MFR.qdf$species%in%selectTax,as.character(MFR.qdf$species),ifelse(MFR.qdf$genus%in%selectGenus,"Others.in.Genus","Others.above.Genus"))

MFR.stat3 <- ddply(MFR.qdf%>%filter(idents>99.5),c("SID","rand","selectTax"),summarise,count=sum(count))
MFR.stat3 <- merge(ddply(MFR.stat3,c("SID","rand"),transform, pct=count/sum(count)*100),
                  fungi.copy.num[,c("species","Tval")],by.x="selectTax",by.y="species",all.x=T)
MFR.stat3$Tval[which(is.na(MFR.stat3$Tval))] <- 0
MFR.stat3$pct.delta <- MFR.stat3$pct-MFR.stat3$Tval
ggplot(MFR.stat3,aes(x=SID,y=pct,fill=factor(rand))) +geom_bar(stat="identity",position="dodge") +facet_grid(.~selectTax)
ggplot(MFR.stat3,aes(x=SID,y=pct,fill=factor(rand))) + geom_bar(aes(y=Tval),fill=NA,color="grey50",stat="identity",position="dodge") +
  geom_bar(stat="identity",position="dodge",alpha=.5) +facet_grid(.~selectTax)

MFR.mean.df <- ddply(MFR.stat3,c("selectTax"),summarise,sd=sd(pct.delta),pct=mean(pct.delta))
MFR.mean.df <- MFR.mean.df%>%filter(selectTax%in%selectTax)
colnames(MFR.mean.df)[1] <- "selectTax"


MFR.mean.df$selectTax <- as.factor(as.character(MFR.mean.df$selectTax))

MFR.mean.df <- MFR.mean.df[order(MFR.mean.df$pct),]
MFR.mean.df$selectTax <- reorder(MFR.mean.df$selectTax,MFR.mean.df$pct,mean)
ggplot(MFR.mean.df,aes(x=selectTax,y=pct,fill=selectTax)) + geom_hline(yintercept = 0,linetype=2,color="grey50",size=2) +
  geom_bar(stat = 'identity', position = 'dodge') + geom_errorbar(aes(ymax=pct+sd,ymin=pct-sd),width=.4) + coord_flip() +
  ylab("delta of quantificatin \nover theoretical value (%)") + guides(fill=F)

```


#FIN.