---
title: "JAN20 DATA ASSESSMENT: annotation part"
author: "Chao"
date: "3/22/2020"
output: 
  html_notebook:
    code_fold: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(reshape2)
library(ggplot2)
library(plyr)
library(dplyr)
library(RColorBrewer)
library(configr)
library(tools)
library(stringi)

md5 <- function(f,m){
  if(tools::md5sum(f)==m){
    warning("MD5 check passed.\n")
    return(f)
  }else{
    stop("MD5 check failed. Please check the version and content!\n")
  }
}
```

# Background

Samples involved:

| ID | lane | barcode | Kindom | RCA | Location    |
| -- | ---- | ------- | -------| --- | ----------- |
| U1 | FF1  | 1,17    | Bac    | RCA | Urine       |
| S1 | FF1  | 3,4     | Bac    | RCA | Soil spot A |
| O1 | FF1  | 5,22    | Bac    | RCA | Soil spot B |
| M1 | FF1  | 7       | Bac    | RCA | Mock        |
| Z1 | FF1  | 8       | Bac    | RCA | ZYMO 8      |

Zymo mock strains:

| Species                  | Genomic DNA | 16S Only1 | 16S & 18S1 | Genome Copy2 | Cell Number3 | Gram | 16/18 Copy Num |
| ------------------------ | ----------- | --------- | ---------- | ------------ | ------------ | ---- | -------------- |
| Pseudomonas aeruginosa   | 12          | 4.2       | 3.6        | 6.1          | 6.1          | -    | 4              |
| Escherichia coli         | 12          | 10.1      | 8.9        | 8.5          | 8.5          | -    | 7              |
| Salmonella enterica      | 12          | 10.4      | 9.1        | 8.7          | 8.8          | -    | 7              |
| Lactobacillus fermentum  | 12          | 18.4      | 16.1       | 21.6         | 21.9         | +    | 5              |
| Enterococcus faecalis    | 12          | 9.9       | 8.7        | 14.6         | 14.6         | +    | 4              |
| Staphylococcus aureus    | 12          | 15.5      | 13.6       | 15.2         | 15.3         | +    | 6              |
| Listeria monocytogenes   | 12          | 14.1      | 12.4       | 13.9         | 13.9         | +    | 6              |
| Bacillus subtilis        | 12          | 17.4      | 15.3       | 10.3         | 10.3         | +    | 10             |
| Saccharomyces cerevisiae | 2           | NA        | 9.3        | 0.57         | 0.29         | Y    | 109            |
| Cryptococcus neoformans  | 2           | NA        | 3.3        | 0.37         | 0.18         | Y    | 60             |

# template to build tree
```{bash, eval=FALSE}
#rRNA predict
barrnap --kingdom bac --threads 4 --reject 0.1 SAM/Z1/CLIP/all.clip.rename.clust.merge.clust.m5.fa \
--outseq SAM/Z1/CLIP/all.clip.rename.clust.merge.clust.m5.barrnap.fa &> SAM/Z1/CLIP/all.clip.rename.clust.merge.clust.m5.barrnap.log
grep -A1 "16S" SAM/Z1/CLIP/all.clip.rename.clust.merge.clust.m5.barrnap > SAM/Z1/CLIP/all.clip.rename.clust.merge.clust.m5.barrnap.16S.fa
vsearch --threads 16 --cluster_fast SAM/Z1/CLIP/all.clip.rename.clust.merge.clust.m5.barrnap.16S.fa --strand both --fasta_width 0 --sizeout \
--centroids SAM/Z1/CLIP/all.clip.rename.clust.merge.clust.m5.barrnap.16S.clust.fa --uc SAM/Z1/CLIP/all.clip.rename.clust.merge.clust.m5.barrnap.16S.clust.uc --id .98
vsearch --sortbysize SAM/Z1/CLIP/all.clip.rename.clust.merge.clust.m5.barrnap.16S.clust.fa --fasta_width 0 --minsize 2 --sizeout --output SAM/Z1/CLIP/all.clip.rename.clust.merge.clust.m5.barrnap.16S.clust.size2.fa
perl -e 'while(<>){if($_=~/^>16S_rRNA\:\:(\d+\_\d+\.\d+\_\d+\_\d+)_BI.*size=(\d+)$/){print ">$1_$2\n"}else{print}}' SAM/Z1/CLIP/all.clip.rename.clust.merge.clust.m5.barrnap.16S.clust.size2.fa \
> SAM/Z1/CLIP/all.clip.rename.clust.merge.clust.m5.barrnap.16S.clust.size2.rename.fa
cat $LFR/Source/REF/zymo/D6305.ssrRNA.fasta SAM/Z1/CLIP/all.clip.rename.clust.merge.clust.m5.barrnap.16S.clust.size2.rename.fa > SAM/Z1/CLIP/all.clip.rename.clust.merge.clust.m5.barrnap.16S.clust.size2.ref.fa
mafft --thread 16 SAM/Z1/CLIP/all.clip.rename.clust.merge.clust.m5.barrnap.16S.clust.size2.ref.fa > SAM/Z1/CLIP/all.clip.rename.clust.merge.clust.m5.barrnap.16S.clust.size2.ref.mafft.fa
sed 's/^.*::/>/;s/:/./g;s/;size=/_/' SAM/Z1/CLIP/all.clip.rename.clust.merge.clust.m5.barrnap.16S.clust.size2.ref.mafft.fa > SAM/Z1/CLIP/all.clip.rename.clust.merge.clust.m5.barrnap.16S.clust.size2.mafft.tmp.fa
trimal -gappyout -in SAM/Z1/CLIP/all.clip.rename.clust.merge.clust.m5.barrnap.16S.clust.size2.mafft.tmp.fa -out SAM/Z1/CLIP/all.clip.rename.clust.merge.clust.m5.barrnap.16S.clust.size2.mafft.trimal.fa

```



# stat assembly clips
```{bash, eval=FALSE}
for i in Z{1,2,3};do 
 awk -v n=$i '{print n"\t"$0}' SAM/$i/summary.BI.megahit.clip.metadata.tsv;
done > STAT/CLIP/ZBR.clip.metadata.tsv

```


# map clips and OTUs to zymo ref
```{bash,eval=FALSE}
mkdir -p ../../Results/JAN20/STAT/ANNO

metabbq smk --force -j -npk SAM/Z{1,2,3}/CLIP/LOTU.fa
for i in Z{1,2,3};do sed 's/\S\+=//g' SAM/$i/CLIP/LOTU.log|awk -v n=$i '{print n" "$0}' ;done > STAT/CLIP/ZBR.LOTU.log
for i in Z{1,2,3};do grep "^S" SAM/$i/CLIP/all.clust.uc |sed 's/_/\t/g;s/\S\+=//g'\
 | awk -v n=$i '{print n"\t"$9"\t"$10"\t"$11"\t"$12"\t"$13"\t"$2}';
done > STAT/CLIP/ZBR.LOTU.centroids.info

for i in Z{1,2,3};do grep "^H" SAM/$i/CLIP/all.clust.uc | awk -v n=$i '{print n"\t"$2"\t"$4"\t"$8"\t"$9"\t"$10}';
done > STAT/CLIP/ZBR.LOTU.members.info

#closed mock ref mapping
metabbq smk --force -j -npk SAM/Z{1,2,3}/ANNO/clip2MockRef.clip.anno
for i in Z{1,2,3};do awk -v n=$i '{print n"\t"$0}' SAM/$i/ANNO/clip2MockRef.clip.anno;done > STAT/ANNO/clip2MockRef.clip.anno
awk '{if(p!=$2){print $0};p=$2}' STAT/ANNO/clip2MockRef.clip.anno > STAT/ANNO/clip2MockRef.top1.anno
for i in Z{1,2,3};do awk -v n=$i '{print n"\t"$0}' SAM/$i/ANNO/clip2MockRef.bead.anno;done > STAT/ANNO/clip2MockRef.bead.anno

metabbq smk --force -j -npk SAM/Z{1,2,3}/ANNO/LOTU.to.MockRef.bead.anno
for i in Z{1,2,3};do awk -v n=$i '{print n"\t"$0}' SAM/$i/ANNO/LOTU.to.MockRef.bead.anno;done > STAT/ANNO/ZBR.LOTU.to.MockRef.bead.anno

#open silva ref mapping
metabbq smk -j -npk SAM/Z{1,2,3}/ANNO/LOTU.map.merge.anno
for i in Z{1,2,3};do sed 's/\S\+=//g' SAM/$i/ANNO/LOTU.map.merge.anno|awk -v n=$i '{print n"\t"$0}' ;done > STAT/ANNO/ZBR.LOTU.map.merge.anno

#SNV
metabbq smk --force -j -pk SAM/Z{1,2,3}/CLIP/clips2REF.bcf.tsv
for i in Z{1,2,3};do awk -v n=$i '{print n"\t"$0}' SAM/$i/CLIP/clips2REF.bcf.tsv;done > STAT/CLIP/ZBR.clips2REF.bcf.tsv


#quantification
metabbq smk -j -npk SAM/Z{2,3}/PROFILE/LOTU.bwa.prop

# write to tmp/mock.LOTU.quantification.sh
SAM=$1
j=$2
mkdir -p $SAM/PROFILE/RAND
bwa mem -t 8 -h 10000 $SAM/CLIP/all.LOTU.fa $SAM/clean/fastp.sort.{1,2}_$j.fq.gz \
| samtools view -b |tee $SAM/PROFILE/RAND/all.LOTU.bwa.map_$j.bam \
| samtools view | metabbq beadStat sam -i - -o $SAM/PROFILE/RAND/LOTU.bwa.stat_$j -v
metabbq beadStat sam2b -i $SAM/PROFILE/RAND/LOTU.bwa.stat_$j -o $SAM/PROFILE/RAND/LOTU.bwa.sbb1_$j -v
#

for i in {1,2,3} ;do SAM=SAM/Z$i; for j in {00..02};do echo sh tmp/mock.LOTU.quantification.sh $SAM $j \&;done;done > tmp/mock.batch.quantification.sh

mkdir -p STAT/PROFILE
for i in {1,2,3} ;do SAM=Z$i; for j in {00..02};do 
awk -v s=$SAM -v r=$j '{print s"\t"r"\t"$1"\t"$2"\t"$3}' SAM/$SAM/PROFILE/RAND/LOTU.bwa.sbb1_$j.p;
done;done > STAT/PROFILE/ZBR.LOTU.bwa.sbb1_rand3


```




Load data:
```{r}
metadata <- read.csv("../../Results/JAN20/metadata.csv")

ZBR.LOTU.info <- read.table(md5("../../Results/JAN20/STAT/CLIP/ZBR.LOTU.log","b10505df928124f0d4d80586f7ce47ac"),sep=" ", 
    col.names = c("SID","LOTU","size","S75","CL","MAXLEN","MEANLEN","SBP","P1D","D9P","COR","PMR"))

ZBR.LOTU.centroids.info <- read.table(md5("../../Results/JAN20/STAT/CLIP/ZBR.LOTU.centroids.info",
  "97a02fd00947dafdaeda5c3c65ba6ddf"), sep="\t",
  col.names = c("SID","BID","kmer","BNO","flag","multi","cluster"))

ZBR.LOTU.members.info <- read.table(md5("../../Results/JAN20/STAT/CLIP/ZBR.LOTU.members.info",
  "b85aa7e1272cc5a741ad4cb4dd28eb57"), sep="\t",
  col.names = c("SID","CID","ident","CIART","member","centroid"))

ZBR.clip2closeRef.top1.anno <- read.table(md5("../../Results/JAN20/STAT/ANNO/clip2MockRef.top1.anno",
                                "325268f87f50515c5ee2d1b62da2dd6e"),
                           sep="\t", col.names =c("SID","CLIP","tax","idents","lens","mis","gap","qs","qe","ss","se","eval","bit","qlen")
)

ZBR.mock.anno <- read.table(md5("../../Results/JAN20/STAT/ANNO/ZBR.LOTU.to.MockRef.bead.anno",
                                "ea648ae5671a5b67ef838f2a87dd2bd5"),
                           sep="\t", col.names = c("SID","LOTU","BLen","hits","hits.2","tax.clips","tax.pieces","tax",
                 "ident","qMapLen","mismatch","gap","bitScore","bitScore2nd",
                 "sMapPieces","sMapLen","MAPDESC","strands","qMaps","sMaps"
))

ZBR.silva.anno <- read.table(md5("../../Results/JAN20/STAT/ANNO/ZBR.LOTU.map.merge.anno",
      "5d079afbd752d1f2de3ee7fb2d2c1e52"), sep="\t",comment.char = "",
   col.names = c("SID","LOTU","tax","idents","lens","mis","gap","bit","qlen","qcov","scov","scov.ITS","scov.SSU","scov.LSU","taxID","rank","taxPath")
)

ZBR.clip2REF.bcf <- read.table(md5("../../Results/JAN20/STAT/CLIP/ZBR.clips2REF.bcf.tsv",
  "844ff306eaa5985685ef6885d32fb5fa"), sep="\t",col.names = c("SID","CHR","POS","REF","ALT","QUAL","TYPE","IDV","DP","DR","DA","BQR","BQA","MQR","MQA","MDR","MDA","BQB","MQSB"))

ZBR.prop <- read.table(md5("../../Results/JAN20/STAT/PROFILE/ZBR.LOTU.bwa.sbb1_rand3","12564411f41155ab37b0a07f92d0c57c"),sep="\t",
                      col.names = c("SID","rand","count0","count","LOTU"))
#

```














#FIN.