---
title: "JAN20 DATA ASSESSMENT: annotation part"
author: "Chao"
date: "3/22/2020"
output: 
  html_notebook:
    code_fold: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(reshape2)
library(ggplot2)
library(plyr)
library(dplyr)
library(RColorBrewer)
library(configr)
library(tools)
library(stringi)

#functions

#https://www.biostars.org/p/9335/
matcher <- function(pattern, x) {

  ind = gregexpr(pattern, x)[[1]]
  start = as.numeric(ind)
  if(length(start)==1 && start <1){
    return(NA)
  }else{
    end = start + attr(ind, "match.length") - 2
    res <- as.numeric(apply(cbind(start,end), 1, function(y) substr(x, start=y[1], stop=y[2])))
    res[which(is.na(res))] <- 1
    return(res)
  }
}

doone <- function(c, cigar) {
  pat <- paste("\\d*", c , sep="")
  sum(as.numeric(matcher(pat, cigar)), na.rm=T)
}


## takes a cigar string and parses it, not very fast but...
cigarsums <- function(cigar, chars=c("M","N","D","I","S","H", "P", "X", "=")) {
   sapply (chars, doone, cigar)
}


n50 <- function(x){
  x1 <- rev(sort(x))
  cumsum.x1 <- cumsum(x1)
  find.n50 <- which(cumsum.x1 >= sum(x1)/2)
  return(x1[find.n50[1]])
}

md5 <- function(f,m){
  if(tools::md5sum(f)==m){
    warning("MD5 check passed.\n")
    return(f)
  }else{
    stop("MD5 check failed. Please check the version and content!\n")
  }
}

getMapPropFun <- function(d){
  uOrder <- c("ALL","SSU","ITS1","5.8S","ITS2","LSU")
  ulist <- data.frame(key=c("ALL","SSU","16S","ITS","ITS1","5.8S","ITS2","LSU","23S"),
                      ind=c(1,2,2,3,3,4,5,6,6))
  getM <- stri_match_all_regex(d,"(ALL|23S|LSU|ITS|ITS1|ITS2|5.8S|SSU|16S)\\((\\d+)%\\)")
  getN <- length(getM)
  res <- matrix(0,ncol = 6,nrow = getN)
  for(i in 1:length(getM)){
    geti <- getM[[i]]
    for(j in 1:nrow(geti)){
      res[i,ulist$ind[which(ulist$key==geti[j,2])]] <- geti[j,3]
    }
  }
  res <- matrix(as.numeric(res),ncol=6)
  colnames(res) <- uOrder
  return(res)
}
```

# Background

Samples involved:

| ID | lane | barcode | Kindom | RCA | Location    |
| -- | ---- | ------- | -------| --- | ----------- |
| U1 | FF1  | 1,17    | Bac    | RCA | Urine       |
| S1 | FF1  | 3,4     | Bac    | RCA | Soil spot A |
| O1 | FF1  | 5,22    | Bac    | RCA | Soil spot B |
| M1 | FF1  | 7       | Bac    | RCA | Mock        |
| Z1 | FF1  | 8       | Bac    | RCA | ZYMO 8      |

Zymo mock strains:

| Species                  | Genomic DNA | 16S Only1 | 16S & 18S1 | Genome Copy2 | Cell Number3 | Gram | 16/18 Copy Num |
| ------------------------ | ----------- | --------- | ---------- | ------------ | ------------ | ---- | -------------- |
| Pseudomonas aeruginosa   | 12          | 4.2       | 3.6        | 6.1          | 6.1          | -    | 4              |
| Escherichia coli         | 12          | 10.1      | 8.9        | 8.5          | 8.5          | -    | 7              |
| Salmonella enterica      | 12          | 10.4      | 9.1        | 8.7          | 8.8          | -    | 7              |
| Lactobacillus fermentum  | 12          | 18.4      | 16.1       | 21.6         | 21.9         | +    | 5              |
| Enterococcus faecalis    | 12          | 9.9       | 8.7        | 14.6         | 14.6         | +    | 4              |
| Staphylococcus aureus    | 12          | 15.5      | 13.6       | 15.2         | 15.3         | +    | 6              |
| Listeria monocytogenes   | 12          | 14.1      | 12.4       | 13.9         | 13.9         | +    | 6              |
| Bacillus subtilis        | 12          | 17.4      | 15.3       | 10.3         | 10.3         | +    | 10             |
| Saccharomyces cerevisiae | 2           | NA        | 9.3        | 0.57         | 0.29         | Y    | 109            |
| Cryptococcus neoformans  | 2           | NA        | 3.3        | 0.37         | 0.18         | Y    | 60             |

# stat assembly clips
```{bash, eval=FALSE}
for i in {S,O}{1,2,3};do 
 awk -v n=$i '{print n"\t"$0}' SAM/$i/summary.BI.megahit.clip.metadata.tsv;
done > STAT/CLIP/SBR.clip.metadata.tsv
md5sum STAT/CLIP/SBR.bead.metadata.tsv
awk '!/#skip/{if($2!=pB){if(FNR>1){print $1"\t"pB"\t"blen"\t"ctg"\t"clips};blen=$10;ctg=1;clips=1}else{blen=blen+$10;clips=clips+1;if($4!=pc){ctg=ctg+1}};pc=$4;pB=$2}END{print $1"\t"pB"\t"blen"\t"ctg"\t"clips}' STAT/CLIP/SBR.clip.metadata.tsv > STAT/CLIP/SBR.bead.metadata.tsv
md5sum STAT/CLIP/SBR.bead.metadata.tsv
```


# map clips and OTUs to zymo ref
```{bash,eval=FALSE}
#mkdir -p ../../Results/JAN20/STAT/ANNO

# Subunits detection
metabbq smk -j -npk SAM/{S,O}{1,2,3}/CLIP/all.barrnap.sum.tsv
for i in {S,O}{1,2,3};do awk -v n=$i '{print n"\t"$0}' SAM/$i/CLIP/all.barrnap.sum.tsv;done > STAT/CLIP/SBR.all.barrnap.sum.tsv


#open silva ref mapping
metabbq smk -j -npk SAM/{S,O}{1,2,3}/ANNO/CLIP.map.merge.bead.anno
for i in {S,O}{1,2,3};do awk -v n=$i '{print n"\t"$0}' SAM/$i/ANNO/CLIP.map.merge.bead.anno;done > STAT/ANNO/SBR.CLIP.map.merge.bead.anno

metabbq smk -j -npk SAM/{S,O}{1,2,3}/ANNO/LOTU.map.merge.anno
for i in {S,O}{1,2,3};do sed 's/\S\+=//g' SAM/$i/ANNO/LOTU.map.merge.anno|awk -v n=$i '{print n"\t"$0}' ;done > STAT/ANNO/SBR.LOTU.map.merge.anno


#SNV
metabbq smk --force -j -pk SAM/{S,O}{1,2,3}/CLIP/clips2REF.bcf.tsv
for i in {S,O}{1,2,3};do awk -v n=$i '{print n"\t"$0}' SAM/$i/CLIP/clips2REF.bcf.tsv;done > STAT/CLIP/SBR.clips2REF.bcf.tsv

#diversity
for i in {S,O}{1,2,3};do metabbq diversityEst -e 0 -s $i -t 100 -d SAM/$i -m SAM/SBR -o SAM/$i/CLIP/divEst.tsv & done
for i in {S,O}{1,2,3};do awk -v n=$i '{print n"\t"$0}' SAM/$i/CLIP/divEst.tsv;done > STAT/CLIP/SBR.divEst.tsv

```




Load data:
```{r}
metadata <- read.csv("../../Results/JAN20/metadata.csv")
SBR.bead.metadata.df <- read.table(
  md5("../../Results/JAN20/STAT/CLIP/SBR.bead.metadata.tsv","8dfae78865bbefa3bbb3be0f0806fff1"),
  sep="\t",col.names = c("SID","BID","BLen","contigs","clips")
)
levels(SBR.bead.metadata.df$SID)[which(levels(SBR.bead.metadata.df$SID)%in%c("O2","S2"))] <- c("S2","O2")

SBR.barrnap.gff3.tsv <- read.table(
  md5("../../Results/JAN20/STAT/CLIP/SBR.all.barrnap.sum.tsv","6b25b249c1a02f26039acc6c6a4be062"),
  col.names = c("SID","BID","len16S","pct16S","len18S","pct18S","len5.8S","pct5.8S","len23S","pct23S","len28S","pct28S","unit")
)
levels(SBR.barrnap.gff3.tsv$SID)[which(levels(SBR.barrnap.gff3.tsv$SID)%in%c("O2","S2"))] <- c("S2","O2")

SBR.silva.bead.anno <- read.table(
  md5("../../Results/JAN20/STAT/ANNO/SBR.CLIP.map.merge.bead.anno","268c21c818055b51b3a51f5396961056"),comment.char = "",sep="\t", quote = "", 
  fill = NA,col.names = c("SID","BINFO","tax","ident","mapLen","mismatch","gap","qs","qe","ss","se","eval","bitscore","qlen","slen","strands","qDesc","sDesc","covs","TopAnno","TANum","TALv","score1st","score2nd","ident1st","ident2nd","TANames","hybCt","hybLen","hybPct")
)
levels(SBR.silva.bead.anno$SID)[which(levels(SBR.silva.bead.anno$SID)%in%c("O2","S2"))] <- c("S2","O2")

# SBR.silva.LOTU.anno <- read.table(md5("../../Results/JAN20/STAT/ANNO/SBR.LOTU.map.merge.anno",
#       "5d079afbd752d1f2de3ee7fb2d2c1e52"), sep="\t",comment.char = "",
#    col.names = c("SID","LOTU","tax","idents","lens","mis","gap","bit","qlen","qcov","scov","scov.ITS","scov.SSU","scov.LSU","taxID","rank","taxPath")
# )
# 
SBR.clip2REF.bcf <- read.table(md5("../../Results/JAN20/STAT/CLIP/SBR.clips2REF.bcf.tsv",
  "844ff306eaa5985685ef6885d32fb5fa"), sep="\t",col.names = c("SID","CHR","POS","REF","ALT","QUAL","TYPE","IDV","DP","DR","DA","BQR","BQA","MQR","MQA","MDR","MDA","BQB","MQSB"))
# 
# SBR.prop <- read.table(md5("../../Results/JAN20/STAT/PROFILE/SBR.LOTU.bwa.sbb1_rand3","12564411f41155ab37b0a07f92d0c57c"),sep="\t",
#                       col.names = c("SID","rand","count0","count","LOTU"))
# #
```


# bead info
```{r}
SBR.bdmeta.df <- merge(metadata,SBR.bead.metadata.df,by="SID")
ggplot(SBR.bdmeta.df,aes(x=BLen,color=SID)) + geom_density() + xlim(c(0,6000))
```

```{r}
ggplot(SBR.bdmeta.df,aes(x=BLen,y=clips)) +
  stat_density_2d(aes(fill = stat(nlevel)), geom = "polygon") +
  scale_fill_gradientn(colours=rev(brewer.pal(11,"Spectral"))) +
  facet_wrap(~SID,ncol=3)
```

# subunits detection
```{r}
SBR.bac.subunit.df <- SBR.barrnap.gff3.tsv[,c(1:4,7:10,13)]
SBR.bac.subunit.df$unit <- factor(SBR.bac.subunit.df$unit,levels = seq(0,7))
SBR.subunits.df <- merge(SBR.bdmeta.df,SBR.bac.subunit.df,by=c("SID","BID"),all.x=T)
SBR.subunits.df$unit[which(is.na(SBR.subunits.df$unit))] <- 0
levels(SBR.subunits.df$unit) <- c("None","SSU","5S","SSU+5S","LSU","SSU+LSU","LSU+5S","ALL")
SBR.subunits.df$SAM <- substr(SBR.subunits.df$SID,1,1)

ggplot(SBR.subunits.df,aes(x=BLen)) +geom_histogram(aes(fill=unit),position="stack",binwidth=100) +
  geom_density(aes(x=BLen,y=..density..*(-4e6),color=unit,linetype=SAM),size=1,alpha=.7) +
  geom_vline(xintercept = 900,linetype=2,size=1) +
  scale_fill_manual(values=c("grey50",brewer.pal(11,"Spectral")[c(1,5,9,10,11)])) +
  scale_color_manual(values=c("grey50",brewer.pal(11,"Spectral")[c(1,5,9,10,11)])) + theme_bw() +
  #theme(legend.position = c(0.99,.9),legend.justification = c(1,1)) +
  scale_x_continuous(breaks=seq(0,10000,1000),limits=c(-1,6000))

pct.unit <- ddply(SBR.subunits.df,"unit",summarise,count=length(BID))
pct.unit$pct <- pct.unit$count/sum(pct.unit$count)
pct.unit$lab <- sprintf("%.2f%%",100*pct.unit$pct)
ggplot(pct.unit,aes(x="X",y=pct,fill=unit)) +geom_bar(stat="identity") + coord_polar(theta = 'y')+ 
  geom_text(aes(y=cumsum(rev(pct)) - rev(pct)/2, x = sum(pct)*1.5, label = rev(lab))) +
  scale_fill_manual(values=c("grey50",brewer.pal(11,"Spectral")[c(1,2,3,8,9,10,7,11)])) +
  labs(x = '', y = '', title = '') +
  theme_classic() + theme(axis.text = element_blank())

ggplot(SBR.subunits.df%>%filter(len16S>0),aes(x=BLen,y=pct16S)) +
  stat_density_2d(aes(fill = stat(nlevel)), geom = "polygon") + 
  scale_fill_gradientn(colours=rev(brewer.pal(11,"Spectral"))) + 
  geom_abline(intercept = 0, slope = 1,linetype=2) +
  facet_wrap(~SAM,ncol=2) + ggtitle("Polygon between 16S coverage and bead len")

ggplot(SBR.subunits.df%>%filter(len23S>0),aes(x=BLen,y=pct23S)) +
  stat_density_2d(aes(fill = stat(nlevel)), geom = "polygon") + 
  scale_fill_gradientn(colours=rev(brewer.pal(11,"Spectral"))) + 
  geom_abline(intercept = 0, slope = 1,linetype=2) +
  facet_wrap(~SAM,ncol=2) + ggtitle("Polygon between 23S coverage and bead len")
```

# silva mapping annotation
```{r}

getMapPropFun2 <- function(d){
  uOrder <- c("ALL","SSU","ITS1","5.8S","ITS2","LSU")
  ulist <- data.frame(key=c("ALL","SSU","16S","ITS","ITS1","5.8S","ITS2","LSU","23S"),
                      ind=c(1,2,2,3,3,4,5,6,6))
  getM <- stri_match_all_regex(d,"(ALL|23S|LSU|ITS|ITS1|ITS2|5.8S|SSU|16S)\\((\\d+)\\)")
  getN <- length(getM)
  res <- matrix(0,ncol = 6,nrow = getN)
  for(i in 1:length(getM)){
    geti <- getM[[i]]
    for(j in 1:nrow(geti)){
      res[i,ulist$ind[which(ulist$key==geti[j,2])]] <- geti[j,3]
    }
  }
  res <- matrix(as.numeric(res),ncol=6)
  colnames(res) <- uOrder
  return(res)
}
sbinfo <- matrix(unlist(strsplit(as.character(SBR.silva.bead.anno$BINFO),"|",fixed = T)),ncol=4,byrow = T)
sbinfodf <- data.frame(BID=sbinfo[,1],clips=as.numeric(sbinfo[,3]),hits=as.numeric(sbinfo[,4]))
covdf <- getMapPropFun2(as.character(SBR.silva.bead.anno$covs))

SBR.slv.bead.df <- merge(SBR.bdmeta.df,cbind(sbinfodf[,-2],SBR.silva.bead.anno[,-c(2,16:19)],covdf[,c(2,6)]),by=c("SID","BID"),all.x=T)
SBR.slv.bead.df$unit <- factor(ifelse(SBR.slv.bead.df$SSU>0,1,0) + ifelse(SBR.slv.bead.df$LSU>0,2,0), levels=c(0,2,1,3))
SBR.slv.bead.df$unit[which(is.na(SBR.slv.bead.df$unit))] <- 0
levels(SBR.slv.bead.df$unit) <- c("Missing","LSU","SSU","BOTH")
```

```{r}
n50(SBR.slv.bead.df$qlen)

SBR.slv.bead.df$identG <- ifelse(SBR.slv.bead.df$ident == 100,100,ifelse(SBR.slv.bead.df$ident >=99.5,99.5,
       ifelse(SBR.slv.bead.df$ident <97,96,round(SBR.slv.bead.df$ident-0.5,0))))

SBR.slv.bead.df$identG[which(is.na(SBR.slv.bead.df$identG))] <- 0

pct.identG <- ddply(SBR.slv.bead.df,"identG",summarise,count=length(BID))
pct.identG$pct <- pct.identG$count/sum(pct.identG$count)
pct.identG$lab <- sprintf("%.2f%%",100*pct.identG$pct)
pct.identG$identG <- factor(pct.identG$identG,levels=c(0,96,97,98,99,99.5,100))
ggplot(pct.identG,aes(x="X",y=pct,fill=identG)) +geom_bar(stat="identity") + coord_polar(theta = 'y')+ 
  geom_text(aes(y=cumsum(rev(pct)) - rev(pct)/2, x = sum(pct)*1.55, label = rev(lab))) +
  scale_fill_brewer(palette="Blues") +labs(x = '', y = '', title = '') +
  theme_classic() + theme(axis.text = element_blank())


ggplot(SBR.slv.bead.df,aes(x=qlen)) +geom_histogram(aes(fill=unit),position="stack",binwidth=100) +
  geom_density(data=SBR.slv.bead.df%>%filter(unit!="partial(<50%)"),
               aes(x=qlen,y=..density..*(-2e6),color=unit,linetype=SID),size=1) +
  scale_fill_manual(values=c("grey50",brewer.pal(11,"Spectral")[c(3,8,11)])) +
  scale_color_manual(values=brewer.pal(11,"Spectral")[c(3,8,11)]) + theme_bw() +
  #theme(legend.position = c(0.99,.9),legend.justification = c(1,1)) +
  scale_x_continuous(breaks=seq(0,10000,1000),limits=c(0,6000))


pct.unit <- ddply(SBR.slv.bead.df,c("SID","unit"),summarise,count=length(BID))
pct.unit <- ddply(pct.unit,c("SID"),transform,pct=count/sum(count))
pct.unit$lab <- sprintf("%.2f%%",100*pct.unit$pct)
ggplot(pct.unit,aes(x="X",y=pct,fill=unit)) +geom_bar(stat="identity") + coord_polar(theta = 'y')+ 
  #geom_text(aes(y=cumsum(rev(pct)) - rev(pct)/2, x = sum(pct)*.3, label = rev(lab))) +
  scale_fill_manual(values=c("grey50",brewer.pal(11,"Spectral")[c(1,3,8,10,11)])) +
  labs(x = '', y = '', title = '') + facet_wrap(~SID,ncol=3) +
  theme_bw() + theme(axis.text = element_blank())

```


```{r}
combine.unit.df <- merge(SBR.subunits.df,SBR.slv.bead.df[,c(1,2,35,36)],by=c("SID","BID"))
combine.unit.df$len16S[which(is.na(combine.unit.df$len16S))] <- 0
combine.unit.df$len23S[which(is.na(combine.unit.df$len23S))] <- 0
combine.unit.df$SSU[which(is.na(combine.unit.df$SSU))] <- 0
combine.unit.df$LSU[which(is.na(combine.unit.df$LSU))] <- 0
combine.unit.df$foundSSU <- ifelse(combine.unit.df$len16S>0,1,0) + ifelse(combine.unit.df$SSU>0,2,0)
#combine.unit.df$foundSSU[which(is.na(combine.unit.df$foundSSU))] <- 0
combine.unit.df$foundLSU <- ifelse(combine.unit.df$len23S>0,1,0) + ifelse(combine.unit.df$LSU>0,2,0)
#combine.unit.df$foundLSU[which(is.na(combine.unit.df$foundLSU))] <- 0
combine.unit.df$foundUNIT <- as.factor(
  ifelse(combine.unit.df$len16S>0,1,0) + 
  ifelse(combine.unit.df$SSU>0,2,0) + 
  ifelse(combine.unit.df$len23S>0,4,0) + 
  ifelse(combine.unit.df$LSU>0,8,0)
)
levels(combine.unit.df$foundUNIT) <- c("Unknown","SSU(predicted)","SSU(aligned)","SSU(cross validated)",
                                       "LSU(predicted)","Both(predicted)","Both(a&p)","Both(c&p)",
                                       "LSU(aligned)","Both(p&a)","Both(aligned)","Both(c&a)","LSU(cross validated)",
                                       "Both(c&p)","Both(c&a)","Both(cross validated")
levels(combine.unit.df$foundUNIT) <- c("Unknown","SSU","SSU","SSU",
                                       "LSU","Both","Both","Both",
                                       "LSU","Both","Both","Both","LSU",
                                       "Both","Both","Both")
ggplot(combine.unit.df,aes(x=BLen)) +geom_histogram(aes(fill=foundUNIT),position="stack",binwidth=100) +
  geom_density(aes(x=BLen,y=..density..*(-6e6),color=SID),size=1) +
  scale_fill_manual(values=c("grey50",brewer.pal(11,"Spectral")[c(1,3,5,7,9,11,1,3,5,7,9,11,1)])) +
  scale_color_manual(values=c("grey50",brewer.pal(11,"Spectral")[c(1,3,5,7,9,11,1,3,5,7,9,11,1)])) + theme_bw() +
  #theme(legend.position = c(0.99,.9),legend.justification = c(1,1)) +
  scale_x_continuous(breaks=seq(0,10000,1000),limits=c(0,5000))


```

```{r}
combine.unit.df$foundSSU <- as.factor(combine.unit.df$foundSSU)
levels(combine.unit.df$foundSSU) <- c("Unknown","predicted by barrnap","predicted by ITSx","cross validated","aligned to reference","cross validated","cross validated","cross validated")
source.SSU <- ddply(combine.unit.df,"foundSSU",summarise,count=length(BID))
source.SSU$pct <- source.SSU$count/sum(source.SSU$count)
source.SSU$lab <- sprintf("%.2f%%",100*source.SSU$pct)

combine.unit.df$foundLSU <- as.factor(combine.unit.df$foundLSU)
levels(combine.unit.df$foundLSU) <- c("Unknown","predicted by barrnap","predicted by ITSx","cross validated","aligned to reference","cross validated","cross validated","cross validated")
source.LSU <- ddply(combine.unit.df,"foundLSU",summarise,count=length(BID))
source.LSU$pct <- source.LSU$count/sum(source.LSU$count)
source.LSU$lab <- sprintf("%.2f%%",100*source.LSU$pct)

ggplot(source.SSU,aes(x="X",y=pct,fill=foundSSU)) +geom_bar(stat="identity") + coord_polar(theta = 'y')+ 
  geom_text(aes(y=cumsum(rev(pct)) - rev(pct)/2, x = sum(pct)*1.55, label = rev(lab))) +
  scale_fill_manual(values=c("grey50",brewer.pal(11,"Spectral")[c(1,2,3,5,7,9,11)])) +
  labs(x = '', y = '', title = '') +
  theme_classic() + theme(axis.text = element_blank())

ggplot(source.LSU,aes(x="X",y=pct,fill=foundLSU)) +geom_bar(stat="identity") + coord_polar(theta = 'y')+ 
  geom_text(aes(y=cumsum(rev(pct)) - rev(pct)/2, x = sum(pct)*1.55, label = rev(lab))) +
  scale_fill_manual(values=c("grey50",brewer.pal(11,"Spectral")[c(1,2,3,5,7,9,11)])) +
  labs(x = '', y = '', title = '') +
  theme_classic() + theme(axis.text = element_blank())
```

# Assess TopAnno
```{r}
tmp.bead.df <- SBR.slv.bead.df%>%filter(!is.na(TopAnno))
tmp.bead.df <- ddply(tmp.bead.df,c("SID","TopAnno"),transform,ATCt=length(BID))
tmp.bead.df <- ddply(tmp.bead.df,c("SID"),transform,ATWeigth=ATCt/length(BID))
ggplot(SBR.slv.bead.df,aes(x=qlen,fill=TopAnno)) + geom_histogram(position="stack") + xlim(c(0,4000))
ggplot(SBR.slv.bead.df,aes(x=qlen,fill=TopAnno)) + geom_histogram(position="fill") + xlim(c(0,4000))

ggplot(tmp.bead.df,aes(x=BLen,color=TopAnno)) + geom_density(aes(y=..density..)) + xlim(c(0,4000))

ggplot(SBR.slv.bead.df,aes(x=qlen,y=ident)) + 
  stat_density_2d(aes(fill = stat(nlevel)), geom = "polygon") + 
  scale_fill_gradientn(colours=rev(brewer.pal(11,"Spectral"))) +
  ggtitle("Polygon between qlen and ident") +
  facet_grid(TALv~TopAnno,scale="free")
```


# Quantification

cluster same samples' clips to get LOTU:
```{bash}
# merge LOTUs
mkdir -p SAM/SBR/ANNO
for i in {S,O}{1,2,3};do 
  awk -v n=$i '/>BI/{gsub(">",">"n)}{print}' SAM/$i/summary.BI.megahit.clip.all.fasta;
done > SAM/SBR/summary.BI.megahit.clip.all.fasta
for i in {S,O}{1,2,3};do 
  awk -v n=$i '{print n$0}' SAM/$i/ANNO/CLIP.map.merge.bead.anno;
done > SAM/SBR/ANNO/CLIP.map.merge.bead.anno

# add anno
metabbq IO clade2tree -i SAM/SBR/CLIP/id90def4.clust.fa -a SAM/SBR/ANNO/CLIP.map.merge.bead.anno \
 -d $LFR/Source/REF/KRAKEN2/rDNA -s SAM/SBR/CLIP/id90def4.clade.uc -o SAM/SBR/KRAKEN/BEAD.ANNO -v

#tmp change
rename kreport2 old.kreoprt2 SAM/{S,O}{1,2,3}/KRAKEN/reads2BEAD.kreport2

metabbq smk -j -npk SAM/SBR/KRAKEN/db/hash.k2d

awk -F "\t" '{split($1,a,";");print a[1]"\t"$2}' SAM/SBR/KRAKEN/db/taxonomy/paths.txt > SAM/SBR/KRAKEN/db/data/topRank.map

#if use bracken
bracken-build -d SAM/SBR/KRAKEN/db -t 8

####
# mapping to merged db
for i in {S,O}{1,2,3,7,8,9};do
mkdir -p SAM/$i/KRAKEN
ln -s ../../SBR/KRAKEN/db SAM/$i/KRAKEN/SBR;
done

#use bracken
metabbq smk --config kk2_db="SBR" -j -npk SAM/{S,O}{1,2,3,7,8,9}/KRAKEN/reads2SBR.bracken
for i in {S,O}{1,2,3,7,8,9};do 
  awk -v n=$i 'FNR>1{print n"\t"$0}' SAM/$i/KRAKEN/reads2SBR.k2.1.bracken;
done > STAT/KRAKEN/SBR.reads2SBR.bracken

for i in {S,O}{1,2,3,7,8,9};do for j in {G,P};do
  bracken -d SAM/$i/KRAKEN/SBR -i SAM/$i/KRAKEN/reads2SBR.k2.report -l $j -o SAM/$i/KRAKEN/reads2SBR.$j.bracken -w SAM/$i/KRAKEN/reads2SBR.$j.breport2 ;
done & done
for j in {G,P};do for i in {S,O}{1,2,3,7,8,9};do 
  awk -v n=$i 'FNR>1{print n"\t"$0}' SAM/$i/KRAKEN/reads2SBR.$j.bracken
done > STAT/KRAKEN/SBR.reads2SBR.$j.bracken;done

for i in {S,O}{1,2,3,7,8,9};do 
  awk -v n=$i '{print n"\t"$5}' SAM/$i/KRAKEN/reads2SBR.k2.bead;
done|uniq -c > STAT/KRAKEN/SBR.reads2SBR.bead.stat

# custom db mapping

# count dbs' species counts
for i in {SILVA138SSU,SILVA132LSU,UNITE,rDNA};do
awk -F '\t' '{print "1\t"$2"\t"$3"\t"$1"\t"$1}' $LFR/Source/REF/KRAKEN2/$i/taxonomy/paths.txt \
| metabbq IO sumprf -r species -i - -o $LFR/Source/REF/KRAKEN2/$i/data/all.species.count
awk -F '\t' '/Bacteria;/{print "1\t"$2"\t"$3"\t"$1"\t"$1}' $LFR/Source/REF/KRAKEN2/$i/taxonomy/paths.txt \
| metabbq IO sumprf -r species -i - -o $LFR/Source/REF/KRAKEN2/$i/data/bac.species.count
awk -F '\t' '/Eukaryota;/{print "1\t"$2"\t"$3"\t"$1"\t"$1}' $LFR/Source/REF/KRAKEN2/$i/taxonomy/paths.txt \
| metabbq IO sumprf -r species -i - -o $LFR/Source/REF/KRAKEN2/$i/data/euk.species.count
awk -F '\t' '/Fungi;/{print "1\t"$2"\t"$3"\t"$1"\t"$1}' $LFR/Source/REF/KRAKEN2/$i/taxonomy/paths.txt \
| metabbq IO sumprf -r species -i - -o $LFR/Source/REF/KRAKEN2/$i/data/fgi.species.count
done
#export
for i in {SILVA138SSU,SILVA132LSU,UNITE,rDNA};do for rank in {all,bac,euk,fgi};do
awk -v db=$i -v rk=$rank 'END{print db"\t"rk"\t"FNR}' $LFR/Source/REF/KRAKEN2/$i/data/$rank.species.count;
done;done > STAT/KRAKEN/db.tax.count

# mapping to above dbs
for db in {SILVA138SSU,SILVA132LSU,UNITE,rDNA};do
 for s in {S,O}{1..6};do
 ln -sf $LFR/Source/REF/KRAKEN2/$db SAM/$s/KRAKEN/
 done;
done

for db in {SBR,SBR,SILVA138SSU,SILVA132LSU,UNITE,rDNA};do
 metabbq smk --config kk2_db=$db -j -npk SAM/{S,O}{1..6}/KRAKEN/reads2$db.out
done

for i in {S,O}{1..6};do 
  line=`wc -l SAM/$i/KRAKEN/reads2UNITE.out|cut -d " " -f1`;
  mkdir -p SAM/$i/KRAKEN/rand;
  for db in {SILVA138SSU,SILVA132LSU,UNITE,rDNA,SBR,SBR};do
    file="SAM/$i/KRAKEN/reads2$db.out";
    if [ -e $file ];then
      time=`expr $line / 10000000`;
      t=1; while(( $t <= $time ));do
        num=`expr $t \* 10000000`;
        
        for r in {0..9};do
          seed=`expr $line + $t + $r`;
          echo metabbq randomlyPickLines.pl -s $seed -n $num -m $line -v -i $file \| metabbq IO kk2prf -d SAM/$i/KRAKEN/$db -i - -o SAM/$i/KRAKEN/rand/$db.$num.$r.prof;
        done
        let "t++";
      done;
    fi
  done;
done > tmp/batch.stepKK2.prof.sh

for i in {S,O}{1..6};do for j in SAM/$i/KRAKEN/rand/*.prof;
do echo  metabbq IO sumprf -r species -i $j -o ${j/prof/sp.prf};
done;done > tmp/batch.sumprf.sp.sh

for s in {S,O}{1..6};do for i in SAM/$s/KRAKEN/rand/*.sp.prf;do 
perl -e 'exit unless $ARGV[0]=~/SAM\/(\S+)\/KRAKEN\/rand\/(\S+)\.(\d+)\.(\d+)\.sp\.prf$/;($s,$d,$t,$r)=($1,$2,$3,$4); open I,"<$ARGV[0]"; $c=0;while(<I>){
@s=split;if($_=~/UNCLASSIFIED/){$u=$s[0]}else{next if $s[0] < 10;$c++}; $sum += $s[0];
};close I; print join("\t",$s,$d,$t,$r,$c,$sum,$u)."\n" ' $i;
done  > SAM/$s/KRAKEN/rand.tax.count & done

cat SAM/{S,O}{1..6}/KRAKEN/rand.tax.count > STAT/KRAKEN/SCR.rand.tax.count
cat SAM/{S,O}{1..3}/KRAKEN/rand.tax.count > STAT/KRAKEN/SBR.rand.tax.count
cat SAM/{S,O}{4..6}/KRAKEN/rand.tax.count > STAT/KRAKEN/SBR.rand.tax.count

mkdir -p SAM/SBR/ANNO
for i in {S,O}{1,2,3};do 
  awk -v n=$i '/>BI/{gsub(">",">"n)}{print}' SAM/$i/summary.BI.megahit.clip.all.fasta;
done > SAM/SBR/summary.BI.megahit.clip.all.fasta
for i in {S,O}{1,2,3};do 
  awk -v n=$i '{print n$0}' SAM/$i/ANNO/CLIP.map.merge.bead.anno;
done > SAM/SBR/ANNO/CLIP.map.merge.bead.anno

metabbq smk -j -npk SAM/SBR/KRAKEN/db/hash.k2d
#
for i in {S,O}{1,2,3};do 
  ln -s ../../SBR/KRAKEN/db SAM/$i/KRAKEN/SBR;
done




```

**check database taxonomy size**
```{r}
db.tax.count <- read.table(
  md5("../../Results/JAN20/STAT/KRAKEN/db.tax.count","2db30e049ec1f709eb40bef515796816"),
  col.names=c("DB","kingdom","count"))
tmp <- db.tax.count$count[which(db.tax.count$DB=="UNITE")]
db.tax.count$count[which(db.tax.count$DB=="UNITE")] <- tmp[c(1,2,1,4)]
db.tax.count <- db.tax.count%>%filter(count >0)
db.tax.count$DB <- reorder(db.tax.count$DB,db.tax.count$count,max)
levels(db.tax.count$kingdom) <- c("All","Bacteria","Eukaryota","Fungi")
ggplot(db.tax.count,aes(x=DB,y=count,fill=kingdom)) + geom_bar(stat='identity',position="dodge")
```


### Read Bracken
```{r}
SBR.topRank.map <- read.table("../../Results/JAN20/SAM/SBR/KRAKEN/db/data/topRank.map",comment.char = "",
                        sep="\t",col.names=c("domain","taxID"))
```

```{r}
#species
SBR.r2SBR.species.bracken <- read.table(
  md5("../../Results/JAN20/STAT/KRAKEN/SBR.reads2SBR.bracken","1e459f767182acde7252c4018a2f0e81"),
  as.is = T,quote = "",comment.char = "", sep="\t",
  col.names = c("SID","name","taxID","rank","kraken_assigned_reads","added_reads","new_est_reads","fraction_total_reads"))
SBR.r2SBR.species.bracken$SAM <- substr(SBR.r2SBR.species.bracken$SID,1,1)
SBR.r2SBR.species.bracken$tax <- sapply(strsplit(as.character(SBR.r2SBR.species.bracken$name),"__"),"[",2)
SBR.r2SBR.species.bracken$tax <- reorder(SBR.r2SBR.species.bracken$tax,SBR.r2SBR.species.bracken$fraction_total_reads,sum)

SBR.r2SBR.species.bracken <- ddply(
  SBR.r2SBR.species.bracken,c("SID"),transform,
  pct.k=kraken_assigned_reads/sum(kraken_assigned_reads),
  pct.a=added_reads/sum(added_reads),
  pct=new_est_reads/sum(new_est_reads))


top20tax <- rev(levels(reorder(SBR.r2SBR.species.bracken$tax,SBR.r2SBR.species.bracken$pct,sum)))[1:20]

SBR.r2SBR.species.b.count <- ddply(SBR.r2SBR.species.bracken,c("SID"),summarise,observed=length(which(new_est_reads>10)))

SBR.r2SBR.species.bracken$tax20 <- ifelse(
  as.character(SBR.r2SBR.species.bracken$tax)%in%top20tax,as.character(SBR.r2SBR.species.bracken$tax),"Others")
SBR.r2SBR.species.bracken$tax20[which(SBR.r2SBR.species.bracken$name=="UNCLASSIFIED")] <- "UNCLASSIFIED"

SBR.r2SBR.species.b.stat <- ddply(SBR.r2SBR.species.bracken,c("SAM","SID","tax20"),summarise,pct=sum(pct))
SBR.r2SBR.species.b.stat$taxOrder <- reorder(SBR.r2SBR.species.b.stat$tax20,SBR.r2SBR.species.b.stat$pct,max)

```


```{r}
ggplot(SBR.r2SBR.species.b.stat,aes(x=taxOrder,y=pct,fill=SID)) + geom_bar(stat="identity",position="dodge") + coord_flip() + facet_grid(.~SAM)
ggplot(SBR.r2SBR.species.b.stat,aes(x=taxOrder,y=pct,fill=SID)) + geom_bar(stat="identity",position="dodge") + coord_flip() + facet_grid(.~SAM) + scale_y_log10()
ggplot(SBR.r2SBR.species.b.stat,aes(x=SID,y=pct,fill=taxOrder)) + geom_bar(stat="identity",position="fill")
ggplot(SBR.r2SBR.species.b.stat%>%filter(!tax20%in%c("Others","bacterium")),aes(x=SID,y=pct,fill=taxOrder)) + 
  geom_bar(stat="identity",position="stack") +facet_grid(.~SAM,scale="free")

```


### PCA
```{r}
SBR.r2SBR.b.sp.more <- merge(SBR.r2SBR.species.bracken,SBR.topRank.map,by="taxID",all.x=T)

SBR.sp.count.mtx <- dcast(SBR.r2SBR.b.sp.more,taxID+domain~SID,value.var = "kraken_assigned_reads",fill = 0)
row.kept <- which(apply(SBR.sp.count.mtx[,-c(1:2)],1,min)>10)

SBR.sp.prf.mtx <- dcast(SBR.r2SBR.b.sp.more,taxID+domain+name~SID,value.var = "pct",fill = 0)
rownames(SBR.sp.prf.mtx) <- SBR.sp.prf.mtx$taxID
taxName <- SBR.sp.prf.mtx$name
taxType <- sapply(strsplit(taxName,"__"),"[",1)
SBR.sp.cut.mtx <- SBR.sp.prf.mtx[row.kept,]
SBR.sp.cut.mtx.pca <- prcomp(t(as.matrix(SBR.sp.cut.mtx[,-c(1:3)])),center = T,scale. = T)
summary(SBR.sp.cut.mtx.pca)

SBR.sp.known.mtx <- as.matrix(SBR.sp.prf.mtx[intersect(row.kept,which(taxType=="species")),-c(1:3)])
SBR.sp.known.mtx.pca <- prcomp(t(SBR.sp.known.mtx),center = T)
SBR.sp.unknown.mtx <- as.matrix(SBR.sp.prf.mtx[intersect(row.kept,which(taxType!="species")),-c(1:3)])
SBR.sp.unknown.mtx.pca <- prcomp(t(SBR.sp.unknown.mtx),center = T)
#SBR.sp.ori.A.mtx <- as.matrix(SBR.sp.cut.mtx[which(SBR.sp.cut.mtx$domain=="domain__Archaea"),-c(1:3)])
#SBR.sp.ori.A.mtx.pca <- prcomp(t(SBR.sp.ori.A.mtx),center = T,scale. = T)
SBR.sp.ori.E.mtx <- as.matrix(SBR.sp.cut.mtx[which(SBR.sp.cut.mtx$domain=="domain__Eukaryota"),-c(1:3)])
SBR.sp.ori.E.mtx.pca <- prcomp(t(SBR.sp.ori.E.mtx),center = T,scale. = T)
SBR.sp.ori.B.mtx <- as.matrix(SBR.sp.cut.mtx[which(SBR.sp.cut.mtx$domain=="domain__Bacteria"),-c(1:3)])
SBR.sp.ori.B.mtx.pca <- prcomp(t(SBR.sp.ori.E.mtx),center = T,scale. = T)

SBR.sp.mtx.pca.df <- data.frame(SID=rownames(SBR.sp.cut.mtx.pca$x),SBR.sp.cut.mtx.pca$x)
ggplot(SBR.sp.mtx.pca.df,aes(x=PC1,y=PC2,color=SID)) +geom_point(size=3) + geom_text(aes(label=SID))

SBR.sp.multi.pca.df <- rbind(
  data.frame(type="ALL",SID=rownames(SBR.sp.cut.mtx.pca$x),SBR.sp.cut.mtx.pca$x[,1:5]),
  data.frame(type="known",SID=rownames(SBR.sp.known.mtx.pca$x),SBR.sp.known.mtx.pca$x[,1:5]),
  data.frame(type="unknown",SID=rownames(SBR.sp.unknown.mtx.pca$x),SBR.sp.unknown.mtx.pca$x[,1:5])
)

ggplot(SBR.sp.multi.pca.df,aes(x=PC1,y=PC2,color=SID)) +geom_point(size=3) + 
  geom_text(aes(label=SID)) + facet_wrap(~type,ncol=3,scale="free")

```

```{r}
pheatmap(cor(SBR.sp.cut.mtx[,-c(1:3)],method = "s"),display_numbers = T,cluster_rows = F,cluster_cols = F)
pheatmap(cor(SBR.sp.known.mtx,method = "s"),display_numbers = T,cluster_rows = F,cluster_cols = F)
pheatmap(cor(SBR.sp.unknown.mtx,method = "s"),display_numbers = T,cluster_rows = F,cluster_cols = F)

```


**QQ plot**
```{r}
SBR.r2SBR.b.sp.more$rep <- as.factor(substr(SBR.r2SBR.b.sp.more$SID,2,2))
SBR.r2SBR.b.sp.more$treat <- ifelse(SBR.r2SBR.b.sp.more$rep%in%c(1,2,3),"RCA","NoRCA")
levels(SBR.r2SBR.b.sp.more$rep) <- c(1,2,3,1,2,3)
#SBR.stat.sp.df <- ddply(SBR.r2SBR.bracken,c("SAM","treat","tax","rep","rank"),summarise,pct=mean(fraction_total_reads))
#SBR.stat.sp.dfs <- dcast(SBR.stat.sp.df,SAM+tax+rank+rep~treat,value.var = "fraction_total_reads")

#SBR.sp.cor <- ddply(SBR.stat.sp.dfs%>%filter(NoRCA+RCA>0),c("SAM","rank","rep"),summarise,spearman=cor(NoRCA,RCA,method="s"))

SBR.stat.sp.df2 <- SBR.r2SBR.b.sp.more
SBR.stat.sp.df2$rep <- as.numeric(SBR.stat.sp.df2$rep)
SBR.stat.sp.df2$rep[which(SBR.stat.sp.df2$rep>3)] <- SBR.stat.sp.df2$rep[which(SBR.stat.sp.df2$rep>3)] - 3
SBR.stat.sp.df2s <- dcast(SBR.stat.sp.df2,SAM+domain+rank+rep+tax~treat,value.var = "pct",fun.aggregate = sum,fill = 0)
ggplot(SBR.stat.sp.df2s,aes(x=RCA,y=NoRCA,color=factor(rep))) +geom_point() +
  geom_abline(intercept = c(-1,0,1),slope=1,linetype=2) + facet_grid(SAM~rep) +
  scale_x_log10() + scale_y_log10() 

```

```{r}
SBR.stat.sp.df2r <- dcast(SBR.stat.sp.df2,SAM+domain+rank+rep+tax~treat,value.var = "pct",fun.aggregate = sum,fill = 0)
ggplot(SBR.stat.sp.df2r,aes(x=RCA,y=NoRCA,color=factor(rep))) +geom_point() +
  geom_abline(intercept = c(-1,0,1),slope=1,linetype=2) + facet_grid(SAM~rep) +
  scale_x_log10() + scale_y_log10() 
```

```{r}
SBR.stat.sp.df3s <- dcast(SBR.stat.sp.df2,SAM+rank+tax+treat~rep,value.var = "pct",fun.aggregate = sum)
ggplot(SBR.stat.sp.df3s,aes(x=`1`,y=`2`,color=factor(treat))) +geom_point() +
  geom_abline(intercept = c(-1,0,1),slope=1,linetype=2) + facet_grid(treat~SAM) +
  scale_x_log10() + scale_y_log10() 
```


```{r}
ddply(SBR.stat.sp.df2s,c("SAM","rep"),summarise,
      sumRCAwith0NoRCA=sum(RCA[which(NoRCA==0)]),
      sumNoRCAwith0RCA=sum(NoRCA[which(RCA==0)]))
```

# old

# observed taxonomies

```{r}
SBCR.rand.tax.count <- read.table(
  md5("../../Results/JAN20/STAT/KRAKEN/SCR.rand.tax.count","83fd7b6030ae999c48cfc6866bc607ea"),
  sep="\t",col.names = c("SID","db","size","rep","count","beads","unclassified"))
levels(SBCR.rand.tax.count$SID)[which(levels(SBCR.rand.tax.count$SID)%in%c("S2","O2"))] <- c("S2","O2")
SBCR.rand.tax.count$SID <- factor(as.character(SBCR.rand.tax.count$SID),levels=paste0(rep(c("O","S"),each=6),rep(seq(1,6),2)))
SBCR.rand.tax.count$SAM <- substr(SBCR.rand.tax.count$SID,1,1)
SBCR.rand.tax.count$SREP <- substr(SBCR.rand.tax.count$SID,2,1)
SBCR.rand.tax.count$KDM <- ifelse(grepl("1|2|3",SBCR.rand.tax.count$SID),"BAC","FUNGI")

ggplot(SBCR.rand.tax.count%>%filter(count>0&db!=""&!is.na(size)),aes(x=size,y=count,color=db,fill=factor(rep))) + 
  geom_line(size=1) + geom_point() + facet_wrap(~SID,ncol=3)
ggplot(SBCR.rand.tax.count%>%filter(count>0),aes(x=size,y=count,color=factor(rep))) + geom_line() + geom_point() + 
  facet_grid(db~SID,scale="free") + scale_x_continuous(breaks = seq(0,2e9,2e7),labels = seq(0,2000,20))

SBCR.rand.tax.count.stat <- ddply(
  SBCR.rand.tax.count%>%filter(count>0&db!=""&!is.na(size)),c("SID","SAM","KDM","db","size"),summarise,sd=sd(count),count=mean(count))

ggplot(SBCR.rand.tax.count.stat,aes(x=factor(size),y=count,fill=db)) + geom_boxplot(width=1) + 
  #geom_hline(data=db.tax.count,aes(yintercept = count,color=kingdom),linetype=2) +
  facet_grid(.~KDM,scale="free") + 
  scale_x_discrete(breaks = seq(0,2e9,2e7),labels = seq(0,2000,20))

ggplot(SBCR.rand.tax.count%>%filter(count>0&db!=""&!is.na(size)),aes(x=size,y=count,color=db)) + 
  geom_line(aes(group=interaction(SID,db,rep)),alpha=.5) + geom_point(aes(shape=SAM)) + 
  facet_grid(KDM~.,scale="free") +theme_bw() + 
  xlab("# reads") + ylab("Observed Species and Unknown OTUs")
```

```{r}
ggplot(SBCR.rand.tax.count%>%filter(count>0&db!=""&!is.na(size)),aes(x=size,y=count,color=SID)) + 
  geom_line(aes(group=interaction(SID,db,rep)),alpha=.5) + geom_point(aes(shape=SAM)) + 
  #geom_smooth() +
  #geom_hline(data=db.tax.count,aes(yintercept = count,color=kingdom),linetype=2) +
  facet_grid(KDM~SREP,scale="free")
```

```{r}
ggplot(SBCR.rand.tax.count,aes(x=size,y=unclassified/beads,color=db)) + 
  geom_line(aes(group=interaction(SID,db,rep)),alpha=.5) + geom_point(aes(shape=SAM)) + 
  facet_grid(SREP~KDM,scale="free")
```

#FIN.