---
title: "JAN20 DATA ASSESSMENT: mock data"
author: "Chao"
date: "3/22/2020"
output: 
  html_notebook:
    code_fold: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(reshape2)
library(ggplot2)
library(plyr)
library(dplyr)
library(RColorBrewer)
library(configr)
library(tools)
library(stringi)
md5 <- function(f,m){
  if(tools::md5sum(f)==m){
    warning("MD5 check passed.\n")
    return(f)
  }else{
    stop("MD5 check failed. Please check the version and content!\n")
  }
}

# Function
#https://www.biostars.org/p/9335/
matcher <- function(pattern, x) {

  ind = gregexpr(pattern, x)[[1]]
  start = as.numeric(ind)
  if(length(start)==1 && start <1){
    return(NA)
  }else{
    end = start + attr(ind, "match.length") - 2
    res <- as.numeric(apply(cbind(start,end), 1, function(y) substr(x, start=y[1], stop=y[2])))
    res[which(is.na(res))] <- 1
    return(res)
  }
}

doone <- function(c, cigar) {
  pat <- paste("\\d*", c , sep="")
  sum(as.numeric(matcher(pat, cigar)), na.rm=T)
}


## takes a cigar string and parses it, not very fast but...
cigarsums <- function(cigar, chars=c("M","N","D","I","S","H", "P", "X", "=")) {
   sapply (chars, doone, cigar)
}


n50 <- function(x){
  x1 <- rev(sort(x))
  cumsum.x1 <- cumsum(x1)
  find.n50 <- which(cumsum.x1 >= sum(x1)/2)
  return(x1[find.n50[1]])
}
```
# Scope
In this script:  

1. In order to get a better alignment for phylogenetic analysis, only clips with adaptors found were retatined

# Background

Samples involved:

| ID | lane | barcode | Kindom | RCA | Location    |
| -- | ---- | ------- | -------| --- | ----------- |
| U1 | FF1  | 1,17    | Bac    | RCA | Urine       |
| S1 | FF1  | 3,4     | Bac    | RCA | Soil spot A |
| O1 | FF1  | 5,22    | Bac    | RCA | Soil spot B |
| M1 | FF1  | 7       | Bac    | RCA | Mock        |
| Z1 | FF1  | 8       | Bac    | RCA | ZYMO 8      |

Zymo mock strains:

| Species                  | Genomic DNA | 16S Only1 | 16S & 18S1 | Genome Copy2 | Cell Number3 | Gram | 16/18 Copy Num |
| ------------------------ | ----------- | --------- | ---------- | ------------ | ------------ | ---- | -------------- |
| Pseudomonas aeruginosa   | 12          | 4.2       | 3.6        | 6.1          | 6.1          | -    | 4              |
| Escherichia coli         | 12          | 10.1      | 8.9        | 8.5          | 8.5          | -    | 7              |
| Salmonella enterica      | 12          | 10.4      | 9.1        | 8.7          | 8.8          | -    | 7              |
| Lactobacillus fermentum  | 12          | 18.4      | 16.1       | 21.6         | 21.9         | +    | 5              |
| Enterococcus faecalis    | 12          | 9.9       | 8.7        | 14.6         | 14.6         | +    | 4              |
| Staphylococcus aureus    | 12          | 15.5      | 13.6       | 15.2         | 15.3         | +    | 6              |
| Listeria monocytogenes   | 12          | 14.1      | 12.4       | 13.9         | 13.9         | +    | 6              |
| Bacillus subtilis        | 12          | 17.4      | 15.3       | 10.3         | 10.3         | +    | 10             |
| Saccharomyces cerevisiae | 2           | NA        | 9.3        | 0.57         | 0.29         | Y    | 109            |
| Cryptococcus neoformans  | 2           | NA        | 3.3        | 0.37         | 0.18         | Y    | 60             |

# Filter clips
```{bash,eval=FALSE}
mkdir -p ../../Results/JAN20/STAT/ADCLIPS




```




Load data:
```{r}
metadata <- read.csv("../../Results/JAN20/metadata.csv")

ZBR.LOTU.info <- read.table("../../Results/JAN20/STAT/CLIP/ZS.clust.log",sep=" ", 
                           col.names = c("SID","LOTU","size","S75","CL","MAXLEN","MEANLEN"))

tools::md5sum("../../Results/JAN20/STAT/CLIP/ZS.LOTU.to.MockRef.bead.anno")
ZBR.mock.anno <- read.table(md5("../../Results/JAN20/STAT/CLIP/ZS.LOTU.to.MockRef.bead.anno","07242ba03c87f09ebb6e7bf5a549806b"),
                           sep="\t", col.names = c("SID","LOTU","BLen","hits","hits.2","tax.clips","tax.pieces","tax",
                 "ident","qMapLen","mismatch","gap","bitScore","bitScore2nd",
                 "sMapPieces","sMapLen","MAPDESC","strands","qMaps","sMaps"
))

tools::md5sum("../../Results/JAN20/STAT/CLIP/all.LOTU.map.merge.anno")
ZBR.silva.anno <- read.table(
  md5("../../Results/JAN20/STAT/CLIP/all.LOTU.map.merge.anno","e0d9b11d22f10fe7fdd3faf035911f91"), sep="\t",
   col.names = c("SID","LOTU","tax","idents","lens","mis","gap","bit","qlen","qcov","scov","scov.ITS","scov.SSU","scov.LSU","taxID","rank","taxPath")
)


ZBR.prop <- read.table("../../Results/JAN20/STAT/CLIP/rand3.LOTU.bwa.sbb1.p",sep="\t",
                      col.names = c("SID","rand","count0","count","LOTU"))
#

```


#Curation
```{r}
ZBR.badf <- merge(metadata,ZBR.mock.anno,by="SID")
#badf$delta.clips <- badf$clips-badf$tax.clips
getMapPropFun <- function(d){
  uOrder <- c("ALL","SSU","ITS1","5.8S","ITS2","LSU")
  ulist <- data.frame(key=c("ALL","SSU","16S","ITS","ITS1","5.8S","ITS2","LSU","23S"),
                      ind=c(1,2,2,3,3,4,5,6,6))
  getM <- stri_match_all_regex(d,"(ALL|23S|LSU|ITS|ITS1|ITS2|5.8S|SSU|16S)\\((\\d+)%\\)")
  getN <- length(getM)
  res <- matrix(0,ncol = 6,nrow = getN)
  for(i in 1:length(getM)){
    geti <- getM[[i]]
    for(j in 1:nrow(geti)){
      res[i,ulist$ind[which(ulist$key==geti[j,2])]] <- geti[j,3]
    }
  }
  res <- matrix(as.numeric(res),ncol=6)
  colnames(res) <- uOrder
  return(res)
}
ZBR.badf <- cbind(ZBR.badf,getMapPropFun(ZBR.badf$MAPDESC))
ZBR.badf$mainUnit <- ifelse(ZBR.badf$SSU>ZBR.badf$LSU,"SSU","LSU")
ZBR.badf$unit <- as.factor(ifelse(ZBR.badf$SSU>25,1,0) + ifelse(ZBR.badf$LSU>25,2,0)) 
levels(ZBR.badf$unit) <- c("partial(<25%)","SSU","LSU","Both")
ZBR.badfm <- merge(ZBR.LOTU.info%>%filter(S75>0),ZBR.badf,by=c("SID","LOTU"),all.x=T)
```

basic:
```{r}
n50(ZBR.badfm$CL)
ggplot(ZBR.badfm,aes(x=qMapLen,fill=SID)) +geom_histogram(position="dodge") + scale_x_log10()
ggplot(ZBR.badfm,aes(x=ident,fill=SID)) +geom_histogram(position="dodge") + scale_x_log10()

ZBR.badfm$S75.2 <- 2^round(log2(ZBR.badfm$S75),0)

ggplot(ZBR.badfm,aes(x=S75,fill=SID)) +geom_histogram(position="dodge") +
  scale_x_log10() + annotation_logticks(side="b")
ggplot(ZBR.badfm,aes(y=ident,x=factor(S75.2),fill=SID)) +geom_boxplot(aes(fill=SID)) +
  scale_y_continuous(breaks=seq(94,102,1),limits=c(94,100))

ggplot(ZBR.badfm,aes(x=CL,fill=unit)) +geom_histogram(position="stack",binwidth=50)  + facet_grid(.~SID)
ggplot(ZBR.badfm%>%filter(CL>499),aes(x=CL,fill=unit)) +geom_histogram(position="stack",binwidth=100) +
  theme(legend.position = c(0.99,.9),legend.justification = c(1,1))
```


```{r}
ZBR.badfm$ident5 <- round(ZBR.badfm$ident+0.49,)-0.5
ZBR.badfm$ident5 <- ifelse(ZBR.badfm$ident5<95,94.5,ZBR.badfm$ident5)

ggplot(ZBR.badfm,aes(x=ident5)) + geom_bar(aes(fill=unit),stat="count",position="stack") + 
  scale_x_continuous(breaks=seq(94,102,1),limits=c(94,100)) +
  coord_flip() + facet_grid(.~SID)

badf.stat2 <- ddply(
  ddply(ZBR.badfm%>%filter(CL>499),c("SID","ident5"),summarise,count=length(SID)),
  c("ident5"),summarise,mean.count=mean(count),sd=sd(count))
badf.stat3 <- ddply(
  ddply(ZBR.badfm%>%filter(CL>499),c("SID","ident5","unit"),summarise,count=length(SID)),
  c("ident5","unit"),summarise,mean.count=mean(count))

ggplot(badf.stat3,aes(x=ident5,y=mean.count)) + geom_bar(aes(fill=unit),stat="identity",position="stack") + 
  geom_errorbar(data=badf.stat2,aes(ymax=mean.count+sd,ymin=mean.count-sd),width=.1) + 
  scale_x_continuous(breaks=seq(94,102,1),limits=c(94,100)) +
  coord_flip()
```



compare annotation : mock ref vs. silva
```{r}
levels(ZBR.badfm$tax) <- sub("_"," ",levels(ZBR.badfm$tax))
ZBR.compare.anno <- merge(ZBR.badfm,ZBR.silva.anno,by=c("SID","LOTU"),all.x=T)
checkTaxConsitRank <- function(d){
  tax.x <- unlist(strsplit(as.character(d$tax.x)," "))
  tax.y <- unlist(strsplit(as.character(d$tax.y)," "))
  checkLv <- ifelse(tax.x[1]==tax.y[1],1,0) + ifelse(tax.x[2]==tax.y[2],2,0)
  return(checkLv)
}
ZBR.compare.anno$consitLv <- as.factor((ddply(ZBR.compare.anno[,c("SID","LOTU","tax.x","tax.y")],c("SID","LOTU"),checkTaxConsitRank))$V1)
levels(ZBR.compare.anno$consitLv) <- c("different","genus","species")
```

```{r}
ZBR.compare.anno$hitDB <- as.factor(ifelse(ZBR.compare.anno$scov.SSU==0,0,1) + ifelse(ZBR.compare.anno$scov.LSU==0,0,2))
levels(ZBR.compare.anno$hitDB) <- c("SILVA.SSU","SILVA.LSU","SILVA.LSU+SSU")

ZBR.compare.anno$reachLimit <- as.factor(ifelse(ZBR.compare.anno$scov.ITS>99,T,ifelse(ZBR.compare.anno$scov.SSU>99,T,ifelse(ZBR.compare.anno$scov.LSU>99,T,0))))

ggplot(ZBR.compare.anno,aes(x=ident,y=idents,color=consitLv)) + geom_point()
ggplot(ZBR.compare.anno,aes(x=idents-ident,color=consitLv)) + geom_density()

ggplot(ZBR.compare.anno,aes(x=qMapLen,y=lens,color=consitLv)) + geom_point()
ggplot(ZBR.compare.anno,aes(x=bitScore,y=bit,color=consitLv)) + geom_point(alpha=.5)
ggplot(ZBR.compare.anno,aes(x=bitScore,y=bit,color=hitDB,shape=reachLimit)) +
  geom_abline(slope = 1,intercept = 0, linetype=2,alpha=.5,size=2,color="grey50") +
  geom_point(alpha=.5) + scale_shape_manual(values=c(16,4))
  
ggplot(ZBR.compare.anno%>%filter(!is.na(consitLv)),aes(x=idents,color=consitLv)) + geom_density(size=1,alpha=.5) +
  theme(legend.position = c(0.1,.99),legend.justification = c(0,1)) + facet_grid(hitDB~.)
#ggplot(ZBR.compare.anno,aes(x=qlen,fill=hitDB)) + geom_bar() +facet_grid(.~consitLv) #+ theme(axis.text.x = element_text(angle=90))
ggplot(ZBR.compare.anno,aes(x=bitScore,y=bit,color=mainUnit)) + geom_point(alpha=.5) +facet_grid(.~consitLv)

```

completness of subunits
```{r}
ZBR.stat.comp <- melt(ZBR.compare.anno%>%filter(!is.na(consitLv)),measure.vars = c("ALL","ITS1","SSU","LSU"),variable.name = "subunits", value.name = "completness")
ZBR.stat.comp$qlen.seg <- round(ZBR.stat.comp$qlen/500,0)*500
ggplot(ZBR.stat.comp%>%filter(qlen.seg>400&qlen.seg<3500),aes(x=completness,fill=subunits)) + geom_histogram(position="dodge") +
  facet_grid(qlen.seg~.,scale="free")

```

comapre with mock ref annotation
```{r}
fungi.copy.num <- data.frame(species=c("Pseudomonas aeruginosa","Escherichia coli","Salmonella enterica","Lactobacillus fermentum",
  "Enterococcus faecalis","Staphylococcus aureus","Listeria monocytogenes","Bacillus subtilis","Saccharomyces cerevisiae","Cryptococcus neoformans"),
  TC16S=c(4.2,10.1,10.4,18.4,9.9,15.5,14.1,17.4,NA,NA),
  TC16S18S=c(3.6,8.9,9.1,16.1,8.7,13.6,12.4,15.3,9.3,3.3),copyNum=c(4,7,7,5,4,6,6,10,109,60))
fungi.copy.num$genus <- sapply(strsplit(as.character(fungi.copy.num$species)," "),"[",1)
selectTax <- c("Pseudomonas aeruginosa","Escherichia coli","Salmonella enterica","Lactobacillus fermentum",
  "Enterococcus faecalis","Staphylococcus aureus","Listeria monocytogenes","Bacillus subtilis","Saccharomyces cerevisiae","Cryptococcus neoformans")
selectGenus <- sapply(strsplit(selectTax," "),"[",1)

ZBR.qdf <- merge(ZBR.prop,ZBR.compare.anno,by=c("SID","LOTU"),all.x=T)
ZBR.qdf$species <- sapply(strsplit(as.character(ZBR.qdf$taxPath),";"),"[",7)
ZBR.qdf$genus <- sapply(strsplit(as.character(ZBR.qdf$taxPath),";"),"[",6)
ZBR.qdf$selectTax <- ifelse(ZBR.qdf$genus%in%selectGenus,as.character(ZBR.qdf$genus),"Others.above.Genus")
ZBR.qdf$checkLv <- ifelse(ZBR.qdf$species%in%selectTax,"Same Sp.",ifelse(ZBR.qdf$genus%in%selectGenus,"Same Genus","Others.above.Genus"))

ZBR.qdf$subunit <- as.factor(ifelse(ZBR.qdf$scov.ITS>25,1,0)+ifelse(ZBR.qdf$scov.SSU>25,2,0)+ifelse(ZBR.qdf$scov.LSU>25,4,0))
ZBR.stat <- ddply(ZBR.qdf%>%filter(LOTU!='*'),c("SID","rand","tax.x"),summarise,count=sum(count))
ZBR.stat <- ddply(ZBR.stat,c("SID","rand"),transform,pct=count/sum(count)*100)
ZBR.stat1 <- merge(ZBR.stat,fungi.copy.num[,c("species","TC16S")],by.x="tax.x",by.y="species",all.x=T)
ggplot(ZBR.stat1%>%filter(SID!="Z3"),aes(x=tax.x,y=pct,fill=factor(rand))) + 
  geom_bar(aes(y=TC16S),fill=NA,color="grey50",stat="identity",position="dodge") +
  geom_bar(alpha=.5,stat = 'identity', position = 'dodge') +
  facet_grid(SID~.) + coord_flip()

ZBR.stat2 <- ddply(ZBR.qdf%>%filter(ident>99.5),c("SID","rand","tax.x"),summarise,count=sum(count))
ZBR.stat2 <- ddply(ZBR.stat2,c("SID","rand"),transform,pct=count/sum(count)*100)
ZBR.stat2 <- merge(ZBR.stat2,fungi.copy.num[,c("species","TC16S")],by.x="tax.x",by.y="species",all.x=T)
ggplot(ZBR.stat2%>%filter(SID!="Z3"),aes(x=tax.x,y=pct,fill=factor(rand))) + 
  geom_bar(aes(y=TC16S),fill=NA,color="grey50",stat="identity",position="dodge") +
  geom_bar(alpha=.5,stat = 'identity', position = 'dodge') +
  facet_grid(SID~.) + coord_flip()


```

```{r}
ZBR.qdf$selectTax <- ifelse(ZBR.qdf$species%in%selectTax,as.character(ZBR.qdf$species),ifelse(ZBR.qdf$genus%in%selectGenus,"Others.in.Genus","Others.above.Genus"))

ZBR.stat3 <- ddply(ZBR.qdf%>%filter(idents>99.5),c("SID","rand","selectTax"),summarise,count=sum(count))
ZBR.stat3 <- merge(ddply(ZBR.stat3,c("SID","rand"),transform, pct=count/sum(count)*100),
                  fungi.copy.num[,c("species","TC16S")],by.x="selectTax",by.y="species",all.x=T)
ZBR.stat3$TC16S[which(is.na(ZBR.stat3$TC16S))] <- 0
ZBR.stat3$pct.delta <- ZBR.stat3$pct-ZBR.stat3$TC16S
ggplot(ZBR.stat3,aes(x=SID,y=pct,fill=factor(rand))) +geom_bar(stat="identity",position="dodge") +facet_grid(.~selectTax)
ggplot(ZBR.stat3,aes(x=SID,y=pct,fill=factor(rand))) + geom_bar(aes(y=TC16S),fill=NA,color="grey50",stat="identity",position="dodge") +
  geom_bar(stat="identity",position="dodge",alpha=.5) +facet_grid(.~selectTax)

ZBR.mean.df <- ddply(ZBR.stat3,c("selectTax"),summarise,sd=sd(pct.delta),pct=mean(pct.delta))
ZBR.mean.df <- ZBR.mean.df%>%filter(selectTax%in%selectTax)
colnames(ZBR.mean.df)[1] <- "selectTax"


ZBR.mean.df$selectTax <- as.factor(as.character(ZBR.mean.df$selectTax))

ZBR.mean.df <- ZBR.mean.df[order(ZBR.mean.df$pct),]
ZBR.mean.df$selectTax <- reorder(ZBR.mean.df$selectTax,ZBR.mean.df$pct,mean)
ggplot(ZBR.mean.df,aes(x=selectTax,y=pct,fill=selectTax)) + geom_hline(yintercept = 0,linetype=2,color="grey50",size=2) +
  geom_bar(stat = 'identity', position = 'dodge') + geom_errorbar(aes(ymax=pct+sd,ymin=pct-sd),width=.4) + coord_flip() +
  ylab("delta of quantificatin \nover theoretical value (%)") + guides(fill=F)

```

```{r}

ZBR.stat3 <- ddply(ZBR.qdf%>%filter(idents>99.5),c("SID","rand","checkLv","selectTax"),summarise,count=sum(count))
ZBR.stat3 <- merge(ddply(ZBR.stat3,c("SID","rand"),transform, pct=count/sum(count)*100),
                  fungi.copy.num[,c("genus","TC16S")],by.x="selectTax",by.y="genus",all.x=T)
ZBR.stat3$TC16S[which(is.na(ZBR.stat3$TC16S))] <- 0
ZBR.stat3$pct.delta <- ZBR.stat3$pct-ZBR.stat3$TC16S
ggplot(ZBR.stat3,aes(x=SID,y=pct,fill=factor(rand))) +geom_bar(stat="identity",position="dodge") +facet_grid(.~selectTax)
ggplot(ZBR.stat3,aes(x=SID,y=pct,fill=factor(rand))) + geom_bar(aes(y=TC16S),fill=NA,color="grey50",stat="identity",position="dodge") +
  geom_bar(stat="identity",position="dodge",alpha=.5) +facet_grid(.~selectTax)

ZBR.mean.df <- ddply(ZBR.stat3,c("selectTax"),summarise,sd=sd(pct.delta),pct=mean(pct.delta))

ZBR.mean.df1 <- ddply(ddply(ddply(ZBR.qdf%>%filter(idents>99.5),c("SID","rand","checkLv","selectTax"),summarise,count=sum(count)),
                          c("SID","rand","checkLv"),transform, pct=count/sum(count)*100),
                    c("selectTax","checkLv"),summarise,pct=mean(pct))
ZBR.mean.df1 <- merge(ZBR.mean.df1,fungi.copy.num[,c("genus","TC16S")],by.x="selectTax",by.y="genus",all.x=T)
ZBR.mean.df1$pct.delta <- ZBR.mean.df1$pct-ZBR.mean.df1$TC16S

#ZBR.mean.df <- ZBR.mean.df%>%filter(selectTax%in%selectTax)

ZBR.mean.df$selectTax <- as.factor(as.character(ZBR.mean.df$selectTax))

ZBR.mean.df <- ZBR.mean.df[order(ZBR.mean.df$pct),]
ZBR.mean.df$selectTax <- reorder(ZBR.mean.df$selectTax,ZBR.mean.df$pct,mean)
ggplot(ZBR.mean.df,aes(x=selectTax,y=pct,fill=selectTax)) + geom_hline(yintercept = 0,linetype=2,color="grey50",size=2) +
  geom_bar(data=ZBR.mean.df1,stat = 'identity', position = 'stack') + geom_errorbar(aes(ymax=pct+sd,ymin=pct-sd),width=.4) + coord_flip() +
  ylab("delta of quantification over theoretical value (%)")

selectTax.list <- selectTax
ggplot(ZBR.filter.df.ref%>%filter(selectTax%in%selectTax.list),aes(x=rand,y=pct,fill=selectTax)) + geom_bar(stat = 'identity', position = 'fill')

```

```{r}
fungi.copy.num <- data.frame(species=c("Pseudomonas aeruginosa","Escherichia coli","Salmonella enterica","Lactobacillus fermentum",
  "Enterococcus faecalis","Staphylococcus aureus","Listeria monocytogenes","Bacillus subtilis","Saccharomyces cerevisiae","Cryptococcus neoformans"),
  TC16S=c(4.2,10.1,10.4,18.4,9.9,15.5,14.1,17.4,NA,NA),
  TC16S18S=c(3.6,8.9,9.1,16.1,8.7,13.6,12.4,15.3,9.3,3.3),copyNum=c(4,7,7,5,4,6,6,10,109,60))

ZS.qdf$selectTax <- ifelse(ZS.qdf$species%in%selectTax,as.character(ZS.qdf$species),ifelse(ZS.qdf$genus%in%selectGenus,as.character(ZS.qdf$genus),"Others"))
ZS.stat3 <- ddply(ZS.qdf%>%filter(idents>99.5),c("SID","rand","selectTax"),summarise,count=sum(count))
ZS.stat3 <- merge(ddply(ZS.stat3,c("SID","rand"),transform, pct=count/sum(count)*100),
                  fungi.copy.num[,c("species","TC16S")],by.x="selectTax",by.y="species",all.x=T)
ZS.stat3$pct.delta <- ZS.stat3$pct-ZS.stat3$TC16S
ggplot(ZS.stat3,aes(x=SID,y=pct,fill=factor(rand))) +geom_bar(stat="identity",position="dodge") +facet_grid(.~selectTax)
ggplot(ZS.stat3,aes(x=SID,y=pct,fill=factor(rand))) + geom_bar(aes(y=TC16S),fill=NA,color="grey50",stat="identity",position="dodge") +
  geom_bar(stat="identity",position="dodge",alpha=.5) +facet_grid(.~selectTax)

ZS.mean.df <- ddply(ZS.stat3,c("selectTax"),summarise,sd=sd(pct.delta),pct=mean(pct.delta))
ZS.mean.df <- ZS.mean.df%>%filter(selectTax%in%selectTax)
colnames(ZS.mean.df)[1] <- "selectTax"


ZS.mean.df$selectTax <- as.factor(as.character(ZS.mean.df$selectTax))

ZS.mean.df <- ZS.mean.df[order(ZS.mean.df$pct),]
ZS.mean.df$selectTax <- reorder(ZS.mean.df$selectTax,ZS.mean.df$pct,mean)
ggplot(ZS.mean.df,aes(x=selectTax,y=pct,fill=selectTax)) + geom_hline(yintercept = 0,linetype=2,color="grey50",size=2) +
  geom_bar(stat = 'identity', position = 'dodge') + geom_errorbar(aes(ymax=pct+sd,ymin=pct-sd),width=.4) + coord_flip() +
  ylab("delta of quantification over theoretical value (%)")

selectTax.list <- selectTax
ggplot(ZS.filter.df.ref%>%filter(selectTax%in%selectTax.list),aes(x=rand,y=pct,fill=selectTax)) + geom_bar(stat = 'identity', position = 'fill')


```


```{r}
ZS.qdf$selectTax <- ifelse(ZS.qdf$species%in%selectTax,as.character(ZS.qdf$species),ifelse(ZS.qdf$genus%in%selectGenus,as.character(ZS.qdf$genus),"Others"))
ZS.stat4 <- ddply(ZS.qdf%>%filter(idents>99.5&!is.na(consitLv)),c("SID","rand","consitLv","selectTax"),summarise,count=sum(count))
ZS.stat4 <- merge(ddply(ZS.stat4,c("SID","rand","consitLv"),transform, pct=count/sum(count)*100),
                  fungi.copy.num[,c("species","TC16S")],by.x="selectTax",by.y="species",all.x=T)
ZS.stat4$pct.delta <- ZS.stat4$pct-ZS.stat4$TC16S

ggplot(ZS.stat4,aes(x=selectTax,y=pct,fill=interaction(rand,SID))) +geom_bar(stat="identity",position="dodge") +facet_grid(consitLv~.,scale="free") +
  coord_flip()

ZS.mean.df.4 <- ddply(ZS.stat4,c("selectTax","consitLv"),summarise,sd=sd(pct.delta),pct=mean(pct.delta))
ZS.mean.df.4 <- ZS.mean.df.4%>%filter(selectTax%in%selectTax)

ZS.mean.df.4$selectTax <- as.factor(as.character(ZS.mean.df.4$selectTax))

ZS.mean.df.4 <- ZS.mean.df.4[order(ZS.mean.df.4$pct),]
ZS.mean.df.4$selectTax <- reorder(ZS.mean.df.4$selectTax,ZS.mean.df.4$pct,mean)
ggplot(ZS.mean.df.4,aes(x=selectTax,y=pct,fill=selectTax)) + geom_hline(yintercept = 0,linetype=2,color="grey50",size=2) +
  geom_bar(stat = 'identity', position = 'dodge') + geom_errorbar(aes(ymax=pct+sd,ymin=pct-sd),width=.4) + coord_flip() +
  facet_grid(.~consitLv) + ylab("delta of quantification over theoretical value (%)")

```


```{r}
ZS.filter.df.ref2 <- merge(fungi.copy.num.pair[which(!is.na(fungi.copy.num.pair$count)),c("selectTax","count")],
                          ZS.stat[which(ZS.stat$subunit!="*"),select.col],by=c("selectTax"))
#ZS.filter.df.ref2$pct.STANDARD <- ZS.filter.df.ref2$count.x
ZS.filter.df.ref2 <- ddply(ZS.filter.df.ref2,c("SID","rand","subunit"),transform,pct=count.y/sum(count.y)*100)
ZS.filter.df.ref2$pct.delta <- ZS.filter.df.ref2$pct - ZS.filter.df.ref2$count.x
ggplot(ZS.filter.df.ref2,aes(x=SID,y=pct.delta,fill=selectTax,group=selectTax)) + 
  geom_boxplot()+facet_grid(.~selectTax)

ZS.filter.df.ref2$method <- ifelse(ZS.filter.df.ref2$SID%in%c("Z1","Z2","Z3"),"RCA","LFR")
ZS.filter.df.ref2.sum <- ddply(ZS.filter.df.ref2,c("selectTax","method","subunit"),summarise,delta.mean=mean(pct.delta),pct.se=sd(pct.delta))
ggplot(ZS.filter.df.ref2.sum,aes(x=subunit,y=delta.mean,fill=method)) +geom_bar(stat="identity",position="dodge") + 
  geom_errorbar(aes(ymax=delta.mean+pct.se,ymin=delta.mean-pct.se),width=.2) + 
  theme(axis.text.x = element_text(angle = -45,hjust=0)) + ylab("Delta of percentage (%)") +
  facet_grid(.~selectTax)
```


```{r}
#Z2.anno2 <- merge(Z2.faID,Z2.anno,by="LOTU",all.x=T)
Z2.prop2 <- merge(Z,Z2.anno2,by="LOTU")
Z2.stat2 <- ddply(Z2.prop2,"rTax",summarise,count=sum(count))
top20 <- as.character(Z2.stat2$rTax[rev(order(Z2.stat2$count))[1:13]])
Z2.stat2$tax20 <- as.character(Z2.stat2$rTax)
Z2.stat2$tax20 <- ifelse(Z2.stat2$tax20%in%top20,Z2.stat2$tax20,"Others")
ggplot(Z2.stat2,aes(x="1",y=count,fill=tax20)) + geom_bar(stat = 'identity', position = 'stack')

Z2.filter.df.ref2 <- merge(Z2.stat2,fungi.copy.num,by.x="rTax",by.y="species")
Z2.filter.df.ref2$prop <- Z2.filter.df.ref2$count/sum(Z2.filter.df.ref2$count)*100
Z2.filter.df.ref3 <- melt(Z2.filter.df.ref2,measure.vars = c("prop","TC16S"),variable.name = "group",value.name = "pct")
ggplot(Z2.filter.df.ref3,aes(x=group,y=pct,fill=tax20)) + geom_bar(stat = 'identity', position = 'stack')

```


# FUNGI

#close ref
## map clips and OTUs to zymo ref
```{bash,eval=FALSE}


metabbq smk --configfile config.F.yaml -j -npk SAM/M{4,5,6}/CLIP/all.LOTU.to.MockRef.bead.anno
for i in M{4,5,6};do sed 's/\S\+=//g' SAM/$i/CLIP/all.clust.log|awk -v n=$i '{print n" "$0}' ;done > STAT/CLIP/MFR.clust.log
for i in M{4,5,6};do awk -v n=$i '{print n"\t"$0}' SAM/$i/CLIP/all.LOTU.to.MockRef.bead.anno;done > STAT/CLIP/MFR.LOTU.to.MockRef.bead.anno

metabbq smk --configfile config.F.yaml -j -npk SAM/M{4,5,6}/CLIP/all.LOTU.map.merge.anno
for i in M{4,5,6};do sed 's/\S\+=//g' SAM/$i/CLIP/all.LOTU.map.merge.anno|awk -v n=$i '{print n"\t"$0}' ;done > STAT/CLIP/MFR.LOTU.map.merge.anno


for i in {4,5,6} ;do SAM=SAM/M$i; for j in {00..02};do echo sh tmp/mock.LOTU.quantification.sh $SAM $j \&;done;done >> tmp/mock.batch.quantification.sh

for i in {4,5,6} ;do SAM=M$i; for j in {00..02};do 
awk -v s=$SAM -v r=$j '{print s"\t"r"\t"$1"\t"$2"\t"$3}' SAM/$SAM/CLIP/RAND/all.LOTU.bwa.sbb1_$j.p;
done;done > STAT/CLIP/MFR.rand3.LOTU.bwa.sbb1.p

```




Load data:
```{r}
metadata <- read.csv("../../Results/JAN20/metadata.csv")

MFR.LOTU.info <- read.table("../../Results/JAN20/STAT/CLIP/MFR.clust.log",sep=" ", 
                           col.names = c("SID","LOTU","size","S75","CL","MAXLEN","MEANLEN"))

MFR.mock.anno <- read.table("../../Results/JAN20/STAT/CLIP/MFR.LOTU.to.MockRef.bead.anno",sep="\t", col.names = c("SID","LOTU","BLen","hits","hits.2","tax.clips","tax.pieces","tax",
                 "ident","qMapLen","mismatch","gap","bitScore","bitScore2nd",
                 "sMapPieces","sMapLen","MAPDESC","strands","qMaps","sMaps"
))

tools::md5sum("../../Results/JAN20/STAT/CLIP/MFR.LOTU.map.merge.anno")
MFR.silva.anno <- read.table(
  md5("../../Results/JAN20/STAT/CLIP/MFR.LOTU.map.merge.anno","d3831ecadfd6294082069a1a35e17a42"), sep="\t",
   col.names = c("SID","LOTU","tax","idents","lens","mis","gap","bit","qlen","qcov","scov","scov.ITS","scov.SSU","scov.LSU","taxID","rank","taxPath")
)


MFR.prop <- read.table("../../Results/JAN20/STAT/CLIP/MFR.rand3.LOTU.bwa.sbb1.p",sep="\t",
                      col.names = c("SID","rand","count0","count","LOTU"))


```

#Curation
```{r}
MFR.badf <- merge(metadata,MFR.mock.anno,by="SID")

MFR.badf <- cbind(MFR.badf,getMapPropFun(MFR.badf$MAPDESC))
MFR.badf$mainUnit <- ifelse(MFR.badf$SSU>MFR.badf$LSU,"SSU","LSU")
MFR.badf$unit <- as.factor(ifelse(MFR.badf$SSU>25,1,0) + ifelse(MFR.badf$LSU>25,2,0) + ifelse(MFR.badf$ITS1+MFR.badf$ITS2>50,4,0))
levels(MFR.badf$unit) <- c("partial(<25%)","SSU","LSU","Both-ITS","ITS","SSU+ITS","LSU+ITS","All")
MFR.badfm <- merge(MFR.LOTU.info%>%filter(S75>0),MFR.badf,by=c("SID","LOTU"),all.x=T)
```

basic:
```{r}
n50(MFR.badfm$CL)
ggplot(MFR.badf,aes(x=qMapLen,fill=SID)) +geom_histogram(position="dodge") + scale_x_log10()
ggplot(MFR.badf,aes(x=ident,fill=SID)) +geom_histogram(position="dodge") + scale_x_log10()

MFR.badfm$S75.2 <- 2^round(log2(MFR.badfm$S75),0)

ggplot(MFR.badfm,aes(x=S75,fill=SID)) +geom_histogram(position="dodge") +
  scale_x_log10() + annotation_logticks(side="b")
ggplot(MFR.badfm,aes(y=ident,x=factor(S75.2),fill=SID)) +geom_boxplot(aes(fill=SID)) +
  scale_y_continuous(breaks=seq(94,102,1),limits=c(94,100))

ggplot(MFR.badfm,aes(x=CL,fill=unit)) +geom_histogram(position="stack",binwidth=50)  + facet_grid(.~SID)
ggplot(MFR.badfm%>%filter(CL>499),aes(x=CL,fill=unit)) +geom_histogram(position="stack",binwidth=100) +
  theme(legend.position = c(0.99,.9),legend.justification = c(1,1))
```


```{r}
MFR.badfm$ident5 <- round(MFR.badfm$ident+0.49,)-0.5
MFR.badfm$ident5 <- ifelse(MFR.badfm$ident5<95,94.5,MFR.badfm$ident5)

ggplot(MFR.badfm,aes(x=ident5)) + geom_bar(aes(fill=unit),stat="count",position="stack") + 
  scale_x_continuous(breaks=seq(94,102,1),limits=c(94,100)) +
  coord_flip() + facet_grid(.~SID)

MFR.badf.stat2 <- ddply(
  ddply(MFR.badfm%>%filter(CL>499),c("SID","ident5"),summarise,count=length(SID)),
  c("ident5"),summarise,mean.count=mean(count),sd=sd(count))
MFR.badf.stat3 <- ddply(
  ddply(MFR.badfm%>%filter(CL>499),c("SID","ident5","unit"),summarise,count=length(SID)),
  c("ident5","unit"),summarise,mean.count=mean(count))

ggplot(MFR.badf.stat3,aes(x=ident5,y=mean.count)) + geom_bar(aes(fill=unit),stat="identity",position="stack") + 
  geom_errorbar(data=MFR.badf.stat2,aes(ymax=mean.count+sd,ymin=mean.count-sd),width=.1) + 
  scale_x_continuous(breaks=seq(94,102,1),limits=c(94,100)) +
  coord_flip()
```



compare annotation : mock ref vs. silva
```{r}
levels(MFR.badfm$tax) <- gsub("_"," ",levels(MFR.badfm$tax))
levels(MFR.silva.anno$tax) <- gsub("_"," ",levels(MFR.silva.anno$tax))

MFR.compare.anno <- merge(MFR.badfm,MFR.silva.anno,by=c("SID","LOTU"),all.x=T)

MFR.compare.anno$consitLv <- as.factor((ddply(MFR.compare.anno[,c("SID","LOTU","tax.x","tax.y")],c("SID","LOTU"),checkTaxConsitRank))$V1)
levels(MFR.compare.anno$consitLv) <- c("different","genus","species")
```

```{r}
MFR.compare.anno$hitDB <- as.factor(ifelse(MFR.compare.anno$scov.ITS==0,0,1)  + ifelse(MFR.compare.anno$scov.SSU==0,0,2) + 
  ifelse(MFR.compare.anno$scov.LSU==0,0,4))
levels(MFR.compare.anno$hitDB) <- c("UNITE.ITS","SILVA.SSU","SILVA.LSU","SILVA.LSU+UNITE.ITS")

MFR.compare.anno$reachLimit <- as.factor(ifelse(MFR.compare.anno$scov.ITS>99,T,ifelse(MFR.compare.anno$scov.SSU>99,T,ifelse(MFR.compare.anno$scov.LSU>99,T,0))))

ggplot(MFR.compare.anno,aes(x=ident,y=idents,color=consitLv)) + geom_point()
ggplot(MFR.compare.anno,aes(x=idents-ident,color=consitLv)) + geom_density()

ggplot(MFR.compare.anno,aes(x=qMapLen,y=lens,color=consitLv)) + geom_point()
ggplot(MFR.compare.anno,aes(x=bitScore,y=bit,color=factor(reachLimit))) + geom_point(alpha=.5)
ggplot(MFR.compare.anno,aes(x=bitScore,y=bit,color=factor(hitDB))) + geom_point(alpha=.5)
ggplot(MFR.compare.anno,aes(x=bitScore,y=bit,color=hitDB,shape=reachLimit)) +
  geom_abline(slope = 1,intercept = 0, linetype=2,alpha=.5,size=2,color="grey50") +
  geom_point(alpha=.5) + scale_shape_manual(values=c(16,4))

ggplot(MFR.compare.anno%>%filter(!is.na(consitLv)),aes(x=idents,color=consitLv)) + geom_density(size=1,alpha=.5) +
  theme(legend.position = c(0.9,.98),legend.justification = c(1,1)) + facet_grid(hitDB~.)
ggplot(MFR.compare.anno,aes(x=mainUnit,fill=mainUnit)) + geom_bar() +facet_grid(.~consitLv) 
ggplot(MFR.compare.anno,aes(x=bitScore,y=bit,color=mainUnit,linetype=reachLimit)) + geom_point(alpha=.5) +facet_grid(.~consitLv)

```

completness of subunits
```{r}
ZS.stat.comp <- melt(ZBR.compare.anno%>%filter(!is.na(consitLv)),measure.vars = c("ALL","ITS1","SSU","LSU"),variable.name = "subunits", value.name = "completness")
ZS.stat.comp$qlen.seg <- round(ZS.stat.comp$qlen/500,0)*500
ggplot(ZS.stat.comp%>%filter(qlen.seg>400&qlen.seg<3500),aes(x=completness,fill=subunits)) + geom_histogram(position="dodge") +
  facet_grid(qlen.seg~.,scale="free")

```

comapre with mock ref annotation
```{r}
fungi.copy.num <- as.data.frame(table(MFR.qdf$tax.x))
colnames(fungi.copy.num) <- c("species","Tval")
fungi.copy.num$Tval <- 100/7
fungi.copy.num$genus <- sapply(strsplit(as.character(fungi.copy.num$species)," "),"[",1)
selectTax <- as.character(fungi.copy.num$species)
selectGenus <- sapply(strsplit(selectTax," "),"[",1)

MFR.qdf <- merge(MFR.prop,MFR.compare.anno,by=c("SID","LOTU"),all.x=T)
MFR.qdf$species <- MFR.qdf$tax.y
MFR.qdf$genus <- sapply(strsplit(as.character(MFR.qdf$taxPath),";"),"[",6)
MFR.qdf$selectTax <- ifelse(MFR.qdf$species%in%selectTax,as.character(MFR.qdf$species),"Others")
MFR.qdf$checkLv <- ifelse(MFR.qdf$species%in%selectTax,"Same Sp.",ifelse(MFR.qdf$genus%in%selectGenus,"Same Genus","Others.above.Genus"))

MFR.qdf$subunit <- as.factor(ifelse(MFR.qdf$scov.ITS>25,1,0)+ifelse(MFR.qdf$scov.SSU>25,2,0)+ifelse(MFR.qdf$scov.LSU>25,4,0))
MFR.stat <- ddply(MFR.qdf%>%filter(LOTU!='*'),c("SID","rand","tax.x"),summarise,count=sum(count))
MFR.stat <- ddply(MFR.stat,c("SID","rand"),transform,pct=count/sum(count)*100)
MFR.stat1 <- merge(MFR.stat,fungi.copy.num[,c("species","TC16S")],by.x="tax.x",by.y="species",all.x=T)
ggplot(MFR.stat1%>%filter(SID!="Z3"),aes(x=tax.x,y=pct,fill=factor(rand))) + 
  geom_bar(aes(y=TC16S),fill=NA,color="grey50",stat="identity",position="dodge") +
  geom_bar(alpha=.5,stat = 'identity', position = 'dodge') +
  facet_grid(SID~.) + coord_flip()

MFR.stat2 <- ddply(MFR.qdf%>%filter(ident>99.5),c("SID","rand","tax.x"),summarise,count=sum(count))
MFR.stat2 <- ddply(MFR.stat2,c("SID","rand"),transform,pct=count/sum(count)*100)
MFR.stat2 <- merge(MFR.stat2,fungi.copy.num[,c("species","TC16S")],by.x="tax.x",by.y="species",all.x=T)
ggplot(MFR.stat2%>%filter(SID!="Z3"),aes(x=tax.x,y=pct,fill=factor(rand))) + 
  geom_bar(aes(y=TC16S),fill=NA,color="grey50",stat="identity",position="dodge") +
  geom_bar(alpha=.5,stat = 'identity', position = 'dodge') +
  facet_grid(SID~.) + coord_flip()


```

```{r}
MFR.qdf$selectTax <- ifelse(MFR.qdf$species%in%selectTax,as.character(MFR.qdf$species),ifelse(MFR.qdf$genus%in%selectGenus,"Others.in.Genus","Others.above.Genus"))

MFR.stat3 <- ddply(MFR.qdf%>%filter(idents>99.5),c("SID","rand","selectTax"),summarise,count=sum(count))
MFR.stat3 <- merge(ddply(MFR.stat3,c("SID","rand"),transform, pct=count/sum(count)*100),
                  fungi.copy.num[,c("species","Tval")],by.x="selectTax",by.y="species",all.x=T)
MFR.stat3$Tval[which(is.na(MFR.stat3$Tval))] <- 0
MFR.stat3$pct.delta <- MFR.stat3$pct-MFR.stat3$Tval
ggplot(MFR.stat3,aes(x=SID,y=pct,fill=factor(rand))) +geom_bar(stat="identity",position="dodge") +facet_grid(.~selectTax)
ggplot(MFR.stat3,aes(x=SID,y=pct,fill=factor(rand))) + geom_bar(aes(y=Tval),fill=NA,color="grey50",stat="identity",position="dodge") +
  geom_bar(stat="identity",position="dodge",alpha=.5) +facet_grid(.~selectTax)

MFR.mean.df <- ddply(MFR.stat3,c("selectTax"),summarise,sd=sd(pct.delta),pct=mean(pct.delta))
MFR.mean.df <- MFR.mean.df%>%filter(selectTax%in%selectTax)
colnames(MFR.mean.df)[1] <- "selectTax"


MFR.mean.df$selectTax <- as.factor(as.character(MFR.mean.df$selectTax))

MFR.mean.df <- MFR.mean.df[order(MFR.mean.df$pct),]
MFR.mean.df$selectTax <- reorder(MFR.mean.df$selectTax,MFR.mean.df$pct,mean)
ggplot(MFR.mean.df,aes(x=selectTax,y=pct,fill=selectTax)) + geom_hline(yintercept = 0,linetype=2,color="grey50",size=2) +
  geom_bar(stat = 'identity', position = 'dodge') + geom_errorbar(aes(ymax=pct+sd,ymin=pct-sd),width=.4) + coord_flip() +
  ylab("delta of quantificatin \nover theoretical value (%)") + guides(fill=F)

```


#FIN.